{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "62042f97-a76e-4b34-805f-3f1ac81e7288",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import rpy2\n",
    "import logging\n",
    "import warnings\n",
    "import anndata2ri\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import decoupler as dc\n",
    "from anndata import AnnData\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "from matplotlib.pyplot import rcParams\n",
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "447b6d1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.8.0\n",
      "scanpy      1.9.3\n",
      "-----\n",
      "PIL                         9.5.0\n",
      "anndata2ri                  1.1\n",
      "appnope                     0.1.3\n",
      "asttokens                   NA\n",
      "backcall                    0.2.0\n",
      "cffi                        1.15.1\n",
      "comm                        0.1.3\n",
      "cycler                      0.10.0\n",
      "cython_runtime              NA\n",
      "dateutil                    2.8.2\n",
      "debugpy                     1.6.7\n",
      "decorator                   5.1.1\n",
      "decoupler                   1.4.0\n",
      "executing                   1.2.0\n",
      "google                      NA\n",
      "h5py                        3.9.0\n",
      "igraph                      0.10.4\n",
      "ipykernel                   6.23.2\n",
      "ipywidgets                  8.0.6\n",
      "jedi                        0.18.2\n",
      "jinja2                      3.1.2\n",
      "joblib                      1.2.0\n",
      "kiwisolver                  1.4.4\n",
      "leidenalg                   0.9.1\n",
      "llvmlite                    0.39.1\n",
      "louvain                     0.8.0\n",
      "markupsafe                  2.1.3\n",
      "matplotlib                  3.7.1\n",
      "mpl_toolkits                NA\n",
      "natsort                     8.4.0\n",
      "numba                       0.56.4\n",
      "numexpr                     2.8.4\n",
      "numpy                       1.23.5\n",
      "packaging                   23.1\n",
      "pandas                      2.0.2\n",
      "parso                       0.8.3\n",
      "pexpect                     4.8.0\n",
      "pickleshare                 0.7.5\n",
      "pkg_resources               NA\n",
      "platformdirs                3.7.0\n",
      "prompt_toolkit              3.0.38\n",
      "psutil                      5.9.5\n",
      "ptyprocess                  0.7.0\n",
      "pure_eval                   0.2.2\n",
      "pyarrow                     12.0.1\n",
      "pycparser                   2.21\n",
      "pydev_ipython               NA\n",
      "pydevconsole                NA\n",
      "pydevd                      2.9.5\n",
      "pydevd_file_utils           NA\n",
      "pydevd_plugins              NA\n",
      "pydevd_tracing              NA\n",
      "pygments                    2.15.1\n",
      "pyparsing                   3.1.0\n",
      "pytz                        2023.3\n",
      "pytz_deprecation_shim       NA\n",
      "rpy2                        3.5.12\n",
      "scipy                       1.10.1\n",
      "session_info                1.0.0\n",
      "setuptools                  68.0.0\n",
      "six                         1.16.0\n",
      "sklearn                     1.2.2\n",
      "sparse                      0.14.0\n",
      "stack_data                  0.6.2\n",
      "texttable                   1.6.7\n",
      "threadpoolctl               3.1.0\n",
      "tornado                     6.3.2\n",
      "tqdm                        4.65.0\n",
      "traitlets                   5.9.0\n",
      "typing_extensions           NA\n",
      "tzlocal                     NA\n",
      "wcwidth                     0.2.6\n",
      "yaml                        6.0\n",
      "zmq                         25.1.0\n",
      "zoneinfo                    NA\n",
      "-----\n",
      "IPython             8.14.0\n",
      "jupyter_client      8.2.0\n",
      "jupyter_core        5.3.1\n",
      "-----\n",
      "Python 3.10.6 (main, Oct 24 2022, 11:04:34) [Clang 12.0.0 ]\n",
      "macOS-10.16-x86_64-i386-64bit\n",
      "-----\n",
      "Session information updated at 2023-11-22 10:13\n"
     ]
    }
   ],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "#rcParams['figure.figsize']=(4,4) #rescale figures\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "30cd9808",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "data": {
      "text/plain": [
       "In addition: Warning messages:\n",
       "1: package ‘DESeq2’ was built under R version 4.3.1 \n",
       "2: package ‘S4Vectors’ was built under R version 4.3.1 \n",
       "3: package ‘IRanges’ was built under R version 4.3.1 \n",
       "4: package ‘GenomeInfoDb’ was built under R version 4.3.1 \n",
       "5: package ‘MatrixGenerics’ was built under R version 4.3.1 \n",
       "6: In checkMatrixPackageVersion() :\n",
       "  Package version inconsistency detected.\n",
       "TMB was built with Matrix version 1.6.0\n",
       "Current Matrix version is 1.6.1.1\n",
       "Please re-install 'TMB' from source using install.packages('TMB', type = 'source') or ask CRAN for a binary version of 'TMB' matching CRAN's 'Matrix' package\n",
       "7: In checkDepPackageVersion(dep_pkg = \"TMB\") :\n",
       "  Package version inconsistency detected.\n",
       "glmmTMB was built with TMB version 1.9.3\n",
       "Current TMB version is 1.9.6\n",
       "Please re-install glmmTMB from source or restore original ‘TMB’ package (see '?reinstalling' for more information)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "suppressPackageStartupMessages({\n",
    "    library(reticulate)\n",
    "    library(ggplot2)\n",
    "    library(tidyr)\n",
    "    library(dplyr)\n",
    "    library(purrr)\n",
    "    library(Seurat)\n",
    "    library(tibble)\n",
    "    library(magrittr) \n",
    "    library(forcats)\n",
    "    library(Matrix)\n",
    "    library(stats)\n",
    "    library(tester)\n",
    "    library(Seurat)\n",
    "    library(methods)\n",
    "    library(matrixStats)\n",
    "    library(edgeR)\n",
    "    library(DESeq2)\n",
    "    library(limma)\n",
    "    library(pbmcapply)\n",
    "    library(parallel)\n",
    "    library(lmerTest)\n",
    "    library(lme4)\n",
    "    library(glmmTMB)\n",
    "    library(blme)\n",
    "# needs to be run every time you start R and want to use %>%\n",
    "})\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "d40204d0-56d5-4984-a2af-1075b64098f1",
   "metadata": {},
   "source": [
    "## **Prepare data**\n",
    "\n",
    "Now, we load the preprocessed and annotated data for downstream analysis."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3bc72a63",
   "metadata": {},
   "source": [
    "Please set `get_cell_types=True` if, `cell_type` column is absent or contains celltype annotation not of the form\n",
    "\n",
    "- `Excitatory`, `Inhibitory`, `Astrocyte`, `Oligodendrocyte`, `OPC`, `Microglia`, `Endothelial`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e763f1ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_prefix = 'leng_etc'\n",
    "get_cell_types = True\n",
    "\n",
    "adata_annot = sc.read_h5ad(f'../data/raw/{save_prefix}/{save_prefix}_raw_anndata.h5ad')\n",
    "adata_annot.obs_names_make_unique()\n",
    "adata_annot.var_names_make_unique()\n",
    "\n",
    "if 'counts' not in adata_annot.layers.keys():\n",
    "    adata_annot.layers['counts'] = adata_annot.X.copy()\n",
    "else:\n",
    "    adata_annot.X = adata_annot.layers['counts'].copy()\n",
    "    \n",
    "del adata_annot.obsm, adata_annot.layers, adata_annot.varm, adata_annot.uns, adata_annot.obsp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c741d75e",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltypes = [\"Excitatory\", \"Inhibitory\", \"Astrocyte\", \"Microglia\", \"Oligodendrocyte\", \"OPC\", 'Endothelial'] \n",
    "\n",
    "mapping = {'leng_etc':\n",
    "           \n",
    "           {'Exc': 'Excitatory', \n",
    "            'Inh': 'Inhibitory', \n",
    "            'Astro': 'Astrocyte',\n",
    "            'Endo': 'Endothelial', \n",
    "            'Micro': 'Microglia', \n",
    "            'OPC': 'OPC', \n",
    "            'Oligo': 'Oligodendrocyte'},\n",
    "\n",
    "           'leng_sfg':\n",
    "           \n",
    "           {'Exc': 'Excitatory', \n",
    "            'Inh': 'Inhibitory', \n",
    "            'Astro': 'Astrocyte',\n",
    "            'Endo': 'Endothelial', \n",
    "            'Micro': 'Microglia', \n",
    "            'OPC': 'OPC', \n",
    "            'Oligo': 'Oligodendrocyte'},\n",
    "            \n",
    "            'seaad_mtg':\n",
    "           \n",
    "           {'Excitatory': 'Excitatory', \n",
    "            'Inhibitory': 'Inhibitory',\n",
    "            'Astrocyte': 'Astrocyte',\n",
    "            'Microglia': 'Microglia', \n",
    "            'Endothelial': 'Endothelial', \n",
    "            'OPC': 'OPC',\n",
    "            'Oligodendrocyte': 'Oligodendrocyte'},\n",
    "            }\n",
    "\n",
    "cell_column = {'leng_etc': 'clusterCellType',\n",
    "               'leng_sfg': 'clusterCellType',\n",
    "               'seaad_mtg': 'cell_type',\n",
    "               }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "093d77f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "if get_cell_types:\n",
    "    adata_annot.obs['cell_type'] = adata_annot.obs[cell_column[save_prefix]].map(mapping[save_prefix])\n",
    "    adata_annot = adata_annot[adata_annot.obs.cell_type.isin(celltypes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fb234cad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 42528 × 33694\n",
       "    obs: 'SampleID', 'PatientID', 'BrainRegion', 'BraakStage', 'SampleBatch', 'nUMI', 'nGene', 'initialClusterAssignments', 'seurat.clusters', 'clusterAssignment', 'clusterCellType', 'cell_type'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_annot"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "53847a6f",
   "metadata": {},
   "source": [
    "## **Additional Parametersa**\n",
    "\n",
    "Now we specify other related information\n",
    "\n",
    "Specify the following:\n",
    "\n",
    "- `metadata`: Path to metadata. Metadata must contain a column called `pathology.group` with the only unique groups being `no`, `early`, and `late`.\n",
    "\n",
    "- `map_meta`: whether to map metadata to obtain `pathology.group`. If False, it will be assumed that `pathology.group` exist in `adata.obs`\n",
    "\n",
    "- `test_names`: List of the different test names of interest.\n",
    "\n",
    "- `save_prefix`: Prefix for saving critical files. preferably chosen to be in the format `{source name}_{brain region}`. e.g `mathys_pfc`\n",
    "\n",
    "- `subject_id`: Name of the column containing Subject/Patient ID in metadata and `.obs`\n",
    "\n",
    "- `covariates`: This should be a list of additional confounding covariates (`not including pathology.group`) e.g `Sex`, `Sample Batch`, `Age` and other factors that might not be of interest but might have an effect on the pathological status. If you would not like to include any cofounders, please set `covaraites = ['None']`.\n",
    "\n",
    "- `filter_genes`: Specifies whether to filter genes using `gene_celltype_threshold` before before performing differential expression tests`\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a1ba83ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "map_meta = True\n",
    "filter_genes = \"TRUE\"\n",
    "\n",
    "\n",
    "subject_ids_for_study = {'leng_sfg': 'PatientID',\n",
    "                        'leng_etc': 'PatientID',\n",
    "                        'seaad_mtg': 'Donor ID'}\n",
    "\n",
    "subject_id = subject_ids_for_study[save_prefix]     # for leng this is `PatientID` for mathys is 'Subject', and allen is 'individualID'\n",
    "gene_celltype_threshold = 0.10                      # determines number of cells the gene must be expressed in \n",
    "\n",
    "\n",
    "covaraites_for_study = {'leng_sfg': ['None'],  # ['ageDeath.cat',],\n",
    "                        'leng_etc': ['None'],  # ['ageDeath.cat',]\n",
    "                        'seaad_mtg': ['None'],\n",
    "                        }\n",
    "\n",
    "covariates = covaraites_for_study[save_prefix]                               # list of covariates to be accounted for in regression.\n",
    "\n",
    "test_names = ['early_vs_no', 'late_vs_early', 'late_vs_no', 'ad_vs_no']\n",
    "metadata = f'../data/raw/{save_prefix}/{save_prefix}_metadata.csv'  # path to metadata\n",
    "\n",
    "deg_methods_to_run = robjects.ListVector({\n",
    "                                        'single_cell_methods': \"FALSE\",\n",
    "                                        'pseudo_bulk_methods': \"TRUE\",\n",
    "                                        'mixed_model_methods': \"FALSE\",\n",
    "                                        })\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e772597d",
   "metadata": {},
   "source": [
    "## **Methods**\n",
    "\n",
    "Here we implement a modified version of the R package, `Libra`, provided by [**Squair et al. 2021**](https://www.nature.com/articles/s41467-021-25960-2), which implementing all methods for DE analysis discussed in the study within a consistent interface. \n",
    "The source package is available from [GitHub](https://github.com/neurorestore/Libra) and as Supplementary Software. \n",
    "\n",
    "The source package implements a total of `22 unique differential expression methods` that can all be accessed from a single function. These methods are listed below\n",
    "\n",
    "By default the pipeline will use a pseudobulk approach, implementing the `edgeR` package with a likelihood ratio test (LRT) null hypothesis testing framework. Each of the 22 tests can be accessed through three key variables of the `run_de` function: `de_family`, `de_method`, and `de_type`. Their precise access arguments are summarized in the below table.\n",
    "\n",
    "| Method | de_family | de_method | de_type |\n",
    "|--------|-----------|-----------|---------|\n",
    "Wilcoxon Rank-Sum test | singlecell | wilcox | |\n",
    "Likelihood ratio test | singecell | bimod | |\n",
    "Student's t-test | singlecell | t | |\n",
    "Negative binomial linear model | singlecell | negbinom | |\n",
    "Logistic regression | singlecell | LR | |\n",
    "MAST | singlecell | MAST | |\n",
    "edgeR-LRT | pseudobulk | edgeR | LRT\n",
    "edgeR-QLF | pseudobulk | edgeR | QLF\n",
    "DESeq2-LRT | pseudobulk | DESeq2 | LRT\n",
    "**DESeq2-Wald** | **pseudobulk** | **DESeq2** | **Wald**\n",
    "limma-trend | pseudobulk | limma | trend\n",
    "limma-voom | pseudobulk | limma | voom\n",
    "Linear mixed model | mixedmodel | linear | Wald\n",
    "Linear mixed model-LRT | mixedmodel | linear | LRT\n",
    "Negative binomial generalized linear mixed model | mixedmodel | negbinom | Wald\n",
    "Negative binomial generalized linear mixed model-LRT | mixedmodel | negbinom | LRT\n",
    "Negative binomial generalized linear mixed model with offset | mixedmodel | negbinom_offset | Wald\n",
    "Negative binomial generalized linear mixed model with offset-LRT | mixedmodel | negbinom_offset | LRT\n",
    "Poisson generalized linear mixed model | mixedmodel | poisson | Wald\n",
    "Poisson generalized linear mixed model-LRT | mixedmodel | poisson | LRT\n",
    "Poisson generalized linear mixed model with offset | mixedmodel | poisson_offset | Wald\n",
    "Poisson generalized linear mixed model with offset-LRT | mixedmodel | poisson_offset | LRT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2de08892",
   "metadata": {},
   "outputs": [],
   "source": [
    "single_cell_methods = ['wilcox', 'MAST']\n",
    "pseudo_bulk_methods = ['DESeq2-Wald']\n",
    "mixed_model_methods = ['poisson-LRT', 'negbinom-LRT']"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "edb67647",
   "metadata": {},
   "source": [
    "## **Map Metadata**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d3301215",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta = pd.read_csv(metadata, encoding_errors='ignore')\n",
    "meta = meta.astype(str)\n",
    "mapping = dict(zip(meta[subject_id], meta['pathology.group']))\n",
    "adata_annot.obs['pathology.group'] = adata_annot.obs[subject_id].map(mapping)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "866e1208",
   "metadata": {},
   "source": [
    "## **Loading data into memory**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c09d6bfb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filtering genes in Excitatory...\n",
      "filtering genes in Inhibitory...\n",
      "filtering genes in Astrocyte...\n",
      "filtering genes in Microglia...\n",
      "filtering genes in Oligodendrocyte...\n",
      "filtering genes in OPC...\n"
     ]
    }
   ],
   "source": [
    "pdata = dc.get_pseudobulk(adata_annot,\n",
    "                          sample_col=subject_id,\n",
    "                          groups_col='cell_type',\n",
    "                          layer=None,\n",
    "                          mode='sum',\n",
    "                          min_cells=0,\n",
    "                          min_counts=0\n",
    "                         )\n",
    "pdata\n",
    "genes_to_keep = {}\n",
    "\n",
    "celltypes = ['Excitatory', 'Inhibitory', 'Astrocyte', 'Microglia', 'Oligodendrocyte', 'OPC']\n",
    "\n",
    "for cell_type in celltypes:\n",
    "    print(f'filtering genes in {cell_type}...')\n",
    "    genes_to_keep[cell_type] = dc.filter_by_prop(pdata[pdata.obs['cell_type'] == cell_type].copy(), min_prop=gene_celltype_threshold)\n",
    "\n",
    "genes_to_keep_list = robjects.ListVector(genes_to_keep)\n",
    "\n",
    "del pdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3a679ec8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "class: SingleCellExperiment \n",
      "dim: 33694 42528 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B\n",
      "rowData names(0):\n",
      "colnames(42528): EC2_AAACCTGAGGATGCGT EC2_AAACCTGAGTCAATAG ...\n",
      "  EC10_TTTGTCATCTATCGCC EC10_TTTGTCATCTCTGCTG\n",
      "colData names(13): SampleID PatientID ... cell_type pathology.group\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "[1] \"loaded data into memory for recursive use\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i adata_annot -i genes_to_keep_list -i deg_methods_to_run -i celltypes\n",
    "\n",
    "print(adata_annot)\n",
    "\n",
    "print('loaded data into memory for recursive use')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c10e5b17",
   "metadata": {},
   "source": [
    "#### **Pseduo-bulk Differential Expression Analysis with custom scripts adapted from [**Squair et al. 2021**](https://www.nature.com/articles/s41467-021-25960-2)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "66f7e120",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"--------------------------------\"\n",
      "[1] \"Estimating DEGs in Excitatory...\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Excitatory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Excitatory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Excitatory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Excitatory\"\n",
      "[1] \"........................\"\n",
      "[1] \"--------------------------------\"\n",
      "[1] \"Estimating DEGs in Inhibitory...\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Inhibitory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Inhibitory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Inhibitory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Inhibitory\"\n",
      "[1] \"........................\"\n",
      "[1] \"--------------------------------\"\n",
      "[1] \"Estimating DEGs in Astrocyte...\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Astrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Astrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Astrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Astrocyte\"\n",
      "[1] \"........................\"\n",
      "[1] \"--------------------------------\"\n",
      "[1] \"Estimating DEGs in Microglia...\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Microglia\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Microglia\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Microglia\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Microglia\"\n",
      "[1] \"........................\"\n",
      "[1] \"--------------------------------\"\n",
      "[1] \"Estimating DEGs in Oligodendrocyte...\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Oligodendrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Oligodendrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Oligodendrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Oligodendrocyte\"\n",
      "[1] \"........................\"\n",
      "[1] \"--------------------------------\"\n",
      "[1] \"Estimating DEGs in OPC...\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"OPC\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"OPC\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"OPC\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"OPC\"\n",
      "[1] \"........................\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "converting counts to integer mode\n",
       "estimating size factors\n",
       "estimating dispersions\n",
       "gene-wise dispersion estimates\n",
       "mean-dispersion relationship\n",
       "final dispersion estimates\n",
       "fitting model and testing\n",
       "In addition: Warning message:\n",
       "package ‘scuttle’ was built under R version 4.3.1 \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i covariates -i test_names -i subject_id -i pseudo_bulk_methods -i gene_celltype_threshold -o pseudobulk_degs\n",
    "\n",
    "library(scuttle)\n",
    "\n",
    "source('../scripts/functions/deg_functions/run_de.R')\n",
    "source('../scripts/functions/deg_functions/pseudobulk_de.R')\n",
    "source('../scripts/functions/deg_functions/check_inputs.R')\n",
    "source('../scripts/functions/deg_functions/to_pseudobulk.R')\n",
    "source('../scripts/functions/deg_functions/mixedmodel_de.R')\n",
    "source('../scripts/functions/deg_functions/singlecell_de.R')\n",
    "\n",
    "if ('None' %in% covariates){\n",
    "    latent_vars = NULL\n",
    "} else {\n",
    "    latent_vars = covariates\n",
    "}\n",
    "\n",
    "\n",
    "if (eval(deg_methods_to_run[['pseudo_bulk_methods']])){ \n",
    "    \n",
    "    pseudobulk_degs <- list()\n",
    "    for (celltype in celltypes){\n",
    "\n",
    "        pseudobulk_degs[[celltype]] <- list()\n",
    "\n",
    "        # filter out celltypes\n",
    "        print('--------------------------------')\n",
    "        print(paste0('Estimating DEGs in ', celltype, '...'))\n",
    "        sce_cell <- adata_annot[, adata_annot$cell_type == celltype]\n",
    "\n",
    "        # Subset the `SingleCellExperiment` object to include only genes with > 0.1 detection rate/frequency\n",
    "        sce_cell <- sce_cell[genes_to_keep_list[[celltype]], ]\n",
    "                            \n",
    "        meta = colData(sce_cell)\n",
    "\n",
    "        for (test_name in test_names){\n",
    "\n",
    "            sce <- sce_cell\n",
    "\n",
    "            pseudobulk_degs[[celltype]][[test_name]] <- list()\n",
    "            \n",
    "            ref_level = 'no'\n",
    "            meta$pathology.group <- factor(meta$pathology.group)\n",
    "            meta$pathology.group <- relevel(meta$pathology.group, ref = ref_level)\n",
    "\n",
    "            for(de_method in pseudo_bulk_methods){\n",
    "                print('----------------------------------')\n",
    "                print(paste0('Obtaining statistics for ', toupper(test_name), ' with pseudo-bulk method ', toupper(de_method)))\n",
    "                pseudobulk_degs[[celltype]][[test_name]][[de_method]] = run_de(sce,\n",
    "                                                                                meta = meta,\n",
    "                                                                                replicate_col = ifelse(grepl(\" \", subject_id), gsub(\" \", \".\", subject_id), subject_id),\n",
    "                                                                                cell_type_col = 'cell_type',\n",
    "                                                                                label_col = 'pathology.group',\n",
    "                                                                                latent_vars = latent_vars,\n",
    "                                                                                test_name = test_name,\n",
    "                                                                                ref_level = ref_level,\n",
    "                                                                                min_cells = 3,\n",
    "                                                                                min_reps = 2,\n",
    "                                                                                min_features = 0,\n",
    "                                                                                de_family = 'pseudobulk',\n",
    "                                                                                de_method = strsplit(de_method, \"-\")[[1]][1],\n",
    "                                                                                de_type = strsplit(de_method, \"-\")[[1]][2],\n",
    "                                                                                n_threads = 2)\n",
    "            }\n",
    "        }\n",
    "\n",
    "        print('........................')\n",
    "            \n",
    "    }\n",
    "}   else{\n",
    "    print(\"Not evaluating DEGs with Pseudobulk Methods.\")\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "6bf927c1",
   "metadata": {},
   "source": [
    "##### **Save Results**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "00b0b482",
   "metadata": {},
   "outputs": [],
   "source": [
    "if eval(dict(zip(deg_methods_to_run.names, list(deg_methods_to_run)))['pseudo_bulk_methods'][0].capitalize()):\n",
    "    for test_name in test_names:\n",
    "\n",
    "        fig_dir = f'../results/{test_name}/{save_prefix}/DEG/'\n",
    "\n",
    "        if not os.path.exists(fig_dir):\n",
    "            os.makedirs(fig_dir)\n",
    "\n",
    "        for deg_method in pseudo_bulk_methods:\n",
    "            with pd.ExcelWriter(fig_dir+f'{deg_method}_degs.xlsx') as writer:\n",
    "                for cell_type in celltypes:\n",
    "                    \n",
    "                    df = pseudobulk_degs[cell_type][test_name][deg_method].sort_values(by='p_val_adj')\n",
    "                    df['abs_logFC'] = abs(df['avg_logFC'])\n",
    "                    df['direction'] = df['avg_logFC'].apply(lambda x: \"up\" if x>0 else \"down\")\n",
    "                    df.to_excel(writer, sheet_name=cell_type, na_rep='NA')\n",
    "else:\n",
    "    print('No result saved for Pseudobulk Methods')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ae412e1f",
   "metadata": {},
   "source": [
    "### **Mixed Models**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5a7d8857",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Not evaluating DEGs with Mixed-model Methods.\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i mixed_model_methods -o mixedmodel_degs\n",
    "\n",
    "if ('None' %in% covariates){\n",
    "    latent_vars = NULL\n",
    "} else {\n",
    "    latent_vars = covariates\n",
    "}\n",
    "\n",
    "if (eval(deg_methods_to_run[['mixed_model_methods']])){ \n",
    "    mixedmodel_degs <- list()\n",
    "    for (celltype in celltypes){\n",
    "\n",
    "        mixedmodel_degs[[celltype]] <- list()\n",
    "\n",
    "        # filter out celltypes\n",
    "        print('--------------------------------')\n",
    "        print(paste0('Estimating DEGs in ', celltype, '...'))\n",
    "        sce_cell <- adata_annot[, adata_annot$cell_type == celltype]\n",
    "\n",
    "        assay(sce_cell, 'counts') <- assay(sce_cell)\n",
    "\n",
    "        # Calculate the detection rate/frequency for each gene\n",
    "        detection_rate <- rowMeans(counts(sce_cell) > 0)\n",
    "        # Subset the `SingleCellExperiment` object to include only genes with >0.1 detection rate/frequency\n",
    "        sce_cell <- sce_cell[detection_rate > gene_celltype_threshold, ]\n",
    "        \n",
    "        meta = colData(sce_cell)\n",
    "\n",
    "        for (test_name in test_names){\n",
    "            # print('---------------------------------')\n",
    "            # print(paste0('Running differential expression test for: ', test_name))\n",
    "            sce <- sce_cell\n",
    "\n",
    "            mixedmodel_degs[[celltype]][[test_name]] <- list()\n",
    "\n",
    "            if (test_name == 'ad_vs_no'){\n",
    "                    sce$pathology.group = ifelse(sce$pathology.group == \"no\", \"no\", \"ad\")\n",
    "                    ref_level <- tail(strsplit('ad_vs_no', \"_vs_\")[[1]], n = 1)\n",
    "                } else if (test_name %in% c('late_vs_no', 'late_vs_early', 'early_vs_no')){\n",
    "                    sce <- sce[, colData(sce)$pathology.group %in% c(strsplit(test_name, \"_vs_\")[[1]])]\n",
    "                    ref_level <- tail(strsplit(test_name, \"_vs_\")[[1]], n = 1)\n",
    "                }\n",
    "            \n",
    "            meta$pathology.group <- factor(meta$pathology.group)\n",
    "            meta$pathology.group <- relevel(meta$pathology.group, ref = ref_level)\n",
    "\n",
    "            for(de_method in mixed_model_methods){\n",
    "                print('----------------------------------')\n",
    "                print(paste0('Obtaining statistics for ', toupper(test_name), ' with single-cell method ', toupper(de_method)))\n",
    "                mixedmodel_degs[[celltype]][[test_name]][[de_method]] = run_de(sce,\n",
    "                                                                                meta = meta,\n",
    "                                                                                replicate_col = ifelse(grepl(\" \", subject_id), gsub(\" \", \".\", subject_id), subject_id),\n",
    "                                                                                cell_type_col = 'cell_type',\n",
    "                                                                                label_col = 'pathology.group',\n",
    "                                                                                latent_vars = latent_vars,\n",
    "                                                                                min_cells = 3,\n",
    "                                                                                min_reps = 2,\n",
    "                                                                                min_features = 0,\n",
    "                                                                                de_family = 'mixedmodel',\n",
    "                                                                                de_method = strsplit(de_method, \"-\")[[1]][1],\n",
    "                                                                                de_type = strsplit(de_method, \"-\")[[1]][2],\n",
    "                                                                                n_threads = 2)\n",
    "            }\n",
    "        }\n",
    "        print('........................')\n",
    "    }\n",
    "}    else{\n",
    "\n",
    "    print(\"Not evaluating DEGs with Mixed-model Methods.\")\n",
    "    mixedmodel_degs = NULL\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "fa3f7229",
   "metadata": {},
   "source": [
    "##### **Save Results**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2106faf6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No result saved for Mixed-model Methods\n"
     ]
    }
   ],
   "source": [
    "if eval(dict(zip(deg_methods_to_run.names, list(deg_methods_to_run)))['mixed_model_methods'][0].capitalize()):\n",
    "    for test_name in test_names:\n",
    "\n",
    "        fig_dir = f'../results/{test_name}/{save_prefix}/DEG/'\n",
    "\n",
    "        if not os.path.exists(fig_dir):\n",
    "            os.makedirs(fig_dir)\n",
    "\n",
    "        for deg_method in pseudo_bulk_methods:\n",
    "            with pd.ExcelWriter(fig_dir+f'{deg_method}_degs.xlsx') as writer:\n",
    "                for cell_type in celltypes:\n",
    "                    \n",
    "                    df = mixedmodel_degs[cell_type][test_name][deg_method].sort_values(by='p_val_adj')\n",
    "                    df['abs_logFC'] = abs(df['avg_logFC'])\n",
    "                    df['direction'] = df['avg_logFC'].apply(lambda x: \"up\" if x>0 else \"down\")\n",
    "                    df.to_excel(writer, sheet_name=cell_type, na_rep='NA')\n",
    "else:\n",
    "    print('No result saved for Mixed-model Methods')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4974ce28",
   "metadata": {},
   "source": [
    "### **Single-cell Models**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "36127617",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Not evaluating DEGs with Single-cell Methods.\"\n"
     ]
    }
   ],
   "source": [
    "%%R  -i single_cell_methods -o singlecell_degs\n",
    "\n",
    "\n",
    "if ('None' %in% covariates){\n",
    "    latent_vars = NULL\n",
    "} else {\n",
    "    latent_vars = covariates\n",
    "}\n",
    "\n",
    "if (eval(deg_methods_to_run[['single_cell_methods']])){ \n",
    "    singlecell_degs <- list()\n",
    "    for (celltype in celltypes){\n",
    "\n",
    "        singlecell_degs[[celltype]] <- list()\n",
    "\n",
    "        # filter out celltypes\n",
    "        print('--------------------------------')\n",
    "        print(paste0('Estimating DEGs in ', celltype, '...'))\n",
    "        sce_cell <- adata_annot[, adata_annot$cell_type == celltype]\n",
    "\n",
    "        # Calculate the detection rate/frequency for each gene\n",
    "        detection_rate <- rowMeans(counts(sce_cell) > 0)\n",
    "        # Subset the `SingleCellExperiment` object to include only genes with >0.1 detection rate/frequency\n",
    "        sce_cell <- sce_cell[detection_rate > gene_celltype_threshold, ]\n",
    "        meta = colData(sce_cell)\n",
    "\n",
    "        for (test_name in test_names){\n",
    "            sce <- sce_cell\n",
    "\n",
    "            singlecell_degs[[celltype]][[test_name]] <- list()\n",
    "\n",
    "            if (test_name == 'ad_vs_no'){\n",
    "                    sce$pathology.group = ifelse(sce$pathology.group == \"no\", \"no\", \"ad\")\n",
    "                    ref_level <- tail(strsplit('ad_vs_no', \"_vs_\")[[1]], n = 1)\n",
    "                } else if (test_name %in% c('late_vs_no', 'late_vs_early', 'early_vs_no')){\n",
    "                    sce <- sce[, colData(sce)$pathology.group %in% c(strsplit(test_name, \"_vs_\")[[1]])]\n",
    "                    ref_level <- tail(strsplit(test_name, \"_vs_\")[[1]], n = 1)\n",
    "                }\n",
    "            \n",
    "            meta$pathology.group <- factor(meta$pathology.group)\n",
    "            meta$pathology.group <- relevel(meta$pathology.group, ref = ref_level)\n",
    "\n",
    "            for(de_method in single_cell_methods){\n",
    "                print('----------------------------------')\n",
    "                print(paste0('Obtaining statistics for ', toupper(test_name), ' with single-cell method ', toupper(de_method)))\n",
    "                singlecell_degs[[celltype]][[test_name]][[de_method]] = run_de(sce,\n",
    "                                                                                meta = meta,\n",
    "                                                                                replicate_col = ifelse(grepl(\" \", subject_id), gsub(\" \", \".\", subject_id), subject_id),\n",
    "                                                                                cell_type_col = 'cell_type',\n",
    "                                                                                label_col = 'pathology.group',\n",
    "                                                                                latent_vars = latent_vars,\n",
    "                                                                                min_cells = 3,\n",
    "                                                                                min_reps = 2,\n",
    "                                                                                min_features = 0,\n",
    "                                                                                de_family = 'singlecell',\n",
    "                                                                                de_method = de_method,\n",
    "                                                                                de_type = NULL,\n",
    "                                                                                n_threads = 2)\n",
    "            }\n",
    "        }\n",
    "\n",
    "        print('........................')\n",
    "    }\n",
    "\n",
    "}    else{\n",
    "\n",
    "    print(\"Not evaluating DEGs with Single-cell Methods.\")\n",
    "    singlecell_degs = NULL\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b18d571a",
   "metadata": {},
   "source": [
    "##### **Save Results**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "02c5cbb5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No result saved for Mixed-model Methods\n"
     ]
    }
   ],
   "source": [
    "if eval(dict(zip(deg_methods_to_run.names, list(deg_methods_to_run)))['single_cell_methods'][0].capitalize()):\n",
    "    for test_name in test_names:\n",
    "        fig_dir = f'../results/{test_name}/{save_prefix}/DEG/'\n",
    "\n",
    "        if not os.path.exists(fig_dir):\n",
    "            os.makedirs(fig_dir)\n",
    "\n",
    "        for deg_method in pseudo_bulk_methods:\n",
    "            with pd.ExcelWriter(fig_dir+f'{deg_method}_degs.xlsx') as writer:\n",
    "                for cell_type in celltypes:\n",
    "                    \n",
    "                    df = singlecell_degs[cell_type][test_name][deg_method].sort_values(by='p_val_adj')\n",
    "                    df['abs_logFC'] = abs(df['avg_logFC'])\n",
    "                    df['direction'] = df['avg_logFC'].apply(lambda x: \"up\" if x>0 else \"down\")\n",
    "                    df.to_excel(writer, sheet_name=cell_type, na_rep='NA')\n",
    "else:\n",
    "    print('No result saved for Mixed-model Methods')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "2f414a000012e1980f5fd1119ed6b9bf8cc73b51e37b52eb52a16ff857effd1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
