{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "62042f97-a76e-4b34-805f-3f1ac81e7288",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Global seed set to 0\n",
      "/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/pytorch_lightning/utilities/warnings.py:53: LightningDeprecationWarning: pytorch_lightning.utilities.warnings.rank_zero_deprecation has been deprecated in v1.6 and will be removed in v1.8. Use the equivalent function from the pytorch_lightning.utilities.rank_zero module instead.\n",
      "  new_rank_zero_deprecation(\n",
      "/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/pytorch_lightning/utilities/warnings.py:58: LightningDeprecationWarning: The `pytorch_lightning.loggers.base.rank_zero_experiment` is deprecated in v1.7 and will be removed in v1.9. Please use `pytorch_lightning.loggers.logger.rank_zero_experiment` instead.\n",
      "  return new_rank_zero_deprecation(*args, **kwargs)\n",
      "/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/phenograph/cluster.py:13: DeprecationWarning: Please use `spmatrix` from the `scipy.sparse` namespace, the `scipy.sparse.base` namespace is deprecated.\n",
      "  from scipy.sparse.base import spmatrix\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import scvi\n",
    "import scgen\n",
    "import rpy2\n",
    "import scib\n",
    "import json\n",
    "import torch\n",
    "import anndata\n",
    "import logging\n",
    "import warnings\n",
    "import scanorama\n",
    "import anndata2ri\n",
    "import matplotlib\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import seaborn as sb\n",
    "import scrublet as scr\n",
    "import doubletdetection\n",
    "from anndata import AnnData\n",
    "from tabnanny import verbose\n",
    "import matplotlib.pyplot as plt\n",
    "from os import PathLike, fspath\n",
    "import rpy2.robjects as robjects\n",
    "from scipy.sparse import csr_matrix\n",
    "from rpy2.robjects import pandas2ri\n",
    "from matplotlib.pyplot import rcParams\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from sklearn.model_selection import train_test_split\n",
    "from pytorch_lightning.loggers import TensorBoardLogger\n",
    "from rpy2.robjects.conversion import localconverter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c36a2bac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sys_dpi(width, height, diag):\n",
    "    '''\n",
    "    obtain dpi of system\n",
    "    \n",
    "    w: width in pixels (if unsure, go vist `whatismyscreenresolution.net`)\n",
    "    h: height in pixels\n",
    "    d: diagonal in inches\n",
    "    '''\n",
    "    w_inches = (diag**2/ (1 + height**2/width**2))**0.5\n",
    "    return round(width/w_inches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "447b6d1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/rpy2/robjects/pandas2ri.py:356: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/rpy2/robjects/numpy2ri.py:245: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/rpy2/robjects/conversion.py:28: DeprecationWarning: The use of converter in module rpy2.robjects.conversion is deprecated. Use rpy2.robjects.conversion.get_conversion() instead of rpy2.robjects.conversion.converter.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.8.0\n",
      "scanpy      1.9.1\n",
      "-----\n",
      "OpenSSL                     22.0.0\n",
      "PIL                         9.2.0\n",
      "absl                        NA\n",
      "adjustText                  NA\n",
      "anndata2ri                  1.1\n",
      "annoy                       NA\n",
      "appnope                     0.1.2\n",
      "asttokens                   NA\n",
      "astunparse                  1.6.3\n",
      "attr                        21.4.0\n",
      "backcall                    0.2.0\n",
      "beta_ufunc                  NA\n",
      "binom_ufunc                 NA\n",
      "boto3                       1.26.32\n",
      "botocore                    1.29.32\n",
      "bottleneck                  1.3.5\n",
      "brotli                      NA\n",
      "certifi                     2022.09.24\n",
      "cffi                        1.15.1\n",
      "chex                        0.1.5\n",
      "cloudpickle                 2.2.0\n",
      "colorama                    0.4.4\n",
      "contextlib2                 NA\n",
      "cryptography                38.0.1\n",
      "cycler                      0.10.0\n",
      "cython_runtime              NA\n",
      "dask                        2022.11.0\n",
      "dateutil                    2.8.2\n",
      "debugpy                     1.5.1\n",
      "decorator                   5.1.1\n",
      "defusedxml                  0.7.1\n",
      "deprecate                   0.3.2\n",
      "deprecated                  1.2.13\n",
      "docrep                      0.3.2\n",
      "doubletdetection            4.2\n",
      "entrypoints                 0.4\n",
      "executing                   0.8.3\n",
      "fbpca                       NA\n",
      "flax                        0.6.1\n",
      "fsspec                      2022.10.0\n",
      "google                      NA\n",
      "h5py                        3.7.0\n",
      "hypergeom_ufunc             NA\n",
      "igraph                      0.10.2\n",
      "intervaltree                NA\n",
      "ipykernel                   6.15.2\n",
      "ipython_genutils            0.2.0\n",
      "ipywidgets                  7.6.5\n",
      "jax                         0.3.24\n",
      "jaxlib                      0.3.24\n",
      "jedi                        0.18.1\n",
      "jinja2                      3.1.2\n",
      "jmespath                    1.0.1\n",
      "joblib                      1.1.1\n",
      "kiwisolver                  1.4.2\n",
      "leidenalg                   0.9.0\n",
      "llvmlite                    0.39.1\n",
      "louvain                     0.8.0\n",
      "markupsafe                  2.1.1\n",
      "matplotlib                  3.5.3\n",
      "ml_collections              NA\n",
      "mpl_toolkits                NA\n",
      "msgpack                     1.0.3\n",
      "mudata                      0.1.2\n",
      "multipledispatch            0.6.0\n",
      "natsort                     7.1.1\n",
      "nbinom_ufunc                NA\n",
      "ncf_ufunc                   NA\n",
      "numba                       0.56.3\n",
      "numexpr                     2.8.3\n",
      "numpy                       1.23.3\n",
      "numpyro                     0.10.1\n",
      "opt_einsum                  v3.3.0\n",
      "optax                       0.1.3\n",
      "packaging                   21.3\n",
      "pandas                      1.4.4\n",
      "parso                       0.8.3\n",
      "pexpect                     4.8.0\n",
      "phenograph                  1.5.7\n",
      "pickleshare                 0.7.5\n",
      "pkg_resources               NA\n",
      "prompt_toolkit              3.0.20\n",
      "psutil                      5.9.0\n",
      "ptyprocess                  0.7.0\n",
      "pure_eval                   0.2.2\n",
      "pyarrow                     11.0.0\n",
      "pydev_ipython               NA\n",
      "pydevconsole                NA\n",
      "pydevd                      2.6.0\n",
      "pydevd_concurrency_analyser NA\n",
      "pydevd_file_utils           NA\n",
      "pydevd_plugins              NA\n",
      "pydevd_tracing              NA\n",
      "pygments                    2.13.0\n",
      "pyparsing                   3.0.9\n",
      "pyro                        1.8.2+2e3bd02\n",
      "pytorch_lightning           1.7.7\n",
      "pytz                        2022.1\n",
      "pytz_deprecation_shim       NA\n",
      "rich                        NA\n",
      "rpy2                        3.5.10\n",
      "scanorama                   1.7.2\n",
      "scgen                       2.1.0\n",
      "scib                        1.0.4\n",
      "scipy                       1.9.3\n",
      "scrublet                    NA\n",
      "scvi                        0.19.0\n",
      "seaborn                     0.11.2\n",
      "session_info                1.0.0\n",
      "setuptools                  65.6.3\n",
      "six                         1.16.0\n",
      "sklearn                     1.1.3\n",
      "sortedcontainers            2.4.0\n",
      "sparse                      0.13.0\n",
      "stack_data                  0.2.0\n",
      "statsmodels                 0.13.5\n",
      "tensorboard                 2.10.1\n",
      "texttable                   1.6.4\n",
      "threadpoolctl               2.2.0\n",
      "tlz                         0.12.0\n",
      "toolz                       0.12.0\n",
      "torch                       1.12.1\n",
      "torchmetrics                0.10.2\n",
      "tornado                     6.2\n",
      "tqdm                        4.64.1\n",
      "traitlets                   5.1.1\n",
      "tree                        0.1.7\n",
      "typing_extensions           NA\n",
      "tzlocal                     NA\n",
      "urllib3                     1.26.12\n",
      "wcwidth                     0.2.5\n",
      "wrapt                       1.14.1\n",
      "yaml                        5.4.1\n",
      "zmq                         23.2.0\n",
      "zoneinfo                    NA\n",
      "-----\n",
      "IPython             8.6.0\n",
      "jupyter_client      7.3.5\n",
      "jupyter_core        4.11.2\n",
      "notebook            6.4.12\n",
      "-----\n",
      "Python 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:41:22) [Clang 13.0.1 ]\n",
      "macOS-13.0.1-arm64-arm-64bit\n",
      "-----\n",
      "Session information updated at 2023-04-18 19:59\n"
     ]
    }
   ],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# # Automatically convert rpy2 outputs to pandas dataframes\n",
    "# pandas2ri.activate()\n",
    "# anndata2ri.activate()\n",
    "# %load_ext rpy2.ipython\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "rcParams['figure.dpi'] = get_sys_dpi(1512, 982, 14.125)\n",
    "#rcParams['figure.figsize']=(4,4) #rescale figures\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "30cd9808",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    }
   ],
   "source": [
    "%%R\n",
    "suppressPackageStartupMessages({\n",
    "    library(reticulate)\n",
    "    library(ggplot2)\n",
    "    library(tidyr)\n",
    "    library(dplyr)\n",
    "    library(purrr)\n",
    "    library(Seurat)\n",
    "    library(tibble)\n",
    "    library(magrittr) # needs to be run every time you start R and want to use %>%\n",
    "})\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1a6a687-9397-4cce-809f-0bca4c0072f6",
   "metadata": {},
   "source": [
    "## Table of contents:\n",
    "\n",
    "  * <a href=#Reading>1. Reading in the data</a>\n",
    "  * <a href=#Preprocessing>2. Systematic differential analysis of gene expression</a>"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "10be21f0-000d-48d7-9e97-4051ec9da0bd",
   "metadata": {},
   "source": [
    "# **1. Reading in the data**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "d40204d0-56d5-4984-a2af-1075b64098f1",
   "metadata": {},
   "source": [
    "### **Prepare data**\n",
    "\n",
    "Now, we load the preprocessed and annotated data for downstream analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e763f1ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_prefix = 'mathys_pfc'\n",
    "\n",
    "adata_annot = sc.read_h5ad(f'../data/processed/{save_prefix}/{save_prefix}_mapped_anndata.h5ad')\n",
    "adata_annot.X = csr_matrix(adata_annot.layers['counts'].toarray())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6263d762",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 65619 × 16590\n",
       "    obs: 'projid', 'fastq', 'Subject', 'sample', 'libraryid', 'study', 'age_death', 'educ', 'msex_x', 'gpath_x', 'amyloid_x', 'plaq_n_x', 'cogdx_x', 'pathologic diagnosis of AD', 'amyloid_y', 'plaq_n_y', 'nft', 'tangles', 'cogn_global_lv', 'gpath_y', 'gpath_3neocort', 'amyloid.group', 'caa_4gp', 'ceradsc', 'braaksc', 'niareagansc', 'cogdx_y', 'msex_y', 'pathology.group', 'sampleid', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'total_counts_mt', 'pct_counts_mt', 'n_genes', 'doublet_score', 'predicted_doublet', 'louvain_0.5', 'louvain_1.0', 'cell_type'\n",
       "    var: 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: 'Subject_colors', 'amyloid.group_colors', 'braaksc_colors', 'cell_type_colors', 'dendrogram_louvain_0.5', 'hvg', 'log1p', 'louvain', 'louvain_0.5_colors', 'louvain_1.0_colors', 'msex_x_colors', 'neighbors', 'ov_t_test_rank_genes_0.5', 'pathology.group_colors', 'pca', 'predicted_doublet_colors', 'scrublet', 't_test_rank_genes_0.5', 'tsne', 'umap'\n",
       "    obsm: 'X_pca', 'X_tsne', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts', 'log'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_annot"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "53847a6f",
   "metadata": {},
   "source": [
    "### **Prepare metadata**\n",
    "\n",
    "Now we specify other related information\n",
    "\n",
    "Specify the following:\n",
    "\n",
    "- `metadata`: Path to metadata. Metadata must contain a column called `pathology.group` with the only unique groups being `no`, `early`, and `late`.\n",
    "\n",
    "- `map_meta`: whether to map metadata to obtain `pathology.group`. If False, it will be assumed that `pathology.group` exist in `adata.obs`\n",
    "\n",
    "- `reference_group`: Name of control group in metadata. This should ideally be `no`, representing the control group \n",
    "\n",
    "- `test_group`: A list of the name of the group(s) in metadata/ `.obs` that should be treated as the test groups.d \n",
    "\n",
    "- `save_prefix`: Prefix for saving critical files. preferably chosen to be in the format `{source name}_{brain region}`. e.g `mathys_pfc`\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a1ba83ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = f'../data/raw/{save_prefix}/{save_prefix}_metadata.csv'\n",
    "reference_group = 'no'                  # name of the control group in metadata \n",
    "test_groups = ['late', 'early']         # list of covariates to be accounted for in regression.\n",
    "filter_genes = True\n",
    "gene_celltype_threshold = 0.05          # determines number of cells the gene must be expressed in \n",
    "test_name = '_n_'.join(test_groups)+'_vs_'+reference_group"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a5d8a01b",
   "metadata": {},
   "source": [
    "## 2.4 Systematic differential analysis of gene expression\n",
    "\n",
    "[**Hansruedi Mathys et. al.**](https://doi.org/10.1038/s41586-019-1195-2) compared gene expression levels between `AD-pathology and no-pathology individuals in a cell type manner. The differential expression analysis was assessed using two tests. \n",
    "\n",
    "- **First**, a cell-level analysis was performed using the Wilcoxon rank-sum test and FDR multiple-testing correction (`FDR-adjusted p-values`). \n",
    "\n",
    "- **Second**, a Poisson mixed model accounting for the individual of origin for nuclei and for unwanted sources of variability was performed using the R packages `lme4` and `RUV-seq`, respectively.\n",
    "\n",
    "\n",
    "Next, we use the ` Wilcoxon rank-sum test` in `scapany.tl.rank_genes_group` comparing `AD-pathology` group to `no-pathology` such that the log foldchange is ;\n",
    "\n",
    "$$ Log_{2} ({Mean\\ Gene\\ Expression\\ in\\ AD\\ category\\ of\\ Cell\\ Type\\ x \\over Mean\\ Gene\\ Expression\\ in\\ Normal\\ category\\ of\\ Cell\\ Type\\ x})$$\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c568575c",
   "metadata": {},
   "source": [
    "##### Group Cells (Pathology / No-Pathology)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "12fa36cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "if disease_group in \n",
    "adata_annot.obs['disease_group'] = None\n",
    "adata_annot.obs.loc[adata_annot.obs['pathology.group'] == 'no-pathology', 'disease_group'] = 'no-pathology'\n",
    "adata_annot.obs.loc[adata_annot.obs['pathology.group'].isin(['early-pathology', 'late-pathology']), 'disease_group'] = 'AD-pathology'"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3830d142",
   "metadata": {},
   "source": [
    "##### Load DEGs from Cell-type Differential Expression with Wilcoxon Rank-Sum Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "64fddc81",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_annot.uns['log1p'] = {'base': None}\n",
    "\n",
    "adata_sub = {}\n",
    "degs_t_test = {}\n",
    "for cell_type in adata_annot.obs.cell_type.unique():\n",
    "\n",
    "    # adata_sub[cell_type] = adata_annot[adata_annot.obs.cell_type==cell_type]\n",
    "    # adata_sub[cell_type].X = adata_sub[cell_type].layers['counts'].toarray()\n",
    "    degs_t_test[cell_type] = pd.read_excel('../results/ad_vs_no/t_test_degs.xlsx', \n",
    "                                   sheet_name=cell_type)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "0f84051c",
   "metadata": {},
   "source": [
    "Reported DEGs\n",
    "\n",
    "Next, we load the reported DEGs from [**Hansruedi Mathys et. al.**](https://doi.org/10.1038/s41586-019-1195-2) for comparison in a later step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f184cab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "mathys_degs = {}\n",
    "for key in adata_annot.obs.cell_type.unique():\n",
    "    try:\n",
    "        mathys_degs[key] = pd.read_excel(f'../data/raw/mathys_pfc_from_paper/degs/ad_vs_no/{key.lower()}_degs.xlsx')\n",
    "        mathys_degs[key].rename({'Unnamed: 0': 'names'}, inplace=True)\n",
    "        mathys_degs[key]['IndModel.absFC'] = np.abs(mathys_degs[key]['IndModel.FC'])\n",
    "\n",
    "    except FileNotFoundError:\n",
    "        continue"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5b076c9",
   "metadata": {},
   "source": [
    "The authors assessed the consistency of DEGs detected using the cell-level analysis model (`Wilcoxon rank-sum test`) with those obtained with the Poisson mixed model (`lme4` combined with `RUV-seq`) by comparing the directionality and rank of DEGs in the two models. Considering the technical challenges, of using these packages, we instead benchmark the `rank-sum test` results against `MAST`, a generalized linear model for modelling scRNA-seq data in R, developed by [**Finak, G. et. al. 2015**](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0844-5). In addition to accounting for the the sample (`Subject`) of origin for each nuclei, we include the cellular detection rate, the fraction of genes expressed in a cell as a sources of unwanted variation/variability. Sveral benchmarking studies have shown that MAST is comparitvely better than other methods for scRNA-seq differential expression testing and is a commonly used tool in scRNA-seq analysis. Thus, We perform differential testing using the `MAST package`. \n",
    "\n",
    "Since MAST is only available in R, we convert our AnnData object into an R object via `anndata2ri`. MAST requires its own data input format, `SingleCellAssy` instead of the `SingleCellExperiment` object produced by the `anndata2ri` conversion. So, to run MAST we thus first put the data into the SingleCellExperiment format, then convert the SCE object into MAST's expected SingleCellAssay (sca) object. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6da78c91",
   "metadata": {},
   "source": [
    "#### MAST implementation details\n",
    "\n",
    "Here, we perform differential testing for AD-pathology vs No-pathology `disease_group`s in each of the cell clusters. Since `MAST` incorporates a zero-inflated negative binomial model tests for differential expression using a hurdle model, it can be very computationally intensive for large data sizes. To remedy this, `we randomly split each data in a stratified manner to inlude <5000 cells across both conditions` while retaining the relative proportion of cells in both test groups (AD-pathology vs No-pathology). This is expected to retain the expression patterns in both tests groups, while speeding up computation\n",
    "\n",
    "In the generalized linear mixed model (specified and fit with the `zlm()` function), we include the test covariate `disease_group` (AD-pathology vs No-pathology), the `Subject` ID (ROS1 -- ROS48), and the number of genes in the cells, reculated from the log normalized values stored in `adata.layers['log']`. The `Subject` IDs are included as random effects to account for unwanted variability/effects due to individual nuclei orgin, mouse-specific that may confound our results. The number of genes is added to fit the technical variability.\n",
    "\n",
    "To test for differences over the `disease_group` covariate we perform a likelihood ratio test (in the `summary()` function call after fitting the model). \n",
    "\n",
    "In post-processing we correct for multiple testing using a Benjamini-Hochberg FDR correction (function `p.adjust()`) and map the Ensembl Gene IDs to gene symbols which are easier to read and interpret."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "abaffa9e",
   "metadata": {},
   "source": [
    "### **Run MAST on Excitatory**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c30c5afc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 7487 × 16590\n",
       "    obs: 'projid', 'fastq', 'Subject', 'sample', 'libraryid', 'study', 'age_death', 'educ', 'msex_x', 'gpath_x', 'amyloid_x', 'plaq_n_x', 'cogdx_x', 'pathologic diagnosis of AD', 'amyloid_y', 'plaq_n_y', 'nft', 'tangles', 'cogn_global_lv', 'gpath_y', 'gpath_3neocort', 'amyloid.group', 'caa_4gp', 'ceradsc', 'braaksc', 'niareagansc', 'cogdx_y', 'msex_y', 'pathology.group', 'sampleid', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'total_counts_mt', 'pct_counts_mt', 'n_genes', 'doublet_score', 'predicted_doublet', 'louvain_0.5', 'louvain_1.0', 'cell_type', 'disease_group'\n",
       "    var: 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: 'Subject_colors', 'amyloid.group_colors', 'braaksc_colors', 'dendrogram_louvain_0.5', 'hvg', 'log1p', 'louvain', 'louvain_0.5_colors', 'louvain_1.0_colors', 'msex_x_colors', 'neighbors', 'ov_t_test_rank_genes_0.5', 'pathology.group_colors', 'pca', 'predicted_doublet_colors', 'scrublet', 't_test_rank_genes_0.5', 'tsne', 'umap'\n",
       "    obsm: 'X_pca', 'X_tsne', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts', 'log'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create new Anndata object for use in MAST with non-batch corrected data as before\n",
    "\n",
    "adata_exc = adata_annot[adata_annot.obs.cell_type=='Excitatory']\n",
    "# adata_exc.X = adata_exc.layers['counts'].toarray()\n",
    "adata_exc.obs['n_genes'] = (adata_exc.X > 0).sum(1) # recompute number of genes expressed per cell\n",
    "\n",
    "# split data into smaller batches stratified such that the proportion of test groups remains the same\n",
    "# This is used to avoid kernel crashes over too large data\n",
    "\n",
    "strata = adata_exc.obs.disease_group\n",
    "\n",
    "adata_train, adata_test = train_test_split(adata_exc, test_size=0.25, train_size=None, random_state=None,\n",
    "                                            shuffle=True, stratify=strata)\n",
    "# adata_test.X = adata_test.X.toarray()\n",
    "# adata_train.X = adata_train.X.toarray()\n",
    "\n",
    "adata_exc = adata_test\n",
    "\n",
    "del adata_train, adata_test\n",
    "\n",
    "adata_exc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8beae31d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/pandas2ri.py:351: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/numpy2ri.py:226: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/conversion.py:28: DeprecationWarning: The use of converter in module rpy2.robjects.conversion is deprecated. Use rpy2.robjects.conversion.get_conversion() instead of rpy2.robjects.conversion.converter.\n",
      "  warnings.warn(\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/conversion.py:28: DeprecationWarning: The use of py2rpy in module rpy2.robjects.conversion is deprecated. Use rpy2.robjects.conversion.get_conversion() instead of rpy2.robjects.conversion.converter.\n",
      "  warnings.warn(\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/anndata2ri/py2r.py:37: NotConvertedWarning: Conversion 'py2rpy' not defined for objects of type '<class 'NoneType'>'\n",
      "  warn(str(e), NotConvertedWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"running hurdle model on Excitatory cells...\"\n",
      "[1] \"Finished...\"\n",
      "[1] \"--------------------------------------\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i adata_exc -o exc_degs\n",
    "\n",
    "library(Seurat)\n",
    "\n",
    "print(\"running hurdle model on Excitatory cells...\")\n",
    "cur_seurat_obj <- as.Seurat(adata_exc, counts = \"counts\", data = 'log', verbose=TRUE)\n",
    "Idents(cur_seurat_obj) <- cur_seurat_obj$disease_group\n",
    "cur_seurat_obj$amyloid <- as.numeric(cur_seurat_obj$amyloid_x)\n",
    "\n",
    "exc_degs <- FindMarkers(object = cur_seurat_obj, ident.1='AD-pathology', ident.2='no-pathology', only.pos = FALSE, \n",
    "                        min.pct = 0.0, test.use = \"MAST\", logfc.threshold = 0.00, verbose=TRUE,)\n",
    "\n",
    "exc_degs$diff <- exc_degs$pct.1 - exc_degs$pct.2\n",
    "exc_degs$FDR <- exc_degs$p_val_adj\n",
    "print(\"Finished...\")\n",
    "print(\"--------------------------------------\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bc2ddde0",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata_exc"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "d50877b1",
   "metadata": {},
   "source": [
    "### **Run MAST on Inhibitory**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2ebff745",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 3886 × 16590\n",
       "    obs: 'projid', 'fastq', 'Subject', 'sample', 'libraryid', 'study', 'age_death', 'educ', 'msex_x', 'gpath_x', 'amyloid_x', 'plaq_n_x', 'cogdx_x', 'pathologic diagnosis of AD', 'amyloid_y', 'plaq_n_y', 'nft', 'tangles', 'cogn_global_lv', 'gpath_y', 'gpath_3neocort', 'amyloid.group', 'caa_4gp', 'ceradsc', 'braaksc', 'niareagansc', 'cogdx_y', 'msex_y', 'pathology.group', 'sampleid', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'total_counts_mt', 'pct_counts_mt', 'n_genes', 'doublet_score', 'predicted_doublet', 'louvain_0.5', 'louvain_1.0', 'cell_type', 'disease_group'\n",
       "    var: 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: 'Subject_colors', 'amyloid.group_colors', 'braaksc_colors', 'dendrogram_louvain_0.5', 'hvg', 'log1p', 'louvain', 'louvain_0.5_colors', 'louvain_1.0_colors', 'msex_x_colors', 'neighbors', 'ov_t_test_rank_genes_0.5', 'pathology.group_colors', 'pca', 'predicted_doublet_colors', 'scrublet', 't_test_rank_genes_0.5', 'tsne', 'umap'\n",
       "    obsm: 'X_pca', 'X_tsne', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts', 'log'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create new Anndata object for use in MAST with non-batch corrected data as before\n",
    "\n",
    "adata_inh = adata_annot[adata_annot.obs.cell_type=='Inhibitory']\n",
    "#adata_inh.X = adata_sub['Inhibitory'].layers['counts']\n",
    "adata_inh.obs['n_genes'] = (adata_inh.X > 0).sum(1) # recompute number of genes expressed per cell\n",
    "\n",
    "# split data into smaller batches stratified such that the proportion of test groups remains the same\n",
    "# This is used to avoid kernel crashes over too large data\n",
    "\n",
    "strata = adata_inh.obs.disease_group\n",
    "\n",
    "adata_train, adata_test = train_test_split(adata_inh, test_size=0.5, train_size=None, random_state=None,\n",
    "                                            shuffle=True, stratify=strata)\n",
    "# adata_test.X = adata_test.X.toarray()\n",
    "# adata_train.X = adata_train.X.toarray()\n",
    "\n",
    "adata_inh = adata_test\n",
    "\n",
    "del adata_train, adata_test\n",
    "\n",
    "adata_inh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "826ef28b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/pandas2ri.py:351: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/numpy2ri.py:226: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/conversion.py:28: DeprecationWarning: The use of converter in module rpy2.robjects.conversion is deprecated. Use rpy2.robjects.conversion.get_conversion() instead of rpy2.robjects.conversion.converter.\n",
      "  warnings.warn(\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/conversion.py:28: DeprecationWarning: The use of py2rpy in module rpy2.robjects.conversion is deprecated. Use rpy2.robjects.conversion.get_conversion() instead of rpy2.robjects.conversion.converter.\n",
      "  warnings.warn(\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/anndata2ri/py2r.py:37: NotConvertedWarning: Conversion 'py2rpy' not defined for objects of type '<class 'NoneType'>'\n",
      "  warn(str(e), NotConvertedWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"------------------------------------------\"\n",
      "[1] \"running hurdle model on Inhibitory cells...\"\n",
      "[1] \"Finished...\"\n",
      "[1] \"--------------------------------------\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i adata_inh -o inh_degs\n",
    "\n",
    "print(\"------------------------------------------\")\n",
    "print(\"running hurdle model on Inhibitory cells...\")\n",
    "cur_seurat_obj <- as.Seurat(adata_inh, counts = \"counts\", data = \"log\")\n",
    "Idents(cur_seurat_obj) <- cur_seurat_obj$disease_group\n",
    "\n",
    "inh_degs <- FindMarkers(object = cur_seurat_obj, ident.1='AD-pathology', ident.2='no-pathology', only.pos = FALSE, \n",
    "                        min.pct = 0.0, test.use = \"MAST\", logfc.threshold = 0.00, verbose=TRUE,)\n",
    "\n",
    "inh_degs$diff <- inh_degs$pct.1 - inh_degs$pct.2\n",
    "inh_degs$FDR <- inh_degs$p_val_adj\n",
    "\n",
    "\n",
    "print(\"Finished...\")\n",
    "print(\"--------------------------------------\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ed00461f",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata_inh"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "193afa09",
   "metadata": {},
   "source": [
    "### **Run MAST on Oligodendrocytes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0256f6d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 4883 × 16590\n",
       "    obs: 'projid', 'fastq', 'Subject', 'sample', 'libraryid', 'study', 'age_death', 'educ', 'msex_x', 'gpath_x', 'amyloid_x', 'plaq_n_x', 'cogdx_x', 'pathologic diagnosis of AD', 'amyloid_y', 'plaq_n_y', 'nft', 'tangles', 'cogn_global_lv', 'gpath_y', 'gpath_3neocort', 'amyloid.group', 'caa_4gp', 'ceradsc', 'braaksc', 'niareagansc', 'cogdx_y', 'msex_y', 'pathology.group', 'sampleid', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'total_counts_mt', 'pct_counts_mt', 'n_genes', 'doublet_score', 'predicted_doublet', 'louvain_0.5', 'louvain_1.0', 'cell_type', 'disease_group'\n",
       "    var: 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: 'Subject_colors', 'amyloid.group_colors', 'braaksc_colors', 'dendrogram_louvain_0.5', 'hvg', 'log1p', 'louvain', 'louvain_0.5_colors', 'louvain_1.0_colors', 'msex_x_colors', 'neighbors', 'ov_t_test_rank_genes_0.5', 'pathology.group_colors', 'pca', 'predicted_doublet_colors', 'scrublet', 't_test_rank_genes_0.5', 'tsne', 'umap'\n",
       "    obsm: 'X_pca', 'X_tsne', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts', 'log'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create new Anndata object for use in MAST with non-batch corrected data as before\n",
    "\n",
    "adata_oli = adata_annot[adata_annot.obs.cell_type=='Oligodendrocyte']\n",
    "#adata_oli.X = adata_sub['Oligodendrocyte'].layers['counts']\n",
    "adata_oli.obs['n_genes'] = (adata_oli.X > 0).sum(1) # recompute number of genes expressed per cell\n",
    "\n",
    "# split data into smaller batches stratified such that the proportion of test groups remains the same\n",
    "\n",
    "strata = adata_oli.obs.disease_group\n",
    "adata_train, adata_test = train_test_split(adata_oli, test_size=0.25, train_size=None, random_state=None,\n",
    "                                            shuffle=True, stratify=strata)\n",
    "\n",
    "# adata_test.X = adata_test.X.toarray()\n",
    "# adata_train.X = adata_train.X.toarray()\n",
    "\n",
    "adata_oli = adata_test\n",
    "\n",
    "del adata_train, adata_test\n",
    "\n",
    "adata_oli"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1c12165e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/pandas2ri.py:351: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/numpy2ri.py:226: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/conversion.py:28: DeprecationWarning: The use of converter in module rpy2.robjects.conversion is deprecated. Use rpy2.robjects.conversion.get_conversion() instead of rpy2.robjects.conversion.converter.\n",
      "  warnings.warn(\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/conversion.py:28: DeprecationWarning: The use of py2rpy in module rpy2.robjects.conversion is deprecated. Use rpy2.robjects.conversion.get_conversion() instead of rpy2.robjects.conversion.converter.\n",
      "  warnings.warn(\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/anndata2ri/py2r.py:37: NotConvertedWarning: Conversion 'py2rpy' not defined for objects of type '<class 'NoneType'>'\n",
      "  warn(str(e), NotConvertedWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"------------------------------------------\"\n",
      "[1] \"running hurdle model on Oligodendrocytes...\"\n",
      "[1] \"Finished...\"\n",
      "[1] \"--------------------------------------\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i adata_oli -o oli_degs\n",
    "\n",
    "print(\"------------------------------------------\")\n",
    "print(\"running hurdle model on Oligodendrocytes...\")\n",
    "cur_seurat_obj <- as.Seurat(adata_oli, counts = \"counts\", data = \"log\")\n",
    "Idents(cur_seurat_obj) <- cur_seurat_obj$disease_group\n",
    "\n",
    "oli_degs <- FindMarkers(object = cur_seurat_obj, ident.1='AD-pathology', ident.2='no-pathology', only.pos = FALSE, \n",
    "                        min.pct = 0.0, test.use = \"MAST\", logfc.threshold = 0.00, verbose=TRUE,)\n",
    "\n",
    "oli_degs$diff <- oli_degs$pct.1 - oli_degs$pct.2\n",
    "oli_degs$FDR <- oli_degs$p_val_adj\n",
    "print(\"Finished...\")\n",
    "print(\"--------------------------------------\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a3fc69a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata_oli"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2fadaa2b",
   "metadata": {},
   "source": [
    "### **Run MAST on Astrocytes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "36f9aa7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 3470 × 16590\n",
       "    obs: 'projid', 'fastq', 'Subject', 'sample', 'libraryid', 'study', 'age_death', 'educ', 'msex_x', 'gpath_x', 'amyloid_x', 'plaq_n_x', 'cogdx_x', 'pathologic diagnosis of AD', 'amyloid_y', 'plaq_n_y', 'nft', 'tangles', 'cogn_global_lv', 'gpath_y', 'gpath_3neocort', 'amyloid.group', 'caa_4gp', 'ceradsc', 'braaksc', 'niareagansc', 'cogdx_y', 'msex_y', 'pathology.group', 'sampleid', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'total_counts_mt', 'pct_counts_mt', 'n_genes', 'doublet_score', 'predicted_doublet', 'louvain_0.5', 'louvain_1.0', 'cell_type', 'disease_group'\n",
       "    var: 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: 'Subject_colors', 'amyloid.group_colors', 'braaksc_colors', 'dendrogram_louvain_0.5', 'hvg', 'log1p', 'louvain', 'louvain_0.5_colors', 'louvain_1.0_colors', 'msex_x_colors', 'neighbors', 'ov_t_test_rank_genes_0.5', 'pathology.group_colors', 'pca', 'predicted_doublet_colors', 'scrublet', 't_test_rank_genes_0.5', 'tsne', 'umap'\n",
       "    obsm: 'X_pca', 'X_tsne', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts', 'log'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_ast = adata_annot[adata_annot.obs.cell_type=='Astrocyte']\n",
    "#adata_ast.X = adata_sub['Astrocyte'].layers['counts'].toarray()\n",
    "adata_ast.obs['n_genes'] = (adata_ast.X > 0).sum(1) # recompute number of genes expressed per cell\n",
    "\n",
    "adata_ast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bfc2febe",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/pandas2ri.py:351: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/numpy2ri.py:226: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/conversion.py:28: DeprecationWarning: The use of converter in module rpy2.robjects.conversion is deprecated. Use rpy2.robjects.conversion.get_conversion() instead of rpy2.robjects.conversion.converter.\n",
      "  warnings.warn(\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/conversion.py:28: DeprecationWarning: The use of py2rpy in module rpy2.robjects.conversion is deprecated. Use rpy2.robjects.conversion.get_conversion() instead of rpy2.robjects.conversion.converter.\n",
      "  warnings.warn(\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/anndata2ri/py2r.py:37: NotConvertedWarning: Conversion 'py2rpy' not defined for objects of type '<class 'NoneType'>'\n",
      "  warn(str(e), NotConvertedWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"------------------------------------------\"\n",
      "[1] \"running hurdle model on Astrocytes...\"\n",
      "[1] \"Finished...\"\n",
      "[1] \"--------------------------------------\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i adata_ast -o ast_degs\n",
    "\n",
    "print(\"------------------------------------------\")\n",
    "print(\"running hurdle model on Astrocytes...\")\n",
    "cur_seurat_obj <- as.Seurat(adata_ast, counts = \"counts\", data = \"log\")\n",
    "Idents(cur_seurat_obj) <- cur_seurat_obj$disease_group\n",
    "\n",
    "ast_degs <- FindMarkers(object = cur_seurat_obj, ident.1='AD-pathology', ident.2='no-pathology', only.pos = FALSE, \n",
    "                        min.pct = 0.0, test.use = \"MAST\", logfc.threshold = 0.00, verbose=TRUE,)\n",
    "\n",
    "ast_degs$diff <- ast_degs$pct.1 - ast_degs$pct.2\n",
    "ast_degs$FDR <- ast_degs$p_val_adj\n",
    "print(\"Finished...\")\n",
    "print(\"--------------------------------------\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ade51aeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata_ast"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b66bc824",
   "metadata": {},
   "source": [
    "### **Run MAST on Endothelial Cells**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d0ddfebc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 322 × 16590\n",
       "    obs: 'projid', 'fastq', 'Subject', 'sample', 'libraryid', 'study', 'age_death', 'educ', 'msex_x', 'gpath_x', 'amyloid_x', 'plaq_n_x', 'cogdx_x', 'pathologic diagnosis of AD', 'amyloid_y', 'plaq_n_y', 'nft', 'tangles', 'cogn_global_lv', 'gpath_y', 'gpath_3neocort', 'amyloid.group', 'caa_4gp', 'ceradsc', 'braaksc', 'niareagansc', 'cogdx_y', 'msex_y', 'pathology.group', 'sampleid', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'total_counts_mt', 'pct_counts_mt', 'n_genes', 'doublet_score', 'predicted_doublet', 'louvain_0.5', 'louvain_1.0', 'cell_type', 'disease_group'\n",
       "    var: 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: 'Subject_colors', 'amyloid.group_colors', 'braaksc_colors', 'dendrogram_louvain_0.5', 'hvg', 'log1p', 'louvain', 'louvain_0.5_colors', 'louvain_1.0_colors', 'msex_x_colors', 'neighbors', 'ov_t_test_rank_genes_0.5', 'pathology.group_colors', 'pca', 'predicted_doublet_colors', 'scrublet', 't_test_rank_genes_0.5', 'tsne', 'umap'\n",
       "    obsm: 'X_pca', 'X_tsne', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts', 'log'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_end = adata_annot[adata_annot.obs.cell_type=='Endothelial']\n",
    "#adata_end.X = adata_sub['Endothelial'].layers['counts'].toarray()\n",
    "adata_end.obs['n_genes'] = (adata_end.X > 0).sum(1) # recompute number of genes expressed per cell\n",
    "\n",
    "adata_end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "99959ee0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/pandas2ri.py:351: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/numpy2ri.py:226: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/conversion.py:28: DeprecationWarning: The use of converter in module rpy2.robjects.conversion is deprecated. Use rpy2.robjects.conversion.get_conversion() instead of rpy2.robjects.conversion.converter.\n",
      "  warnings.warn(\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/conversion.py:28: DeprecationWarning: The use of py2rpy in module rpy2.robjects.conversion is deprecated. Use rpy2.robjects.conversion.get_conversion() instead of rpy2.robjects.conversion.converter.\n",
      "  warnings.warn(\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/anndata2ri/py2r.py:37: NotConvertedWarning: Conversion 'py2rpy' not defined for objects of type '<class 'NoneType'>'\n",
      "  warn(str(e), NotConvertedWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"------------------------------------------\"\n",
      "[1] \"running hurdle model on Endothelial cells...\"\n",
      "[1] \"Finished...\"\n",
      "[1] \"--------------------------------------\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i adata_end -o end_degs \n",
    "\n",
    "\n",
    "print(\"------------------------------------------\")\n",
    "print(\"running hurdle model on Endothelial cells...\")\n",
    "cur_seurat_obj <- as.Seurat(adata_end, counts = \"counts\", data = \"log\")\n",
    "Idents(cur_seurat_obj) <- cur_seurat_obj$disease_group\n",
    "\n",
    "end_degs <- FindMarkers(object = cur_seurat_obj, ident.1='AD-pathology', ident.2='no-pathology', only.pos = FALSE, \n",
    "                        min.pct = 0.0, test.use = \"MAST\", logfc.threshold = 0.00, verbose=TRUE,)\n",
    "\n",
    "end_degs$diff <- end_degs$pct.1 - end_degs$pct.2\n",
    "end_degs$FDR <- end_degs$p_val_adj\n",
    "print(\"Finished...\")\n",
    "print(\"--------------------------------------\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e60196f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata_end"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1a4823eb",
   "metadata": {},
   "source": [
    "### **Run MAST on Microglia Cells**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "cd6f837f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1961 × 16590\n",
       "    obs: 'projid', 'fastq', 'Subject', 'sample', 'libraryid', 'study', 'age_death', 'educ', 'msex_x', 'gpath_x', 'amyloid_x', 'plaq_n_x', 'cogdx_x', 'pathologic diagnosis of AD', 'amyloid_y', 'plaq_n_y', 'nft', 'tangles', 'cogn_global_lv', 'gpath_y', 'gpath_3neocort', 'amyloid.group', 'caa_4gp', 'ceradsc', 'braaksc', 'niareagansc', 'cogdx_y', 'msex_y', 'pathology.group', 'sampleid', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'total_counts_mt', 'pct_counts_mt', 'n_genes', 'doublet_score', 'predicted_doublet', 'louvain_0.5', 'louvain_1.0', 'cell_type', 'disease_group'\n",
       "    var: 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: 'Subject_colors', 'amyloid.group_colors', 'braaksc_colors', 'dendrogram_louvain_0.5', 'hvg', 'log1p', 'louvain', 'louvain_0.5_colors', 'louvain_1.0_colors', 'msex_x_colors', 'neighbors', 'ov_t_test_rank_genes_0.5', 'pathology.group_colors', 'pca', 'predicted_doublet_colors', 'scrublet', 't_test_rank_genes_0.5', 'tsne', 'umap'\n",
       "    obsm: 'X_pca', 'X_tsne', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts', 'log'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_mic = adata_annot[adata_annot.obs.cell_type=='Microglia']\n",
    "#adata_mic.X = adata_sub['Microglia'].layers['counts'].toarray()\n",
    "adata_mic.obs['n_genes'] = (adata_mic.X > 0).sum(1) # recompute number of genes expressed per cell\n",
    "\n",
    "adata_mic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4952730f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/pandas2ri.py:351: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/numpy2ri.py:226: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/conversion.py:28: DeprecationWarning: The use of converter in module rpy2.robjects.conversion is deprecated. Use rpy2.robjects.conversion.get_conversion() instead of rpy2.robjects.conversion.converter.\n",
      "  warnings.warn(\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/conversion.py:28: DeprecationWarning: The use of py2rpy in module rpy2.robjects.conversion is deprecated. Use rpy2.robjects.conversion.get_conversion() instead of rpy2.robjects.conversion.converter.\n",
      "  warnings.warn(\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/anndata2ri/py2r.py:37: NotConvertedWarning: Conversion 'py2rpy' not defined for objects of type '<class 'NoneType'>'\n",
      "  warn(str(e), NotConvertedWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"------------------------------------------\"\n",
      "[1] \"running hurdle model on Microglia...\"\n",
      "[1] \"Finished...\"\n",
      "[1] \"--------------------------------------\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i adata_mic -o mic_degs\n",
    "\n",
    "print(\"------------------------------------------\")\n",
    "print(\"running hurdle model on Microglia...\")\n",
    "cur_seurat_obj <- as.Seurat(adata_mic, counts = \"counts\", data = \"log\")\n",
    "Idents(cur_seurat_obj) <- cur_seurat_obj$disease_group\n",
    "\n",
    "mic_degs <- FindMarkers(object = cur_seurat_obj, ident.1='AD-pathology', ident.2='no-pathology', only.pos = FALSE, \n",
    "                        min.pct = 0.0, test.use = \"MAST\", logfc.threshold = 0.00, verbose=TRUE,)\n",
    "\n",
    "mic_degs$diff <- mic_degs$pct.1 - mic_degs$pct.2\n",
    "mic_degs$FDR <- mic_degs$p_val_adj\n",
    "\n",
    "print(\"Finished...\")\n",
    "print(\"--------------------------------------\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b2562dbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata_mic"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "29e4b8d8",
   "metadata": {},
   "source": [
    "### **Run MAST on OPCs**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "eb1b5071",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2618 × 16590\n",
       "    obs: 'projid', 'fastq', 'Subject', 'sample', 'libraryid', 'study', 'age_death', 'educ', 'msex_x', 'gpath_x', 'amyloid_x', 'plaq_n_x', 'cogdx_x', 'pathologic diagnosis of AD', 'amyloid_y', 'plaq_n_y', 'nft', 'tangles', 'cogn_global_lv', 'gpath_y', 'gpath_3neocort', 'amyloid.group', 'caa_4gp', 'ceradsc', 'braaksc', 'niareagansc', 'cogdx_y', 'msex_y', 'pathology.group', 'sampleid', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'total_counts_mt', 'pct_counts_mt', 'n_genes', 'doublet_score', 'predicted_doublet', 'louvain_0.5', 'louvain_1.0', 'cell_type', 'disease_group'\n",
       "    var: 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: 'Subject_colors', 'amyloid.group_colors', 'braaksc_colors', 'dendrogram_louvain_0.5', 'hvg', 'log1p', 'louvain', 'louvain_0.5_colors', 'louvain_1.0_colors', 'msex_x_colors', 'neighbors', 'ov_t_test_rank_genes_0.5', 'pathology.group_colors', 'pca', 'predicted_doublet_colors', 'scrublet', 't_test_rank_genes_0.5', 'tsne', 'umap'\n",
       "    obsm: 'X_pca', 'X_tsne', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts', 'log'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_opc = adata_annot[adata_annot.obs.cell_type=='OPC']\n",
    "#adata_opc.X = adata_sub['OPC'].layers['counts'].toarray()\n",
    "adata_opc.obs['n_genes'] = (adata_opc.X > 0).sum(1) # recompute number of genes expressed per cell\n",
    "\n",
    "adata_opc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d0f7c4b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/pandas2ri.py:351: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/numpy2ri.py:226: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/conversion.py:28: DeprecationWarning: The use of converter in module rpy2.robjects.conversion is deprecated. Use rpy2.robjects.conversion.get_conversion() instead of rpy2.robjects.conversion.converter.\n",
      "  warnings.warn(\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/conversion.py:28: DeprecationWarning: The use of py2rpy in module rpy2.robjects.conversion is deprecated. Use rpy2.robjects.conversion.get_conversion() instead of rpy2.robjects.conversion.converter.\n",
      "  warnings.warn(\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/anndata2ri/py2r.py:37: NotConvertedWarning: Conversion 'py2rpy' not defined for objects of type '<class 'NoneType'>'\n",
      "  warn(str(e), NotConvertedWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"------------------------------------------\"\n",
      "[1] \"running hurdle model on OPCs...\"\n",
      "[1] \"Finished...\"\n",
      "[1] \"--------------------------------------\"\n"
     ]
    }
   ],
   "source": [
    "%%R  -i adata_opc -o opc_degs\n",
    "\n",
    "print(\"------------------------------------------\")\n",
    "print(\"running hurdle model on OPCs...\")\n",
    "cur_seurat_obj <- as.Seurat(adata_opc, counts = \"counts\", data = \"log\")\n",
    "Idents(cur_seurat_obj) <- cur_seurat_obj$disease_group\n",
    "\n",
    "opc_degs <- FindMarkers(object = cur_seurat_obj, ident.1='AD-pathology', ident.2='no-pathology', only.pos = FALSE, \n",
    "                        min.pct = 0.0, test.use = \"MAST\", logfc.threshold = 0.00, verbose=TRUE,)\n",
    "\n",
    "opc_degs$diff <- opc_degs$pct.1 - opc_degs$pct.2\n",
    "opc_degs$FDR <- opc_degs$p_val_adj\n",
    "\n",
    "print(\"Finished...\")\n",
    "print(\"--------------------------------------\")\n",
    "\n",
    "rm(cur_seurat_obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5d9e8d4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata_opc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fff8a68e",
   "metadata": {},
   "source": [
    "Run Mast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "af19a208",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -i adata_exc -i adata_inh -i adata_oli -i adata_mic -i adata_ast -i adata_opc -i adata_end -o exc_degs \n",
    "# #-o inh_degs -o oli_degs -o ast_degs -o mic_degs -o opc_degs -o end_degs \n",
    "\n",
    "# library(MAST)\n",
    "# library(parallel)\n",
    "\n",
    "# print('parsing data...')\n",
    "\n",
    "# #Convert SingleCellExperiment to SingleCellAssay type as required by MAST\n",
    "# exc_sca <- SceToSingleCellAssay(adata_exc, class=\"SingleCellAssay\")\n",
    "# inh_sca <- SceToSingleCellAssay(adata_inh, class=\"SingleCellAssay\")\n",
    "# oli_sca <- SceToSingleCellAssay(adata_oli, class=\"SingleCellAssay\")\n",
    "# ast_sca <- SceToSingleCellAssay(adata_ast, class=\"SingleCellAssay\")\n",
    "# mic_sca <- SceToSingleCellAssay(adata_mic, class=\"SingleCellAssay\")\n",
    "# opc_sca <- SceToSingleCellAssay(adata_opc, class=\"SingleCellAssay\")\n",
    "# end_sca <- SceToSingleCellAssay(adata_end, class=\"SingleCellAssay\")\n",
    "\n",
    "# print('parsing done...')\n",
    "# rm(adata_exc, adata_inh, adata_oli, adata_ast, adata_mic, adata_end, adata_opc)\n",
    "\n",
    "# print(\"Deleted unused variables...\")\n",
    "\n",
    "# #Scale Gene detection rate\n",
    "# colData(exc_sca)$n_genes = scale(colData(exc_sca)$n_genes)\n",
    "# colData(inh_sca)$n_genes = scale(colData(inh_sca)$n_genes)\n",
    "# colData(oli_sca)$n_genes = scale(colData(oli_sca)$n_genes)\n",
    "# colData(ast_sca)$n_genes = scale(colData(ast_sca)$n_genes)\n",
    "# colData(mic_sca)$n_genes = scale(colData(mic_sca)$n_genes)\n",
    "# colData(opc_sca)$n_genes = scale(colData(opc_sca)$n_genes)\n",
    "# colData(end_sca)$n_genes = scale(colData(end_sca)$n_genes)\n",
    "\n",
    "# print('Filtering non-expressed genes...')\n",
    "\n",
    "# exc_sca_filt = exc_sca[rowSums(assay(exc_sca)) != 0, ]\n",
    "# inh_sca_filt = inh_sca[rowSums(assay(inh_sca)) != 0, ]\n",
    "# oli_sca_filt = oli_sca[rowSums(assay(oli_sca)) != 0, ]\n",
    "# mic_sca_filt = mic_sca[rowSums(assay(mic_sca)) != 0, ]\n",
    "# ast_sca_filt = ast_sca[rowSums(assay(ast_sca)) != 0, ]\n",
    "# end_sca_filt = end_sca[rowSums(assay(end_sca)) != 0, ]\n",
    "# opc_sca_filt = opc_sca[rowSums(assay(opc_sca)) != 0, ]\n",
    "\n",
    "# rm(exc_sca, oli_sca, mic_sca, ast_sca, end_sca, opc_sca, inh_sca)\n",
    "\n",
    "# #Define & run hurdle model excitatory cells\n",
    "# print(\"-------------------------------------------\")\n",
    "# print(\"running hurdle model on excitatory cells...\")\n",
    "\n",
    "# cond<-factor(colData(exc_sca_filt)$disease_group)\n",
    "# cond<-relevel(cond,\"no-pathology\")\n",
    "# colData(exc_sca_filt)$disease_group<-cond\n",
    "\n",
    "# # zlmCond <- zlm(formula = ~disease_group + Subject + n_genes, sca=sca_filt)\n",
    "# zlmCond <- zlm(formula = ~disease_group + n_genes, sca=exc_sca_filt)\n",
    "\n",
    "# print(\"finished hurdle model\")\n",
    "# print(\"Estimating summary\")\n",
    "# summaryCond <- summary(zlmCond, doLRT='disease_groupAD-pathology')\n",
    "# summaryDt <- summaryCond$datatable\n",
    "\n",
    "# result <- merge(summaryDt[contrast=='disease_groupAD-pathology' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "#                     summaryDt[contrast=='disease_groupAD-pathology' & component=='logFC', .(primerid, coef)],\n",
    "#                     by='primerid') #logFC coefficients\n",
    "\n",
    "# #Correct for multiple testing (FDR correction) and filtering\n",
    "# result[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')]\n",
    "# exc_degs = result\n",
    "# exc_degs = exc_degs[order(exc_degs$FDR),]\n",
    "# print(\"Finished...\")\n",
    "# print(\"--------------------------------\")\n",
    "# rm(exc_sca_filt, zlmCond, summaryCond, summaryDt, result)\n",
    "\n",
    "\n",
    "# #Define & run hurdle model inhibitory cells\n",
    "# print(\"-------------------------------------------\")\n",
    "# print(\"running hurdle model on inhibitory cells...\")\n",
    "# # zlmCond <- zlm(formula = ~disease_group + Subject + n_genes, sca=sca_filt)\n",
    "# zlmCond <- zlm(formula = ~disease_group + n_genes, sca=inh_sca_filt)\n",
    "\n",
    "# print(\"finished hurdle model\")\n",
    "# print(\"Estimating summary\")\n",
    "# summaryCond <- summary(zlmCond, doLRT='disease_groupno-pathology')\n",
    "# summaryDt <- summaryCond$datatable\n",
    "\n",
    "# result <- merge(summaryDt[contrast=='disease_groupno-pathology' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "#                     summaryDt[contrast=='disease_groupno-pathology' & component=='logFC', .(primerid, coef)],\n",
    "#                     by='primerid') #logFC coefficients\n",
    "\n",
    "# #Correct for multiple testing (FDR correction) and filtering\n",
    "# result[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')]\n",
    "# inh_degs = result\n",
    "# inh_degs = inh_degs[order(inh_degs$FDR),]\n",
    "# print(\"Finished...\")\n",
    "# print(\"--------------------------------\")\n",
    "# rm(inh_sca_filt, zlmCond, summaryCond, summaryDt, result)\n",
    "\n",
    "\n",
    "\n",
    "# #Define & run hurdle model oligodendrocytes\n",
    "# print(\"------------------------------------------\")\n",
    "# print(\"running hurdle model on oligodendrocyte...\")\n",
    "# # zlmCond <- zlm(formula = ~disease_group + Subject + n_genes, sca=sca_filt)\n",
    "# zlmCond <- zlm(formula = ~disease_group + n_genes, sca=oli_sca_filt)\n",
    "\n",
    "# print(\"finished hurdle model\")\n",
    "# print(\"Estimating summary\")\n",
    "# summaryCond <- summary(zlmCond, doLRT='disease_groupno-pathology')\n",
    "# summaryDt <- summaryCond$datatable\n",
    "\n",
    "# result <- merge(summaryDt[contrast=='disease_groupno-pathology' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "#                     summaryDt[contrast=='disease_groupno-pathology' & component=='logFC', .(primerid, coef)],\n",
    "#                     by='primerid') #logFC coefficients\n",
    "\n",
    "# #Correct for multiple testing (FDR correction) and filtering\n",
    "# result[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')]\n",
    "# oli_degs = result\n",
    "# oli_degs = oli_degs[order(oli_degs$FDR),]\n",
    "# print(\"Finished...\")\n",
    "# print(\"--------------------------------\")\n",
    "# rm(oli_sca_filt, zlmCond, summaryCond, summaryDt, result)\n",
    "\n",
    "\n",
    "\n",
    "# #Define & run hurdle model astrocytes\n",
    "# print(\"--------------------------------------\")\n",
    "# print(\"running hurdle model on astrocytes...\")\n",
    "# # zlmCond <- zlm(formula = ~disease_group + Subject + n_genes, sca=sca_filt)\n",
    "# zlmCond <- zlm(formula = ~disease_group + n_genes, sca=ast_sca_filt)\n",
    "\n",
    "# print(\"finished hurdle model\")\n",
    "# print(\"Estimating summary\")\n",
    "# summaryCond <- summary(zlmCond, doLRT='disease_groupno-pathology')\n",
    "# summaryDt <- summaryCond$datatable\n",
    "\n",
    "# result <- merge(summaryDt[contrast=='disease_groupno-pathology' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "#                     summaryDt[contrast=='disease_groupno-pathology' & component=='logFC', .(primerid, coef)],\n",
    "#                     by='primerid') #logFC coefficients\n",
    "\n",
    "# #Correct for multiple testing (FDR correction) and filtering\n",
    "# result[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')]\n",
    "# ast_degs = result\n",
    "# ast_degs = ast_degs[order(ast_degs$FDR),]\n",
    "# print(\"Finished...\")\n",
    "# print(\"--------------------------------\")\n",
    "# rm(ast_sca_filt, zlmCond, summaryCond, summaryDt, result)\n",
    "\n",
    "\n",
    "# #Define & run hurdle model microglia\n",
    "# print(\"------------------------------------\")\n",
    "# print(\"running hurdle model on microglia...\")\n",
    "# # zlmCond <- zlm(formula = ~disease_group + Subject + n_genes, sca=sca_filt)\n",
    "# zlmCond <- zlm(formula = ~disease_group + n_genes, sca=mic_sca_filt)\n",
    "\n",
    "# print(\"finished hurdle model\")\n",
    "# print(\"Estimating summary\")\n",
    "# summaryCond <- summary(zlmCond, doLRT='disease_groupno-pathology')\n",
    "# summaryDt <- summaryCond$datatable\n",
    "\n",
    "# result <- merge(summaryDt[contrast=='disease_groupno-pathology' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "#                     summaryDt[contrast=='disease_groupno-pathology' & component=='logFC', .(primerid, coef)],\n",
    "#                     by='primerid') #logFC coefficients\n",
    "\n",
    "# #Correct for multiple testing (FDR correction) and filtering\n",
    "# result[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')]\n",
    "# mic_degs = result\n",
    "# mic_degs = mic_degs[order(mic_degs$FDR),]\n",
    "# print(\"Finished...\")\n",
    "# print(\"--------------------------------\")\n",
    "# rm(mic_sca_filt, zlmCond, summaryCond, summaryDt, result)\n",
    "\n",
    "\n",
    "# #Define & run hurdle model OPC\n",
    "# print(\"--------------------------------\")\n",
    "# print(\"running hurdle model on OPCs...\")\n",
    "# # zlmCond <- zlm(formula = ~disease_group + Subject + n_genes, sca=sca_filt)\n",
    "# zlmCond <- zlm(formula = ~disease_group + n_genes, sca=opc_sca_filt)\n",
    "\n",
    "# print(\"finished hurdle model\")\n",
    "# print(\"Estimating summary\")\n",
    "# summaryCond <- summary(zlmCond, doLRT='disease_groupno-pathology')\n",
    "# summaryDt <- summaryCond$datatable\n",
    "\n",
    "# result <- merge(summaryDt[contrast=='disease_groupno-pathology' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "#                     summaryDt[contrast=='disease_groupno-pathology' & component=='logFC', .(primerid, coef)],\n",
    "#                     by='primerid') #logFC coefficients\n",
    "\n",
    "# #Correct for multiple testing (FDR correction) and filtering\n",
    "# result[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')]\n",
    "# opc_degs = result\n",
    "# opc_degs = opc_degs[order(opc_degs$FDR),]\n",
    "# print(\"Finished...\")\n",
    "# print(\"--------------------------------\")\n",
    "# rm(opc_sca_filt, zlmCond, summaryCond, summaryDt, result)\n",
    "\n",
    "\n",
    "# #Define & run hurdle model endothelial\n",
    "# print(\"--------------------------------------\")\n",
    "# print(\"running hurdle model on endothelial...\")\n",
    "# # zlmCond <- zlm(formula = ~disease_group + Subject + n_genes, sca=sca_filt)\n",
    "# zlmCond <- zlm(formula = ~disease_group + n_genes, sca=end_sca_filt)\n",
    "\n",
    "# print(\"finished hurdle model\")\n",
    "# print(\"Estimating summary\")\n",
    "# summaryCond <- summary(zlmCond, doLRT='disease_groupno-pathology')\n",
    "# summaryDt <- summaryCond$datatable\n",
    "\n",
    "# result <- merge(summaryDt[contrast=='disease_groupno-pathology' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "#                     summaryDt[contrast=='disease_groupno-pathology' & component=='logFC', .(primerid, coef)],\n",
    "#                     by='primerid') #logFC coefficients\n",
    "\n",
    "# #Correct for multiple testing (FDR correction) and filtering\n",
    "# result[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')]\n",
    "# end_degs = result\n",
    "# end_degs = end_degs[order(end_degs$FDR),]\n",
    "# print(\"Finished...\")\n",
    "# print(\"--------------------------------\")\n",
    "# rm(end_sca_filt, zlmCond, summaryCond, summaryDt, result)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "6037e17f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -i adata_exc -i adata_inh -i adata_oli -i adata_mic -i adata_ast -i adata_opc -i adata_end  \\\n",
    "#     -o exc_degs -o inh_degs -o oli_degs -o ast_degs -o mic_degs -o opc_degs -o end_degs \n",
    "\n",
    "# library(Seurat)\n",
    "\n",
    "# print(\"running hurdle model on Excitatory cells...\")\n",
    "# cur_seurat_obj <- as.Seurat(adata_exc, counts = \"counts\", data = \"log\")\n",
    "# Idents(cur_seurat_obj) <- cur_seurat_obj$disease_group\n",
    "# cur_seurat_obj$amyloid <- as.numeric(cur_seurat_obj$amyloid_x)\n",
    "\n",
    "# exc_degs <- FindMarkers(object = cur_seurat_obj, ident.1='AD-pathology', ident.2='no-pathology', only.pos = FALSE, \n",
    "#                         min.pct = 0.0, test.use = \"MAST\", logfc.threshold = 0.00, verbose=TRUE,)\n",
    "\n",
    "# exc_degs$diff <- exc_degs$pct.1 - exc_degs$pct.2\n",
    "# exc_degs$FDR <- exc_degs$p_val_adj\n",
    "# print(\"Finished...\")\n",
    "# print(\"--------------------------------------\")\n",
    "\n",
    "\n",
    "# print(\"------------------------------------------\")\n",
    "# print(\"running hurdle model on Inhibitory cells...\")\n",
    "# cur_seurat_obj <- as.Seurat(adata_inh, counts = \"counts\", data = \"log\")\n",
    "# Idents(cur_seurat_obj) <- cur_seurat_obj$disease_group\n",
    "\n",
    "# inh_degs <- FindMarkers(object = cur_seurat_obj, ident.1='AD-pathology', ident.2='no-pathology', only.pos = FALSE, \n",
    "#                         min.pct = 0.0, test.use = \"MAST\", logfc.threshold = 0.00, verbose=TRUE,)\n",
    "\n",
    "# inh_degs$diff <- inh_degs$pct.1 - inh_degs$pct.2\n",
    "# inh_degs$FDR <- inh_degs$p_val_adj\n",
    "# print(\"Finished...\")\n",
    "# print(\"--------------------------------------\")\n",
    "\n",
    "\n",
    "# print(\"------------------------------------------\")\n",
    "# print(\"running hurdle model on Oligodendrocytes...\")\n",
    "# cur_seurat_obj <- as.Seurat(adata_oli, counts = \"counts\", data = \"log\")\n",
    "# Idents(cur_seurat_obj) <- cur_seurat_obj$disease_group\n",
    "\n",
    "# oli_degs <- FindMarkers(object = cur_seurat_obj, ident.1='AD-pathology', ident.2='no-pathology', only.pos = FALSE, \n",
    "#                         min.pct = 0.0, test.use = \"MAST\", logfc.threshold = 0.00, verbose=TRUE,)\n",
    "\n",
    "# oli_degs$diff <- oli_degs$pct.1 - oli_degs$pct.2\n",
    "# oli_degs$FDR <- oli_degs$p_val_adj\n",
    "# print(\"Finished...\")\n",
    "# print(\"--------------------------------------\")\n",
    "\n",
    "# print(\"------------------------------------------\")\n",
    "# print(\"running hurdle model on Astrocytes...\")\n",
    "# cur_seurat_obj <- as.Seurat(adata_ast, counts = \"counts\", data = \"log\")\n",
    "# Idents(cur_seurat_obj) <- cur_seurat_obj$disease_group\n",
    "\n",
    "# ast_degs <- FindMarkers(object = cur_seurat_obj, ident.1='AD-pathology', ident.2='no-pathology', only.pos = FALSE, \n",
    "#                         min.pct = 0.0, test.use = \"MAST\", logfc.threshold = 0.00, verbose=TRUE,)\n",
    "\n",
    "# ast_degs$diff <- ast_degs$pct.1 - ast_degs$pct.2\n",
    "# ast_degs$FDR <- ast_degs$p_val_adj\n",
    "# print(\"Finished...\")\n",
    "# print(\"--------------------------------------\")\n",
    "\n",
    "\n",
    "# print(\"------------------------------------------\")\n",
    "# print(\"running hurdle model on Microglia...\")\n",
    "# cur_seurat_obj <- as.Seurat(adata_mic, counts = \"counts\", data = \"log\")\n",
    "# Idents(cur_seurat_obj) <- cur_seurat_obj$disease_group\n",
    "\n",
    "# mic_degs <- FindMarkers(object = cur_seurat_obj, ident.1='AD-pathology', ident.2='no-pathology', only.pos = FALSE, \n",
    "#                         min.pct = 0.0, test.use = \"MAST\", logfc.threshold = 0.00, verbose=TRUE,)\n",
    "\n",
    "# mic_degs$diff <- mic_degs$pct.1 - mic_degs$pct.2\n",
    "# mic_degs$FDR <- mic_degs$p_val_adj\n",
    "\n",
    "# print(\"Finished...\")\n",
    "# print(\"--------------------------------------\")\n",
    "\n",
    "\n",
    "# print(\"------------------------------------------\")\n",
    "# print(\"running hurdle model on Endothelial cells...\")\n",
    "# cur_seurat_obj <- as.Seurat(adata_end, counts = \"counts\", data = \"log\")\n",
    "# Idents(cur_seurat_obj) <- cur_seurat_obj$disease_group\n",
    "\n",
    "# end_degs <- FindMarkers(object = cur_seurat_obj, ident.1='AD-pathology', ident.2='no-pathology', only.pos = FALSE, \n",
    "#                         min.pct = 0.0, test.use = \"MAST\", logfc.threshold = 0.00, verbose=TRUE,)\n",
    "\n",
    "# end_degs$diff <- end_degs$pct.1 - end_degs$pct.2\n",
    "# end_degs$FDR <- end_degs$p_val_adj\n",
    "# print(\"Finished...\")\n",
    "# print(\"--------------------------------------\")\n",
    "\n",
    "\n",
    "\n",
    "# print(\"------------------------------------------\")\n",
    "# print(\"running hurdle model on OPCs...\")\n",
    "# cur_seurat_obj <- as.Seurat(adata_opc, counts = \"counts\", data = \"log\")\n",
    "# Idents(cur_seurat_obj) <- cur_seurat_obj$disease_group\n",
    "\n",
    "# opc_degs <- FindMarkers(object = cur_seurat_obj, ident.1='AD-pathology', ident.2='no-pathology', only.pos = FALSE, \n",
    "#                         min.pct = 0.0, test.use = \"MAST\", logfc.threshold = 0.00, verbose=TRUE,)\n",
    "\n",
    "# opc_degs$diff <- opc_degs$pct.1 - opc_degs$pct.2\n",
    "# opc_degs$FDR <- opc_degs$p_val_adj\n",
    "\n",
    "# print(\"Finished...\")\n",
    "# print(\"--------------------------------------\")\n",
    "\n",
    "# rm(cur_seurat_obj)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "fcccd251",
   "metadata": {},
   "source": [
    "#### **Prep MAST results for all cell types**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "6572d9e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "degs_mast = {}\n",
    "degs_mast['Excitatory'] = exc_degs.sort_values(by='FDR')\n",
    "degs_mast['Inhibitory'] = inh_degs.sort_values(by='FDR')\n",
    "degs_mast['Oligodendrocyte'] = oli_degs.sort_values(by='FDR')\n",
    "degs_mast['Astrocyte'] = ast_degs.sort_values(by='FDR')\n",
    "degs_mast['OPC'] = opc_degs.sort_values(by='FDR')\n",
    "degs_mast['Microglia'] = mic_degs.sort_values(by='FDR')\n",
    "degs_mast['Endothelial'] = end_degs.sort_values(by='FDR')\n",
    "\n",
    "for key in degs_mast.keys():\n",
    "    degs_mast[key].reset_index(inplace=True)\n",
    "    degs_mast[key].rename(columns={\"index\": \"names\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90c66c6f",
   "metadata": {},
   "source": [
    "#### Consensus of Differential Expressed Transcripts\n",
    "\n",
    "The consistency of DEGs detected using the cell-level `Wilcoxon rank-sum` analysis model with those obtained with the `MAST` generalized mixed model and the `scANVI` deep generative model was assessed by comparing the directionality and rank of DEGs in the two models. Consistency in directionality was measured by counting the fraction of the top 1,000 DEGs (ranked by FDR scores) detected in cell-level analysis that showed consistent direction in the deep generative/mixed model.\n",
    "\n",
    "For analyses involving DEG counts, [**Hansuredi et. al. 2019**](https://www.nature.com/articles/s41586-019-1195-2) used only genes that were significantly supported by both models using the criteria\n",
    "\n",
    "- `FDR-corrected P < 0.01 in a two sided Wilcoxon-rank sum test`,\n",
    "- `absolute log2(mean gene expression in AD category x/mean gene expression in AD category y) > 0.25`,\n",
    "- `FDR-corrected P < 0.05 in the Poisson mixed model` \n",
    "\n",
    "Motivated by [**Zhou. et. al. 2020**](https://doi.org/10.1038/s41591-019-0695-9) who used estimated DEGs between conditions using the `MAST algorithm of the Seurat package in R`, we include the following filtering criteria for the \n",
    "\n",
    "- `DEGs in the generalized mixed effects model are obatined by filtering genes for log2(fold change) > 0.1, P < 0.05`. \n",
    "- `DEGs in the deep generative model are obatined by filtering genes for`  $$\\ln(Bayes\\ Factor) > 2.3 $$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "58d73849",
   "metadata": {},
   "outputs": [],
   "source": [
    "mast_overlap = {}\n",
    "mast_consistency  = {}\n",
    "\n",
    "for key in adata_annot.obs.cell_type.unique():\n",
    "    mast_overlap[key] = pd.merge(degs_t_test[key], degs_mast[key], how='outer', on=['names'])\n",
    "\n",
    "    mast_consistency[key] = len(list(set(list(degs_t_test[key]['names'][:1000])) &\n",
    "                            set(mast_overlap[key][((mast_overlap[key]['avg_log2FC']>0) & (mast_overlap[key]['logfoldchanges']>0))|\n",
    "                            ((mast_overlap[key]['avg_log2FC']<0) & (mast_overlap[key]['logfoldchanges']<0))].names)))/1000\n",
    "\n",
    "    mast_overlap[key]['bulk_log2FC'] = np.log2((mast_overlap[key]['pct.1']+0.005)/\n",
    "                                            (mast_overlap[key]['pct.2']+0.005)) + mast_overlap[key]['avg_log2FC']\n",
    "\n",
    "    mast_overlap[key]['p_val_adj_DEGs.Ind.Model'] = np.NaN\n",
    "\n",
    "    mast_overlap[key]['DEGs.Ind.Model'] = False\n",
    "    mast_overlap[key].loc[(mast_overlap[key]['pvals_adj'] < 0.01) & \n",
    "                        (mast_overlap[key]['abs_logfoldchanges'] > 0.25), 'DEGs.Ind.Model'] = True\n",
    "\n",
    "    pvals = mast_overlap[key].loc[mast_overlap[key]['DEGs.Ind.Model'], 'p_val_adj'].tolist()\n",
    "                        \n",
    "    try:\n",
    "        mast_overlap[key].loc[mast_overlap[key]['DEGs.Ind.Model'],\n",
    "                        'p_val_adj_DEGs.Ind.Model'] = multipletests(pvals, alpha=0.05, method='fdr_bh')[1]\n",
    "        \n",
    "        mast_overlap[key]['DEGs.Ind.Mixed.Model'] = False\n",
    "        mast_overlap[key].loc[(mast_overlap[key]['p_val_adj_DEGs.Ind.Model'] < 0.05), 'DEGs.Ind.Mixed.Model'] = True\n",
    "        \n",
    "    except ZeroDivisionError:\n",
    "        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "bc2f07f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>names</th>\n",
       "      <th>scores</th>\n",
       "      <th>pvals</th>\n",
       "      <th>pvals_adj</th>\n",
       "      <th>logfoldchanges</th>\n",
       "      <th>abs_logfoldchanges</th>\n",
       "      <th>Direction_t_test</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>RASGEF1B</td>\n",
       "      <td>30.155712</td>\n",
       "      <td>9.027398e-200</td>\n",
       "      <td>1.428766e-195</td>\n",
       "      <td>1.758242</td>\n",
       "      <td>1.758242</td>\n",
       "      <td>up</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>LINGO1</td>\n",
       "      <td>29.114767</td>\n",
       "      <td>2.334517e-186</td>\n",
       "      <td>1.847420e-182</td>\n",
       "      <td>1.433463</td>\n",
       "      <td>1.433463</td>\n",
       "      <td>up</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>SLC26A3</td>\n",
       "      <td>24.047064</td>\n",
       "      <td>8.959428e-128</td>\n",
       "      <td>4.726696e-124</td>\n",
       "      <td>1.057443</td>\n",
       "      <td>1.057443</td>\n",
       "      <td>up</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>BEX1</td>\n",
       "      <td>-23.102089</td>\n",
       "      <td>4.411363e-118</td>\n",
       "      <td>1.396373e-114</td>\n",
       "      <td>-0.872032</td>\n",
       "      <td>0.872032</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>SPARCL1</td>\n",
       "      <td>-23.105984</td>\n",
       "      <td>4.031128e-118</td>\n",
       "      <td>1.396373e-114</td>\n",
       "      <td>-1.169944</td>\n",
       "      <td>1.169944</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15822</th>\n",
       "      <td>15822</td>\n",
       "      <td>ZNF580</td>\n",
       "      <td>-0.100866</td>\n",
       "      <td>9.196568e-01</td>\n",
       "      <td>9.998724e-01</td>\n",
       "      <td>-1.700520</td>\n",
       "      <td>1.700520</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15823</th>\n",
       "      <td>15823</td>\n",
       "      <td>SSX2</td>\n",
       "      <td>-0.100854</td>\n",
       "      <td>9.196663e-01</td>\n",
       "      <td>9.998724e-01</td>\n",
       "      <td>-1.212336</td>\n",
       "      <td>1.212336</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15824</th>\n",
       "      <td>15824</td>\n",
       "      <td>RPEL1</td>\n",
       "      <td>-0.100728</td>\n",
       "      <td>9.197667e-01</td>\n",
       "      <td>9.998724e-01</td>\n",
       "      <td>-0.302013</td>\n",
       "      <td>0.302013</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15825</th>\n",
       "      <td>15825</td>\n",
       "      <td>DMRTC2</td>\n",
       "      <td>-0.109230</td>\n",
       "      <td>9.130199e-01</td>\n",
       "      <td>9.998724e-01</td>\n",
       "      <td>-0.307460</td>\n",
       "      <td>0.307460</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15826</th>\n",
       "      <td>15826</td>\n",
       "      <td>TMPRSS11D</td>\n",
       "      <td>-0.262470</td>\n",
       "      <td>7.929593e-01</td>\n",
       "      <td>9.998724e-01</td>\n",
       "      <td>-0.443457</td>\n",
       "      <td>0.443457</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15827 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0      names     scores          pvals      pvals_adj  \\\n",
       "0               0   RASGEF1B  30.155712  9.027398e-200  1.428766e-195   \n",
       "1               1     LINGO1  29.114767  2.334517e-186  1.847420e-182   \n",
       "2               2    SLC26A3  24.047064  8.959428e-128  4.726696e-124   \n",
       "3               3       BEX1 -23.102089  4.411363e-118  1.396373e-114   \n",
       "4               4    SPARCL1 -23.105984  4.031128e-118  1.396373e-114   \n",
       "...           ...        ...        ...            ...            ...   \n",
       "15822       15822     ZNF580  -0.100866   9.196568e-01   9.998724e-01   \n",
       "15823       15823       SSX2  -0.100854   9.196663e-01   9.998724e-01   \n",
       "15824       15824      RPEL1  -0.100728   9.197667e-01   9.998724e-01   \n",
       "15825       15825     DMRTC2  -0.109230   9.130199e-01   9.998724e-01   \n",
       "15826       15826  TMPRSS11D  -0.262470   7.929593e-01   9.998724e-01   \n",
       "\n",
       "       logfoldchanges  abs_logfoldchanges Direction_t_test  \n",
       "0            1.758242            1.758242               up  \n",
       "1            1.433463            1.433463               up  \n",
       "2            1.057443            1.057443               up  \n",
       "3           -0.872032            0.872032             down  \n",
       "4           -1.169944            1.169944             down  \n",
       "...               ...                 ...              ...  \n",
       "15822       -1.700520            1.700520             down  \n",
       "15823       -1.212336            1.212336             down  \n",
       "15824       -0.302013            0.302013             down  \n",
       "15825       -0.307460            0.307460             down  \n",
       "15826       -0.443457            0.443457             down  \n",
       "\n",
       "[15827 rows x 8 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "degs_t_test['Excitatory']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "067fb3b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize empty dictionaries to store the results\n",
    "mast_overlap = {}\n",
    "mast_consistency = {}\n",
    "\n",
    "# Loop through each cell type\n",
    "for key in adata_annot.obs.cell_type.unique():\n",
    "    # Merge DEGs detected by t-test and MAST algorithms based on their gene names\n",
    "    mast_overlap[key] = pd.merge(degs_t_test[key], degs_mast[key], how='outer', on=['names'])\n",
    "\n",
    "    # Calculate the consistency in directionality of the top 1000 DEGs between t-test and MAST algorithms\n",
    "    mast_consistency[key] = len(list(set(list(degs_t_test[key]['names'][:1000])) &\n",
    "                            set(mast_overlap[key][((mast_overlap[key]['avg_log2FC']>0) & (mast_overlap[key]['logfoldchanges']>0))|\n",
    "                            ((mast_overlap[key]['avg_log2FC']<0) & (mast_overlap[key]['logfoldchanges']<0))].names)))/1000\n",
    "\n",
    "    # Calculate the bulk log2FC of each gene by taking the log2 ratio of the percentage of cells expressing the gene in two conditions\n",
    "    # and adding the average log2FC calculated from MAST and t-test algorithms\n",
    "    mast_overlap[key]['bulk_log2FC'] = np.log2((mast_overlap[key]['pct.1']+0.005)/\n",
    "                                            (mast_overlap[key]['pct.2']+0.005)) + mast_overlap[key]['avg_log2FC']\n",
    "\n",
    "    # Initialize columns for adjusted p-values and DEG identification in the merged table\n",
    "    mast_overlap[key]['p_val_adj_DEGs.Ind.Model'] = np.NaN\n",
    "    mast_overlap[key]['DEGs.Ind.Model'] = False\n",
    "    \n",
    "    # Mark genes as DEGs if they meet certain criteria based on FDR-corrected p-values and absolute log2 fold change\n",
    "    mast_overlap[key].loc[(mast_overlap[key]['pvals_adj'] < 0.01) & \n",
    "                        (mast_overlap[key]['abs_logfoldchanges'] > 0.25), 'DEGs.Ind.Model'] = True\n",
    "\n",
    "    # Get the p-values of identified DEGs\n",
    "    pvals = mast_overlap[key].loc[mast_overlap[key]['DEGs.Ind.Model'], 'p_val_adj'].tolist()\n",
    "                        \n",
    "    try:\n",
    "        # Apply FDR correction to the p-values of identified DEGs\n",
    "        mast_overlap[key].loc[mast_overlap[key]['DEGs.Ind.Model'],\n",
    "                        'p_val_adj_DEGs.Ind.Model'] = multipletests(pvals, alpha=0.05, method='fdr_bh')[1]\n",
    "        \n",
    "        # Identify DEGs in the mixed model based on the FDR-corrected p-values\n",
    "        mast_overlap[key]['DEGs.Ind.Mixed.Model'] = False\n",
    "        mast_overlap[key].loc[(mast_overlap[key]['p_val_adj_DEGs.Ind.Model'] < 0.05), 'DEGs.Ind.Mixed.Model'] = True\n",
    "        \n",
    "    except ZeroDivisionError:\n",
    "        # Skip if no DEGs are found\n",
    "        continue\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c63fd2dd",
   "metadata": {},
   "source": [
    "View some DEGs from mathys data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "d0bdf0ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>IndModel.adj.pvals</th>\n",
       "      <th>no.pathology.mean</th>\n",
       "      <th>pathology.mean</th>\n",
       "      <th>IndModel.FC</th>\n",
       "      <th>MixedModel.z</th>\n",
       "      <th>MixedModel.p</th>\n",
       "      <th>DEGs.Ind.Model</th>\n",
       "      <th>DEGs.Ind.Mix.models</th>\n",
       "      <th>IndModel.absFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RASGEF1B</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.195138</td>\n",
       "      <td>5.950515</td>\n",
       "      <td>0.897136</td>\n",
       "      <td>6.127535</td>\n",
       "      <td>8.925109e-10</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.897136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NGFRAP1</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.837262</td>\n",
       "      <td>1.160450</td>\n",
       "      <td>-0.662872</td>\n",
       "      <td>-11.221532</td>\n",
       "      <td>3.196845e-29</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.662872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LINGO1</td>\n",
       "      <td>6.259506e-298</td>\n",
       "      <td>2.884947</td>\n",
       "      <td>4.381583</td>\n",
       "      <td>0.602907</td>\n",
       "      <td>5.689687</td>\n",
       "      <td>1.272728e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.602907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BEX1</td>\n",
       "      <td>8.203347e-283</td>\n",
       "      <td>1.582531</td>\n",
       "      <td>1.039826</td>\n",
       "      <td>-0.605891</td>\n",
       "      <td>-9.953163</td>\n",
       "      <td>2.442933e-23</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.605891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SLC26A3</td>\n",
       "      <td>2.121809e-276</td>\n",
       "      <td>1.750697</td>\n",
       "      <td>3.262202</td>\n",
       "      <td>0.897916</td>\n",
       "      <td>6.540229</td>\n",
       "      <td>6.142477e-11</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.897916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13004</th>\n",
       "      <td>TNFRSF25</td>\n",
       "      <td>7.630976e-03</td>\n",
       "      <td>0.238967</td>\n",
       "      <td>0.312709</td>\n",
       "      <td>0.388009</td>\n",
       "      <td>4.637332</td>\n",
       "      <td>3.529353e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.388009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13025</th>\n",
       "      <td>RP11-618P17.4</td>\n",
       "      <td>8.239413e-03</td>\n",
       "      <td>0.310741</td>\n",
       "      <td>0.410157</td>\n",
       "      <td>0.400463</td>\n",
       "      <td>4.307534</td>\n",
       "      <td>1.650848e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.400463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13068</th>\n",
       "      <td>FZD1</td>\n",
       "      <td>8.869463e-03</td>\n",
       "      <td>0.011419</td>\n",
       "      <td>0.020506</td>\n",
       "      <td>0.844692</td>\n",
       "      <td>4.687224</td>\n",
       "      <td>2.769360e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.844692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13105</th>\n",
       "      <td>PDZD4</td>\n",
       "      <td>9.460903e-03</td>\n",
       "      <td>0.254717</td>\n",
       "      <td>0.398181</td>\n",
       "      <td>0.644528</td>\n",
       "      <td>8.728511</td>\n",
       "      <td>2.580469e-18</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.644528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13120</th>\n",
       "      <td>BCAR1</td>\n",
       "      <td>9.861804e-03</td>\n",
       "      <td>0.190384</td>\n",
       "      <td>0.256937</td>\n",
       "      <td>0.432499</td>\n",
       "      <td>8.244239</td>\n",
       "      <td>1.662150e-16</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.432499</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>756 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Unnamed: 0  IndModel.adj.pvals  no.pathology.mean  pathology.mean  \\\n",
       "0           RASGEF1B        0.000000e+00           3.195138        5.950515   \n",
       "1            NGFRAP1        0.000000e+00           1.837262        1.160450   \n",
       "2             LINGO1       6.259506e-298           2.884947        4.381583   \n",
       "3               BEX1       8.203347e-283           1.582531        1.039826   \n",
       "4            SLC26A3       2.121809e-276           1.750697        3.262202   \n",
       "...              ...                 ...                ...             ...   \n",
       "13004       TNFRSF25        7.630976e-03           0.238967        0.312709   \n",
       "13025  RP11-618P17.4        8.239413e-03           0.310741        0.410157   \n",
       "13068           FZD1        8.869463e-03           0.011419        0.020506   \n",
       "13105          PDZD4        9.460903e-03           0.254717        0.398181   \n",
       "13120          BCAR1        9.861804e-03           0.190384        0.256937   \n",
       "\n",
       "       IndModel.FC  MixedModel.z  MixedModel.p  DEGs.Ind.Model  \\\n",
       "0         0.897136      6.127535  8.925109e-10            True   \n",
       "1        -0.662872    -11.221532  3.196845e-29            True   \n",
       "2         0.602907      5.689687  1.272728e-08            True   \n",
       "3        -0.605891     -9.953163  2.442933e-23            True   \n",
       "4         0.897916      6.540229  6.142477e-11            True   \n",
       "...            ...           ...           ...             ...   \n",
       "13004     0.388009      4.637332  3.529353e-06            True   \n",
       "13025     0.400463      4.307534  1.650848e-05            True   \n",
       "13068     0.844692      4.687224  2.769360e-06            True   \n",
       "13105     0.644528      8.728511  2.580469e-18            True   \n",
       "13120     0.432499      8.244239  1.662150e-16            True   \n",
       "\n",
       "       DEGs.Ind.Mix.models  IndModel.absFC  \n",
       "0                     True        0.897136  \n",
       "1                     True        0.662872  \n",
       "2                     True        0.602907  \n",
       "3                     True        0.605891  \n",
       "4                     True        0.897916  \n",
       "...                    ...             ...  \n",
       "13004                 True        0.388009  \n",
       "13025                 True        0.400463  \n",
       "13068                 True        0.844692  \n",
       "13105                 True        0.644528  \n",
       "13120                 True        0.432499  \n",
       "\n",
       "[756 rows x 10 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mathys_degs['Excitatory'][(mathys_degs['Excitatory']['DEGs.Ind.Model'])&\n",
    "                           (mathys_degs['Excitatory']['DEGs.Ind.Mix.models'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "166d56dd",
   "metadata": {},
   "source": [
    "View some DEGs from re-analysis "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "48671b12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>names</th>\n",
       "      <th>scores</th>\n",
       "      <th>pvals</th>\n",
       "      <th>pvals_adj</th>\n",
       "      <th>logfoldchanges</th>\n",
       "      <th>abs_logfoldchanges</th>\n",
       "      <th>Direction_t_test</th>\n",
       "      <th>p_val</th>\n",
       "      <th>avg_log2FC</th>\n",
       "      <th>pct.1</th>\n",
       "      <th>pct.2</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>diff</th>\n",
       "      <th>FDR</th>\n",
       "      <th>bulk_log2FC</th>\n",
       "      <th>p_val_adj_DEGs.Ind.Model</th>\n",
       "      <th>DEGs.Ind.Model</th>\n",
       "      <th>DEGs.Ind.Mixed.Model</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GFAP</td>\n",
       "      <td>9.283654</td>\n",
       "      <td>1.637632e-20</td>\n",
       "      <td>2.129904e-16</td>\n",
       "      <td>1.128408</td>\n",
       "      <td>1.128408</td>\n",
       "      <td>up</td>\n",
       "      <td>7.630713e-23</td>\n",
       "      <td>0.316605</td>\n",
       "      <td>0.559</td>\n",
       "      <td>0.411</td>\n",
       "      <td>1.265935e-18</td>\n",
       "      <td>0.148</td>\n",
       "      <td>1.265935e-18</td>\n",
       "      <td>0.755717</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>NRXN1</td>\n",
       "      <td>-8.830657</td>\n",
       "      <td>1.040630e-18</td>\n",
       "      <td>6.767220e-15</td>\n",
       "      <td>-1.276099</td>\n",
       "      <td>1.276099</td>\n",
       "      <td>down</td>\n",
       "      <td>5.718907e-13</td>\n",
       "      <td>-0.142494</td>\n",
       "      <td>0.898</td>\n",
       "      <td>0.942</td>\n",
       "      <td>9.487666e-09</td>\n",
       "      <td>-0.044</td>\n",
       "      <td>9.487666e-09</td>\n",
       "      <td>-0.211132</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.0</td>\n",
       "      <td>PRKG1</td>\n",
       "      <td>8.440312</td>\n",
       "      <td>3.164897e-17</td>\n",
       "      <td>1.029066e-13</td>\n",
       "      <td>0.988118</td>\n",
       "      <td>0.988118</td>\n",
       "      <td>up</td>\n",
       "      <td>3.192618e-19</td>\n",
       "      <td>0.213602</td>\n",
       "      <td>0.761</td>\n",
       "      <td>0.685</td>\n",
       "      <td>5.296554e-15</td>\n",
       "      <td>0.076</td>\n",
       "      <td>5.296554e-15</td>\n",
       "      <td>0.364350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3.0</td>\n",
       "      <td>TENM2</td>\n",
       "      <td>-8.466842</td>\n",
       "      <td>2.521362e-17</td>\n",
       "      <td>1.029066e-13</td>\n",
       "      <td>-0.734385</td>\n",
       "      <td>0.734385</td>\n",
       "      <td>down</td>\n",
       "      <td>1.081682e-16</td>\n",
       "      <td>-0.458044</td>\n",
       "      <td>0.498</td>\n",
       "      <td>0.639</td>\n",
       "      <td>1.794511e-12</td>\n",
       "      <td>-0.141</td>\n",
       "      <td>1.794511e-12</td>\n",
       "      <td>-0.814546</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4.0</td>\n",
       "      <td>RASGEF1B</td>\n",
       "      <td>7.965983</td>\n",
       "      <td>1.639156e-15</td>\n",
       "      <td>4.263772e-12</td>\n",
       "      <td>0.476628</td>\n",
       "      <td>0.476628</td>\n",
       "      <td>up</td>\n",
       "      <td>3.981474e-14</td>\n",
       "      <td>0.509415</td>\n",
       "      <td>0.629</td>\n",
       "      <td>0.512</td>\n",
       "      <td>6.605266e-10</td>\n",
       "      <td>0.117</td>\n",
       "      <td>6.605266e-10</td>\n",
       "      <td>0.803734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6394</th>\n",
       "      <td>6394.0</td>\n",
       "      <td>MMAB</td>\n",
       "      <td>-0.689976</td>\n",
       "      <td>4.902093e-01</td>\n",
       "      <td>9.998465e-01</td>\n",
       "      <td>-0.416674</td>\n",
       "      <td>0.416674</td>\n",
       "      <td>down</td>\n",
       "      <td>2.127330e-01</td>\n",
       "      <td>-0.039396</td>\n",
       "      <td>0.050</td>\n",
       "      <td>0.064</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.014</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.366560</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6416</th>\n",
       "      <td>6416.0</td>\n",
       "      <td>UBE2J1</td>\n",
       "      <td>-0.713720</td>\n",
       "      <td>4.754006e-01</td>\n",
       "      <td>9.998465e-01</td>\n",
       "      <td>-0.203220</td>\n",
       "      <td>0.203220</td>\n",
       "      <td>down</td>\n",
       "      <td>8.506776e-02</td>\n",
       "      <td>-0.027932</td>\n",
       "      <td>0.068</td>\n",
       "      <td>0.083</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.015</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.297539</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6430</th>\n",
       "      <td>6430.0</td>\n",
       "      <td>JUP</td>\n",
       "      <td>-0.699657</td>\n",
       "      <td>4.841416e-01</td>\n",
       "      <td>9.998465e-01</td>\n",
       "      <td>-0.592525</td>\n",
       "      <td>0.592525</td>\n",
       "      <td>down</td>\n",
       "      <td>7.678922e-02</td>\n",
       "      <td>-0.038278</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.040</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.014</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.575935</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6474</th>\n",
       "      <td>6474.0</td>\n",
       "      <td>TMTC2</td>\n",
       "      <td>-0.661018</td>\n",
       "      <td>5.086006e-01</td>\n",
       "      <td>9.998465e-01</td>\n",
       "      <td>0.053248</td>\n",
       "      <td>0.053248</td>\n",
       "      <td>up</td>\n",
       "      <td>1.571967e-02</td>\n",
       "      <td>-0.079056</td>\n",
       "      <td>0.498</td>\n",
       "      <td>0.528</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.030</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.162633</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6501</th>\n",
       "      <td>6501.0</td>\n",
       "      <td>DYNC1I1</td>\n",
       "      <td>-0.677187</td>\n",
       "      <td>4.982873e-01</td>\n",
       "      <td>9.998465e-01</td>\n",
       "      <td>-0.427949</td>\n",
       "      <td>0.427949</td>\n",
       "      <td>down</td>\n",
       "      <td>8.080137e-02</td>\n",
       "      <td>-0.051140</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.051</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.015</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.500943</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>797 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0     names    scores         pvals     pvals_adj  \\\n",
       "0            0.0      GFAP  9.283654  1.637632e-20  2.129904e-16   \n",
       "1            1.0     NRXN1 -8.830657  1.040630e-18  6.767220e-15   \n",
       "2            2.0     PRKG1  8.440312  3.164897e-17  1.029066e-13   \n",
       "3            3.0     TENM2 -8.466842  2.521362e-17  1.029066e-13   \n",
       "4            4.0  RASGEF1B  7.965983  1.639156e-15  4.263772e-12   \n",
       "...          ...       ...       ...           ...           ...   \n",
       "6394      6394.0      MMAB -0.689976  4.902093e-01  9.998465e-01   \n",
       "6416      6416.0    UBE2J1 -0.713720  4.754006e-01  9.998465e-01   \n",
       "6430      6430.0       JUP -0.699657  4.841416e-01  9.998465e-01   \n",
       "6474      6474.0     TMTC2 -0.661018  5.086006e-01  9.998465e-01   \n",
       "6501      6501.0   DYNC1I1 -0.677187  4.982873e-01  9.998465e-01   \n",
       "\n",
       "      logfoldchanges  abs_logfoldchanges Direction_t_test         p_val  \\\n",
       "0           1.128408            1.128408               up  7.630713e-23   \n",
       "1          -1.276099            1.276099             down  5.718907e-13   \n",
       "2           0.988118            0.988118               up  3.192618e-19   \n",
       "3          -0.734385            0.734385             down  1.081682e-16   \n",
       "4           0.476628            0.476628               up  3.981474e-14   \n",
       "...              ...                 ...              ...           ...   \n",
       "6394       -0.416674            0.416674             down  2.127330e-01   \n",
       "6416       -0.203220            0.203220             down  8.506776e-02   \n",
       "6430       -0.592525            0.592525             down  7.678922e-02   \n",
       "6474        0.053248            0.053248               up  1.571967e-02   \n",
       "6501       -0.427949            0.427949             down  8.080137e-02   \n",
       "\n",
       "      avg_log2FC  pct.1  pct.2     p_val_adj   diff           FDR  \\\n",
       "0       0.316605  0.559  0.411  1.265935e-18  0.148  1.265935e-18   \n",
       "1      -0.142494  0.898  0.942  9.487666e-09 -0.044  9.487666e-09   \n",
       "2       0.213602  0.761  0.685  5.296554e-15  0.076  5.296554e-15   \n",
       "3      -0.458044  0.498  0.639  1.794511e-12 -0.141  1.794511e-12   \n",
       "4       0.509415  0.629  0.512  6.605266e-10  0.117  6.605266e-10   \n",
       "...          ...    ...    ...           ...    ...           ...   \n",
       "6394   -0.039396  0.050  0.064  1.000000e+00 -0.014  1.000000e+00   \n",
       "6416   -0.027932  0.068  0.083  1.000000e+00 -0.015  1.000000e+00   \n",
       "6430   -0.038278  0.026  0.040  1.000000e+00 -0.014  1.000000e+00   \n",
       "6474   -0.079056  0.498  0.528  1.000000e+00 -0.030  1.000000e+00   \n",
       "6501   -0.051140  0.036  0.051  1.000000e+00 -0.015  1.000000e+00   \n",
       "\n",
       "      bulk_log2FC  p_val_adj_DEGs.Ind.Model  DEGs.Ind.Model  \\\n",
       "0        0.755717                       NaN            True   \n",
       "1       -0.211132                       NaN            True   \n",
       "2        0.364350                       NaN            True   \n",
       "3       -0.814546                       NaN            True   \n",
       "4        0.803734                       NaN            True   \n",
       "...           ...                       ...             ...   \n",
       "6394    -0.366560                       NaN           False   \n",
       "6416    -0.297539                       NaN           False   \n",
       "6430    -0.575935                       NaN           False   \n",
       "6474    -0.162633                       NaN           False   \n",
       "6501    -0.500943                       NaN           False   \n",
       "\n",
       "      DEGs.Ind.Mixed.Model  \n",
       "0                    False  \n",
       "1                    False  \n",
       "2                    False  \n",
       "3                    False  \n",
       "4                    False  \n",
       "...                    ...  \n",
       "6394                 False  \n",
       "6416                 False  \n",
       "6430                 False  \n",
       "6474                 False  \n",
       "6501                 False  \n",
       "\n",
       "[797 rows x 19 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mast_overlap['Astrocyte'][(mast_overlap['Excitatory']['DEGs.Ind.Model'])&\n",
    "                           (mast_overlap['Excitatory']['DEGs.Ind.Mixed.Model'])]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "21d83925",
   "metadata": {},
   "source": [
    "### **Save Data**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "11cfe608",
   "metadata": {},
   "outputs": [],
   "source": [
    "with pd.ExcelWriter(\"../results/ad_vs_no/mixed_model_degs.xlsx\") as writer:\n",
    "    for cell_type in adata_annot.obs.cell_type.unique():\n",
    "        mast_overlap[cell_type].to_excel(writer, sheet_name=cell_type, na_rep='NA')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "2f414a000012e1980f5fd1119ed6b9bf8cc73b51e37b52eb52a16ff857effd1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
