{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62042f97-a76e-4b34-805f-3f1ac81e7288",
   "metadata": {},
   "outputs": [],
   "source": [
    "import rpy2\n",
    "import logging\n",
    "import warnings\n",
    "import anndata2ri\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from tabnanny import verbose\n",
    "from gsva_prep import prep_gsva\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "from matplotlib.pyplot import rcParams\n",
    "from functions import pathway_analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c36a2bac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sys_dpi(width, height, diag):\n",
    "    '''\n",
    "    obtain dpi of system\n",
    "    \n",
    "    w: width in pixels (if unsure, go vist `whatismyscreenresolution.net`)\n",
    "    h: height in pixels\n",
    "    d: diagonal in inches\n",
    "    '''\n",
    "    w_inches = (diag**2/ (1 + height**2/width**2))**0.5\n",
    "    return round(width/w_inches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "447b6d1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "rcParams['figure.dpi'] = get_sys_dpi(1512, 982, 14.125)\n",
    "#rcParams['figure.figsize']=(4,4) #rescale figures\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f8aeaa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "suppressPackageStartupMessages({\n",
    "    library(WGCNA)\n",
    "    library(Matrix)\n",
    "    library(viridis)\n",
    "    library(harmony)\n",
    "    library(ggpubr)\n",
    "    library(tictoc)\n",
    "    library(RColorBrewer)\n",
    "    library(Hmisc)\n",
    "    library(corrplot)\n",
    "    library(grid)\n",
    "    library(gridExtra)\n",
    "    library(igraph)\n",
    "    library(ggrepel)\n",
    "    library(readxl)\n",
    "    library(conflicted)\n",
    "\n",
    "\n",
    "\n",
    "    # single-cell analysis package\n",
    "    library(Seurat)\n",
    "\n",
    "    # plotting and data science packages\n",
    "    library(tidyverse)\n",
    "    library(cowplot)\n",
    "    library(patchwork)\n",
    "\n",
    "    # co-expression network analysis packages:\n",
    "    library(WGCNA)\n",
    "    library(hdWGCNA)\n",
    "\n",
    "    # gene enrichment packages\n",
    "    library(enrichR)\n",
    "    library(GeneOverlap)\n",
    "\n",
    "\n",
    "    library(GSEABase)\n",
    "    library(GSVA) \n",
    "# needs to be run every time you start R and want to use %>%\n",
    "})\n",
    "\n",
    "# using the cowplot theme for ggplot\n",
    "theme_set(theme_cowplot())\n",
    "\n",
    "# set random seed for reproducibility\n",
    "set.seed(12345)\n",
    "\n",
    "# optionally enable multithreading\n",
    "# enableWGCNAThreads(nThreads = 4)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "baf04c5b",
   "metadata": {},
   "source": [
    "## **Prepare data**\n",
    "\n",
    "Now, we load the preprocessed and annotated data for downstream analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cca84d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_prefix = 'leng_sfg'\n",
    "map_meta = True\n",
    "deg_method =  'DESeq2-Wald'\n",
    "test_names = ['late_vs_early', 'early_vs_no', 'late_vs_no', 'ad_vs_no']\n",
    "filter_genes = \"TRUE\"\n",
    "subject_ids_for_study = {'leng_sfg': 'PatientID',\n",
    "                        'leng_etc': 'PatientID',\n",
    "                        'seaad_mtg': 'individualID'}\n",
    "\n",
    "subject_id = subject_ids_for_study[save_prefix]     # for leng this is `PatientID` for mathys is 'Subject', and allen is 'individualID'\n",
    "gene_celltype_threshold = 0.10          # determines number of cells the gene must be expressed in \n",
    "covariates = ['None']                   # list of covariates to be accounted for in regression.\n",
    "gene_selection = 'custom'               # specifies the gene selection method when setting up seurat object for WGCNA. The th\n",
    "celltypes = [\"Excitatory\", \"Inhibitory\", \"Astrocyte\", \"Microglia\", \"Oligodendrocyte\", \"OPC\"]\n",
    "metadata = f'../data/raw/{save_prefix}/{save_prefix}_metadata.csv' \n",
    "meta = pd.read_csv(metadata, encoding_errors='ignore')\n",
    "\n",
    "gene_selection = 'custom'   # specifies the gene selection method when setting up seurat object for WGCNA. The th\n",
    "                            # Posible values are \"custom\", \"fraction\", \"variable\"\n",
    "                            # If custom, a list of genes must be passed.\n",
    "\n",
    "gene_set_select = 'diff_exp'      # If gene_selection = 'custom'. This specifies how to obtain the list of\n",
    "                                  # genes to pass into `SetupForWGCNA`. # The posible values are \"diff_exp\", \"overlap\", \"all\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6626ed1",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i subject_id -i gene_celltype_threshold -i celltypes -i test_names -i save_prefix -i gene_selection -i gene_set_select\n",
    "\n",
    "# Function to handle failure in data loading\n",
    "handle_loading_failure <- function(cell_type, seurat_obj) {\n",
    "    cat(paste0('Could not load data for ', toupper(cell_type), '\\n'))\n",
    "    cat(paste0(toupper(cell_type), ' dropped from experiment\\n'))\n",
    "  \n",
    "    # Remove the failed cell_type from the list\n",
    "    seurat_obj <- seurat_obj[names(seurat_obj) != cell_type]\n",
    "  \n",
    "    # Update the modified seurat_obj in the global environment\n",
    "    assign(\"seurat_obj\", seurat_obj, envir = .GlobalEnv)\n",
    "}\n",
    "\n",
    "seurat_obj <- list()\n",
    "\n",
    "for (cell_type in celltypes) {\n",
    "  \n",
    "    cat(paste0('Loading data for hdWGCNA Experiment in ', toupper(cell_type), '\\n'))\n",
    "  \n",
    "    # Use tryCatch to load data and handle errors or messages\n",
    "    tryCatch({\n",
    "        # Attempt to read the RDS file into the list\n",
    "        seurat_obj[[cell_type]] <- readRDS(paste0(\"../results/hdWGCNA/SeuratObject/\", save_prefix, '/', cell_type, '_hdWGCNA_object.rds'))\n",
    "        print(seurat_obj[[cell_type]])\n",
    "        print('............')\n",
    "  \n",
    "    }, error = function(e){\n",
    "        # Error handling: Remove the cell_type from the list and print an error message\n",
    "        handle_loading_failure(cell_type, seurat_obj)\n",
    "\n",
    "    }, message = function(m){\n",
    "        handle_loading_failure(cell_type, seurat_obj)\n",
    "    })\n",
    "}\n",
    "\n",
    "cat(\"Loaded available data\\n\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "10be21f0-000d-48d7-9e97-4051ec9da0bd",
   "metadata": {},
   "source": [
    "# **hdWGCNA Basic Visualization**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "03dc3f32",
   "metadata": {},
   "source": [
    "Here we visualize some results using hdWGCNA, and we employ some of Seurat’s built-in plotting tools to visualize our hdWGCNA results.\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "368c8716",
   "metadata": {},
   "source": [
    "### **Module Feature Plots**\n",
    "\n",
    "hdWGCNA includes the ModuleFeaturePlot function to consruct FeaturePlots for each co-expression module colored by each module’s uniquely assigned color."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5865383",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/FeaturePlots/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "  print(paste0('Module Feature Plots in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "     \n",
    "  # make a featureplot of hMEs for each module\n",
    "  plot_list <- ModuleFeaturePlot(\n",
    "    seurat_obj[[cell_type]],\n",
    "    features='hMEs', # plot the hMEs\n",
    "    order=TRUE, # order so the points with highest hMEs are on top\n",
    "    # raster=TRUE, \n",
    "    # raster_dpi=400, \n",
    "    # alpha=1,\n",
    "    # restrict_range=FALSE, \n",
    "    # raster_scale=0.25,\n",
    "  )\n",
    "\n",
    "  plot_list <- lapply(1:length(plot_list), function(x){\n",
    "    plot_list[[x]] + NoLegend() + theme(plot.title=element_text(face='plain', vjust=0.25), plot.margin=margin(c(0,0,0,0)))\n",
    "  })\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, \"_featureplot_MEs.pdf\"), height=10, width=5)\n",
    "  print(wrap_plots(plot_list, ncol=3))\n",
    "  dev.off()\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2e09999a",
   "metadata": {},
   "source": [
    "We can also plot the hub gene signature score using the same function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07825465",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/FeaturePlots/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "  print(paste0('Module Feature Plots for HUB GENES in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "     \n",
    "  # make a featureplot of hMEs for each module\n",
    "  plot_list <- ModuleFeaturePlot(\n",
    "    seurat_obj[[cell_type]],\n",
    "    features='scores',  # plot the hub gene scores\n",
    "    order='shuffle', # order so cells are shuffled\n",
    "    ucell = TRUE, # depending on Seurat vs UCell for gene scoring\n",
    "    # raster=TRUE, \n",
    "    # raster_dpi=400, \n",
    "    # alpha=1,\n",
    "    # restrict_range=FALSE, \n",
    "    # raster_scale=0.25,\n",
    "  )\n",
    "\n",
    "  plot_list <- lapply(1:length(plot_list), function(x){\n",
    "    plot_list[[x]] + NoLegend() + theme(plot.title=element_text(face='plain', vjust=0.25), plot.margin=margin(c(0,0,0,0)))\n",
    "  })\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, \"_featureplot_hubs.pdf\"), height=10, width=5)\n",
    "  print(wrap_plots(plot_list, ncol=3))\n",
    "  dev.off()\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "219d5785",
   "metadata": {},
   "source": [
    "### **Module Correlations**\n",
    "\n",
    "hdWGCNA includes the ModuleCorrelogram function to visualize the correlation between each module based on their hMEs, MEs, or hub gene scores using the R package corrplot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ac695b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/ModuleCorrelations/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "    print(paste0('Module Correlation in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "    # plot module correlagram\n",
    "    pdf(paste0(fig_dir, cell_type, \"_correlogram.pdf\"), height=10, width=5)\n",
    "    ModuleCorrelogram(seurat_obj[[cell_type]])\n",
    "    dev.off()\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3cf396dd",
   "metadata": {},
   "source": [
    "### **Seurat plotting functions**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dd017bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "mods <- list()\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "    # get hMEs from seurat object\n",
    "    MEs <- GetMEs(seurat_obj[[cell_type]], harmonized=TRUE)\n",
    "    mods[[cell_type]] <- colnames(MEs); \n",
    "    mods[[cell_type]] <- mods[[cell_type]][mods[[cell_type]] != 'grey']\n",
    "    \n",
    "    \n",
    "    # add hMEs to Seurat meta-data:\n",
    "    seurat_obj[[cell_type]]@meta.data <- cbind(seurat_obj[[cell_type]]@meta.data, MEs)\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c1c37b40",
   "metadata": {},
   "source": [
    "Now we can easily use Seurat’s DotPlot function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f45a3f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i seurat_obj -i mods -o seurat_obj\n",
    "\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/DotPlot/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('DotPlot for average expression of MEs in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # plot with Seurat's DotPlot function\n",
    "  p <- DotPlot(seurat_obj[[cell_type]], features=mods[[cell_type]], group.by = 'cell_type')\n",
    "\n",
    "  # flip the x/y axes, rotate the axis labels, and change color scheme:\n",
    "  p <- p +\n",
    "    coord_flip() +\n",
    "    RotatedAxis() +\n",
    "    scale_color_gradient2(high='red', mid='grey95', low='blue')\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, \"_dotplot.pdf\"), height=10, width=5)\n",
    "  print(p)\n",
    "  dev.off()\n",
    "\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b1b75ca0",
   "metadata": {},
   "source": [
    "Here is another example where we use Seurat’s VlnPlot function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c23ebda",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "# for (cell_type in names(seurat_obj)){\n",
    "\n",
    "#   print(paste0('ViolinPlot for average expression of MEs in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "#   # Plot INH-M4 hME using Seurat VlnPlot function\n",
    "#   p <- VlnPlot(\n",
    "#     seurat_obj,\n",
    "#     features = as.vector(mods[[cell_type]]),\n",
    "#     group.by = 'cell_type',\n",
    "#     pt.size = 0 # don't show actual data points\n",
    "#   )\n",
    "\n",
    "#   # add box-and-whisker plots on top:\n",
    "#   p <- p + geom_boxplot(width=.25, fill='white')\n",
    "\n",
    "#   # change axis labels and remove legend:\n",
    "#   p <- p + xlab('') + ylab('hME') + NoLegend()\n",
    "\n",
    "#   # plot output\n",
    "#   pdf(paste0(\"../results/hdWGCNA/ViolinPlot/\", save_prefix, '_', cell_type, \"_violinplot.pdf\"), height=10, width=5)\n",
    "#   print(p)\n",
    "#   dev.off()\n",
    "\n",
    "# }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8246a79f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # convert nested list of Seurat object into Rpy2 object \n",
    "# seurat_obj =  robjects.ListVector(\n",
    "#                         {\n",
    "#                             cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "#                             for cell_type in seurat_obj.keys()\n",
    "#                         }\n",
    "#                     )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b657e6a1",
   "metadata": {},
   "source": [
    "## **Overlap With AD Genes**\n",
    "\n",
    "Next on the co-expression UMAP, we show&quantify the overlap between hub genes that have a known association with AD using the AD databases obtained from the [**Open Targets Platform**](https://platform.opentargets.org/disease/MONDO_0004975/associations), [**KEGG Alzheimer's pathway**](https://www.genome.jp/pathway/hsa05010) and [**Harmonizome (Mayaanlab)**](https://maayanlab.cloud/Harmonizome/gene_set/Alzheimer+Disease/dbGAP+Gene-Trait+Associations) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "964ea985",
   "metadata": {},
   "outputs": [],
   "source": [
    "KEGG_paths = pathway_analyses.read_pathways('../data/pathway_databases/KEGG_2019_Human.txt')\n",
    "KEGG_genes = [gene for gene in list(KEGG_paths[(KEGG_paths[0].str.startswith('Alzheimer disease'))].iloc[:, 1:].values[0]) if str(gene)!=\"nan\"]\n",
    "mayaanlab_genes = pd.read_csv('../data/pathway_databases/AD_genes.csv').Symbol.to_list()\n",
    "otp_genes = pd.read_csv('../data/pathway_databases/MONDO_0004975-associated-diseases.tsv', sep='\\t')\n",
    "otp_genes = otp_genes[otp_genes.textMining!='No data'].symbol.to_list()\n",
    "\n",
    "AD_genes = list(set(set(KEGG_genes).union(set(mayaanlab_genes), set(otp_genes))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61d42932",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i AD_genes\n",
    "\n",
    "conflicts_prefer(base::intersect)\n",
    "\n",
    "# load modules\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/GeneOverlap/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Estimating AD gene overlap in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  modules <- GetModules(seurat_obj[[cell_type]])\n",
    "  mods <- levels(modules$module)\n",
    "\n",
    "  AD_genes <- intersect(as.vector(AD_genes), (GetWGCNAGenes(seurat_obj[[cell_type]])))\n",
    " \n",
    "  # genome.size <- dim(seurat_obj[[cell_type]])[1]\n",
    "  genome.size <- nrow(modules)\n",
    "  # genome.size <- length(GetWGCNAGenes(seurat_obj[[cell_type]]))\n",
    "  # genome.size <- length(c(GetWGCNAGenes(seurat_obj[[cell_type]]), AD_genes))\n",
    "\n",
    "  overlap_df <- do.call(rbind, lapply(mods, function(cur_mod){\n",
    "\n",
    "  cur_genes <- modules %>% subset(module == cur_mod) %>% .$gene_name\n",
    "\n",
    "  cur_overlap <- testGeneOverlap(newGeneOverlap(\n",
    "      cur_genes,\n",
    "      as.vector(AD_genes),\n",
    "      genome.size=genome.size\n",
    "  ))\n",
    "\n",
    "  cur_overlap <- data.frame(\n",
    "    'odds.ratio' = cur_overlap@odds.ratio,\n",
    "    'pval' = cur_overlap@pval,\n",
    "    'Jaccard' = cur_overlap@Jaccard,\n",
    "    'size_intersection' = length(cur_overlap@intersection),\n",
    "    'module' = cur_mod\n",
    "  )\n",
    "\n",
    "    cur_overlap\n",
    "\n",
    "  })) %>% as.data.frame()\n",
    "\n",
    "  overlap_df <- overlap_df %>% mutate(fdr=p.adjust(pval, method='fdr'))\n",
    "  overlap_df <- overlap_df %>% subset(module != 'grey')\n",
    "\n",
    "  ################################################################################\n",
    "  # Plot as a lollipop\n",
    "  ################################################################################\n",
    "\n",
    "  overlap_df$shape <- ifelse(overlap_df$fdr < 0.05, 21, 4)\n",
    "  overlap_df <- overlap_df %>% arrange(odds.ratio, descending=TRUE)\n",
    "  overlap_df$module <- factor(as.character(overlap_df$module), levels=as.character(overlap_df$module))\n",
    "\n",
    "  mod_colors <- dplyr::select(modules, c(module, color)) %>%\n",
    "    distinct\n",
    "  cp <- mod_colors$color; names(cp) <- mod_colors$module\n",
    "\n",
    "  p <- overlap_df %>%\n",
    "    ggplot(aes(y=module, x=odds.ratio, size= size_intersection, color=module)) +\n",
    "    geom_segment(aes(y=module, yend=module, x=0, xend=odds.ratio), size=0.5, color='grey') +\n",
    "    geom_point() +\n",
    "    geom_point(shape=overlap_df$shape, color='black', fill=NA) +\n",
    "    scale_color_manual(values=cp, guide='none') +\n",
    "    ylab('') + xlab(\"Odds ratio\") +\n",
    "    scale_x_continuous(breaks = seq(0, 19, 2)) +\n",
    "    labs(size='Size\\nintersection') +\n",
    "    ggtitle('Overlap with AD genes') +\n",
    "\n",
    "    theme(\n",
    "      panel.border = element_rect(size=1, color='black', fill=NA),\n",
    "      axis.line.y = element_blank(),\n",
    "      axis.line.x = element_blank(),\n",
    "      plot.title = element_text(hjust=0.5, face='plain')\n",
    "    )\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, '_AD_overlap.pdf'), width=4, height=3.5)\n",
    "  print(p)\n",
    "  dev.off()\n",
    "}\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4cf51263",
   "metadata": {},
   "source": [
    "## **Overlap With DEGs**\n",
    "\n",
    "We quantify the overlap between hub genes and DEGs obtained from previous analysis and saved at `../results/{test_name}/{save_prefix}_{deg_method}_degs.xlsx`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd63a2f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i deg_method\n",
    "\n",
    "\n",
    "# load modules\n",
    "conflicts_prefer(base::intersect)\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/GeneOverlap/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (test in c('ad_vs_no')){\n",
    "  for (cell_type in names(seurat_obj)){\n",
    "\n",
    "    print(paste0('Estimating AD gene overlap in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    wgcna_genes <- GetWGCNAGenes(seurat_obj[[cell_type]])\n",
    "    # hub_genes <- GetHubGenes(seurat_obj[[cell_type]], 25)\n",
    "    degs <- read_excel(paste0('../results/', test, '/', save_prefix, '/DEG/', deg_method, '_degs.xlsx'), sheet = cell_type)\n",
    "    degs <- intersect(GetWGCNAGenes(seurat_obj[[cell_type]]), \n",
    "                      subset(degs, p_val_adj<0.01 & abs_logFC>0.25)$gene)\n",
    "\n",
    "    modules <- GetModules(seurat_obj[[cell_type]])\n",
    "    mods <- levels(modules$module)\n",
    "\n",
    "    # genome.size <- dim(seurat_obj[[cell_type]])[1]\n",
    "    genome.size <- nrow(modules)\n",
    "    # genome.size <- length(GetWGCNAGenes(seurat_obj[[cell_type]]))\n",
    "    # genome.size <- length(c(GetWGCNAGenes(seurat_obj[[cell_type]]), AD_genes))\n",
    "\n",
    "    overlap_df <- do.call(rbind, lapply(mods, function(cur_mod){\n",
    "\n",
    "    cur_genes <- modules %>% subset(module == cur_mod) %>% .$gene_name \n",
    "    # cur_genes <- intersect(cur_genes, GetHubGenes(seurat_obj[[cell_type]], 25))\n",
    "    \n",
    "    cur_overlap <- testGeneOverlap(newGeneOverlap(\n",
    "        cur_genes,\n",
    "        as.vector(degs),\n",
    "        genome.size=genome.size\n",
    "    ))\n",
    "\n",
    "    cur_overlap <- data.frame(\n",
    "      'odds.ratio' = cur_overlap@odds.ratio,\n",
    "      'pval' = cur_overlap@pval,\n",
    "      'Jaccard' = cur_overlap@Jaccard,\n",
    "      'size_intersection' = length(cur_overlap@intersection),\n",
    "      'module' = cur_mod\n",
    "    )\n",
    "\n",
    "      cur_overlap\n",
    "\n",
    "    })) %>% as.data.frame()\n",
    "\n",
    "    overlap_df <- overlap_df %>% mutate(fdr=p.adjust(pval, method='fdr'))\n",
    "    overlap_df <- overlap_df %>% subset(module != 'grey')\n",
    "\n",
    "    ################################################################################\n",
    "    # Plot as a lollipop\n",
    "    ################################################################################\n",
    "\n",
    "    overlap_df$shape <- ifelse(overlap_df$fdr < 0.05, 21, 4)\n",
    "    overlap_df <- overlap_df %>% arrange(odds.ratio, descending=TRUE)\n",
    "    overlap_df$module <- factor(as.character(overlap_df$module), levels=as.character(overlap_df$module))\n",
    "\n",
    "    mod_colors <- dplyr::select(modules, c(module, color)) %>%\n",
    "      distinct\n",
    "    cp <- mod_colors$color; names(cp) <- mod_colors$module\n",
    "\n",
    "    p <- overlap_df %>%\n",
    "      ggplot(aes(y=module, x=odds.ratio, size= size_intersection, color=module)) +\n",
    "      geom_segment(aes(y=module, yend=module, x=0, xend=odds.ratio), size=0.5, color='grey') +\n",
    "      geom_point() +\n",
    "      geom_point(shape=overlap_df$shape, color='black', fill=NA) +\n",
    "      scale_color_manual(values=cp, guide='none') +\n",
    "      ylab('') + xlab(\"Odds ratio\") +\n",
    "      scale_x_continuous(breaks = seq(0, 19, 2)) +\n",
    "      labs(size='Size\\nintersection') +\n",
    "      ggtitle('Overlap with DEGs') +\n",
    "\n",
    "      theme(\n",
    "        panel.border = element_rect(size=1, color='black', fill=NA),\n",
    "        axis.line.y = element_blank(),\n",
    "        axis.line.x = element_blank(),\n",
    "        plot.title = element_text(hjust=0.5, face='plain')\n",
    "      )\n",
    "\n",
    "    pdf(paste0(fig_dir, test, '_', cell_type, '_DEG_overlap.pdf'), width=4, height=3.5)\n",
    "    print(p)\n",
    "    dev.off()\n",
    "  }\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3bc33b36",
   "metadata": {},
   "source": [
    "# **Save Seurat Object**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "431d9ac4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "dat_dir = paste0(\"../results/hdWGCNA/SeuratObject/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(dat_dir)) {\n",
    "  dir.create(dat_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "    print(paste0('Saving hdWGCNA object in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    saveRDS(seurat_obj[[cell_type]], file=paste0(dat_dir, cell_type, '_hdWGCNA_object.rds'))\n",
    "\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "2f414a000012e1980f5fd1119ed6b9bf8cc73b51e37b52eb52a16ff857effd1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
