{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "62042f97-a76e-4b34-805f-3f1ac81e7288",
   "metadata": {},
   "outputs": [],
   "source": [
    "import rpy2\n",
    "import scipy\n",
    "import logging\n",
    "import warnings\n",
    "import anndata2ri\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import seaborn as sb\n",
    "import decoupler as dc\n",
    "import scrublet as scr\n",
    "import decoupler as dc\n",
    "from scipy import sparse\n",
    "from anndata import AnnData\n",
    "from tabnanny import verbose\n",
    "import matplotlib.pyplot as plt\n",
    "from gsva_prep import prep_gsva\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "from typing import Optional, Union\n",
    "from matplotlib.pyplot import rcParams\n",
    "from functions import pathway_analyses\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from sklearn.model_selection import train_test_split\n",
    "from pytorch_lightning.loggers import TensorBoardLogger\n",
    "from rpy2.robjects.conversion import localconverter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c36a2bac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sys_dpi(width, height, diag):\n",
    "    '''\n",
    "    obtain dpi of system\n",
    "    \n",
    "    w: width in pixels (if unsure, go vist `whatismyscreenresolution.net`)\n",
    "    h: height in pixels\n",
    "    d: diagonal in inches\n",
    "    '''\n",
    "    w_inches = (diag**2/ (1 + height**2/width**2))**0.5\n",
    "    return round(width/w_inches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "447b6d1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.8.0\n",
      "scanpy      1.9.3\n",
      "-----\n",
      "PIL                         9.5.0\n",
      "absl                        NA\n",
      "anndata2ri                  1.1\n",
      "appnope                     0.1.3\n",
      "asttokens                   NA\n",
      "attr                        23.1.0\n",
      "backcall                    0.2.0\n",
      "botocore                    1.31.17\n",
      "cffi                        1.15.1\n",
      "comm                        0.1.3\n",
      "cycler                      0.10.0\n",
      "cython_runtime              NA\n",
      "dateutil                    2.8.2\n",
      "debugpy                     1.6.7\n",
      "decorator                   5.1.1\n",
      "decoupler                   1.4.0\n",
      "deprecate                   0.3.2\n",
      "dot_parser                  NA\n",
      "executing                   1.2.0\n",
      "fsspec                      2023.6.0\n",
      "functions                   NA\n",
      "google                      NA\n",
      "gsva_prep                   NA\n",
      "h5py                        3.9.0\n",
      "igraph                      0.10.4\n",
      "ipykernel                   6.23.2\n",
      "ipywidgets                  8.0.6\n",
      "jedi                        0.18.2\n",
      "jinja2                      3.1.2\n",
      "joblib                      1.2.0\n",
      "kiwisolver                  1.4.4\n",
      "leidenalg                   0.9.1\n",
      "llvmlite                    0.39.1\n",
      "louvain                     0.8.0\n",
      "markupsafe                  2.1.3\n",
      "matplotlib                  3.7.1\n",
      "mpl_toolkits                NA\n",
      "mpmath                      1.3.0\n",
      "natsort                     8.4.0\n",
      "numba                       0.56.4\n",
      "numexpr                     2.8.4\n",
      "numpy                       1.23.5\n",
      "opt_einsum                  v3.3.0\n",
      "packaging                   23.1\n",
      "pandas                      2.0.2\n",
      "parso                       0.8.3\n",
      "patsy                       0.5.3\n",
      "pexpect                     4.8.0\n",
      "pickleshare                 0.7.5\n",
      "pkg_resources               NA\n",
      "platformdirs                3.7.0\n",
      "prompt_toolkit              3.0.38\n",
      "psutil                      5.9.5\n",
      "ptyprocess                  0.7.0\n",
      "pure_eval                   0.2.2\n",
      "pyarrow                     12.0.1\n",
      "pycparser                   2.21\n",
      "pydev_ipython               NA\n",
      "pydevconsole                NA\n",
      "pydevd                      2.9.5\n",
      "pydevd_file_utils           NA\n",
      "pydevd_plugins              NA\n",
      "pydevd_tracing              NA\n",
      "pydot                       1.4.2\n",
      "pygments                    2.15.1\n",
      "pyparsing                   3.1.0\n",
      "pytorch_lightning           1.7.7\n",
      "pytz                        2023.3\n",
      "pytz_deprecation_shim       NA\n",
      "rich                        NA\n",
      "rpy2                        3.5.12\n",
      "scipy                       1.10.1\n",
      "scrublet                    NA\n",
      "seaborn                     0.12.2\n",
      "session_info                1.0.0\n",
      "setuptools                  68.0.0\n",
      "six                         1.16.0\n",
      "sklearn                     1.2.2\n",
      "sparse                      0.14.0\n",
      "stack_data                  0.6.2\n",
      "statsmodels                 0.14.0\n",
      "sympy                       1.12\n",
      "tensorboard                 2.13.0\n",
      "texttable                   1.6.7\n",
      "threadpoolctl               3.1.0\n",
      "torch                       2.0.1\n",
      "torchmetrics                0.11.4\n",
      "tornado                     6.3.2\n",
      "tqdm                        4.65.0\n",
      "traitlets                   5.9.0\n",
      "typing_extensions           NA\n",
      "tzlocal                     NA\n",
      "wcwidth                     0.2.6\n",
      "yaml                        6.0\n",
      "zmq                         25.1.0\n",
      "zoneinfo                    NA\n",
      "-----\n",
      "IPython             8.14.0\n",
      "jupyter_client      8.2.0\n",
      "jupyter_core        5.3.1\n",
      "-----\n",
      "Python 3.10.6 (main, Oct 24 2022, 11:04:34) [Clang 12.0.0 ]\n",
      "macOS-10.16-x86_64-i386-64bit\n",
      "-----\n",
      "Session information updated at 2023-09-22 12:15\n"
     ]
    }
   ],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# # Automatically convert rpy2 outputs to pandas dataframes\n",
    "# pandas2ri.activate()\n",
    "# anndata2ri.activate()\n",
    "# %load_ext rpy2.ipython\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "rcParams['figure.dpi'] = get_sys_dpi(1512, 982, 14.125)\n",
    "#rcParams['figure.figsize']=(4,4) #rescale figures\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "10be21f0-000d-48d7-9e97-4051ec9da0bd",
   "metadata": {},
   "source": [
    "# **hdWGCNA in single-cell data**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "0c5d2f8d",
   "metadata": {},
   "source": [
    "Here we use `hdWGCNA` to perform co-expression network analysis on single-cell data. \n",
    "\n",
    "`hdWGCNA` is an R package for performing `weighted gene co-expression network analysis (WGCNA)` in high dimensional transcriptomics data such as single-cell RNA-seq or spatial transcriptomics. hdWGCNA is highly modular and can construct co-expression networks across multi-scale cellular and spatial hierarchies. hdWGNCA identifies robust modules of inerconnected genes, and provides context for these modules through various biological knowledge sources. hdWGCNA requires data formatted as Seurat objects, one of the most ubiquitous formats for single-cell data. \n",
    "\n",
    "`Note: hdWGCNA is under active development, so you may run into errors and small typos. We welcome users to write GitHub issues to report bugs, ask for help, and to request potential enhancements.`\n",
    "\n",
    "Explore the full capabilites of hWGNA in the following publications:\n",
    "\n",
    "- [**Morabito et al. bioRxiv 2022**](https://www.biorxiv.org/content/10.1101/2022.09.22.509094v1)\n",
    "- [**Morabito & Miyoshi et al. Nature Genetics 2021**](https://www.nature.com/articles/s41588-021-00894-z)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1b4733e8",
   "metadata": {},
   "source": [
    "Here, apply hdWGCNA to the processed single-nucleus RNA-seq (snRNA-seq) datasets. These dataset have already been fully processed using a standard single-cell transcritpomics analysis pipelines (please see the following notebooks, `preprocessing_data_integration`, `preprocessing_quality_control`, `preprocessing_cluster_annotation`). \n",
    "\n",
    "\n",
    "If you would like to apply this notebook to your own dataset, you first need to satisfy the following prerequisites:\n",
    "\n",
    "A single-cell or single-nucleus transcriptomics dataset in Seurat format.\n",
    "Normalize the gene expression matrix NormalizeData.\n",
    "Identify highly variable genes VariableFeatures.\n",
    "Scale the normalized expression data ScaleData\n",
    "Perform dimensionality reduction RunPCA and batch correction if needed RunHarmony.\n",
    "Non-linear dimensionality reduction RunUMAP for visualizations.\n",
    "Group cells into clusters (FindNeighbors and FindClusters).\n",
    "An example of running the prerequisite data processing steps can be found in the Seurat Guided Clustering Tutorial.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8b87d243",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "data": {
      "text/plain": [
       "In addition: Warning messages:\n",
       "1: package ‘IRanges’ was built under R version 4.3.1 \n",
       "2: package ‘GSVA’ was built under R version 4.3.1 \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "suppressPackageStartupMessages({\n",
    "    library(WGCNA)\n",
    "    library(Matrix)\n",
    "    library(viridis)\n",
    "    library(harmony)\n",
    "    library(ggpubr)\n",
    "    library(tictoc)\n",
    "    library(RColorBrewer)\n",
    "    library(Hmisc)\n",
    "    library(corrplot)\n",
    "    library(grid)\n",
    "    library(gridExtra)\n",
    "    library(igraph)\n",
    "    library(ggrepel)\n",
    "    library(readxl)\n",
    "    library(conflicted)\n",
    "\n",
    "\n",
    "\n",
    "    # single-cell analysis package\n",
    "    library(Seurat)\n",
    "\n",
    "    # plotting and data science packages\n",
    "    library(tidyverse)\n",
    "    library(cowplot)\n",
    "    library(patchwork)\n",
    "\n",
    "    # co-expression network analysis packages:\n",
    "    library(WGCNA)\n",
    "    library(hdWGCNA)\n",
    "\n",
    "    # gene enrichment packages\n",
    "    library(enrichR)\n",
    "    library(GeneOverlap)\n",
    "\n",
    "\n",
    "    library(GSEABase)\n",
    "    library(GSVA) \n",
    "# needs to be run every time you start R and want to use %>%\n",
    "})\n",
    "\n",
    "# using the cowplot theme for ggplot\n",
    "theme_set(theme_cowplot())\n",
    "\n",
    "# set random seed for reproducibility\n",
    "set.seed(12345)\n",
    "\n",
    "# optionally enable multithreading\n",
    "# enableWGCNAThreads(nThreads = 4)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "af425574",
   "metadata": {},
   "source": [
    "## **Prepare data**\n",
    "\n",
    "Now, we load the preprocessed and annotated data for downstream analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9697f83a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\"counts\" not in layers...\n",
      "analysis requires unnormalized count data...\n",
      "analysis proceesing with data in \"adata.x\"...\n",
      "AnnData object with n_obs × n_vars = 63608 × 33694\n",
      "    obs: 'SampleID', 'PatientID', 'BrainRegion', 'BraakStage', 'SampleBatch', 'nUMI', 'nGene', 'initialClusterAssignments', 'seurat.clusters', 'clusterAssignment', 'clusterCellType', 'cell_type'\n"
     ]
    }
   ],
   "source": [
    "save_prefix = 'leng_sfg'\n",
    "\n",
    "adata_annot = sc.read_h5ad(f'../data/raw/{save_prefix}/{save_prefix}_raw_anndata.h5ad')\n",
    "adata_annot.obs_names_make_unique()\n",
    "adata_annot.var_names_make_unique()\n",
    "\n",
    "try:\n",
    "    adata_annot.X = adata_annot.layers['counts'].copy()\n",
    "except KeyError:\n",
    "    print('\"counts\" not in layers...')\n",
    "    print('analysis requires unnormalized count data...')\n",
    "    print('analysis proceesing with data in \"adata.x\"...')\n",
    "    \n",
    "del adata_annot.obsm, adata_annot.layers, adata_annot.varm, adata_annot.uns, adata_annot.obsp\n",
    "\n",
    "print(adata_annot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d8504101",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_cell_types = True # whether to  to obtain appropriate cell-type columns: True/False\n",
    "map_meta = True\n",
    "deg_method =  'DESeq2-Wald'\n",
    "test_names = ['late_vs_early', 'early_vs_no', 'late_vs_no', 'ad_vs_no']\n",
    "filter_genes = \"TRUE\"\n",
    "subject_ids_for_study = {'leng_sfg': 'PatientID',\n",
    "                        'leng_etc': 'PatientID',\n",
    "                        'seaad_mtg': 'individualID'}\n",
    "\n",
    "subject_id = subject_ids_for_study[save_prefix]     # for leng this is `PatientID` for mathys is 'Subject', and allen is 'individualID'\n",
    "gene_celltype_threshold = 0.10          # determines number of cells the gene must be expressed in \n",
    "covariates = ['None']                   # list of covariates to be accounted for in regression.\n",
    "gene_selection = 'custom'               # specifies the gene selection method when setting up seurat object for WGCNA. The th\n",
    "celltypes = [\"Excitatory\", \"Inhibitory\", \"Astrocyte\", \"Microglia\", \"Oligodendrocyte\", \"OPC\"]\n",
    "metadata = f'../data/raw/{save_prefix}/{save_prefix}_metadata.csv' \n",
    "meta = pd.read_csv(metadata, encoding_errors='ignore')\n",
    "\n",
    "gene_selection = 'custom'   # specifies the gene selection method when setting up seurat object for WGCNA. The th\n",
    "                            # Posible values are \"custom\", \"fraction\", \"variable\"\n",
    "                            # If custom, a list of genes must be passed.\n",
    "\n",
    "gene_set_select = 'diff_exp' # If gene_selection = 'custom'. This specifies how to obtain the list of\n",
    "                             # genes to pass into `SetupForWGCNA`. # The posible values are \"diff_exp\", \"overlap\", \"all\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6640b2ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "mapping = {'leng_etc':\n",
    "           \n",
    "           {'Exc': 'Excitatory', \n",
    "            'Inh': 'Inhibitory', \n",
    "            'Astro': 'Astrocyte',\n",
    "            'Endo': 'Endothelial', \n",
    "            'Micro': 'Microglia', \n",
    "            'OPC': 'OPC', \n",
    "            'Oligo': 'Oligodendrocyte'},\n",
    "\n",
    "           'leng_sfg':\n",
    "           \n",
    "           {'Exc': 'Excitatory', \n",
    "            'Inh': 'Inhibitory', \n",
    "            'Astro': 'Astrocyte',\n",
    "            'Endo': 'Endothelial', \n",
    "            'Micro': 'Microglia', \n",
    "            'OPC': 'OPC', \n",
    "            'Oligo': 'Oligodendrocyte'},\n",
    "            \n",
    "            'seaad_mtg':\n",
    "           \n",
    "           {'Excitatory': 'Excitatory', \n",
    "            'Inhibitory': 'Inhibitory',\n",
    "            'Astrocyte': 'Astrocyte',\n",
    "            'Microglia': 'Microglia', \n",
    "            'Endothelial': 'Endothelial', \n",
    "            'OPC': 'OPC',\n",
    "            'Oligodendrocyte': 'Oligodendrocyte'},\n",
    "            }\n",
    "\n",
    "cell_column = {'leng_etc': 'clusterCellType',\n",
    "               'leng_sfg': 'clusterCellType',\n",
    "               'seaad_mtg': 'cell_type',\n",
    "               }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fa49b63c",
   "metadata": {},
   "outputs": [],
   "source": [
    "if get_cell_types:\n",
    "    adata_annot.obs['cell_type'] = adata_annot.obs[cell_column[save_prefix]].map(mapping[save_prefix])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3888c7a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# map the pathology group to teh subject id in .obs\n",
    "adata_annot = adata_annot[adata_annot.obs.cell_type.isin(celltypes)]\n",
    "adata_annot.obs['pathology.group'] = adata_annot.obs[subject_id].map(dict(zip(meta[subject_id].astype(str), meta['pathology.group'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "839f795a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "class: SingleCellExperiment \n",
      "dim: 33694 62211 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B\n",
      "rowData names(0):\n",
      "colnames(62211): SFG2_AAACCTGAGATGGCGT SFG2_AAACCTGAGCGATCCC ...\n",
      "  SFG10_TTTGTCATCACGGTTA SFG10_TTTGTCATCATACGGT\n",
      "colData names(13): SampleID PatientID ... cell_type pathology.group\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "[1] \"loaded data into memory for recursive use\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i adata_annot -i subject_id -i gene_celltype_threshold -i celltypes -i test_names -i save_prefix -i gene_selection\n",
    "\n",
    "print(adata_annot)\n",
    "print('loaded data into memory for recursive use')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "74c5d20a",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata_annot"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "8a3955a2",
   "metadata": {},
   "source": [
    "## **Set up Seurat object for WGCNA**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2d154d2e",
   "metadata": {},
   "source": [
    "Before running hdWGCNA, we first have to set up the Seurat object. Most of the information computed by hdWGCNA is stored in the Seurat object’s `@misc slot`. \n",
    "\n",
    "The `SetupForWGCNA` function selects the genes that will be used for WGCNA using three different approaches using the `gene_select` parameter:\n",
    "\n",
    "- `variable`: use the genes stored in the Seurat object’s `VariableFeatures`.\n",
    "\n",
    "- `fraction`: use genes that are expressed in a certain fraction of cells for in the whole dataset or in each group of cells, specified by group.by.\n",
    "\n",
    "- `custom`: use genes that are specified in a custom list."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3e7607cd",
   "metadata": {},
   "source": [
    "For our specific case, the `gene_selection` parameter is set to `custom`, we employ the genes present in all the overlapping pathways in a cell type-specific and study-specific manner.\n",
    "\n",
    "To accomplish this, we load the list of overlapping pathways that was previously saved in `../results/{test_name}/{test_name}_pathway_overlap.csv` from the `pathway_meta_analysis.ipynb` notebook. Additionally, we load the genes from the `.gmt` file saved in `../results/gsva/{study}/{cur_celltype}/gs.gmt` from the `pathway_analysis.ipynb` notebook."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c4ddf8bf",
   "metadata": {},
   "source": [
    "##### **Select Gene Set**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "7234ec55",
   "metadata": {},
   "source": [
    "To obtain the gene sets required for the `SetupForWGCNA` function, we consider the `gene_set_select` variable when `gene_select` is set to `custom`. The `gene_set_select` variable determines how the `gene_list` passed into the `SetupForWGCNA` function will be obtained. \n",
    "\n",
    "There are three options available:\n",
    "\n",
    "- `diff_exp`: This option utilizes the genes present in the differentially expressed pathways, previously computed and saved to the `../results/{test_name}/{save_prefix}_differentially_expressed_pathways.csv` file.\n",
    "\n",
    "- `overlap`: Here, the genes used are those expressed in pathways that overlap across different studies. The pathway overlap information is retrieved from the `../results/{test_name}/{test_name}_pathway_overlap.csv` file.\n",
    "\n",
    "- `all`: With this option, all genes in the pathways stored in the `../results/gsva/{save_prefix}/{celltype}/gs.gmt` file are included.\n",
    "\n",
    "Here, we specifically set `gene_select=custom` and `gene_set_select = diff_exp`. However, you can customize the gene selection process for your hdWGCNA experiment.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "703041ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 42\n",
      "[1] 109\n",
      "[1] 134\n",
      "[1] 84\n",
      "[1] 26\n",
      "[1] 23\n",
      "[1] 1449\n",
      "[1] 714\n",
      "[1] 382\n",
      "[1] 97\n",
      "[1] 86\n",
      "[1] 209\n",
      "[1] 1785\n",
      "[1] 1405\n",
      "[1] 393\n",
      "[1] 107\n",
      "[1] 160\n",
      "[1] 359\n",
      "[1] 2034\n",
      "[1] 1410\n",
      "[1] 564\n",
      "[1] 102\n",
      "[1] 196\n",
      "[1] 353\n"
     ]
    }
   ],
   "source": [
    "%%R -o geneSets -i gene_set_select\n",
    "\n",
    "# iterate over celltypes and obtain \n",
    "geneSets <- list()\n",
    "for(test_name in test_names){\n",
    "\n",
    "  geneSets[[test_name]] <- list()\n",
    "  for (cur_celltype in celltypes){\n",
    "    # get gene sets\n",
    "    geneSets[[test_name]][[cur_celltype]] <- getGmt(file.path('../results/gsva/', save_prefix, '/', cur_celltype, 'gs.gmt'))\n",
    "    geneSets[[test_name]][[cur_celltype]] <- geneIds(geneSets[[test_name]][[cur_celltype]])\n",
    "\n",
    "    if (gene_set_select=='overlap'){\n",
    "      pathways <- read.csv(paste0('../results/',test_name, '/', test_name, '_pathway_overlap.csv'))\n",
    "      pathways <- pathways$pathway\n",
    "    }else if (gene_set_select=='diff_exp'){\n",
    "      pathways <- read.csv(paste0('../results/', test_name, '/', save_prefix, '/Data/differentially_expressed_pathways.csv'))\n",
    "      pathways <- subset(pathways, P.Value < 0.05 & celltype == cur_celltype)$pathway\n",
    "    }else if (gene_set_select=='all'){\n",
    "      pathways <- names(geneSets[[test_name]][[cur_celltype]])\n",
    "    }\n",
    "\n",
    "    geneSets[[test_name]][[cur_celltype]] <-  geneSets[[test_name]][[cur_celltype]][names(geneSets[[test_name]][[cur_celltype]]) %in% pathways]\n",
    "    geneSets[[test_name]][[cur_celltype]] <- unique(unlist(geneSets[[test_name]][[cur_celltype]], recursive = FALSE))\n",
    "\n",
    "    print(length(geneSets[[test_name]][[cur_celltype]]))\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f2d4d1b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of geneSets object into Rpy2 object \n",
    "\n",
    "geneSets =  robjects.ListVector(\n",
    "                        {\n",
    "                            test_name: robjects.ListVector(\n",
    "                                    {cell_type: geneSets[test_name][cell_type]\n",
    "                                                 \n",
    "                                    for cell_type in celltypes\n",
    "                                    }\n",
    "                                )\n",
    "                            for test_name in test_names\n",
    "                        }\n",
    "\n",
    "                        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "845fa1c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "An object of class Seurat \n",
      "33694 features across 62211 samples within 1 assay \n",
      "Active assay: originalexp (33694 features, 0 variable features)\n",
      "[1] \"Creating hdWGNA Experiment for EXCITATORY\"\n",
      "An object of class Seurat \n",
      "33694 features across 20301 samples within 1 assay \n",
      "Active assay: originalexp (33694 features, 2000 variable features)\n",
      " 3 dimensional reductions calculated: pca, harmony, umap\n",
      "[1] \"2034 WGCNA Genes\"\n",
      "[1] \"Creating hdWGNA Experiment for INHIBITORY\"\n",
      "An object of class Seurat \n",
      "33694 features across 7964 samples within 1 assay \n",
      "Active assay: originalexp (33694 features, 2000 variable features)\n",
      " 3 dimensional reductions calculated: pca, harmony, umap\n",
      "[1] \"1410 WGCNA Genes\"\n",
      "[1] \"Creating hdWGNA Experiment for ASTROCYTE\"\n",
      "An object of class Seurat \n",
      "33694 features across 8025 samples within 1 assay \n",
      "Active assay: originalexp (33694 features, 2000 variable features)\n",
      " 3 dimensional reductions calculated: pca, harmony, umap\n",
      "[1] \"564 WGCNA Genes\"\n",
      "[1] \"Creating hdWGNA Experiment for MICROGLIA\"\n",
      "An object of class Seurat \n",
      "33694 features across 4093 samples within 1 assay \n",
      "Active assay: originalexp (33694 features, 2000 variable features)\n",
      " 3 dimensional reductions calculated: pca, harmony, umap\n",
      "[1] \"102 WGCNA Genes\"\n",
      "[1] \"Creating hdWGNA Experiment for OLIGODENDROCYTE\"\n",
      "An object of class Seurat \n",
      "33694 features across 18360 samples within 1 assay \n",
      "Active assay: originalexp (33694 features, 2000 variable features)\n",
      " 3 dimensional reductions calculated: pca, harmony, umap\n",
      "[1] \"196 WGCNA Genes\"\n",
      "[1] \"Creating hdWGNA Experiment for OPC\"\n",
      "An object of class Seurat \n",
      "33694 features across 3468 samples within 1 assay \n",
      "Active assay: originalexp (33694 features, 2000 variable features)\n",
      " 3 dimensional reductions calculated: pca, harmony, umap\n",
      "[1] \"353 WGCNA Genes\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')\n",
       "Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')\n",
       "Calculating gene variances\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Calculating feature variances of standardized and clipped values\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |===================================                                   |  50%\r  |                                                                            \r  |======================================================================| 100%\n",
       "PC_ 1 \n",
       "Positive:  CNDP1, SLC1A3, SEPP1, MOBP, SLC5A11, EMX2, SPP1, FGFR3, PDK4, SDC4 \n",
       "\t   ZIC1, GJA1, CTD-2282P23.2, HIF3A, CX3CR1, SLC14A1, VSIG4, APOE, C1QC, APCDD1 \n",
       "\t   ITPKB-AS1, LINC00609, MS4A6A, ZFP36L1, C1QB, AGT, RP11-489O18.1, C1QA, CD74, CSF3R \n",
       "Negative:  MEG3, KALRN, PDE4D, OPCML, AGBL4, STXBP5L, RIMS2, KCNIP4, OXR1, CCSER1 \n",
       "\t   PRKCB, PHACTR1, KCNQ5, PLCB1, GABRB3, DCLK1, MKL2, RYR2, LDB2, SLC4A10 \n",
       "\t   RALYL, GABBR2, FRMPD4, HECW1, ATP2B1, CADPS, THRB, NELL2, FGF12, LRFN5 \n",
       "PC_ 2 \n",
       "Positive:  LSAMP-AS1, PDE5A, AC016831.7, RP11-479O9.4, SPDYE3, RP11-649G15.2, RP1-78O14.1, RP11-160H12.2, MAK, P2RY14 \n",
       "\t   RP11-318K12.2, RP11-778J15.1, RP11-315E17.1, PAPPA2, TIMP3, RP11-138A9.2, RP11-116O11.1, SNAP25-AS1, CTA-276F8.2, RP11-650J17.1 \n",
       "\t   RP11-192H23.6, RP11-317J19.1, RP11-317G6.1, RP11-390N6.1, LINC00903, AC010127.3, RP11-770G2.2, RP11-123M6.2, RP5-1101C3.1, AC016716.2 \n",
       "Negative:  PHYHIP, MT-ND1, SYN1, TSPYL1, GDI1, MT-ND2, FAIM2, PPP2R1A, NRGN, SLC17A7 \n",
       "\t   EEF1A2, MT-CO1, TSPYL4, SNCB, IDS, MT-ATP6, LINC00657, KLC1, TUBB4A, YWHAG \n",
       "\t   CDK5R1, TOLLIP, PSAP, MT-CO3, MT-ND3, NCDN, MT-ND4, TSPAN7, MTPN, HSP90AB1 \n",
       "PC_ 3 \n",
       "Positive:  ERBB4, NXPH1, RP11-123O10.4, RBMS3, PTPRM, ZNF385D, GAD1, GAD2, SOX6, SLC6A1 \n",
       "\t   GRIP1, VWC2, MYO16, SLC6A1-AS1, GRIP2, KCNAB1, BTBD11, ANK1, DNER, KIAA1211 \n",
       "\t   KCNC2, PTCHD4, ALK, GRIK1, ADCY8, NHS, ADAMTS17, TENM1, ARX, XKR4 \n",
       "Negative:  PTPRK, CBLN2, AC011288.2, KCNIP4, AC114765.1, KIAA1211L, LY86-AS1, MLIP, AC067956.1, RP11-191L9.4 \n",
       "\t   TESPA1, NECAB1, SYN3, SATB2, RALYL, LINC01378, MIR137HG, GRM1, NELL2, IQCJ-SCHIP1 \n",
       "\t   PDZD2, ANO3, R3HDM1, ADAMTS3, RGS6, SNED1, DGKB, LINC01197, CA10, LINC01250 \n",
       "PC_ 4 \n",
       "Positive:  SLC1A3, ADGRV1, SLC1A2, TPD52L1, GPC5, ATP1A2, AQP4, RNF219-AS1, GFAP, ZNF98 \n",
       "\t   COL5A3, HIF3A, CST3, LINC00499, PARD3, BMPR1B, RYR3, F3, RFX4, LRIG1 \n",
       "\t   ALDH1L1, GJA1, PREX2, RGS20, CTD-2282P23.2, SLC4A4, GNA14, MT2A, RANBP3L, SLC14A1 \n",
       "Negative:  RP11-123O10.4, NXPH1, GRIP1, ZNF385D, XKR4, KCNC2, VWC2, KCNAB1, GAD2, PTCHD4 \n",
       "\t   GAD1, GRIP2, CNTN5, ZMAT4, SLC44A5, TENM1, PLCXD3, SLIT2, ANK1, GRIK1 \n",
       "\t   MYO16, SGCZ, ZNF804A, ERBB4, CNTNAP5, PTPRM, KIAA1217, GRIK2, BTBD11, SLC35F4 \n",
       "PC_ 5 \n",
       "Positive:  CTC-340A15.2, CTC-535M15.2, FOXP2, PTPRT, CPNE4, IL1RAPL2, RXFP1, CAMK2D, CELF4, KCNB2 \n",
       "\t   RORB, TLL1, RP11-707A18.1, NKAIN2, CLSTN2, ROBO2, DNAH6, PID1, TSHZ2, ARSJ \n",
       "\t   MSC-AS1, DPY19L1, LRRK1, FSTL5, RP11-478B9.1, CCDC85A, CDH8, STX1A, HS3ST4, TOX \n",
       "Negative:  RP11-314P15.2, GNAL, TESPA1, PDZD2, SV2C, CUX2, AC010127.3, CDH9, CBLN2, SYN3 \n",
       "\t   TIMP3, TRPC6, LAMP5, FAM84A, MOXD1, CTXN3, CDH18, CDH12, PCDH8, ITPR2 \n",
       "\t   PTPRO, KCTD16, HOPX, PPARGC1A, CBLN4, MIR34AHG, EPHA6, ESRRG, ONECUT2, GPAT3 \n",
       "Harmony 1/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 2/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 3/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 4/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 5/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 6/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 7/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 8/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony converged after 8 iterations\n",
       "Warning: Invalid name supplied, making object name syntactically valid. New object name is Seurat..ProjectDim.originalexp.harmony; see ?make.names for more details on syntax validity\n",
       "Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric\n",
       "To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'\n",
       "This message will be shown once per session\n",
       "12:23:10 UMAP embedding parameters a = 1.577 b = 0.8951\n",
       "12:23:10 Read 62211 rows and found 30 numeric columns\n",
       "12:23:10 Using Annoy for neighbor search, n_neighbors = 15\n",
       "12:23:10 Building Annoy index with metric = cosine, n_trees = 50\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "12:23:23 Writing NN index file to temp file /var/folders/70/42xwxjh94x5f164djlw1hwpc0000gv/T//Rtmpt1Axm2/file1840c6f4e8f1f\n",
       "12:23:23 Searching Annoy index using 1 thread, search_k = 1500\n",
       "12:23:35 Annoy recall = 99.4%\n",
       "12:23:36 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 15\n",
       "12:23:38 Initializing from normalized Laplacian + noise (using irlba)\n",
       "12:23:50 Commencing optimization for 200 epochs, with 1473046 positive edges\n",
       "Using method 'umap'\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "12:24:21 Optimization finished\n",
       "In addition: Warning messages:\n",
       "1: Quick-TRANSfer stage steps exceeded maximum (= 3110550) \n",
       "2: In SetupForWGCNA(seurat_obj[[cur_cell_type]], gene_select = \"custom\",  :\n",
       "  102 features selected. You may wish to change the settings to include more features in your analysis.\n",
       "3: In SetupForWGCNA(seurat_obj[[cur_cell_type]], gene_select = \"custom\",  :\n",
       "  196 features selected. You may wish to change the settings to include more features in your analysis.\n",
       "4: In SetupForWGCNA(seurat_obj[[cur_cell_type]], gene_select = \"custom\",  :\n",
       "  353 features selected. You may wish to change the settings to include more features in your analysis.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i geneSets\n",
    "\n",
    "seurat_obj <- as.Seurat(adata_annot, counts = \"X\", data = \"X\")\n",
    "\n",
    "print(seurat_obj)\n",
    "\n",
    "# Perform dimensionality reduction and plot\n",
    "\n",
    "seurat_obj <- FindVariableFeatures(seurat_obj)\n",
    "seurat_obj <- ScaleData(seurat_obj)\n",
    "seurat_obj <- RunPCA(seurat_obj)\n",
    "seurat_obj <- RunHarmony(seurat_obj, group.by.vars = subject_id)\n",
    "seurat_obj <- RunUMAP(seurat_obj, reduction='harmony', n.neighbors=15, dims=1:30, min.dist=0.1)\n",
    "\n",
    "p <- DimPlot(seurat_obj, group.by = \"cell_type\", label = TRUE) +\n",
    "     umap_theme() + ggtitle(\"leng\") + NoLegend()\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/UMAP/\")\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "# Save plot to PDF\n",
    "pdf(file = paste0(fig_dir, save_prefix, \"_integrated_umap.pdf\"), width = 4, height = 4, useDingbats = FALSE)\n",
    "print(p)\n",
    "dev.off()\n",
    "\n",
    "# create a hdWGCNA experiment for each celltype\n",
    "seurat_dat <- seurat_obj\n",
    "seurat_obj <- list()\n",
    "\n",
    "\n",
    "for (cur_cell_type in celltypes){\n",
    "\n",
    "  seurat_obj[[cur_cell_type]] <- subset(seurat_dat, cell_type == cur_cell_type)\n",
    "\n",
    "  print(paste0('Creating hdWGNA Experiment for ', toupper(cur_cell_type)))\n",
    "\n",
    "  if (gene_selection == 'custom') {\n",
    "    seurat_obj[[cur_cell_type]] <- SetupForWGCNA(\n",
    "      seurat_obj[[cur_cell_type]],\n",
    "      gene_select = \"custom\",                                            # the gene selection approach\n",
    "      gene_list = as.vector(geneSets[['ad_vs_no']][[cur_cell_type]]),    # list of genes to be included\n",
    "      group.by = 'cell_type',                                            # grouping parameter\n",
    "      wgcna_name = toupper(substr(cur_cell_type, 1, 3))                  # the name of the hdWGCNA experiment\n",
    "    )\n",
    "  } else {\n",
    "    seurat_obj[[cur_cell_type]] <- SetupForWGCNA(\n",
    "      seurat_obj[[cur_cell_type]],\n",
    "      gene_select = \"fraction\",                               # the gene selection approach\n",
    "      fraction = gene_celltype_threshold,                     # fraction of cells for gene inclusion\n",
    "      group.by = 'cell_type',                                 # grouping parameter\n",
    "      wgcna_name = toupper(substr(cur_cell_type, 1, 3))       # the name of the hdWGCNA experiment\n",
    "    )\n",
    "  }\n",
    "\n",
    "  print(seurat_obj[[cur_cell_type]])\n",
    "  print(paste0(length(GetWGCNAGenes(seurat_obj[[cur_cell_type]])), \" WGCNA Genes\"))\n",
    "\n",
    "}\n",
    "\n",
    "rm(seurat_dat)\n",
    "rm(adata_annot)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ccd0e590",
   "metadata": {},
   "source": [
    "## **Construct metacells**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "07813598",
   "metadata": {},
   "source": [
    "In the hdWGCNA pipeline, the initial step involves constructing metacells from a single-cell dataset using the `MetacellsByGroups` function, which aggregates similar cells using the `k-Nearest Neighbors` algorithm. This process is essential for reducing the sparsity of the gene expression matrix and involves grouping by specific parameters like `Sample` and `cell_type` to ensure metacells originate from the same biological sample. The function allows tuning of the number of cells aggregated (k) and the maximum shared overlap, with a min_cells parameter to exclude small groups, thereby optimizing the construction of a metacell expression matrix for subsequent analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e92860fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Constructing MetaCells in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"Constructing MetaCells in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"Constructing MetaCells in hdWGCNA Experiment for ASTROCYTE\"\n",
      "[1] \"Constructing MetaCells in hdWGCNA Experiment for MICROGLIA\"\n",
      "[1] \"Constructing MetaCells in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"Constructing MetaCells in hdWGCNA Experiment for OPC\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Performing log-normalization\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Performing log-normalization\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Performing log-normalization\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Performing log-normalization\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Performing log-normalization\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Performing log-normalization\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "In addition: Warning messages:\n",
       "1: In MetacellsByGroups(seurat_obj = seurat_obj[[cell_type]], group.by = c(\"cell_type\",  :\n",
       "  Removing the following groups that did not meet min_cells: Microglia#4, Microglia#7\n",
       "2: In MetacellsByGroups(seurat_obj = seurat_obj[[cell_type]], group.by = c(\"cell_type\",  :\n",
       "  Removing the following groups that did not meet min_cells: OPC#7\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "# construct metacells in each group for each wgnca experiment\n",
    "\n",
    "for (cell_type in celltypes){\n",
    "\n",
    "  print(paste0('Constructing MetaCells in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "  \n",
    "  seurat_obj[[cell_type]] <- MetacellsByGroups(\n",
    "      \n",
    "    seurat_obj = seurat_obj[[cell_type]],\n",
    "    group.by = c(\"cell_type\", subject_id), # specify the columns in seurat_obj@meta.data to group by\n",
    "    reduction = 'harmony',  # select the dimensionality reduction to perform KNN on\n",
    "    k = 25, # nearest-neighbors parameter\n",
    "    max_shared = 10, # maximum number of shared cells between two metacells\n",
    "    ident.group = 'cell_type', # set the Idents of the metacell seurat object\n",
    "    wgcna_name = toupper(substr(cell_type, 1, 3)),        # the name of the hdWGCNA experiment\n",
    "    )\n",
    "\n",
    "    # normalize metacell expression matrix:\n",
    "  seurat_obj[[cell_type]] <- NormalizeMetacells(seurat_obj[[cell_type]],  wgcna_name = toupper(substr(cell_type, 1, 3)))\n",
    "}\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "fac55be8",
   "metadata": {},
   "source": [
    "## **Co-expression network analysis**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "f03072ae",
   "metadata": {},
   "source": [
    "### **Set up the expression matrix**\n",
    "\n",
    "the `SetDatExpr` function is used to set up the expression matrix for network analysis, involving subsetting the data to include only a specific cell type per run. By default, the metacell expression matrix is utilized `(use_metacells=TRUE)`, though the option to use the single-cell expression matrix is available. This function also allows users to choose the source slot for the expression matrix, accommodating different normalization methods like SCTransform or NormalizeData."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "7830296b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Setting up expression matrix in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"Setting up expression matrix in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"Setting up expression matrix in hdWGCNA Experiment for ASTROCYTE\"\n",
      "[1] \"Setting up expression matrix in hdWGCNA Experiment for MICROGLIA\"\n",
      "[1] \"Setting up expression matrix in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"Setting up expression matrix in hdWGCNA Experiment for OPC\"\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "for (cell_type in celltypes){\n",
    "\n",
    "    print(paste0('Setting up expression matrix in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    seurat_obj[[cell_type]] <- SetDatExpr(\n",
    "      seurat_obj[[cell_type]],\n",
    "      group_name = cell_type, # the name of the group of interest in the group.by column\n",
    "      group.by = 'cell_type', # the metadata column containing the cell type info. This same column should have also been used in MetacellsByGroups\n",
    "      assay = 'originalexp', # using RNA assay\n",
    "      slot = 'data', # using normalized data\n",
    "      wgcna_name = toupper(substr(cell_type, 1, 3)) # the name of the hdWGCNA experiment\n",
    "    )\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a70cc5c4",
   "metadata": {},
   "source": [
    "### **Select soft-power threshold**\n",
    "\n",
    "Next we will select the “soft power threshold”. Selecting the `soft power threshold` is a crucial step, as it influences the construction of the gene-gene correlation adjacency matrix used to infer co-expression relationships between genes. The `TestSoftPowers` function is employed to conduct a parameter sweep for various soft power thresholds, aiding in choosing an appropriate value to ensure the co-expression network approximates a scale-free topology. Additionally, the `PlotSoftPowers`` function is used to visually represent the outcomes of this parameter sweep, to assess the network's structure at different thresholds."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e63133c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Estinmating Soft-Power Threshold in hdWGCNA Experiment for EXCITATORY\"\n",
      "pickSoftThreshold: will use block size 2034.\n",
      " pickSoftThreshold: calculating connectivity for given powers...\n",
      "   ..working on genes 1 through 2034 of 2034\n",
      "   Power SFT.R.sq  slope truncated.R.sq  mean.k. median.k.   max.k.\n",
      "1      1   0.7890 16.000          0.907 1.07e+03  1.07e+03 1160.000\n",
      "2      2   0.5770  6.180          0.898 5.71e+02  5.73e+02  681.000\n",
      "3      3   0.0697  1.520          0.798 3.09e+02  3.08e+02  414.000\n",
      "4      4   0.0185 -0.692          0.908 1.69e+02  1.67e+02  260.000\n",
      "5      5   0.1540 -1.750          0.931 9.35e+01  9.17e+01  167.000\n",
      "6      6   0.3620 -2.570          0.960 5.24e+01  5.05e+01  110.000\n",
      "7      7   0.5330 -2.970          0.942 2.97e+01  2.82e+01   73.500\n",
      "8      8   0.6620 -3.270          0.956 1.70e+01  1.58e+01   49.900\n",
      "9      9   0.7480 -3.400          0.966 9.89e+00  8.91e+00   34.400\n",
      "10    10   0.8020 -3.370          0.958 5.81e+00  5.12e+00   24.100\n",
      "11    12   0.8540 -3.170          0.949 2.08e+00  1.70e+00   12.200\n",
      "12    14   0.8910 -2.880          0.951 7.86e-01  5.72e-01    6.530\n",
      "13    16   0.4300 -3.870          0.390 3.13e-01  1.97e-01    3.640\n",
      "14    18   0.4260 -3.490          0.377 1.32e-01  6.90e-02    2.110\n",
      "15    20   0.4080 -3.170          0.356 5.88e-02  2.51e-02    1.280\n",
      "16    22   0.3930 -2.880          0.333 2.79e-02  9.07e-03    0.797\n",
      "17    24   0.3800 -2.630          0.316 1.40e-02  3.35e-03    0.514\n",
      "18    26   0.3750 -2.430          0.306 7.41e-03  1.28e-03    0.340\n",
      "19    28   0.3650 -2.250          0.296 4.13e-03  4.80e-04    0.231\n",
      "20    30   0.3590 -2.120          0.290 2.41e-03  1.83e-04    0.160\n",
      "  Power   SFT.R.sq      slope truncated.R.sq    mean.k.  median.k.    max.k.\n",
      "1     1 0.78883750 15.9727587      0.9067747 1070.73984 1073.92792 1162.4601\n",
      "2     2 0.57658627  6.1780008      0.8981842  571.43548  572.73615  680.6203\n",
      "3     3 0.06974750  1.5162920      0.7979864  308.82860  307.87998  414.3543\n",
      "4     4 0.01853858 -0.6916295      0.9084724  168.95144  167.12434  260.3603\n",
      "5     5 0.15413220 -1.7543238      0.9312117   93.53992   91.71976  167.4796\n",
      "6     6 0.36206798 -2.5732216      0.9595248   52.40481   50.54391  109.9463\n",
      "[1] \"Estinmating Soft-Power Threshold in hdWGCNA Experiment for INHIBITORY\"\n",
      "pickSoftThreshold: will use block size 1410.\n",
      " pickSoftThreshold: calculating connectivity for given powers...\n",
      "   ..working on genes 1 through 1410 of 1410\n",
      "   Power SFT.R.sq  slope truncated.R.sq  mean.k. median.k.  max.k.\n",
      "1      1    0.565 28.500         0.8300 7.21e+02  7.23e+02 770.000\n",
      "2      2    0.390 10.300         0.9460 3.73e+02  3.75e+02 429.000\n",
      "3      3    0.187  4.320         0.9860 1.95e+02  1.96e+02 244.000\n",
      "4      4    0.015 -0.795         0.9670 1.03e+02  1.03e+02 146.000\n",
      "5      5    0.205 -2.420         0.9430 5.54e+01  5.48e+01  89.700\n",
      "6      6    0.585 -3.910         0.9670 3.00e+01  2.94e+01  56.900\n",
      "7      7    0.770 -4.040         0.9770 1.65e+01  1.59e+01  37.100\n",
      "8      8    0.797 -4.050         0.9300 9.17e+00  8.64e+00  24.700\n",
      "9      9    0.876 -3.610         0.9580 5.17e+00  4.74e+00  16.800\n",
      "10    10    0.914 -3.330         0.9740 2.96e+00  2.61e+00  11.700\n",
      "11    12    0.896 -3.020         0.9140 1.02e+00  8.09e-01   5.910\n",
      "12    14    0.336 -3.950         0.1780 3.78e-01  2.56e-01   3.150\n",
      "13    16    0.911 -2.210         0.8980 1.52e-01  8.31e-02   1.770\n",
      "14    18    0.903 -1.960         0.8850 6.64e-02  2.78e-02   1.050\n",
      "15    20    0.876 -1.800         0.8500 3.16e-02  9.52e-03   0.659\n",
      "16    22    0.244 -2.100         0.0328 1.62e-02  3.29e-03   0.452\n",
      "17    24    0.938 -1.590         0.9310 8.86e-03  1.17e-03   0.335\n",
      "18    26    0.239 -1.830         0.0424 5.12e-03  4.19e-04   0.252\n",
      "19    28    0.371 -2.130         0.3020 3.10e-03  1.55e-04   0.193\n",
      "20    30    0.381 -2.090         0.3180 1.94e-03  5.77e-05   0.149\n",
      "  Power   SFT.R.sq      slope truncated.R.sq   mean.k. median.k.    max.k.\n",
      "1     1 0.56511004 28.5340303      0.8302502 721.04505 722.81957 770.05638\n",
      "2     2 0.38992926 10.2692130      0.9459914 373.26034 374.73680 429.23076\n",
      "3     3 0.18660788  4.3187087      0.9863004 195.41708 196.09677 243.95767\n",
      "4     4 0.01499945 -0.7948735      0.9666698 103.47198 103.46093 145.65710\n",
      "5     5 0.20524040 -2.4181459      0.9425020  55.42252  54.84898  89.72688\n",
      "6     6 0.58537808 -3.9097164      0.9671269  30.04156  29.40730  56.90162\n",
      "[1] \"Estinmating Soft-Power Threshold in hdWGCNA Experiment for ASTROCYTE\"\n",
      "pickSoftThreshold: will use block size 564.\n",
      " pickSoftThreshold: calculating connectivity for given powers...\n",
      "   ..working on genes 1 through 564 of 564\n",
      "   Power SFT.R.sq   slope truncated.R.sq  mean.k. median.k.  max.k.\n",
      "1      1  0.00102  -0.848         0.9220 2.84e+02  2.83e+02 301.000\n",
      "2      2  0.75500 -11.000         0.8110 1.45e+02  1.43e+02 172.000\n",
      "3      3  0.95900  -8.420         0.9480 7.57e+01  7.24e+01 105.000\n",
      "4      4  0.96500  -5.830         0.9800 4.01e+01  3.71e+01  68.100\n",
      "5      5  0.95100  -4.380         0.9570 2.16e+01  1.91e+01  46.600\n",
      "6      6  0.94900  -3.440         0.9550 1.19e+01  9.94e+00  33.200\n",
      "7      7  0.95400  -2.830         0.9610 6.71e+00  5.19e+00  24.500\n",
      "8      8  0.95800  -2.440         0.9580 3.89e+00  2.74e+00  18.600\n",
      "9      9  0.96600  -2.160         0.9630 2.32e+00  1.46e+00  14.500\n",
      "10    10  0.93700  -2.020         0.9190 1.43e+00  7.75e-01  11.500\n",
      "11    12  0.95000  -1.690         0.9360 6.00e-01  2.22e-01   7.480\n",
      "12    14  0.97700  -1.480         0.9720 2.86e-01  6.62e-02   5.080\n",
      "13    16  0.26200  -2.150         0.0856 1.52e-01  2.03e-02   3.560\n",
      "14    18  0.26100  -1.960         0.0900 8.72e-02  6.19e-03   2.550\n",
      "15    20  0.26800  -1.880         0.1020 5.33e-02  1.97e-03   1.850\n",
      "16    22  0.27000  -1.800         0.1100 3.40e-02  6.39e-04   1.370\n",
      "17    24  0.24500  -1.650         0.0309 2.24e-02  2.09e-04   1.020\n",
      "18    26  0.24800  -1.630         0.0344 1.52e-02  7.08e-05   0.770\n",
      "19    28  0.25300  -1.580         0.0413 1.05e-02  2.46e-05   0.586\n",
      "20    30  0.25600  -1.570         0.0447 7.35e-03  8.63e-06   0.449\n",
      "  Power    SFT.R.sq       slope truncated.R.sq   mean.k.  median.k.    max.k.\n",
      "1     1 0.001016574  -0.8479119      0.9223690 284.03074 283.163248 301.12812\n",
      "2     2 0.755259528 -10.9876849      0.8113833 145.48350 142.963740 171.51649\n",
      "3     3 0.959074685  -8.4195908      0.9484562  75.70086  72.423939 104.90320\n",
      "4     4 0.965240591  -5.8262227      0.9795720  40.06447  37.130418  68.13157\n",
      "5     5 0.951497959  -4.3758195      0.9571611  21.60642  19.132084  46.57245\n",
      "6     6 0.949179220  -3.4379689      0.9545731  11.90128   9.942155  33.22422\n",
      "[1] \"Estinmating Soft-Power Threshold in hdWGCNA Experiment for MICROGLIA\"\n",
      "pickSoftThreshold: will use block size 102.\n",
      " pickSoftThreshold: calculating connectivity for given powers...\n",
      "   ..working on genes 1 through 102 of 102\n",
      "   Power SFT.R.sq  slope truncated.R.sq  mean.k. median.k.   max.k.\n",
      "1      1    0.252  66.80        0.07970 5.11e+01  5.12e+01 5.37e+01\n",
      "2      2    0.105   4.96        0.71100 2.61e+01  2.61e+01 2.93e+01\n",
      "3      3    0.167 -21.40        0.07640 1.35e+01  1.34e+01 1.64e+01\n",
      "4      4    0.704  -7.07        0.82000 7.03e+00  6.90e+00 9.38e+00\n",
      "5      5    0.204 -14.40       -0.02190 3.70e+00  3.57e+00 5.47e+00\n",
      "6      6    0.213 -11.70       -0.00905 1.97e+00  1.86e+00 3.25e+00\n",
      "7      7    0.220  -9.63       -0.00239 1.06e+00  1.00e+00 1.97e+00\n",
      "8      8    0.908  -4.19        0.88100 5.75e-01  5.34e-01 1.21e+00\n",
      "9      9    0.877  -3.65        0.84700 3.16e-01  2.86e-01 7.57e-01\n",
      "10    10    0.124  -4.68       -0.02590 1.76e-01  1.54e-01 4.80e-01\n",
      "11    12    0.138  -3.92       -0.00634 5.61e-02  4.48e-02 2.00e-01\n",
      "12    14    0.203  -6.09        0.05270 1.89e-02  1.37e-02 8.74e-02\n",
      "13    16    0.249  -4.95        0.03440 6.69e-03  4.27e-03 3.96e-02\n",
      "14    18    0.243  -4.29        0.02630 2.51e-03  1.33e-03 1.86e-02\n",
      "15    20    0.236  -3.83        0.01820 9.96e-04  4.26e-04 8.94e-03\n",
      "16    22    0.286  -4.52        0.08260 4.17e-04  1.38e-04 4.40e-03\n",
      "17    24    0.172  -2.85       -0.03910 1.84e-04  4.53e-05 2.21e-03\n",
      "18    26    0.172  -2.60       -0.04010 8.47e-05  1.58e-05 1.13e-03\n",
      "19    28    0.170  -2.40       -0.04170 4.05e-05  5.47e-06 5.88e-04\n",
      "20    30    0.167  -2.35       -0.04100 2.00e-05  1.94e-06 3.09e-04\n",
      "  Power  SFT.R.sq      slope truncated.R.sq   mean.k. median.k.    max.k.\n",
      "1     1 0.2517774  66.796134    0.079743411 51.069719 51.234845 53.690412\n",
      "2     2 0.1052943   4.959370    0.710525627 26.097475 26.127152 29.318797\n",
      "3     3 0.1669886 -21.401338    0.076404869 13.473664 13.391470 16.399066\n",
      "4     4 0.7038876  -7.074992    0.819875897  7.026746  6.897696  9.376114\n",
      "5     5 0.2039330 -14.408378   -0.021873575  3.701592  3.569870  5.470854\n",
      "6     6 0.2133051 -11.724367   -0.009045163  1.969784  1.859380  3.253208\n",
      "[1] \"Estinmating Soft-Power Threshold in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "pickSoftThreshold: will use block size 196.\n",
      " pickSoftThreshold: calculating connectivity for given powers...\n",
      "   ..working on genes 1 through 196 of 196\n",
      "   Power SFT.R.sq   slope truncated.R.sq  mean.k. median.k.   max.k.\n",
      "1      1    0.180 162.000         0.0417 9.80e+01  9.81e+01 1.00e+02\n",
      "2      2    0.281 -89.700         0.1310 4.97e+01  4.96e+01 5.35e+01\n",
      "3      3    0.322 -49.900         0.1610 2.54e+01  2.52e+01 2.95e+01\n",
      "4      4    0.329 -31.800         0.1690 1.31e+01  1.29e+01 1.67e+01\n",
      "5      5    0.339 -22.400         0.1720 6.77e+00  6.61e+00 9.71e+00\n",
      "6      6    0.338 -16.500         0.1720 3.54e+00  3.40e+00 5.78e+00\n",
      "7      7    0.337 -12.900         0.1710 1.87e+00  1.76e+00 3.52e+00\n",
      "8      8    0.336 -10.500         0.1690 9.96e-01  9.18e-01 2.19e+00\n",
      "9      9    0.876  -4.020         0.8650 5.37e-01  4.79e-01 1.38e+00\n",
      "10    10    0.895  -3.400         0.8930 2.93e-01  2.51e-01 8.86e-01\n",
      "11    12    0.879  -2.550         0.8780 9.08e-02  6.99e-02 3.80e-01\n",
      "12    14    0.828  -1.960         0.9130 3.01e-02  1.97e-02 1.70e-01\n",
      "13    16    0.775  -1.650         0.9260 1.08e-02  5.66e-03 7.87e-02\n",
      "14    18    0.710  -1.430         0.8610 4.18e-03  1.63e-03 3.76e-02\n",
      "15    20    0.748  -1.280         0.9160 1.76e-03  4.80e-04 2.04e-02\n",
      "16    22    0.166  -1.910        -0.0566 7.97e-04  1.44e-04 1.13e-02\n",
      "17    24    0.803  -1.060         0.9390 3.82e-04  4.36e-05 6.41e-03\n",
      "18    26    0.851  -1.040         0.9380 1.92e-04  1.33e-05 3.67e-03\n",
      "19    28    0.891  -0.989         0.9540 9.99e-05  4.11e-06 2.12e-03\n",
      "20    30    0.911  -0.965         0.9520 5.33e-05  1.30e-06 1.23e-03\n",
      "  Power  SFT.R.sq     slope truncated.R.sq   mean.k. median.k.     max.k.\n",
      "1     1 0.1802568 161.70299      0.0416801 98.049661 98.051601 100.309683\n",
      "2     2 0.2807909 -89.74400      0.1307211 49.687444 49.579496  53.497133\n",
      "3     3 0.3216939 -49.92779      0.1610222 25.371500 25.193203  29.454550\n",
      "4     4 0.3288062 -31.75552      0.1688138 13.054737 12.858189  16.689866\n",
      "5     5 0.3390686 -22.43122      0.1720455  6.770637  6.608841   9.707694\n",
      "6     6 0.3382149 -16.52875      0.1720740  3.541048  3.401192   5.782793\n",
      "[1] \"Estinmating Soft-Power Threshold in hdWGCNA Experiment for OPC\"\n",
      "pickSoftThreshold: will use block size 353.\n",
      " pickSoftThreshold: calculating connectivity for given powers...\n",
      "   ..working on genes 1 through 353 of 353\n",
      "   Power SFT.R.sq  slope truncated.R.sq  mean.k. median.k.   max.k.\n",
      "1      1 6.47e-05 -0.493        0.91700 1.77e+02  1.77e+02 1.84e+02\n",
      "2      2 2.55e-02  4.860        0.93900 9.03e+01  9.03e+01 9.70e+01\n",
      "3      3 1.74e-02  2.190        0.89100 4.65e+01  4.65e+01 5.25e+01\n",
      "4      4 5.09e-02 -2.740        0.70600 2.43e+01  2.41e+01 2.99e+01\n",
      "5      5 4.22e-01 -5.520        0.52200 1.28e+01  1.26e+01 1.78e+01\n",
      "6      6 6.95e-01 -5.190        0.70300 6.81e+00  6.68e+00 1.10e+01\n",
      "7      7 7.83e-01 -5.030        0.73800 3.68e+00  3.54e+00 7.04e+00\n",
      "8      8 8.95e-01 -4.330        0.86500 2.01e+00  1.90e+00 4.62e+00\n",
      "9      9 9.10e-01 -3.890        0.89300 1.11e+00  1.02e+00 3.11e+00\n",
      "10    10 9.57e-01 -3.350        0.96400 6.22e-01  5.51e-01 2.14e+00\n",
      "11    12 9.09e-01 -2.780        0.90600 2.04e-01  1.64e-01 1.07e+00\n",
      "12    14 9.07e-01 -2.320        0.90300 7.12e-02  5.02e-02 5.64e-01\n",
      "13    16 9.24e-01 -1.960        0.92300 2.67e-02  1.57e-02 3.11e-01\n",
      "14    18 2.85e-01 -2.830        0.09030 1.07e-02  5.02e-03 1.77e-01\n",
      "15    20 2.63e-01 -3.190        0.05960 4.66e-03  1.65e-03 1.03e-01\n",
      "16    22 3.04e-01 -2.370        0.18000 2.16e-03  5.46e-04 6.11e-02\n",
      "17    24 2.23e-01 -2.470        0.00401 1.07e-03  1.84e-04 3.68e-02\n",
      "18    26 2.29e-01 -2.310        0.01490 5.54e-04  6.31e-05 2.24e-02\n",
      "19    28 2.77e-01 -2.420        0.12800 2.99e-04  2.19e-05 1.38e-02\n",
      "20    30 2.88e-01 -2.420        0.19900 1.67e-04  7.65e-06 8.54e-03\n",
      "  Power     SFT.R.sq      slope truncated.R.sq    mean.k. median.k.    max.k.\n",
      "1     1 6.471544e-05 -0.4931843      0.9172623 177.261516 177.35618 183.83371\n",
      "2     2 2.547957e-02  4.8647934      0.9386913  90.313694  90.28235  96.98586\n",
      "3     3 1.742094e-02  2.1876698      0.8906334  46.539897  46.48376  52.48436\n",
      "4     4 5.088178e-02 -2.7402848      0.7056966  24.253413  24.09749  29.87632\n",
      "5     5 4.223195e-01 -5.5208513      0.5219432  12.782204  12.63296  17.83665\n",
      "6     6 6.946826e-01 -5.1897920      0.7033315   6.813889   6.67593  11.02635\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "In addition: There were 17 warnings (use warnings() to see them)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/SoftPower/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in celltypes){\n",
    "  \n",
    "  print(paste0('Estinmating Soft-Power Threshold in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # Test different soft powers:\n",
    "  seurat_obj[[cell_type]] <- TestSoftPowers(\n",
    "    seurat_obj[[cell_type]],\n",
    "    networkType = 'signed', # you can also use \"unsigned\" or \"signed hybrid\"\n",
    "    setDatExpr = TRUE,\n",
    "    group.by = 'cell_type',\n",
    "    group_name = cell_type,\n",
    "  )\n",
    "\n",
    "  # plot the results:\n",
    "  plot_list <- PlotSoftPowers(seurat_obj[[cell_type]], \n",
    "                              # wgcna_name = toupper(substr(cell_type, 1, 3)) # the name of the hdWGCNA experiment\n",
    "                              )\n",
    "\n",
    "  # assemble with patchwork\n",
    "  pdf(paste0(fig_dir, cell_type, '_SoftPower.pdf'), width=12, height=8)\n",
    "  print(wrap_plots(plot_list, ncol=2))\n",
    "  dev.off()\n",
    "}\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2ad0b902",
   "metadata": {},
   "source": [
    "For WGCNA and hdWGCNA, the recommended approach is to choose the lowest soft power threshold that achieves a Scale Free Topology Model Fit of 0.8 or higher. If not manually specified, the ConstructNetwork function will automatically select an appropriate soft power threshold during the network construction process."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "7c6251e4",
   "metadata": {},
   "source": [
    "### **Construct co-expression network**\n",
    "\n",
    "We now have everything that we need to construct our co-expression network. Here we use the hdWGCNA function `ConstructNetwork`, which internally utilizes the `WGCNA` function `blockwiseConsensusModules`. While this function offers numerous parameters for advanced customization, default settings are generally effective for many single-cell datasets. Parameters for blockwiseConsensusModules can be directly passed to ConstructNetwork using the same names, allowing for the construction of the co-expression network based on the previously selected soft power threshold."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "14831950",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Constructing co-expression network in hdWGCNA Experiment for EXCITATORY\"\n",
      "Soft power not provided. Automatically using the lowest power that meets 0.8 scale-free topology fit. Using soft_power = 10\n",
      " Calculating consensus modules and module eigengenes block-wise from all genes\n",
      " Calculating topological overlaps block-wise from all genes\n",
      "   Flagging genes and samples with too many missing values...\n",
      "    ..step 1\n",
      "    TOM calculation: adjacency..\n",
      "    ..will not use multithreading.\n",
      "     Fraction of slow calculations: 0.000000\n",
      "    ..connectivity..\n",
      "    ..matrix multiplication (system BLAS)..\n",
      "    ..normalization..\n",
      "    ..done.\n",
      " ..Working on block 1 .\n",
      " ..Working on block 1 .\n",
      " ..merging consensus modules that are too close..\n",
      "[1] \"Constructing co-expression network in hdWGCNA Experiment for INHIBITORY\"\n",
      "Soft power not provided. Automatically using the lowest power that meets 0.8 scale-free topology fit. Using soft_power = 9\n",
      " Calculating consensus modules and module eigengenes block-wise from all genes\n",
      " Calculating topological overlaps block-wise from all genes\n",
      "   Flagging genes and samples with too many missing values...\n",
      "    ..step 1\n",
      "    TOM calculation: adjacency..\n",
      "    ..will not use multithreading.\n",
      "     Fraction of slow calculations: 0.000000\n",
      "    ..connectivity..\n",
      "    ..matrix multiplication (system BLAS)..\n",
      "    ..normalization..\n",
      "    ..done.\n",
      " ..Working on block 1 .\n",
      " ..Working on block 1 .\n",
      " ..merging consensus modules that are too close..\n",
      "[1] \"Constructing co-expression network in hdWGCNA Experiment for ASTROCYTE\"\n",
      "Soft power not provided. Automatically using the lowest power that meets 0.8 scale-free topology fit. Using soft_power = 4\n",
      " Calculating consensus modules and module eigengenes block-wise from all genes\n",
      " Calculating topological overlaps block-wise from all genes\n",
      "   Flagging genes and samples with too many missing values...\n",
      "    ..step 1\n",
      "    TOM calculation: adjacency..\n",
      "    ..will not use multithreading.\n",
      "     Fraction of slow calculations: 0.000000\n",
      "    ..connectivity..\n",
      "    ..matrix multiplication (system BLAS)..\n",
      "    ..normalization..\n",
      "    ..done.\n",
      " ..Working on block 1 .\n",
      " ..Working on block 1 .\n",
      " ..merging consensus modules that are too close..\n",
      "[1] \"Constructing co-expression network in hdWGCNA Experiment for MICROGLIA\"\n",
      "Soft power not provided. Automatically using the lowest power that meets 0.8 scale-free topology fit. Using soft_power = 8\n",
      " Calculating consensus modules and module eigengenes block-wise from all genes\n",
      " Calculating topological overlaps block-wise from all genes\n",
      "   Flagging genes and samples with too many missing values...\n",
      "    ..step 1\n",
      "    TOM calculation: adjacency..\n",
      "    ..will not use multithreading.\n",
      "     Fraction of slow calculations: 0.000000\n",
      "    ..connectivity..\n",
      "    ..matrix multiplication (system BLAS)..\n",
      "    ..normalization..\n",
      "    ..done.\n",
      " ..Working on block 1 .\n",
      " ..Working on block 1 .\n",
      "   ..No significant modules detected in block 1\n",
      "   ..continuing with next block.\n",
      " ..merging consensus modules that are too close..\n",
      " blockwiseConsensusModules: mergeCloseModule failed with this message:\n",
      "       Error in mergeCloseModules(multiExpr, colors[gsg$goodGenes], equalizeQuantiles = equalizeQuantilesForModuleMerging,  : \n",
      "  Error in moduleEigengenes(expr = exprData[[set]]$data, colors = setColors,  : \n",
      "  Color levels are empty. Possible reason: the only color is grey and grey module is excluded from the calculation.\n",
      "\n",
      "  ---> returning unmerged consensus modules\n",
      " multiSetMEs: Calculating module MEs.\n",
      "   Working on set 1 ...\n",
      "[1] \"Constructing co-expression network in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "Soft power not provided. Automatically using the lowest power that meets 0.8 scale-free topology fit. Using soft_power = 9\n",
      " Calculating consensus modules and module eigengenes block-wise from all genes\n",
      " Calculating topological overlaps block-wise from all genes\n",
      "   Flagging genes and samples with too many missing values...\n",
      "    ..step 1\n",
      "    TOM calculation: adjacency..\n",
      "    ..will not use multithreading.\n",
      "     Fraction of slow calculations: 0.000000\n",
      "    ..connectivity..\n",
      "    ..matrix multiplication (system BLAS)..\n",
      "    ..normalization..\n",
      "    ..done.\n",
      " ..Working on block 1 .\n",
      " ..Working on block 1 .\n",
      " No modules detected in block 1 --> continuing with next block.\n",
      " ..merging consensus modules that are too close..\n",
      " blockwiseConsensusModules: mergeCloseModule failed with this message:\n",
      "       Error in mergeCloseModules(multiExpr, colors[gsg$goodGenes], equalizeQuantiles = equalizeQuantilesForModuleMerging,  : \n",
      "  Error in moduleEigengenes(expr = exprData[[set]]$data, colors = setColors,  : \n",
      "  Color levels are empty. Possible reason: the only color is grey and grey module is excluded from the calculation.\n",
      "\n",
      "  ---> returning unmerged consensus modules\n",
      " multiSetMEs: Calculating module MEs.\n",
      "   Working on set 1 ...\n",
      "[1] \"Constructing co-expression network in hdWGCNA Experiment for OPC\"\n",
      "Soft power not provided. Automatically using the lowest power that meets 0.8 scale-free topology fit. Using soft_power = 8\n",
      " Calculating consensus modules and module eigengenes block-wise from all genes\n",
      " Calculating topological overlaps block-wise from all genes\n",
      "   Flagging genes and samples with too many missing values...\n",
      "    ..step 1\n",
      "    TOM calculation: adjacency..\n",
      "    ..will not use multithreading.\n",
      "     Fraction of slow calculations: 0.000000\n",
      "    ..connectivity..\n",
      "    ..matrix multiplication (system BLAS)..\n",
      "    ..normalization..\n",
      "    ..done.\n",
      " ..Working on block 1 .\n",
      " ..Working on block 1 .\n",
      " ..merging consensus modules that are too close..\n",
      "     mergeCloseModules: less than two proper modules.\n",
      "      ..color levels are grey, turquoise\n",
      "      ..there is nothing to merge.\n",
      "[1] \"Excitatory\"      \"Inhibitory\"      \"Astrocyte\"       \"Microglia\"      \n",
      "[5] \"Oligodendrocyte\" \"OPC\"            \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "In addition: Warning messages:\n",
       "1: In ConstructNetwork(seurat_obj[[cell_type]], soft_power = NULL,  :\n",
       "  Overwriting TOM ../results/hdWGCNA/TOM/leng_sfg//EXC_TOM.rda\n",
       "2: In ConstructNetwork(seurat_obj[[cell_type]], soft_power = NULL,  :\n",
       "  Overwriting TOM ../results/hdWGCNA/TOM/leng_sfg//INH_TOM.rda\n",
       "3: In ConstructNetwork(seurat_obj[[cell_type]], soft_power = NULL,  :\n",
       "  Overwriting TOM ../results/hdWGCNA/TOM/leng_sfg//AST_TOM.rda\n",
       "4: In ConstructNetwork(seurat_obj[[cell_type]], soft_power = NULL,  :\n",
       "  Overwriting TOM ../results/hdWGCNA/TOM/leng_sfg//MIC_TOM.rda\n",
       "5: In ConstructNetwork(seurat_obj[[cell_type]], soft_power = NULL,  :\n",
       "  Overwriting TOM ../results/hdWGCNA/TOM/leng_sfg//OLI_TOM.rda\n",
       "6: In ConstructNetwork(seurat_obj[[cell_type]], soft_power = NULL,  :\n",
       "  Overwriting TOM ../results/hdWGCNA/TOM/leng_sfg//OPC_TOM.rda\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "  \n",
    "tom_outdir <-  paste0('../results/hdWGCNA/TOM/', save_prefix, '/')\n",
    "\n",
    "# make output dir for the TOM\n",
    "if(!dir.exists(tom_outdir)){\n",
    "  dir.create(tom_outdir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in celltypes) {\n",
    "\n",
    "  print(paste0('Constructing co-expression network in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "  \n",
    "  tryCatch({\n",
    "    # construct co-expression network:\n",
    "    seurat_obj[[cell_type]] <- ConstructNetwork(\n",
    "      seurat_obj[[cell_type]], \n",
    "      soft_power = NULL, # Set to NULL so that the ConstructNetwork function obtains it automatically (Scale Free Topology Model Fit >= 0.8)\n",
    "      setDatExpr = FALSE,\n",
    "      overwrite_tom = TRUE,\n",
    "      tom_outdir = tom_outdir,\n",
    "      tom_name = toupper(substr(cell_type, 1, 3)) # name of the topological overlap matrix written to disk set as the name of the hdWGCNA experiment \n",
    "    )\n",
    "  }, error = function(e) {\n",
    "    print(paste0('Error encountered while processing ', toupper(cell_type)))\n",
    "    print(paste0(toupper(cell_type), ' dropped from experiment'))\n",
    "    # Code to handle the error condition\n",
    "    seurat_obj <- seurat_obj[names(seurat_obj) != cell_type]\n",
    "    assign(\"seurat_obj\", seurat_obj, envir = .GlobalEnv) # Update the modified seurat_obj in the global environment\n",
    "\n",
    "  },  \n",
    "  message = function(m) {\n",
    "    print(paste0('Error encountered while processing ', toupper(cell_type)))\n",
    "    print(paste0(toupper(cell_type), ' dropped from experiment'))\n",
    "    # Code to handle the error condition\n",
    "    seurat_obj <- seurat_obj[names(seurat_obj) != cell_type]\n",
    "    assign(\"seurat_obj\", seurat_obj, envir = .GlobalEnv) # Update the modified seurat_obj in the global environment\n",
    "\n",
    "  }) \n",
    "  \n",
    "}\n",
    "\n",
    "\n",
    "print(names(seurat_obj))\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "49676183",
   "metadata": {},
   "source": [
    "hdWGCNA also includes a function PlotDendrogram function for visualizing the WGCNA dendrogram, a standard method to display the various co-expression modules derived from network analysis. In this dendrogram, each leaf represents a gene, with colors indicating their module assignments. `Notably, genes in the \"grey\" module, which were not grouped into any specific co-expression module, should be excluded from all subsequent analyses and interpretations.`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "8da7dcca",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/dendrogram/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "tryCatch({\n",
    "    for (cell_type in names(seurat_obj)){\n",
    "        # print(paste0('Constructing co-expression networkd in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "        pdf(paste0(fig_dir, cell_type, \"_dendro.pdf\"), height=2, width=4)\n",
    "        PlotDendrogram(seurat_obj[[cell_type]], main='hdWGCNA Dendrogram')\n",
    "        dev.off()\n",
    "    }\n",
    "}, error = function(e){\n",
    "    NULL\n",
    "})"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "5e1c23ce",
   "metadata": {},
   "source": [
    "### **Module Eigengenes and Connectivity**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "08779263",
   "metadata": {},
   "source": [
    "### Compute harmonized module eigengenes\n",
    "\n",
    "`Module Eigengenes (MEs)` summarize the gene expression profile of co-expression modules in hdWGCNA, calculated by performing principal component analysis (PCA) on each module's subset of the gene expression matrix, with the first principal component as the ME. Given that single-cell data is sensitive to technical artifacts, hdWGCNA's ModuleEigengenes function not only computes MEs but also offers Harmony batch correction to produce `harmonized module eigengenes (hMEs)`. This harmonization is executed with a focus on the sample of origin, as specified by the `group.by.vars` parameter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "df854ccb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Estimating module eigen-genes in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"grey\"\n",
      "[1] \"blue\"\n",
      "[1] \"turquoise\"\n",
      "[1] \"----------------\"\n",
      "[1] \"Estimating module eigen-genes in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"green\"\n",
      "[1] \"turquoise\"\n",
      "[1] \"grey\"\n",
      "[1] \"blue\"\n",
      "[1] \"yellow\"\n",
      "[1] \"brown\"\n",
      "[1] \"----------------\"\n",
      "[1] \"Estimating module eigen-genes in hdWGCNA Experiment for ASTROCYTE\"\n",
      "[1] \"turquoise\"\n",
      "[1] \"grey\"\n",
      "[1] \"blue\"\n",
      "[1] \"----------------\"\n",
      "[1] \"Error encountered while processing MICROGLIA\"\n",
      "[1] \"MICROGLIA dropped from experiment\"\n",
      "[1] \"Error encountered while processing OLIGODENDROCYTE\"\n",
      "[1] \"OLIGODENDROCYTE dropped from experiment\"\n",
      "[1] \"Error encountered while processing OPC\"\n",
      "[1] \"OPC dropped from experiment\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |===================================                                   |  50%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |===================================                                   |  50%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Warning: Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcagrey to pcagrey_\n",
       "Warning: All keys should be one or more alphanumeric characters followed by an underscore '_', setting key to pcagrey_\n",
       "pcagrey_ 1 \n",
       "Positive:  APOE, GFAP, MT2A, BDNF, RAPGEF3, ATP1B2, SLC1A3, ARC, ARRDC2, SOX8 \n",
       "\t   FGF17, CRYAB, PLXND1, SLC6A1, DUSP7, SLC7A11, B2M, DUSP1, ATP1A2, ASIC1 \n",
       "\t   GRM4, FSTL3, NTNG2, ZNF358, CHRM2, ERBB4, DMKN, PLPP3, TFEB, TPD52L1 \n",
       "Negative:  SYT1, CLASP2, MAP2, RBFOX2, GPM6B, MACF1, PI4KA, CAMTA1, SCN8A, SCN2A \n",
       "\t   PAFAH1B1, CCDC88A, DNM3, SH3GL2, ARHGEF7, SLC24A2, SPTBN1, PAK1, RUFY3, NTRK2 \n",
       "\t   CADPS2, ATP8A1, RAB6A, NFASC, DLG1, SORL1, INPP4A, CAMSAP2, TAOK1, MBNL1 \n",
       "pcagrey_ 2 \n",
       "Positive:  TF, HHIP, HOOK2, CPQ, BCO2, NPC1, NEK1, HSPB11, SCN9A, ABHD3 \n",
       "\t   GRIP1, MYO10, SLC16A9, ZNF658, SGK1, DLC1, CCDC66, SCRG1, MYO6, HIP1 \n",
       "\t   SLC25A13, LIMA1, WWTR1, PDE8A, SLC9A9, FGF1, MACF1, BANP, ITCH, DLG1 \n",
       "Negative:  DUSP6, KCNV1, ACTN4, FSCN1, LZTS1, APC2, KCNS1, NAB2, CAMKK2, MAP1S \n",
       "\t   SLC8A2, ATP9A, TESK1, ABCA2, SLC30A3, RAB1B, ADGRB1, ENSA, MYADM, USP7 \n",
       "\t   KCNJ9, TYRO3, CSNK1D, ARC, VPS18, OCRL, SYT7, GPD1L, RTN4RL2, NPTX1 \n",
       "pcagrey_ 3 \n",
       "Positive:  CAMK2D, CPNE4, PTPRT, IL1RAPL2, PID1, FNBP1L, WLS, FARP1, OPRM1, TENM1 \n",
       "\t   KCNAB1, CHRM2, HRK, TRABD2A, CXXC4, TOX, PCDH15, PKD1, SLC8A2, CERS4 \n",
       "\t   GRM8, SYT17, FGFR1, BCL2, CPNE5, NPY1R, KCNC2, KCNH3, GDA, CADPS2 \n",
       "Negative:  FTH1, SYT2, STXBP6, HSPA8, KCNA1, GPAT3, KCNC1, SLC47A1, SLC17A6, GGACT \n",
       "\t   FGF9, BCO2, NAPG, PIP4K2C, PNKD, SAR1A, ST3GAL5, PIP4K2A, BEND6, SUMO1 \n",
       "\t   ENDOD1, PLEKHA2, SULF2, BCAN, HEY1, HHIP, SLC7A1, B2M, CNTN6, KCNS1 \n",
       "pcagrey_ 4 \n",
       "Positive:  PTPRT, FNBP1L, CAMK2D, IL1RAPL2, TOX, WLS, CXXC4, FGFR1, VWC2L, NRSN1 \n",
       "\t   CAMK2G, PID1, TRABD2A, SCN9A, CHRM2, GRIP1, SYT2, FTH1, SULF2, VWC2 \n",
       "\t   HRK, ATP1B2, CPNE4, ST6GALNAC3, PLBD2, CDS1, GRM4, RAB3C, ST8SIA1, AHNAK2 \n",
       "Negative:  KCNIP2, TYRO3, P2RX7, SLC24A4, ARNT2, ZNF365, SLC16A10, ATP6V1H, ROCK2, STXBP6 \n",
       "\t   PITPNM2, LZTS1, CPNE5, TPD52L1, CHKA, DENND3, SLC44A1, ZEB2, CACNA1G, DGKA \n",
       "\t   PTPN1, DPF1, PLEKHA2, ATP11A, SESN1, MTMR2, PCNT, PDE2A, PI4KA, AGPAT4 \n",
       "pcagrey_ 5 \n",
       "Positive:  SULF2, ABCA7, ABCA9, PKD1, CNTN6, NRP1, BEND6, TRPM2, CNKSR3, ADCY8 \n",
       "\t   ACTN4, AGRN, ABCA10, ECM1, SYT2, ACADVL, GPAT3, DYSF, LPCAT4, TTBK2 \n",
       "\t   FGF9, GBA, TBC1D30, RASGRP2, CCDC88A, SLC25A29, PITPNC1, CHST8, SCN9A, NF2 \n",
       "Negative:  NPTX1, RPS3, APOD, GGACT, TXN, MBP, P2RX7, ATP2B4, NPM1, DNAJA1 \n",
       "\t   HSPA8, SUMO1, HSP90B1, ZNF365, MIEN1, CACNA1G, CRYAB, SGK1, NPY1R, FTH1 \n",
       "\t   TRABD2A, ENPP2, MAP2K1, HSPA5, TF, CPNE5, MOBP, DUSP26, PARK7, SLC8A2 \n",
       "Harmony 1/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 2/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 3/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 4/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 5/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 6/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 7/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 8/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony converged after 8 iterations\n",
       "Warning: Invalid name supplied, making object name syntactically valid. New object name is Seurat..ProjectDim.originalexp.harmony; see ?make.names for more details on syntax validity\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Warning: Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcablue to pcablue_\n",
       "Warning: All keys should be one or more alphanumeric characters followed by an underscore '_', setting key to pcablue_\n",
       "pcablue_ 1 \n",
       "Positive:  PLXNB1, MARCKSL1, NEFH, SCNN1D, ARSA, HRH3, BRPF3, LAMTOR2, WFS1, PITPNM1 \n",
       "\t   JUN, CPNE7, WDR1, TRIB2, PFN1, GSTP1, SMPD1, LRRC8D, PSMD10, BHLHB9 \n",
       "\t   CCDC115, NECAB3, DVL1, SCN4B, ARL4D, GGA1, BLOC1S2, AKT1, ARL4C, HRAS \n",
       "Negative:  CHN1, SNAP25, CAMK2A, STXBP1, DNM1, RTN4, CALM1, PREPL, PRNP, PSAP \n",
       "\t   APP, PRKAR1B, UCHL1, DPYSL2, NRGN, NDFIP1, KIF1A, SYP, GPM6A, CALM3 \n",
       "\t   SERINC1, DKK3, PPIA, PEBP1, NPTN, HSP90AB1, ATP2A2, CFL1, CAMK2B, ATP6V1B2 \n",
       "pcablue_ 2 \n",
       "Positive:  TMSB4X, CHCHD10, COX7B, SOD1, DYNLL1, GAP43, ATP6V1F, LAMTOR5, EEF1A1, HSP90AA1 \n",
       "\t   GPX4, PINK1, YWHAH, PIN1, BRK1, FEZ1, ATP6V1G1, REEP5, MT3, BLOC1S2 \n",
       "\t   PRELID1, LAMTOR1, STMN2, NSF, DNAJB11, SCN4B, TMED2, ATP1B1, RTN4, LAMTOR2 \n",
       "Negative:  SLC17A7, NSMF, MTPN, HPCA, SYN1, NRXN2, BRSK2, PPP2R1A, ARFGAP1, AGAP2 \n",
       "\t   DLG4, CDK5R1, ADGRB2, BAIAP3, TPPP, EEF1A2, PLPPR2, PRNP, ARPP19, CAMK2B \n",
       "\t   SEMA6B, GPC1, PPFIA3, PTK2B, LY6H, SYNPO, CLIP3, SLC6A17, KCNF1, SCN3B \n",
       "pcablue_ 3 \n",
       "Positive:  SCN1B, KCNC3, HAPLN4, SCN4B, NEFH, LIMK1, EEF1A2, DIRAS1, CPLX2, CPLX1 \n",
       "\t   PPP2R1A, GAPDH, NECAB3, KCNA2, NCS1, STX1B, NRGN, DYNLL2, HCN2, PCYT2 \n",
       "\t   EGFL7, RAB11B, CAPNS1, PEF1, SYNPO, MLST8, PI4K2A, ACTB, MAPK8IP2, CLU \n",
       "Negative:  GPM6A, SNCA, SCN3B, SCD5, STX1A, APP, SERPINI1, BAIAP3, SLC6A7, LY6H \n",
       "\t   CPNE7, CRMP1, PIK3R1, SCG3, HPCA, ATP6V0A1, NDFIP1, HPCAL4, DNAJB11, GABBR1 \n",
       "\t   MARCKS, DGKZ, ADD1, SYP, PLK2, SYT5, GAP43, DNER, GNAI2, DPYSL2 \n",
       "pcablue_ 4 \n",
       "Positive:  SLC30A10, GPC1, BTBD2, PPFIA3, LIMK1, SRCIN1, SCNN1D, GAP43, SLC6A7, BIN1 \n",
       "\t   DYNC1H1, NECAB3, BRSK1, ATP6V0A1, SNX27, ABHD12, KCNQ2, KCNC3, CAPZB, ARFGAP1 \n",
       "\t   MLST8, DGKZ, HMGCR, DNER, EXTL2, SCN4B, TECR, BHLHB9, GGA1, SLC27A4 \n",
       "Negative:  CAMK2N1, MARCKS, NCDN, HPCAL4, APLP1, HPCA, SERINC1, YWHAH, PRNP, ITM2C \n",
       "\t   ATP6AP1, ACTB, CD81, SYT13, DKK3, SCG2, WDR1, EEF1A1, YWHAB, UBQLN2 \n",
       "\t   TMEM30A, CALM1, RAB3A, JUN, DIRAS2, SCN3B, ITM2B, ALDOA, PSAP, BAIAP3 \n",
       "pcablue_ 5 \n",
       "Positive:  HRH3, CPNE7, HPCA, TRIB2, KCNF1, MARCKSL1, STX1A, MARCKS, FAM3C, ARL4C \n",
       "\t   CHGA, CADM4, SLC6A17, SCOC, DNER, CSPG5, MAZ, SYNGR3, STMN2, ADORA1 \n",
       "\t   SCG2, BNIP3, TPPP, HSP90AB1, GPX4, NDRG4, SLC17A7, INPP5F, PRPS1, DIRAS2 \n",
       "Negative:  CAMK2N1, JUN, PTK2B, PRKCG, NCDN, SYNPO, CAMK2A, CLU, ST8SIA3, B4GALNT1 \n",
       "\t   ADGRB2, APLP1, EEF1A1, RTN4, TSPYL2, KCNC3, DOC2A, PLPPR2, NPTN, DKK3 \n",
       "\t   STARD7, CAMK2B, CPLX2, ARPP19, HAPLN4, YWHAH, GPC1, CALR, MBOAT7, PDK2 \n",
       "Harmony 1/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 2/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 3/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 4/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 5/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 6/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 7/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 8/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 9/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 10/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony converged after 10 iterations\n",
       "Warning: Invalid name supplied, making object name syntactically valid. New object name is Seurat..ProjectDim.originalexp.harmony; see ?make.names for more details on syntax validity\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Warning: Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcaturquoise to pcaturquoise_\n",
       "Warning: All keys should be one or more alphanumeric characters followed by an underscore '_', setting key to pcaturquoise_\n",
       "pcaturquoise_ 1 \n",
       "Positive:  IGFBP5, CNGB1, SLC22A24, RAB11FIP1, BMPR1B, CHAC2, SIRT1, SH2B3, RIN3, ARL4A \n",
       "\t   KCNMB3, SLC6A11, PLA2G4A, KCNG3, TRPC4, ACVR1, ALDH1A2, AVIL, PTCH1, CYP3A5 \n",
       "\t   PDIA5, DDX58, DCDC2, SLC3A1, APAF1, TRPC6, CHRNA7, CAV3, CFAP43, TPTE2 \n",
       "Negative:  RBFOX1, DLG2, KALRN, NRXN1, ANK2, DSCAM, MAGI2, KCNMA1, FGF12, STXBP5L \n",
       "\t   PTPRD, ANK3, KCNQ5, RIMS2, LRRC7, RYR2, AHI1, LDB2, NRCAM, ADGRL3 \n",
       "\t   CADPS, SLC4A10, CACNA1C, ARHGAP26, LSAMP, FMN2, IL1RAPL1, MEF2C, PTK2, HECW1 \n",
       "pcaturquoise_ 2 \n",
       "Positive:  GNAL, TIMP3, KIF13A, TRPC6, KCND2, DOCK2, PTPRO, CUX2, SYNJ2, BMPR1B \n",
       "\t   ACVR1C, PRKCA, PRELID2, SIPA1L1, PPARG, CCDC65, ABLIM1, PRKD1, PRKD3, KCNN2 \n",
       "\t   ROBO1, ARL4A, PIP5K1B, DGKB, SLC35F3, CHAC2, PLXNA4, STEAP2, SLC12A8, SHISA6 \n",
       "Negative:  ROBO2, KCNB2, SLC44A5, PTPRD, EPHA3, PCLO, RAB27B, RIMS1, TRPM3, TRHDE \n",
       "\t   KCNH7, GRM7, NLGN1, ACVR2A, SLC8A3, KCNT2, LRRC4C, TEAD1, ATXN1, NTNG1 \n",
       "\t   ALDH1A2, LRP1B, KCNQ3, HECW1, SHISA9, CACNA1A, NRCAM, MARK1, MYT1L, SRRM4 \n",
       "pcaturquoise_ 3 \n",
       "Positive:  PDXDC1, MAK, FN1, TIMP3, PRKD3, SNTG2, CCDC65, CYP3A5, SLC12A1, CHAC2 \n",
       "\t   CAV3, LRP1B, RAB11A, KCNQ1, NRG1, DGKG, DOCK2, NXN, CTNNA2, SLC3A1 \n",
       "\t   INPP4B, RYR2, TNIK, SDK2, AVIL, TGFB1, STX8, SIL1, VWF, NAIP \n",
       "Negative:  RFX3, CUX2, NTNG1, NTRK3, CNGB1, DGKB, TRHDE, KIRREL3, GRM7, EPHA4 \n",
       "\t   KCND2, CDH13, FBXW7, HOMER1, PPFIA2, PRKCB, GRB14, PRKCA, KCNN2, LMO3 \n",
       "\t   GNAL, PTPRO, SMAD3, SIPA1L1, PTPN13, IL1RAPL1, STXBP5L, KCNH5, RAPGEF5, MEF2A \n",
       "pcaturquoise_ 4 \n",
       "Positive:  MCTP1, CTNNA2, LRP1B, NELL1, INPP4B, SYNJ2, PCDH7, NAV3, NETO1, SNTG2 \n",
       "\t   DGKG, PTPRR, HTR1E, RFTN1, PRKD1, SEMA3A, KCND3, RAB11FIP3, GNPTAB, CDH2 \n",
       "\t   SLC22A10, CAMK1D, SLC35F3, TBC1D1, PRKAG2, TANC2, ATP8B4, CADPS, LMO3, LSAMP \n",
       "Negative:  NTNG1, CUX2, GRM7, LRRC4C, TRHDE, SLIT2, KCNH5, GRB14, SLC25A37, PIP5K1B \n",
       "\t   KCNH7, PRKD3, RFX3, KCNQ1, SCN1A, MAGI2, CYP3A5, MAK, SGPP2, OSBPL5 \n",
       "\t   CFAP221, ABCA6, CUX1, TTN, EVI5, PDXDC1, CNGB1, NLGN1, CHAC2, TGFB1 \n",
       "pcaturquoise_ 5 \n",
       "Positive:  SHISA9, SLC22A10, PTPN13, SLC22A24, CNGB1, TRPC5, SLF1, SNTG2, PPFIA2, ARL4A \n",
       "\t   CDH2, RAB11A, FN1, TRHDE, KIF13A, EXOC6, PRELID2, PDXDC1, CDH13, CCDC65 \n",
       "\t   FUT9, ACTN1, CUX1, TIMP3, MYH9, PTPRR, KCNB2, DGKG, AVIL, MAK \n",
       "Negative:  NRG1, SLC35F3, SCN1A, ASIC2, ABCA6, DCX, PLXNA4, SLC4A8, ADAM23, HOMER1 \n",
       "\t   RIT2, LRP1B, CNTN4, ARHGAP26, PRKD1, DCLK1, PCDH7, ESYT2, ATP8A2, CERS6 \n",
       "\t   SGPP2, STXBP5L, RYR2, HCN1, PIP5K1B, LSAMP, WWP1, ADAM22, EVI5, NETO1 \n",
       "Harmony 1/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 2/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 3/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 4/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 5/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 6/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 7/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 8/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 9/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony converged after 9 iterations\n",
       "Warning: Invalid name supplied, making object name syntactically valid. New object name is Seurat..ProjectDim.originalexp.harmony; see ?make.names for more details on syntax validity\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |===================================                                   |  50%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Warning: Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcagreen to pcagreen_\n",
       "Warning: All keys should be one or more alphanumeric characters followed by an underscore '_', setting key to pcagreen_\n",
       "pcagreen_ 1 \n",
       "Positive:  CLASP2, RTN4, PRKACB, SPTBN1, PTK2, UCHL1, ARHGEF7, ATP6V0A1, KIF3A, STMN2 \n",
       "\t   OCIAD1, CHN1, ATP1A1, CLTC, PGAM1, KIF5A, YWHAH, FLOT1, SYNJ1, SORL1 \n",
       "\t   TTC39B, USP47, THY1, VDAC1, DYNLL1, RFX3, ENO1, MFSD6, SLC44A1, LNPEP \n",
       "Negative:  GFAP, HSPA1A, NEFH, CTDSP2, AGRN, PRKCG, NRP1, HIP1R, RTTN, CADPS2 \n",
       "\t   CRYAB, RRAGD, KANK1, GSN, ABCA7, SYNJ2, TULP3, CAMK1, STX1A, PHKG2 \n",
       "\t   ARHGAP1, CLPX, PICK1, LRPAP1, NAGPA, NCBP1, MDH2, SLIT3, TTC8, SLC44A2 \n",
       "pcagreen_ 2 \n",
       "Positive:  PGAM1, VDAC1, UCHL1, THY1, ENO1, CHN1, ATP6V1E1, KIF5A, DDX3X, PRMT1 \n",
       "\t   ATP1A1, ACTR1B, MDH2, YWHAH, HMGCR, MFSD6, TMEM25, STMN2, PDE2A, ATP6V0A1 \n",
       "\t   KCNAB2, DEF8, STX1A, RAB14, PPT1, KIF3A, SYNJ1, LAMP1, PGK1, AGRN \n",
       "Negative:  PIP4K2A, SLC44A1, GSN, SYNJ2, KANK1, CPQ, SLC39A11, PTK2, CRYAB, CLASP2 \n",
       "\t   ARHGAP1, HIP1R, RFX3, RTTN, EXOC6, TULP3, CEP89, NRP1, CAPRIN2, JAK2 \n",
       "\t   CCP110, TANGO2, DNM2, ARHGEF7, USP47, CDK5RAP3, TRAF3, FTH1, RPS3, MAP3K7 \n",
       "pcagreen_ 3 \n",
       "Positive:  GSN, CRYAB, ABCA2, SYNJ2, PIP4K2A, SLC44A1, SLC38A2, EEF1A1, HSPA1A, SNCA \n",
       "\t   HIP1R, CCP110, PRKCG, ATP6V1E1, AGRN, LAMP1, MDH2, LRPAP1, MT3, TMEM59 \n",
       "\t   KANK1, PGAM1, YWHAH, LAMTOR4, ENO1, RAB14, VDAC1, THY1, DEF8, PDE2A \n",
       "Negative:  CAPRIN2, JAK2, CEP89, TRAF3, TTC8, ATF6, SLIT3, CPQ, RFX3, TANGO2 \n",
       "\t   ARHGEF7, TTC39B, USP47, PTK2, CLASP2, EXOC6, APPL1, PCNT, RAB26, TRPM3 \n",
       "\t   ICE1, NAGPA, DNM2, ABCA7, FLOT1, NCBP1, TULP3, MAP3K7, CLPX, UBE2G2 \n",
       "pcagreen_ 4 \n",
       "Positive:  PRKCG, SNCA, STX1A, EXOC6, PDE2A, CADPS2, NRP1, CHN1, RFX3, AGRN \n",
       "\t   HMGCR, SLIT3, CTDSP2, CAMK1, TRPM3, SSX2IP, TTC8, SLC44A2, RTTN, PDIA6 \n",
       "\t   DNM2, VPS28, PTK2, ICE1, ABCA7, HSPA1A, PICK1, SYNJ2, ARHGEF7, UCHL1 \n",
       "Negative:  CAPRIN2, PIP4K2A, PGK1, TTC39B, PGAM1, MDH2, FLOT1, PRMT1, CRYAB, KCNAB2 \n",
       "\t   DDX3X, ENO1, RHOQ, VDAC1, LNPEP, STMN2, ATP1A1, KIF3A, SLC38A2, NEFH \n",
       "\t   TMEM33, KIF5A, PIN1, RTN4, ATP6V1E1, UBE2G2, PPT1, SLC44A1, SLC39A11, CPQ \n",
       "pcagreen_ 5 \n",
       "Positive:  ABCA2, ABCA7, PIP4K2A, AGRN, SLC44A1, SORL1, GSN, MFSD6, CHN1, DEF8 \n",
       "\t   PCNT, RAB26, STX1A, SYNJ2, TTC39B, TTC8, TMEM25, RHOQ, DNM2, KCNAB2 \n",
       "\t   PTK2, RTTN, SSX2IP, ATF6, ARHGAP1, CCP110, KIF5A, MAP3K7, ACADVL, LNPEP \n",
       "Negative:  EEF1A1, RPS3, DYNLL1, NCBP1, SNCA, YWHAH, LAMTOR4, CPQ, RRAGD, SNX3 \n",
       "\t   CDK5RAP3, NAGPA, PDIA6, CRYAB, FTH1, APPL1, TMEM59, LRPAP1, CEP89, VPS28 \n",
       "\t   TMEM33, NRP1, RAB14, UBE2G2, USP47, TANGO2, TULP3, KANK1, MT3, GFAP \n",
       "Harmony 1/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 2/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 3/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 4/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 5/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 6/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 7/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 8/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 9/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony converged after 9 iterations\n",
       "Warning: Invalid name supplied, making object name syntactically valid. New object name is Seurat..ProjectDim.originalexp.harmony; see ?make.names for more details on syntax validity\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Warning: Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcaturquoise to pcaturquoise_\n",
       "Warning: All keys should be one or more alphanumeric characters followed by an underscore '_', setting key to pcaturquoise_\n",
       "pcaturquoise_ 1 \n",
       "Positive:  MNS1, PLXNB2, NR1D1, ARSA, SLC6A7, LTBP4, SCNN1D, ADAM15, BRAT1, YJEFN3 \n",
       "\t   CXXC1, IFT140, ARL4D, SLC8A2, CBY1, SRC, CDKN1B, TBC1D17, PPIB, NLGN3 \n",
       "\t   GNAI2, KCNC4, ADNP2, KCNH2, DOC2A, SHISA7, AP1S1, ARFRP1, ATAT1, SYNPO \n",
       "Negative:  GNAS, PREPL, MAP2, SNAP25, CAMK2B, STXBP1, STXBP5L, HSP90AB1, MAPT, ACTB \n",
       "\t   CALM1, SNCB, KLC1, MYCBP2, SLC38A1, SCAMP5, NUCKS1, ATP6V1B2, NPTN, KIF1A \n",
       "\t   GAPDH, PSAP, DKK3, SYN1, DYNC1H1, INPP5F, SLC6A17, CANX, CCDC88A, NDRG4 \n",
       "pcaturquoise_ 2 \n",
       "Positive:  STXBP5L, CCDC88A, MYCBP2, TAOK1, HTT, ABCA5, MAPT, MAP2, NCOR2, CAMSAP2 \n",
       "\t   RAB11FIP3, ATP2C1, CCND3, SPOP, ADD1, CBY1, GOLPH3, ATP13A2, SNRNP70, RAMP1 \n",
       "\t   GRIPAP1, DYNC1H1, IFT140, VAPA, DCBLD2, GCC2, ABCB1, SUPV3L1, SYNRG, SGSM2 \n",
       "Negative:  PPP2R1A, SYN1, EEF1A2, DLG4, MTPN, KLC1, PPP1R9B, KCNJ4, SNCB, CDK5R1 \n",
       "\t   ARFGAP1, ACTB, CAMK2B, TAOK2, MAPK8IP2, CHRM1, CDK5R2, SLC7A5, SLC17A7, NSMF \n",
       "\t   HSP90AB1, ADGRB2, RAB11B, ATP1B2, MAZ, FZR1, SLC6A17, ATP6V1B2, NCDN, CFL1 \n",
       "pcaturquoise_ 3 \n",
       "Positive:  ATP1B2, CPLX2, SLC8A2, PIK3R1, KCNC4, PTK2B, ARL4C, CAMK2A, MNS1, ST8SIA3 \n",
       "\t   GNAI2, RIMS3, FASN, DOC2A, PGM2L1, DKK3, KCNB1, GNAO1, TPPP, TSPYL2 \n",
       "\t   SRC, CD81, MBOAT7, SLC17A7, NR1D1, SYT4, ATP6AP1, SLC25A29, EGFL7, SLC6A7 \n",
       "Negative:  NECAB3, LIMK1, SCN1B, MAP2K2, SNCB, DACT3, SLC38A1, KCNJ4, KCNC3, DAB2IP \n",
       "\t   GAS6, NDRG4, KCNIP2, NCS1, HCN2, RIMS4, GOLGA2, STX1B, ZBTB7A, NAPA \n",
       "\t   CHGA, DYNLL2, ADGRB1, SDCBP, SPTBN2, PACSIN1, GAPDH, TPI1, RAB11B, BNIP3 \n",
       "pcaturquoise_ 4 \n",
       "Positive:  SLC17A7, PTK2B, PKD1, ADGRB2, ARFGAP1, BRSK2, PPFIA3, SYNPO, GPC1, SLC6A7 \n",
       "\t   CAMK2A, ADGRB1, SLC25A29, SCNN1D, NLGN3, OGFR, STXBP5L, SYT7, SNRNP70, RAB11FIP3 \n",
       "\t   AGAP2, SRCIN1, SLC30A10, CAMK2B, NSMF, KAT8, ARSA, GAK, TBC1D10B, GGA1 \n",
       "Negative:  ATP6AP1, ALDOA, ARL4C, RAB3A, RNF41, DDB1, ATP1B2, CANX, FKBP4, CD81 \n",
       "\t   EIF4G2, ATP6V1G1, CTSF, NAPB, EI24, OAZ2, YWHAB, ARL8B, CALM3, CALM1 \n",
       "\t   TRIM8, CPLX2, KLHDC3, STXBP1, NUDCD3, VAPB, IRS2, GNAS, DYNLL2, RAMP1 \n",
       "pcaturquoise_ 5 \n",
       "Positive:  PTK2B, SLC17A7, CAMK2A, SYT4, SYT7, CDKN1B, SLC6A7, RIMS3, CCAR2, KCNJ4 \n",
       "\t   AZIN1, RAB3A, CDK5R2, PIK3R1, SYNPO, TPI1, NCDN, FSCN1, ST6GALNAC6, SYT5 \n",
       "\t   CBY1, STXBP5L, SYNRG, YWHAB, VAPB, TMEM9, GCC2, ST8SIA3, ATP6V1B2, STXBP1 \n",
       "Negative:  ATP1B2, SLC29A4, OLFM2, ARL4C, TBC1D10B, SLC7A5, RAMP1, QSOX1, GOLPH3, GPC1 \n",
       "\t   ARL4D, NR1D1, HGS, TESK1, SLC6A8, SLC30A10, CPLX2, NCOR2, DNAJB2, SOCS5 \n",
       "\t   CXXC1, SYT12, ABCB1, KCNQ2, YJEFN3, TBC1D17, ATP6V0D1, SGSM3, SLC25A29, STK11 \n",
       "Harmony 1/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 2/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 3/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 4/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 5/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony converged after 5 iterations\n",
       "Warning: Invalid name supplied, making object name syntactically valid. New object name is Seurat..ProjectDim.originalexp.harmony; see ?make.names for more details on syntax validity\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Warning: Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcagrey to pcagrey_\n",
       "Warning: All keys should be one or more alphanumeric characters followed by an underscore '_', setting key to pcagrey_\n",
       "pcagrey_ 1 \n",
       "Positive:  SLC1A3, FAM107A, SYT2, SPRY2, TRIM14, TRHDE, RAB31, FAM161A, SLC16A2, GNB3 \n",
       "\t   SEMA6A, RASGRP2, WASF2, PTK7, CXXC4, LRRK2, GLTP, PTPN18, GAA, RABL3 \n",
       "\t   NPRL2, RAB27A, WDR35, OPRM1, SERTAD2, FIS1, TRIM27, AXIN1, PIK3CD, SNAP29 \n",
       "Negative:  IL1RAPL1, EIF4G3, APP, MACF1, RIMS2, TANC2, KCNMA1, TBC1D5, CACNA1C, FMN2 \n",
       "\t   HECW2, FAF1, AKAP9, GDAP1, PRKCA, HECTD4, MYO5A, PAFAH1B1, RAPGEF2, RUFY3 \n",
       "\t   SCN2A, GNAQ, GPM6A, OSBPL1A, SGK1, CLASP1, FER, MAP2K5, APBB2, RAB3C \n",
       "pcagrey_ 2 \n",
       "Positive:  TMEM30A, HYOU1, SYT11, ATP2A2, KLC2, SYT13, GABBR1, CX3CL1, ARL6IP5, ACLY \n",
       "\t   SURF4, GOT2, MFN2, GOT1, ATP6V1F, TMED9, VPS4A, SQSTM1, SLC7A14, CDC42 \n",
       "\t   RRAGA, SNX27, SLC1A4, ITM2C, TBC1D9, GPM6A, PINK1, SLC32A1, GPS2, B4GALNT1 \n",
       "Negative:  TF, PAPPA2, IL1RAPL1, SGK1, NXN, TBC1D12, WDR19, HECW2, NEK1, TSHR \n",
       "\t   PLD1, SUMF1, FMN2, ARL17A, TLK1, TBC1D5, ROCK1, TRPM7, RAB3IP, P2RX7 \n",
       "\t   TJP1, SLF1, STX8, ZFYVE16, TMEM165, LEPR, PRKD1, PICALM, APBB2, EIF4G3 \n",
       "pcagrey_ 3 \n",
       "Positive:  ASIC2, KCNMA1, SLC35F3, KCNK1, PRKAG2, CACNA1C, GSAP, SNCG, SLC25A37, MDH1 \n",
       "\t   VMP1, SLC25A27, SLC9A1, TLE1, PPARA, EEA1, FBXW7, EXTL2, NNT, TBC1D8 \n",
       "\t   PIK3CB, RIT2, TBC1D5, SLC7A14, SLC25A13, GDAP1, SYT2, LMNA, CCNI, MYO6 \n",
       "Negative:  SLC8A1, ALK, TOX, SPOCK1, CTNNA2, ITM2C, RAB3C, FUT9, SEMA6A, NETO1 \n",
       "\t   EEPD1, RASGRP1, DGKD, EPHB2, KIAA1109, EPN2, TLK1, OPHN1, HECW2, FAM107A \n",
       "\t   GPM6A, SPOPL, GNAI1, DYNC2H1, SEC14L1, TRHDE, RAB3IP, IL1RAPL1, ADAM10, GAP43 \n",
       "pcagrey_ 4 \n",
       "Positive:  TRHDE, GPM6A, GAP43, NETO1, DAPK1, RIMS2, RAB31, SLC8A1, GPI, NALCN \n",
       "\t   SCN2A, ASIC2, RAB27B, OPRM1, CACNA1C, SEMA6A, GRB10, CTNNA2, RIT2, LRRK2 \n",
       "\t   ATP2B4, SLC7A14, SLF1, KCNMA1, SLC13A3, NLGN4X, RPS6KA3, MAP6, SNX32, PIK3CD \n",
       "Negative:  SGK1, DGKD, TOX, IL1RAPL1, ALK, ULK3, EEPD1, RASGRP1, SLC7A2, PPARD \n",
       "\t   SYNDIG1, EXTL2, SEC14L1, PAPPA2, SLC25A37, GOT1, ANO6, ERP29, ST3GAL5, TLK1 \n",
       "\t   SPOCK1, SLC9A1, ADCY9, SPOPL, ITGB1, SLC25A13, TSHR, PLD1, KIAA1109, MDH1 \n",
       "pcagrey_ 5 \n",
       "Positive:  TF, TMEM165, FIS1, LAMP2, PLD1, SLC25A13, TPP1, SUMO1, P2RX7, CCNI \n",
       "\t   SEMA6A, BBIP1, FBXO7, SLC12A2, BANP, NPM1, DNAJB11, ATP2B4, RAB31, GLTP \n",
       "\t   TRHDE, TJP1, KIF13A, SLC1A3, TSHR, BRK1, PARK7, NXN, MIEN1, HAX1 \n",
       "Negative:  ASIC2, DGKD, KCNMA1, ALK, TBCK, RAB3IP, TOX, TBC1D5, KCNN2, ARHGEF11 \n",
       "\t   PRKCA, SPRY2, RIMS2, TSG101, SHANK1, HECW2, AMPH, IL1RAPL1, PLEKHA8, ACVR1 \n",
       "\t   FER, RAB3GAP2, ODF2, SNX9, CDKL5, TVP23A, GOLGA4, RASGRP2, CACNA1C, TBC1D9 \n",
       "Harmony 1/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 2/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 3/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 4/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 5/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony converged after 5 iterations\n",
       "Warning: Invalid name supplied, making object name syntactically valid. New object name is Seurat..ProjectDim.originalexp.harmony; see ?make.names for more details on syntax validity\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Warning: Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcablue to pcablue_\n",
       "Warning: All keys should be one or more alphanumeric characters followed by an underscore '_', setting key to pcablue_\n",
       "pcablue_ 1 \n",
       "Positive:  BBS4, DRAM2, CFTR, AREL1, ARF4, SEC23IP, HDAC2, IL17D, TGFB1, STX6 \n",
       "\t   HHIP, TRIM13, ZDHHC15, VPS29, CFAP43, RIPK1, IQCG, CHP1, IFT52, TMED4 \n",
       "\t   VPS26A, PGM1, ABCB7, ATP6V0A2, ERAP1, NTN1, SAR1A, KIF27, BCCIP, HMGXB4 \n",
       "Negative:  NRG3, AHI1, ERBB4, LRRC4C, RABGAP1L, PDXDC1, SLC25A12, WWOX, NSF, BCAS3 \n",
       "\t   EXOC4, SPAG16, NTRK3, NLGN1, CACNA1D, KCNH1, PTPRS, SIL1, ARHGAP44, ABLIM1 \n",
       "\t   BMPR2, HERC2, SPIDR, GNAL, SHISA9, CHCHD3, HECW1, APC, TTC17, RAB11A \n",
       "pcablue_ 2 \n",
       "Positive:  ZNF804A, ERBB4, LRRC4C, AHI1, ESYT2, SLC4A8, RAB28, KCNH1, NTRK3, RTN4RL1 \n",
       "\t   NRG3, HERC2, SHISA9, SYP, HECW1, ABLIM1, PHKB, YTHDC2, GNAL, NDRG3 \n",
       "\t   ARNT, NLGN1, TTC17, KCNH7, ATP6V1A, SPIDR, LMBRD1, VPS35, PLCE1, TTC21B \n",
       "Negative:  CRLF3, HHIP, PDXDC1, MAK, MTCP1, SAR1A, MAP2K1, KCNQ1, CCDC65, HSPB11 \n",
       "\t   MYO10, RAB11A, NPC1, SERP1, GBE1, TGFB1, ERAP1, MFN1, BCO2, TCHP \n",
       "\t   DNAJC3, IQCG, PTPRS, BBS4, UGP2, MFF, NTN1, TRIM13, CFTR, STX6 \n",
       "pcablue_ 3 \n",
       "Positive:  ATP1B1, VPS35, SYP, ZEB2, TMCO1, FOXO3, TMED4, IFT52, ATP6V1A, PTEN \n",
       "\t   GRB2, ARF4, MTDH, MARK4, AREL1, RAB1A, LAPTM4B, LMBRD1, ARSB, WWOX \n",
       "\t   VPS29, GFPT1, VPS45, PRKAA2, BCAS3, SUMF2, CHP1, NDRG3, BBS4, PIGU \n",
       "Negative:  ZNF804A, ERBB4, LRRC4C, HECW1, KCNH7, PLCE1, MAK, IFT172, BCO2, SPAG16 \n",
       "\t   NLGN1, GNAL, PDXDC1, TRPC1, MTCP1, RTN4RL1, KCNH1, NRG3, CRLF3, STX6 \n",
       "\t   CHCHD3, HSPB11, KCNQ1, RABGAP1L, NTN1, EXOC4, SIL1, TCHP, KIF27, MFN1 \n",
       "pcablue_ 4 \n",
       "Positive:  ZEB2, ESR2, IFT52, SPIDR, ARSB, HMGXB4, PIBF1, DHRSX, MYO10, KIF27 \n",
       "\t   PHACTR4, RB1, ZDHHC15, PTCH2, CFAP43, TTC17, RABL2A, SMARCA2, NRG3, NPC1 \n",
       "\t   RIPK1, CC2D2A, CHPT1, ACACA, TBK1, TTC21B, DRAM2, NLGN1, SMG1, EXOC4 \n",
       "Negative:  LAPTM4B, ATP1B1, PRKAA1, AREL1, SAR1A, VPS35, PGM1, MTMR4, NSF, ESYT2 \n",
       "\t   UGP2, PTPRS, MTCP1, USO1, TMED4, RALA, SLC4A8, ADAM17, ATP6V1A, BCO2 \n",
       "\t   DNAJC3, RTN4RL1, SYP, PLCE1, STX6, NDRG3, GNAL, GRB2, PRKAG1, ARF4 \n",
       "pcablue_ 5 \n",
       "Positive:  TRIM13, EP300, PHACTR4, ZNF804A, DRAM2, PTPRS, SRGAP2C, AP3B1, ABCB7, SGPL1 \n",
       "\t   CFAP43, RALA, CC2D2A, GFPT1, TGFB1, HSPB11, VPS45, MYO10, HHIP, CFTR \n",
       "\t   ERBB4, SLC30A5, ADAM17, DNAJC3, ATP1B1, NTN1, IDE, SLC23A2, SMG1, TCHP \n",
       "Negative:  PLCE1, CACNA1D, BBS4, DTNBP1, IQCG, ZDHHC15, NPC1, PARL, EPS15L1, ESR2 \n",
       "\t   RTN4RL1, CCDC65, SCFD2, VPS29, AREL1, IFT172, PANK2, KCNQ1, KCNH7, SCFD1 \n",
       "\t   PIGU, SESN1, IFT52, BCO2, PRKAG1, ASXL1, VPS26A, STX6, PGM1, TRPC1 \n",
       "Harmony 1/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 2/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 3/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 4/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 5/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 6/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony converged after 6 iterations\n",
       "Warning: Invalid name supplied, making object name syntactically valid. New object name is Seurat..ProjectDim.originalexp.harmony; see ?make.names for more details on syntax validity\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Warning: Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcayellow to pcayellow_\n",
       "Warning: All keys should be one or more alphanumeric characters followed by an underscore '_', setting key to pcayellow_\n",
       "pcayellow_ 1 \n",
       "Positive:  SHISA8, IGFBP5, RELN, SLC24A4, PM20D2, VIP, CHRM5, ASIC4, ARRDC3, PLK2 \n",
       "\t   ITGAV, CCNG2, CHRM2, TRPC6, SLC12A9, KCNK2, IRS1, CHST15, ADRA1B, GOLGA7 \n",
       "\t   RBBP7, FOXO1, CNR1, CRY1, ASIC1, SLC6A11, HLA-A, ARL13B, PTH2R, TBC1D20 \n",
       "Negative:  ANK3, NRXN1, LRRC7, AKAP6, CSMD3, RIMS1, DYNC1I1, GRIK2, PAK3, KCNQ3 \n",
       "\t   KCNJ3, SLC6A1, PRKCB, CDH13, LRP1B, ADCY2, NLGN4Y, GSK3B, KIDINS220, KCNQ5 \n",
       "\t   NEDD4L, UBASH3B, MAPK8, SREBF2, COBL, KCND3, WWP1, PID1, DCC, SCN3A \n",
       "pcayellow_ 2 \n",
       "Positive:  KIT, NTNG1, NRG1, SLC6A1, PID1, ROR2, EPHA3, KCNH5, COBL, GNG4 \n",
       "\t   WWP1, CDH13, UBASH3B, LRRC7, EGFR, AP1S2, PRKCB, RIMS1, FOXO1, KCNQ3 \n",
       "\t   ABCC1, BCL2, TTC39C, CSMD3, RERG, SLC1A1, PTPRO, DYNC1I1, HTR1E, PLPPR5 \n",
       "Negative:  GRM7, CHRM3, PTPRE, B4GALT6, LRP1B, RYR3, ASIC4, VIP, SLC24A3, KCNT2 \n",
       "\t   KCNK2, SHISA8, CHST15, SYT17, ADRA1B, CCNG2, KCNQ5, SHISA6, CHRNA7, IGFBP5 \n",
       "\t   TRPC6, SCN3A, ASIC1, PTH2R, CHRM5, DCC, CNR1, CHRM2, NLGN4Y, CDH2 \n",
       "pcayellow_ 3 \n",
       "Positive:  KCNT2, VIP, CHRM3, NEDD4L, NRXN1, SLC24A3, SHISA8, ASIC4, RERG, EGFR \n",
       "\t   HTR1E, SCG2, COBL, GRM7, KIT, SLC6A11, AP1S2, NRG1, IGFBP5, ADRA1B \n",
       "\t   RYR3, CCNG2, PID1, TTC39C, NTNG1, CHST15, HLA-A, FOXO1, LRRC7, RXRA \n",
       "Negative:  AKAP6, GSK3B, REEP5, KCNJ3, LRP1B, DYNC1I1, FNBP1L, WAC, CPEB4, MAPK8 \n",
       "\t   ADCY2, KCNQ5, ANK3, ATP6V0E2, KIF5B, FBXL2, PAK3, DHX36, SLC16A14, SCLT1 \n",
       "\t   SCG5, KIDINS220, ARL13B, PTPRE, LRP1, TBC1D20, USP13, KCNQ3, CHRM2, NLGN4Y \n",
       "pcayellow_ 4 \n",
       "Positive:  SHISA8, CCNG2, B4GALT6, GRM7, PRKCB, TTC39C, PPP1R16B, HTR1E, LRP1B, SLC24A3 \n",
       "\t   MAPK8, FOXK1, TBC1D1, ITGAV, ASIC4, RERG, PTPRE, USP13, SLC6A11, DHX36 \n",
       "\t   SLC6A1, SLC24A4, RYR3, VIP, BID, CDH13, UBASH3B, ATP6V0E2, IGFBP5, KCNH5 \n",
       "Negative:  CNR1, CHRNA7, KCNQ5, GRIK2, SHISA6, KCND3, NRXN1, RELN, KCNQ3, SCN3A \n",
       "\t   SCN3B, KCNB2, WWP1, SYT17, CHRM3, CSMD3, PID1, ABCC1, KCNT2, DCC \n",
       "\t   FBXL2, TRPC6, PLK2, ADRA1B, KIDINS220, GNG4, CDH2, ANK3, SLC1A1, AP1S2 \n",
       "pcayellow_ 5 \n",
       "Positive:  KCNQ5, AP1S2, CHRNA7, TGFB2, PTPRE, KCNT2, KCNJ3, FNBP1L, CNR1, ADRA1B \n",
       "\t   WWP1, SYT17, EPHA3, ITGAV, ASIC1, PID1, IRS1, ASIC4, ABCC1, VIP \n",
       "\t   EGFR, NRG1, CCNH, CPEB4, SSR1, KIT, TTC39C, IGFBP5, HSPA5, EPHA4 \n",
       "Negative:  KCNB2, GRIK2, CDH2, SCN3B, TRPC6, PTPRO, SHISA6, CDH13, KCNK2, LRRC7 \n",
       "\t   DCC, GNG4, SLC6A11, RERG, NTNG1, UBASH3B, ADCY2, SLC24A3, LGI1, NEDD4L \n",
       "\t   CHRM2, CHRM3, PAK3, SCN3A, SLC24A4, HTR1E, PPP1R16B, LRP1B, CHST15, ANK3 \n",
       "Harmony 1/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 2/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 3/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 4/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 5/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 6/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 7/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 8/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 9/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 10/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Warning: Invalid name supplied, making object name syntactically valid. New object name is Seurat..ProjectDim.originalexp.harmony; see ?make.names for more details on syntax validity\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Warning: Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcabrown to pcabrown_\n",
       "Warning: All keys should be one or more alphanumeric characters followed by an underscore '_', setting key to pcabrown_\n",
       "pcabrown_ 1 \n",
       "Positive:  ABCG1, SLC6A6, NR1H3, DNAJC27, PDCD6, SCN7A, RAB3B, PTPDC1, GNAI3, RABL2B \n",
       "\t   PTPRT, KCNH3, MAVS, CCNJL, KCNK3, HTR4, GYS1, STS, LGI3, KCNA2 \n",
       "\t   WIF1, KSR1, LGMN, KCNMB4, AGL, SLC36A1, TRAF3IP1, ARL6, IGF2R, WLS \n",
       "Negative:  MAGI2, CADPS, ADGRB3, KCNC2, SLC4A10, SYT1, TENM2, HCN1, ADGRL3, KCND2 \n",
       "\t   PDE4DIP, NTRK2, CACNA1A, SLIT2, KCNAB1, PPFIA2, CAMK1D, SCN1A, WDPCP, SIPA1L1 \n",
       "\t   SLC24A2, ADCY1, STXBP5, TENM1, SCN8A, PCLO, CLEC16A, MKLN1, ST8SIA1, PPARGC1A \n",
       "pcabrown_ 2 \n",
       "Positive:  SULF1, SLC4A4, PPARGC1A, SLC9A9, CUX2, TRPC4, TENM1, SCN9A, SLIT2, CLEC16A \n",
       "\t   ST8SIA1, KCNS3, DOCK11, HTR4, ROR1, KCNAB3, TRPC5, LGR4, GHR, IQCB1 \n",
       "\t   HCN1, CPLX1, GRM8, SLC17A5, KIF16B, BEND6, WDPCP, TENM2, ADCY1, PTPDC1 \n",
       "Negative:  PTPRT, WLS, ADGRL2, SYT1, MCTP1, ST6GALNAC5, NTRK2, LGMN, STXBP6, GALNT11 \n",
       "\t   KCNK3, RAB3B, SGSM1, ST6GALNAC3, ABCA3, ATG7, PIK3CA, PCLO, MARK1, RAPGEF1 \n",
       "\t   PCDH15, SCN1A, RABGAP1, KSR1, KIF3B, RNF157, CACNA1A, KCNAB1, RUBCN, STX12 \n",
       "pcabrown_ 3 \n",
       "Positive:  PTPRT, SLC7A8, KCNAB1, SCN1A, SIPA1L1, ADCY1, TTBK2, SYT1, CACNA1A, KCNK3 \n",
       "\t   KCNC1, KCNAB3, CPLX1, IGF2R, MAVS, FNIP2, ABCA3, LGI3, CAMK2G, SLC39A14 \n",
       "\t   BEND6, SULF1, NTRK2, DNAJC27, RUBCN, MCTP1, GGT7, KCNMB4, HCN1, SLC6A6 \n",
       "Negative:  STXBP6, ST6GALNAC5, ADGRL3, KCND2, TENM3, RAB3B, PCDH15, TENM2, WLS, SCN9A \n",
       "\t   WIF1, ADCY8, SLC36A1, SCN7A, TRPC4, MAGI2, CBLB, PCLO, ITPR1, DOCK11 \n",
       "\t   ADGRL2, SLC24A2, CAMK2D, STXBP5, MARK1, TBC1D30, GHR, LGMN, ST6GALNAC3, PKD2 \n",
       "pcabrown_ 4 \n",
       "Positive:  PTPRT, LGR4, SCN9A, WDPCP, MNAT1, PPFIA2, HTR4, MCTP1, PTPDC1, GHR \n",
       "\t   PDE4DIP, CUX2, ARL6, KCNAB1, ADGRB3, ATG7, SIPA1L1, ST6GALNAC3, TANGO6, IQCB1 \n",
       "\t   PKD2, SCN7A, SLC38A6, RABGAP1, ADGRL2, SLC9A9, KIF16B, SYT1, MKLN1, CAMK1D \n",
       "Negative:  CPLX1, KCNA2, KCNK3, RAB3B, LRP8, LGI3, SERPINI1, SLC7A8, LGMN, STXBP6 \n",
       "\t   STS, KCNAB3, KCNC1, DNAJC27, SLC36A1, SLC24A2, KCNS3, ABCA3, KIF3B, PPARGC1A \n",
       "\t   RNF157, ITPR1, SLC39A14, ST6GALNAC5, ADCY1, SULF1, TTBK2, PCDH15, SLC17A5, KCNH3 \n",
       "pcabrown_ 5 \n",
       "Positive:  CAMK2D, SCN7A, PCLO, ASAP1, SESN3, TBC1D30, WIF1, PPFIA2, MARK1, KCNK3 \n",
       "\t   ITPR1, PDE4DIP, IGF2R, TRAF3IP1, PIK3CA, TRPC4, SLC17A5, FNIP2, RABGAP1, MNAT1 \n",
       "\t   SLC8A3, RPTOR, NLK, SLC7A8, KCNJ6, PTPDC1, ABCA3, KCNA2, RNF157, PDCD6 \n",
       "Negative:  PCDH15, ADGRL2, ST6GALNAC5, PTPRT, TRPC5, KCNAB3, CACNA1A, PPARG, ST6GALNAC3, MCTP1 \n",
       "\t   TENM1, SLIT2, SULF1, TENM2, KCNH3, SCN1A, ROR1, DOCK11, TBC1D4, KCNC2 \n",
       "\t   TENM3, SIPA1L1, HCN1, GHR, SLC24A2, LRP8, CUX2, CADPS, LGMN, CBLB \n",
       "Harmony 1/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 2/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 3/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 4/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 5/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 6/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 7/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 8/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 9/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 10/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony converged after 10 iterations\n",
       "Warning: Invalid name supplied, making object name syntactically valid. New object name is Seurat..ProjectDim.originalexp.harmony; see ?make.names for more details on syntax validity\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |===================================                                   |  50%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Warning: Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcaturquoise to pcaturquoise_\n",
       "Warning: All keys should be one or more alphanumeric characters followed by an underscore '_', setting key to pcaturquoise_\n",
       "pcaturquoise_ 1 \n",
       "Positive:  ARFGAP1, PLK5, NTNG2, GPS1, DLG4, NSMF, SLC38A3, AGRN, PPIE, BRSK2 \n",
       "\t   FSTL3, LAMB2, ANO8, TM7SF2, FBXW5, NR1D1, CTSA, PRKCSH, IFT122, OBSL1 \n",
       "\t   CSF1, S1PR1, ARAP1, GAS6, ADGRB2, ZBTB7A, EPAS1, GAS2L1, AKT2, STYXL1 \n",
       "Negative:  CLU, CST3, ADCY2, ATP1B2, ITM2C, NTRK3, PSAP, HIF3A, GFAP, CHN1 \n",
       "\t   MT3, APOE, SPOCK1, ADD1, CALM1, CD81, FAM107A, ENO1, GNAO1, GPR37L1 \n",
       "\t   APP, ACADVL, PREPL, ATP2A2, ATP6V0A1, SREBF2, GABBR1, GNAS, CANX, MAP2 \n",
       "pcaturquoise_ 2 \n",
       "Positive:  BASP1, SNCA, GRIK2, SLC4A10, STMN2, CAMK2A, APP, MAP2, PREPL, DNM1 \n",
       "\t   SYP, CHN1, CALM1, FBXW7, NCDN, PTK2B, SERPINI1, APLP2, ENO2, PPIA \n",
       "\t   SPOCK1, DNAJC6, GNAS, ATP6V0A1, INPP5F, NPTN, ATP9A, KNDC1, MAP1LC3A, ADAR \n",
       "Negative:  GPR37L1, APOE, CLU, CD81, ITM2C, ADGRG1, ATP1B2, WFS1, MLC1, CST3 \n",
       "\t   SLC1A4, VEGFA, PLXNB1, SCG3, ENO1, S1PR1, ADCYAP1R1, MFGE8, PSAP, TMEM132A \n",
       "\t   RAMP1, RXRA, APC2, SLC38A3, FAM107A, PLXNB2, LAMB2, HIP1R, HIF3A, SREBF1 \n",
       "pcaturquoise_ 3 \n",
       "Positive:  DNM1, PLXNB1, MFGE8, ARFGAP1, NSMF, PTK2B, NCDN, WFS1, CAMK2B, ADGRB2 \n",
       "\t   PSAP, APOE, ADGRB1, DLG4, CAMK2A, TSPYL2, GPR37L1, KLC1, CALM1, NDRG4 \n",
       "\t   TMEM132A, ACTR1A, SERPINI1, NTNG2, BASP1, NCS1, SLC4A10, SLC1A4, AGRN, APC2 \n",
       "Negative:  GFAP, SPOCK1, NTRK3, DAB2IP, SREBF2, FAM20C, MT3, INPP5F, ADCY2, CST3 \n",
       "\t   LMNA, ADD1, RAB34, EPAS1, FAM107A, INSR, STYXL1, EDF1, SHISA5, DNAJC6 \n",
       "\t   NPRL3, LAMP2, FLOT1, DNM2, HIF3A, RPTOR, CDK5RAP3, SMPD4, FLOT2, POM121 \n",
       "pcaturquoise_ 4 \n",
       "Positive:  SLC25A29, SNRNP70, PLK5, GFAP, PLXNB1, PRKCSH, FLNA, DNM2, NSMF, MT3 \n",
       "\t   LAMB2, ARFGAP1, RAMP1, TMEM132A, LMNA, PLXNB2, KIF1A, NTNG2, FAM107A, ZBTB7A \n",
       "\t   GGA1, FAM20C, CSF1, HSP90B1, BRSK2, ARAP1, GAS6, AGRN, SMPD4, WFS1 \n",
       "Negative:  SCG3, ATP1B2, FBXO21, CST3, GPR37L1, CHN1, DNAJC6, NDRG4, DDB1, PLXNA2 \n",
       "\t   RAB34, INSR, ITM2C, TM7SF2, MLC1, S1PR1, SLC1A4, PREPL, PGAM1, KPNB1 \n",
       "\t   ADD1, APLP2, APOE, APC2, STMN2, RAB11B, CANX, HIP1R, ATP6V0E2, CTDSP2 \n",
       "pcaturquoise_ 5 \n",
       "Positive:  RAMP1, MT3, ENO1, SSR1, ITM2C, KLHL21, PGAM1, CLU, CD81, CSF1 \n",
       "\t   CTSB, CANX, HSP90B1, PSAP, PLK5, CST3, APLP2, CALM1, PPIA, ARAP1 \n",
       "\t   ADCYAP1R1, CFL1, GNAS, NPTN, FAM107A, SCARB2, NCDN, ATP9A, ADGRB2, NR1D1 \n",
       "Negative:  NTNG2, ANO8, TNPO2, SLC38A3, GPS1, ABCA2, APC2, PALM, RAB11B, NSMF \n",
       "\t   AKT2, GAK, ARFGAP1, SREBF1, SPOCK1, TM7SF2, PLXNB1, STYXL1, SREBF2, CAPZB \n",
       "\t   CTDSP2, CTSA, FSTL3, GNAO1, SNRNP70, KNDC1, MFGE8, POM121, DLG4, FBXO21 \n",
       "Harmony 1/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 2/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 3/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 4/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 5/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 6/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 7/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 8/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 9/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 10/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Warning: Invalid name supplied, making object name syntactically valid. New object name is Seurat..ProjectDim.originalexp.harmony; see ?make.names for more details on syntax validity\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Warning: Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcagrey to pcagrey_\n",
       "Warning: All keys should be one or more alphanumeric characters followed by an underscore '_', setting key to pcagrey_\n",
       "pcagrey_ 1 \n",
       "Positive:  TFRC, TNC, SPP1, FBXO2, FSTL1, POLR2B, SLC38A1, PSMC6, TIPARP, PSMA1 \n",
       "\t   SIRT2, ABCD2, NDEL1, WDR35, DCTN4, FADS1, UBXN7, ATP6V1D, POLR2A, GOLM1 \n",
       "\t   GNA13, SLC25A17, SLC38A2, BMP7, ARSB, FBXO7, ST3GAL2, DPYSL3, PGK1, ATP6V1E1 \n",
       "Negative:  MACF1, SPARCL1, DNM3, TENM2, NTRK2, IL1RAPL1, KANK1, GNA14, RUFY3, CCDC88A \n",
       "\t   SPTBN1, PPFIA2, PTPRD, ATG7, CAMK1D, ZBTB16, SGIP1, KTN1, BMPR2, CLEC16A \n",
       "\t   ANK3, FBXO11, CBFA2T2, CAMSAP2, TANC2, CUX1, NLGN1, RAPGEF2, ROBO1, COLEC12 \n",
       "pcagrey_ 2 \n",
       "Positive:  GNA14, MACF1, NRP1, COLEC12, CNN3, HIF1A, PTPN9, SPARCL1, SMURF2, SLC6A11 \n",
       "\t   SLC3A2, LGALS3, ITGA6, KANK1, WASF2, NTRK2, WWC1, LRP1, LDLR, SDC2 \n",
       "\t   FSTL1, RUFY3, HLA-A, CLEC16A, AXL, TIPARP, DCUN1D1, PDLIM5, TULP3, AKAP12 \n",
       "Negative:  IL1RAPL1, NLGN1, ANK3, PTPRD, CHRM3, ROBO2, CSMD3, HECW2, ROBO1, DYNC1I1 \n",
       "\t   RYR2, NSF, PPFIA2, TANC2, HECW1, ADCY1, MCTP1, PAK1, TIAM1, DCC \n",
       "\t   RAPGEF2, POLR2F, EPHA4, ARHGAP44, SNAP91, PRKACB, NEDD4L, PIK3R1, SLC38A1, PRKAG2 \n",
       "pcagrey_ 3 \n",
       "Positive:  HSP90AA1, CRYAB, NTRK2, SLC38A2, FBXO2, NEDD4L, ADCY8, SEMA3A, VCAN, EEF1A1 \n",
       "\t   SLC38A1, SPARCL1, SLC17A5, SORL1, SLC16A9, HLA-A, PIK3R1, GNA13, CDH2, AXL \n",
       "\t   RCN1, IGF2R, TIMP3, SDC2, DDX3X, ABCD2, SGIP1, CALR, APPL2, ABCD3 \n",
       "Negative:  COLEC12, GNA14, MACF1, NRP1, LDLR, SLC6A11, IL1RAPL1, TENM2, COBL, RUFY3 \n",
       "\t   BMPR2, PTPN9, POLR2F, SMURF2, TIPARP, ANK3, PTPRD, CAMK1D, WASF2, HECW2 \n",
       "\t   CCDC88A, HIF1A, DENND5A, GNA12, ROBO1, HECW1, LGALS3, AP1S2, RYR2, CLEC16A \n",
       "pcagrey_ 4 \n",
       "Positive:  CRYAB, HSP90AA1, VCAN, DPYSL3, AP1S2, SLC38A1, TNC, LDLR, FBXO2, COLEC12 \n",
       "\t   SLC38A2, POLR2F, ADCY8, MACF1, SPP1, COBL, EEF1A1, RAPGEF1, CNN3, ST6GALNAC3 \n",
       "\t   IL1RAPL1, SNX1, HAP1, NPC1, RUFY3, CTTN, BMPR2, PTPN9, DENND5A, DYNLL1 \n",
       "Negative:  NTRK2, ABCD2, SPARCL1, EPHA4, GRAMD1B, TIMP3, DCUN1D1, TPP1, SNAP91, CLEC16A \n",
       "\t   APPL2, CDH2, ATP6V1H, SDC2, MYO10, SPSB3, ATAD1, HECW1, LMBRD1, CYP7B1 \n",
       "\t   SORL1, SPTBN1, ERCC8, CHRM3, BTRC, IFT88, SLC6A11, CSMD3, RCN1, MTM1 \n",
       "pcagrey_ 5 \n",
       "Positive:  CAMK1D, CHRM3, DPYSL3, ROBO2, TENM2, ADCY1, CSMD3, MCTP1, KIFAP3, RYR2 \n",
       "\t   BMPR2, ARHGAP44, HECW1, TNC, PRKAG2, RAPGEF2, VCAN, SLC38A1, SLC25A17, NSF \n",
       "\t   PPFIA2, DCC, SNAP91, EPHA4, CDKL5, PEX14, PAK1, ADCY9, BANP, TPR \n",
       "Negative:  POLR2F, IL1RAPL1, PSEN1, DNM3, ANK3, FAXDC2, SPP1, COBL, PTPRD, BBS2 \n",
       "\t   NPC1, SNX1, CCDC88A, SLC13A3, NLGN1, DENND5A, SIRT2, UVRAG, MYO10, FBXO7 \n",
       "\t   DYNC1I2, PSME4, IFT88, CYP7B1, DCUN1D1, KLHL5, ZBTB16, SDC2, ST6GALNAC3, HECW2 \n",
       "Harmony 1/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 2/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 3/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony converged after 3 iterations\n",
       "Warning: Invalid name supplied, making object name syntactically valid. New object name is Seurat..ProjectDim.originalexp.harmony; see ?make.names for more details on syntax validity\n",
       "Centering and scaling data matrix\n",
       "\r  |                                                                            \r  |                                                                      |   0%\r  |                                                                            \r  |======================================================================| 100%\n",
       "Warning: Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcablue to pcablue_\n",
       "Warning: All keys should be one or more alphanumeric characters followed by an underscore '_', setting key to pcablue_\n",
       "pcablue_ 1 \n",
       "Positive:  MAGI2, RORA, GPM6A, CTNNA2, LRP1B, ERBB4, ADGRB3, ADGRL3, FMN2, GNAQ \n",
       "\t   FBXL7, BCAS3, WSB1, RAB11A, APC, FAF1, FGFR2, SIPA1L1, LGR4, FUT9 \n",
       "\t   OSBPL1A, SRPK2, LRRC4C, PTN, FOXO1, PRKD1, KIF13A, ITCH, TF, PKN2 \n",
       "Negative:  GTF2F2, RHOQ, AP1G1, ATP6V0E1, NUP88, SEC24A, CYP27A1, ABCA1, NUP214, TBC1D12 \n",
       "\t   FBXL4, FBXL3, NUP58, CHM, CNOT7, CBR1, LAMC1, HSPA8, PSMD5, LIMA1 \n",
       "\t   SEC24B, KIF13B, NUP160, SHISA6, PAQR3, ITGB1, WWP1, GRIP1, SNX14, TENM3 \n",
       "pcablue_ 2 \n",
       "Positive:  IGFBP7, RAB11A, LTBP1, TF, PRSS23, WSB1, STAT3, ITCH, SEC23A, UBA3 \n",
       "\t   ITGB1, FN1, TBC1D12, KIF13B, TULP4, COMMD1, GAN, LIPA, FBXL13, CUL4A \n",
       "\t   DCAF5, FGFR2, DCAF10, PICALM, WWP1, MDM4, SMURF1, VMP1, NUP160, AP2B1 \n",
       "Negative:  TOX, GPM6A, EGFR, CAMK2G, LRP1B, ERBB4, ADGRB3, PTN, RORA, LRRC4C \n",
       "\t   FUT9, ADGRL3, MAGI2, HACD3, CTNNA2, PDK4, EPS15, TMEM108, LGR4, ULK4 \n",
       "\t   WRN, PRKD1, FOXO1, ABI2, SHISA6, SUMF1, RHBDD1, FMN2, RSF1, BCAS3 \n",
       "pcablue_ 3 \n",
       "Positive:  LRRC7, LIPA, GDAP1, TULP4, GRM7, SIPA1L1, PTPRO, GRIP1, ROCK1, TMEM108 \n",
       "\t   KIF13A, KIF13B, DCAF6, KIF2A, USP33, ABI2, ADGRL3, TENM3, PICALM, SRPK2 \n",
       "\t   TBC1D12, LRRC4C, MDM4, LRP6, SUMF1, PAQR3, GTF2F2, ATP6V1A, OSBPL1A, PRKD1 \n",
       "Negative:  IGFBP7, PRSS23, STAT3, LGR4, FGFR2, GAN, PTN, APC, FBXL7, MTMR3 \n",
       "\t   LTBP1, CTNNA2, ITGB1, FOXO1, CAPZA1, ABCA1, GNAQ, ERBB4, RAB11A, RAB31 \n",
       "\t   FN1, NUP160, RORA, DCAF5, PDK4, HACD3, GPM6A, FMN2, EPS15, TOX \n",
       "pcablue_ 4 \n",
       "Positive:  RHOQ, ABCA1, HSPA8, FBXL4, KIF13B, INVS, CBR1, PICALM, ATP6V0E1, SMURF1 \n",
       "\t   TF, TBC1D12, KIF13A, EGFR, FCHSD2, LAMC1, FOXO1, EPS15, FUT9, PDK4 \n",
       "\t   SEC24B, KIF2A, LIPA, MDM4, ADAM10, OSBPL1A, CYP27A1, FBXL7, FMN2, PRKD1 \n",
       "Negative:  TMEM108, LRRC7, PTPRO, GDAP1, ALK, RAB11A, MAP3K13, GRM7, LIMA1, SHISA6 \n",
       "\t   MTMR3, LRRC4C, NUP160, ADGRB3, TENM3, GRIP1, PAQR3, CHM, AP2B1, IGFBP7 \n",
       "\t   PTN, CAPZA1, PRSS23, LGR4, DCAF5, PSMA3, CNOT7, FN1, ATP6V1A, RAB31 \n",
       "pcablue_ 5 \n",
       "Positive:  KIF13B, LIPA, OSBPL1A, ADAM10, TULP4, PICALM, TF, FGFR2, CBR1, SNX9 \n",
       "\t   MAGI2, SUMF1, PTN, GTF2F2, DHX36, CYP27A1, AP2B1, TOX, FBXW11, FBXL7 \n",
       "\t   LGR4, BCAS3, SHISA6, ERBB4, PKN2, WSB1, IGFBP7, LRRC4C, GNAQ, LIMA1 \n",
       "Negative:  RHOQ, DNAJC3, RIC1, CNOT7, LAMC1, NUP88, PAQR3, HSPA8, SIPA1L1, PRKD1 \n",
       "\t   WRN, FMN2, PSMA3, FUT9, TENM3, PTEN, DCAF10, MDM4, COMMD1, CUL5 \n",
       "\t   PTPRO, KIF13A, ATP6V1A, SERP1, AP1G1, ATP6V0E1, LTBP1, FOXO1, GRIP1, WWP1 \n",
       "Harmony 1/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 2/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 3/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 4/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony 5/10\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "Harmony converged after 5 iterations\n",
       "Warning: Invalid name supplied, making object name syntactically valid. New object name is Seurat..ProjectDim.originalexp.harmony; see ?make.names for more details on syntax validity\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "    modules <- GetModules(seurat_obj[[cell_type]])\n",
    "\n",
    "    if (!(length(unique(modules$module))<=2 &  'grey'%in%modules$module)){\n",
    "\n",
    "        print(paste0('Estimating module eigen-genes in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "        \n",
    "        # fneed to run ScaleData first or else harmony throws an error:\n",
    "        seurat_obj[[cell_type]] <- ScaleData(seurat_obj[[cell_type]], features=VariableFeatures(seurat_obj[[cell_type]]))\n",
    "\n",
    "        # compute all MEs in the full single-cell dataset\n",
    "        seurat_obj[[cell_type]] <- ModuleEigengenes(\n",
    "        seurat_obj[[cell_type]],\n",
    "        group.by.vars=subject_id\n",
    "        )\n",
    "        print('----------------')\n",
    "\n",
    "    }   else{    \n",
    "\n",
    "        print(paste0('Error encountered while processing ', toupper(cell_type)))\n",
    "        print(paste0(toupper(cell_type), ' dropped from experiment'))\n",
    "        # Code to handle the error condition\n",
    "        seurat_obj <- seurat_obj[names(seurat_obj) != cell_type]\n",
    "        assign(\"seurat_obj\", seurat_obj, envir = .GlobalEnv) # Update the modified seurat_obj in the global environment\n",
    "\n",
    "    } \n",
    "    \n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "05cd78cf",
   "metadata": {},
   "source": [
    "The ME matrices are stored as a matrix where each row is a cell and each column is a module. This matrix can be extracted from the Seurat object using the GetMEs function, which retrieves the hMEs by default."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "93840f67",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Getting harmonized and non-harmonized MEs in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"Getting harmonized and non-harmonized MEs in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"Getting harmonized and non-harmonized MEs in hdWGCNA Experiment for ASTROCYTE\"\n"
     ]
    }
   ],
   "source": [
    "%%R -o hMEs -o MEs\n",
    "\n",
    "hMEs <- list()\n",
    "MEs <- list()\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    print(paste0('Getting harmonized and non-harmonized MEs in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "    \n",
    "    # harmonized module eigengenes:\n",
    "    hMEs[[cell_type]] <- GetMEs(seurat_obj[[cell_type]])\n",
    "\n",
    "    # module eigengenes:\n",
    "    MEs[[cell_type]] <- GetMEs(seurat_obj[[cell_type]], harmonized=FALSE)\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c011bbb2",
   "metadata": {},
   "source": [
    "### **Compute module connectivity**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e9f2e4ad",
   "metadata": {},
   "source": [
    "We want to focus on `hub genes`, which are highly connected within each module. To assess this, hdWGCNA includes the `ModuleConnectivity` function to calculate the `eigengene-based connectivity (kME)` of each gene in the full single-cell data. This function computes pairwise correlations between genes and module eigengenes. While kME can be calculated for all cells, it is recommended to compute it specifically in the cell type or group initially used in the ConstructNetwork function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "9ebaef5b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Estimating module connectivity in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"Estimating module connectivity in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"Estimating module connectivity in hdWGCNA Experiment for ASTROCYTE\"\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Estimating module connectivity in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # compute eigengene-based connectivity (kME):\n",
    "  seurat_obj[[cell_type]] <- ModuleConnectivity(\n",
    "    seurat_obj[[cell_type]],\n",
    "    group.by = 'cell_type', \n",
    "    group_name = cell_type\n",
    "  )\n",
    "  \n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "95b46026",
   "metadata": {},
   "source": [
    "For convenience, we re-name the hdWGCNA modules to indicate that they are from a cell-type group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a52dff32",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Renaming modules in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"Renaming modules in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"Renaming modules in hdWGCNA Experiment for ASTROCYTE\"\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Renaming modules in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # rename the modules\n",
    "  seurat_obj[[cell_type]] <- ResetModuleNames(\n",
    "    seurat_obj[[cell_type]],\n",
    "    new_name = paste0(toupper(substr(cell_type, 1, 3)), \"-M\"),\n",
    "    wgcna_name = toupper(substr(cell_type, 1, 3)) # the name of the hdWGCNA experiment,\n",
    "  )\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "88ad5a62",
   "metadata": {},
   "source": [
    "We also reset module colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "5e7726f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Resetting module colors in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"Resetting module colors in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"Resetting module colors in hdWGCNA Experiment for ASTROCYTE\"\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "\n",
    "library(MetBrewer)\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Resetting module colors in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "\n",
    "  modules <- GetModules(seurat_obj[[cell_type]])\n",
    "  mods <- levels(modules$module)\n",
    "  mod_colors <- dplyr::select(modules, c(module, color)) %>%\n",
    "    distinct %>% arrange(module) %>% .$color\n",
    "  n_colors <- length(mod_colors) -1\n",
    "\n",
    "  new_colors <- paste0(met.brewer(\"Signac\", n=n_colors,))\n",
    "  new_colors <- sample(new_colors)\n",
    "  seurat_obj[[cell_type]] <- ResetModuleColors(seurat_obj[[cell_type]], new_colors)\n",
    "\n",
    "}\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ae334b97",
   "metadata": {},
   "source": [
    "We can visualize the genes in each module ranked by kME using the PlotKMEs function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "03423498",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[conflicted] Will prefer dplyr::select over any other package.\n",
      "[1] \"Visualize Module Connectivity (kMEs) in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"EXC-M1\"\n",
      "[1] \"EXC-M2\"\n",
      "[1] \"EXC-M1\"\n",
      "[1] \"EXC-M2\"\n",
      "[1] \"Visualize Module Connectivity (kMEs) in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"INH-M1\"\n",
      "[1] \"INH-M2\"\n",
      "[1] \"INH-M3\"\n",
      "[1] \"INH-M4\"\n",
      "[1] \"INH-M5\"\n",
      "[1] \"INH-M1\"\n",
      "[1] \"INH-M2\"\n",
      "[1] \"INH-M3\"\n",
      "[1] \"INH-M4\"\n",
      "[1] \"INH-M5\"\n",
      "[1] \"Visualize Module Connectivity (kMEs) in hdWGCNA Experiment for ASTROCYTE\"\n",
      "[1] \"AST-M1\"\n",
      "[1] \"AST-M2\"\n",
      "[1] \"AST-M1\"\n",
      "[1] \"AST-M2\"\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "library(conflicted)\n",
    "conflicts_prefer(dplyr::select)\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/ModuleConnectivity/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "    print(paste0('Visualize Module Connectivity (kMEs) in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    # plot genes ranked by kME for each module\n",
    "\n",
    "    p <- PlotKMEs(seurat_obj[[cell_type]], ncol=5)\n",
    "\n",
    "    pdf(paste0(fig_dir, cell_type, '_KMEs.pdf'), width=12, height=8)\n",
    "    print(p)\n",
    "    dev.off()\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b8d0658b",
   "metadata": {},
   "source": [
    "### Getting the module assignment table\n",
    "\n",
    "The hdWGCNA package provides the `GetModules` function for convenient access to the module assignment table. This table includes three primary columns: gene_name for the gene symbol or ID, module for the gene's module assignment, and color for the color mapping of each module, which is frequently used in downstream plotting. Additionally, if ModuleConnectivity has been used in the hdWGCNA experiment, the table will feature extra columns showing the kME values for each module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "9121be04",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Obtain Module Assignment Table in hdWGCNA Experiment for EXCITATORY\"\n",
      "         gene_name module   color  kME_grey kME_EXC-M1 kME_EXC-M2\n",
      "ACADVL      ACADVL   grey    grey 0.5326983  0.4393766  0.5220664\n",
      "ADD1          ADD1 EXC-M1 #d8443c 0.6120542  0.5898830  0.5640474\n",
      "ARFGAP1    ARFGAP1 EXC-M1 #d8443c 0.3459760  0.4051022  0.2897552\n",
      "ATP6V0D1  ATP6V0D1 EXC-M1 #d8443c 0.4817774  0.5637797  0.3959944\n",
      "CTDSP2      CTDSP2   grey    grey 0.2870663  0.2668398  0.2715963\n",
      "CXXC1        CXXC1   grey    grey 0.2609811  0.2441866  0.2468051\n",
      "[1] \"Obtain Module Assignment Table in hdWGCNA Experiment for INHIBITORY\"\n",
      "         gene_name module   color kME_INH-M1 kME_INH-M2  kME_grey\n",
      "ACADVL      ACADVL INH-M1 #de597c  0.3727467  0.3579442 0.3734083\n",
      "ADD1          ADD1 INH-M2 #d8443c  0.4691682  0.4891299 0.4844225\n",
      "ARFGAP1    ARFGAP1 INH-M2 #d8443c  0.2373664  0.3069337 0.2105659\n",
      "ATP6V0D1  ATP6V0D1 INH-M2 #d8443c  0.3695820  0.4323699 0.3397629\n",
      "CTDSP2      CTDSP2 INH-M1 #de597c  0.1681184  0.1631429 0.1607690\n",
      "CXXC1        CXXC1 INH-M2 #d8443c  0.1943025  0.2082054 0.1951205\n",
      "[1] \"Obtain Module Assignment Table in hdWGCNA Experiment for ASTROCYTE\"\n",
      "        gene_name module   color kME_AST-M1   kME_grey kME_AST-M2\n",
      "ACADVL     ACADVL AST-M1 #fe9b00 0.29778521 0.27854594 0.25820745\n",
      "ADD1         ADD1 AST-M1 #fe9b00 0.34659899 0.37081287 0.36398561\n",
      "ARFGAP1   ARFGAP1 AST-M1 #fe9b00 0.09799177 0.04174751 0.01815066\n",
      "CTDSP2     CTDSP2 AST-M1 #fe9b00 0.18298649 0.18355315 0.17089674\n",
      "DNAJB11   DNAJB11   grey    grey 0.17307919 0.21704292 0.19136278\n",
      "DNAJC3     DNAJC3 AST-M2 #d8443c 0.21889589 0.27685548 0.29815467\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "    print(paste0('Obtain Module Assignment Table in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    # get the module assignment table:\n",
    "    modules <- GetModules(seurat_obj[[cell_type]])\n",
    "\n",
    "    # show the first 6 columns:\n",
    "    print(head(modules[,1:6]))\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "48d2e707",
   "metadata": {},
   "source": [
    "A table of the top N hub genes sorted by kME can be extracted using the GetHubGenes function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "3a0b8898",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Obtain Hub Genes in hdWGCNA Experiment for EXCITATORY\"\n",
      "  gene_name module       kME\n",
      "1     GAPDH EXC-M1 0.7639032\n",
      "2      ACTB EXC-M1 0.7650383\n",
      "3  HSP90AB1 EXC-M1 0.7664502\n",
      "4     UCHL1 EXC-M1 0.7671371\n",
      "5      NRGN EXC-M1 0.7693697\n",
      "6     CALM3 EXC-M1 0.7752952\n",
      "[1] \"Obtain Hub Genes in hdWGCNA Experiment for INHIBITORY\"\n",
      "  gene_name module       kME\n",
      "1      CLTC INH-M1 0.5729561\n",
      "2      PTK2 INH-M1 0.5753169\n",
      "3     PGAM1 INH-M1 0.5781938\n",
      "4      THY1 INH-M1 0.5795863\n",
      "5     STMN2 INH-M1 0.6037622\n",
      "6    ATP1A1 INH-M1 0.6044947\n",
      "[1] \"Obtain Hub Genes in hdWGCNA Experiment for ASTROCYTE\"\n",
      "  gene_name module       kME\n",
      "1     ITM2C AST-M1 0.3993349\n",
      "2     HIF3A AST-M1 0.3998826\n",
      "3     NTRK3 AST-M1 0.4000449\n",
      "4      APOE AST-M1 0.4185815\n",
      "5     CALM1 AST-M1 0.4201640\n",
      "6    ATP1B2 AST-M1 0.4342709\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "    print(paste0('Obtain Hub Genes in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    # get hub genes\n",
    "    hub_df <- GetHubGenes(seurat_obj[[cell_type]], n_hubs = 10)\n",
    "\n",
    "    print(head(hub_df))\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "27f2d80d",
   "metadata": {},
   "source": [
    "## **Compute hub gene signature scores**\n",
    "\n",
    "Gene scoring analysis is a popular method in single-cell transcriptomics for computing a score for the overall signature of a set of genes.\n",
    "\n",
    "hdWGCNA offers the `ModuleExprScore` function, which computes gene scores for a specified number of genes in each module, employing either the Seurat or UCell algorithm. This gene scoring method serves as an alternative to summarizing module expression, differing from the approach of computing module eigengenes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "e1f93a8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Compute hub gene scores in hdWGCNA Experiment for EXCITATORY\"\n",
      "Selecting by kME_EXC-M1\n",
      "Selecting by kME_EXC-M2\n",
      "[1] \"Compute hub gene scores in hdWGCNA Experiment for INHIBITORY\"\n",
      "Selecting by kME_INH-M1\n",
      "Selecting by kME_INH-M2\n",
      "Selecting by kME_INH-M3\n",
      "Selecting by kME_INH-M4\n",
      "Selecting by kME_INH-M5\n",
      "[1] \"Compute hub gene scores in hdWGCNA Experiment for ASTROCYTE\"\n",
      "Selecting by kME_AST-M1\n",
      "Selecting by kME_AST-M2\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Warning: Invalid name supplied, making object name syntactically valid. New object name is EXC.M1_UCellEXC.M2_UCell; see ?make.names for more details on syntax validity\n",
       "Warning: Invalid name supplied, making object name syntactically valid. New object name is INH.M1_UCellINH.M2_UCellINH.M3_UCellINH.M4_UCellINH.M5_UCell; see ?make.names for more details on syntax validity\n",
       "Warning: Invalid name supplied, making object name syntactically valid. New object name is AST.M1_UCellAST.M2_UCell; see ?make.names for more details on syntax validity\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R \n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "  print(paste0('Compute hub gene scores in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "    \n",
    "  # compute gene scoring for the top 25 hub genes by kME for each module\n",
    "  # with Seurat method\n",
    "  # seurat_obj[[cell_type]] <- ModuleExprScore(\n",
    "  #   seurat_obj[[cell_type]],\n",
    "  #   n_genes = 25,\n",
    "  #   method='Seurat'\n",
    "  # )\n",
    "\n",
    "  # compute gene scoring for the top 25 hub genes by kME for each module\n",
    "  # with UCell method\n",
    "  library(UCell)\n",
    "  seurat_obj[[cell_type]] <- ModuleExprScore(\n",
    "    seurat_obj[[cell_type]],\n",
    "    n_genes = 25,\n",
    "    method='UCell'\n",
    "  )\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3bc33b36",
   "metadata": {},
   "source": [
    "# **Save Seurat Object**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c0739db5",
   "metadata": {},
   "source": [
    "This wraps up the critical analysis steps for hdWGCNA, let's save our output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "431d9ac4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Saving hdWGCNA object in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"Saving hdWGCNA object in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"Saving hdWGCNA object in hdWGCNA Experiment for ASTROCYTE\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj\n",
    "\n",
    "dat_dir = paste0(\"../results/hdWGCNA/SeuratObject/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(dat_dir)) {\n",
    "  dir.create(dat_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "    print(paste0('Saving hdWGCNA object in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    saveRDS(seurat_obj[[cell_type]], file=paste0(dat_dir, cell_type, '_hdWGCNA_object.rds'))\n",
    "\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "2f414a000012e1980f5fd1119ed6b9bf8cc73b51e37b52eb52a16ff857effd1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
