{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62042f97-a76e-4b34-805f-3f1ac81e7288",
   "metadata": {},
   "outputs": [],
   "source": [
    "import rpy2\n",
    "import logging\n",
    "import warnings\n",
    "import anndata2ri\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from tabnanny import verbose\n",
    "import matplotlib.pyplot as plt\n",
    "from gsva_prep import prep_gsva\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "from matplotlib.pyplot import rcParams"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c36a2bac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sys_dpi(width, height, diag):\n",
    "    '''\n",
    "    obtain dpi of system\n",
    "    \n",
    "    w: width in pixels (if unsure, go vist `whatismyscreenresolution.net`)\n",
    "    h: height in pixels\n",
    "    d: diagonal in inches\n",
    "    '''\n",
    "    w_inches = (diag**2/ (1 + height**2/width**2))**0.5\n",
    "    return round(width/w_inches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "447b6d1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "rcParams['figure.dpi'] = get_sys_dpi(1512, 982, 14.125)\n",
    "#rcParams['figure.figsize']=(4,4) #rescale figures\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c6ce1b6b",
   "metadata": {},
   "source": [
    "# **hdWGCNA Enrichment Analysis**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7503bb6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "suppressPackageStartupMessages({\n",
    "    library(WGCNA)\n",
    "    library(Matrix)\n",
    "    library(viridis)\n",
    "    library(harmony)\n",
    "    library(ggpubr)\n",
    "    library(tictoc)\n",
    "    library(RColorBrewer)\n",
    "    library(Hmisc)\n",
    "    library(corrplot)\n",
    "    library(grid)\n",
    "    library(gridExtra)\n",
    "    library(igraph)\n",
    "    library(ggrepel)\n",
    "    library(conflicted)\n",
    "    library(readxl)\n",
    "\n",
    "    # single-cell analysis package\n",
    "    library(Seurat)\n",
    "\n",
    "    # plotting and data science packages\n",
    "    library(tidyverse)\n",
    "    library(cowplot)\n",
    "    library(patchwork)\n",
    "\n",
    "    # co-expression network analysis packages:\n",
    "    library(WGCNA)\n",
    "    library(hdWGCNA)\n",
    "\n",
    "    # gene enrichment packages\n",
    "    library(enrichR)\n",
    "    library(GeneOverlap)\n",
    "\n",
    "\n",
    "    library(GSEABase)\n",
    "    library(GSVA) \n",
    "# needs to be run every time you start R and want to use %>%\n",
    "})\n",
    "\n",
    "# using the cowplot theme for ggplot\n",
    "theme_set(theme_cowplot())\n",
    "\n",
    "# set random seed for reproducibility\n",
    "set.seed(12345)\n",
    "\n",
    "# optionally enable multithreading\n",
    "# enableWGCNAThreads(nThreads = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8504101",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_prefix = 'leng_sfg'\n",
    "map_meta = True\n",
    "deg_method =  'DESeq2-Wald'\n",
    "test_names = ['late_vs_early', 'early_vs_no', 'late_vs_no', 'ad_vs_no']\n",
    "filter_genes = \"TRUE\"\n",
    "subject_ids_for_study = {'leng_sfg': 'PatientID',\n",
    "                        'leng_etc': 'PatientID',\n",
    "                        'seaad_mtg': 'individualID'}\n",
    "\n",
    "subject_id = subject_ids_for_study[save_prefix]     # for leng this is `PatientID` for mathys is 'Subject', and allen is 'individualID'\n",
    "gene_celltype_threshold = 0.10          # determines number of cells the gene must be expressed in \n",
    "covariates = ['None']                   # list of covariates to be accounted for in regression.\n",
    "gene_selection = 'custom'               # specifies the gene selection method when setting up seurat object for WGCNA. The th\n",
    "celltypes = [\"Excitatory\", \"Inhibitory\", \"Astrocyte\", \"Microglia\", \"Oligodendrocyte\", \"OPC\"]\n",
    "metadata = f'../data/raw/{save_prefix}/{save_prefix}_metadata.csv' \n",
    "meta = pd.read_csv(metadata, encoding_errors='ignore')\n",
    "\n",
    "gene_selection = 'custom'   # specifies the gene selection method when setting up seurat object for WGCNA. The th\n",
    "                            # Posible values are \"custom\", \"fraction\", \"variable\"\n",
    "                            # If custom, a list of genes must be passed.\n",
    "\n",
    "\n",
    "gene_set_select = 'diff_exp'      # If gene_selection = 'custom'. This specifies how to obtain the list of\n",
    "                                  # genes to pass into `SetupForWGCNA`. # The posible values are \"diff_exp\", \"overlap\", \"all\""
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "8a3955a2",
   "metadata": {},
   "source": [
    "## **Load Seurat object for Enrichment and Differential Module Eigen Gene Analysis**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "839f795a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i subject_id -i gene_celltype_threshold -i celltypes -i test_names -i save_prefix -i gene_selection -i gene_set_select\n",
    "\n",
    "# Function to handle failure in data loading\n",
    "handle_loading_failure <- function(cell_type, seurat_obj) {\n",
    "    cat(paste0('Could not load data for ', toupper(cell_type), '\\n'))\n",
    "    cat(paste0(toupper(cell_type), ' dropped from experiment\\n'))\n",
    "  \n",
    "    # Remove the failed cell_type from the list\n",
    "    seurat_obj <- seurat_obj[names(seurat_obj) != cell_type]\n",
    "  \n",
    "    # Update the modified seurat_obj in the global environment\n",
    "    assign(\"seurat_obj\", seurat_obj, envir = .GlobalEnv)\n",
    "}\n",
    "\n",
    "seurat_obj <- list()\n",
    "\n",
    "for (cell_type in celltypes) {\n",
    "  \n",
    "    cat(paste0('Loading data for hdWGCNA Experiment in ', toupper(cell_type), '\\n'))\n",
    "  \n",
    "    # Use tryCatch to load data and handle errors or messages\n",
    "    tryCatch({\n",
    "        # Attempt to read the RDS file into the list\n",
    "        seurat_obj[[cell_type]] <- readRDS(paste0(\"../results/hdWGCNA/SeuratObject/\", save_prefix, '/', cell_type, '_hdWGCNA_object.rds'))\n",
    "        print(seurat_obj[[cell_type]])\n",
    "        print('............')\n",
    "  \n",
    "    }, error = function(e){\n",
    "        # Error handling: Remove the cell_type from the list and print an error message\n",
    "        handle_loading_failure(cell_type, seurat_obj)\n",
    "\n",
    "    }, message = function(m){\n",
    "        handle_loading_failure(cell_type, seurat_obj)\n",
    "    })\n",
    "}\n",
    "\n",
    "cat(\"Loaded available data\\n\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b534ca69",
   "metadata": {},
   "source": [
    "# **Enrichment analysis**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4261d643",
   "metadata": {},
   "source": [
    "To gain insights into the biological processes associated with hdWGCNA modules, we perform enrichment tests using the R package enrichR, which accesses a broad spectrum of curated gene lists. Additionally, a gene set overlap analysis is conducted to compare the genes in hdWGCNA modules with marker genes identified using Seuratâ€™s FindAllMarkers function. This dual approach helps in understanding the functional relevance of the identified modules in the broader context of gene expression and biological pathways.\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a77a18cb",
   "metadata": {},
   "source": [
    "## **EnrichR**\n",
    "\n",
    "hdWGCNA includes the function RunEnrichr to compare the set of genes in each module with any of the gene lists hosted by Enrichr.\n",
    "\n",
    "The results of the enrichment tests are stored in the hdWGCNA experiment, so can be easily retrieved for downstream analysis or exporting to external applicaitons like Excel. In the following example, we perform the enrichment test with three Gene Ontology datbases:\n",
    "\n",
    "- `GO_Biological_Process_2021`\n",
    "- `GO_Cellular_Component_2021`\n",
    "- `GO_Molecular_Function_2021`\n",
    "\n",
    "We specifically focus on differentially expressed pathways, consistent with the construction of co-expression networks based on gene programs from these pathways.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f99c0a19",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# dbs <- c('GO_Biological_Process_2021')\n",
    "\n",
    "dbs <- c('GO_Biological_Process_2021', 'GO_Cellular_Component_2021','GO_Molecular_Function_2021')\n",
    "deps <- read.csv(paste0('../results/ad_vs_no/', save_prefix, '/Data/differentially_expressed_pathways.csv'))\n",
    "\n",
    "enrich_df <- list()\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Estimating Module Enrichments in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # enrichr databases to test\n",
    "\n",
    "  # perform enrichment tests\n",
    "  seurat_obj[[cell_type]] <- RunEnrichr(\n",
    "    seurat_obj[[cell_type]],\n",
    "    dbs=dbs, # character vector of enrichr databases to test\n",
    "    max_genes = 500 # number of genes per module to test\n",
    "  )\n",
    "  # retrieve the output table\n",
    "  enrich_df[[cell_type]] <- GetEnrichrTable(seurat_obj[[cell_type]]) %>% subset(P.value < 0.05)\n",
    "  \n",
    "  pathways <- subset(deps, P.Value < 0.05 & celltype == cell_type)$pathway\n",
    "  bio_proc <- subset(enrich_df[[cell_type]], Term %in% pathways)\n",
    "  combined_output <- rbind(subset(enrich_df[[cell_type]], db %in% c('GO_Cellular_Component_2021',\n",
    "                          'GO_Molecular_Function_2021')), \n",
    "                          bio_proc)\n",
    "                          \n",
    "  seurat_obj[[cell_type]] <- SetEnrichrTable(seurat_obj[[cell_type]], combined_output)\n",
    "  enrich_df[[cell_type]] <- GetEnrichrTable(seurat_obj[[cell_type]]) \n",
    "\n",
    "  \n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "f2494516",
   "metadata": {},
   "source": [
    "## **Visualize enrichments**\n",
    "\n",
    "### **EnrichrBarPlot**\n",
    "\n",
    "The enrichment results can be visualized using the `EnrichrBarPlot` function. This function provides a comprehensive summary of the enrichment analysis for each Enrichr database and module, generating a .pdf figure for each module. These figures contain bar plots that display the top N enriched terms. In the provided example, the function is set to plot the top 10 terms for each module, with the output directed to a folder named enrichr_plots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7d94ea7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/Enrichr/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Enrichr Plots for Modules in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # make GO term plots:\n",
    "  EnrichrBarPlot(\n",
    "    seurat_obj[[cell_type]],\n",
    "    outdir = paste0(fig_dir, cell_type, \"_enrichr_plots\"), # name of output directory\n",
    "    n_terms = 10, # number of enriched terms to show (sometimes more show if there are ties!!!)\n",
    "    plot_size = c(5,7), # width, height of the output .pdfs\n",
    "    logscale=TRUE # do you want to show the enrichment as a log scale?\n",
    "  )\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "678b520b",
   "metadata": {},
   "source": [
    "### **EnrichrDotPlot**\n",
    "\n",
    "hdWGCNA includes an additional visualization function for enrichment results, EnrichrDotPlot, which shows the top results for one Enrichr database in each module. In the following example, we plot the top term in the GO_Biological_Process_2021 database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcb62788",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "conflicts_prefer(dplyr::select)\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/Enrichr/\", save_prefix, '/DotPlot/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "    print(paste0('Enrichr Dot-Plots for Modules in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    # enrichr dotplot\n",
    "    p <- EnrichrDotPlot(\n",
    "    seurat_obj[[cell_type]],\n",
    "        mods='all',\n",
    "        database = c('GO_Biological_Process_2021'), \n",
    "        n_terms=3,\n",
    "        break_ties=TRUE,\n",
    "        )\n",
    "        \n",
    "    pdf(paste0(fig_dir, cell_type, \"_enrichr_dotplots.pdf\"), width=10, height=10, useDingbats=FALSE)\n",
    "    print(p)\n",
    "    dev.off()\n",
    "\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3ccd6f93",
   "metadata": {},
   "source": [
    "## **Plot Only Select GO Terms**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "0f048699",
   "metadata": {},
   "source": [
    "Barplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13c4d314",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "\n",
    "pathways_to_annotate = read_excel('../data/pathway_databases/pathways_to_annotate.xlsx', sheet = 1)\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  fig_dir = paste0(\"../results/hdWGCNA/Enrichr/\", save_prefix, '/', cell_type, '_enrichr_plots/')\n",
    "\n",
    "  if (!dir.exists(fig_dir)) {\n",
    "    dir.create(fig_dir, recursive=TRUE)\n",
    "  }\n",
    "\n",
    "  modules <- GetModules(seurat_obj[[cell_type]])\n",
    "  mods <- levels(modules$module)\n",
    "  mods <- mods[mods!='grey']\n",
    "  module_colors <- modules %>% dplyr::select(c(module, color)) %>% distinct()\n",
    "  rownames(module_colors) <- module_colors$module\n",
    "  mod_colors <- module_colors[mods, 'color']\n",
    "  names(mod_colors) <- mods\n",
    "\n",
    "  # helper function to wrap text\n",
    "  wrapText <- function(x, len) {\n",
    "      sapply(x, function(y) paste(strwrap(y, len), collapse = \"\\n\"), USE.NAMES = FALSE)\n",
    "  }\n",
    "\n",
    "  \n",
    "  selected_terms <- subset(enrich_df[[cell_type]], Term %in% pathways_to_annotate$pathway)\n",
    "\n",
    "  # remove GO Term ID\n",
    "  selected_terms$Term <- str_replace(selected_terms$Term, \" \\\\s*\\\\([^\\\\)]+\\\\)\", \"\")\n",
    "\n",
    "  selected_terms <- selected_terms %>%\n",
    "    group_by(module) %>%\n",
    "    arrange(Combined.Score) %>%\n",
    "    slice_max(order_by=Combined.Score, n=10)\n",
    "\n",
    "  selected_terms$wrap <- wrapText(selected_terms$Term, 35)\n",
    "\n",
    "  selected_terms <- selected_terms %>% arrange(Combined.Score) \n",
    "  # selected_terms$wrap <- factor(selected_terms$wrap)\n",
    "  # selected_terms$wrap <- factor(selected_terms$wrap, \n",
    "  #                         levels = unique(selected_terms$wrap[order(selected_terms$Combined.Score)]))\n",
    "  selected_terms$wrap <- factor(selected_terms$wrap, \n",
    "                          levels = unique(selected_terms$wrap[order(selected_terms$module, \n",
    "                                                                    selected_terms$Combined.Score)]))\n",
    "\n",
    "\n",
    "  p <- selected_terms  %>%\n",
    "    ggplot(aes(x=log(Combined.Score), y=wrap, fill=module))+\n",
    "    geom_bar(stat='identity', position='identity', color='white') +\n",
    "    geom_text(aes(label=wrap), x=.1, color='black', size=3.5, hjust='left') +\n",
    "    scale_fill_manual(values=mod_colors) +\n",
    "    ylab('Term') + xlab('log(Enrichment)') +\n",
    "    scale_x_continuous(expand = c(0, 0), limits = c(0, NA)) +\n",
    "    theme(\n",
    "      panel.grid.major=element_blank(),\n",
    "      panel.grid.minor=element_blank(),\n",
    "      legend.title = element_blank(),\n",
    "      axis.ticks.y=element_blank(),\n",
    "      axis.text.y=element_blank(),\n",
    "      axis.line.y=element_blank(),\n",
    "      plot.title = element_text(hjust = 0.5)\n",
    "    )\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, '_selected_GO_terms.pdf'), width= 6, height=30, useDingbats=FALSE)\n",
    "  print(p + facet_wrap(~module, ncol=2, scales='free') + NoLegend())\n",
    "  dev.off()\n",
    "\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "375754cd",
   "metadata": {},
   "source": [
    "Dotplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58091cd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "conflicts_prefer(dplyr::rename)\n",
    "\n",
    "pathways_to_annotate = read_excel('../data/pathway_databases/pathways_to_annotate.xlsx', sheet = 1)\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/Enrichr/\", save_prefix, '/DotPlot/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "    \n",
    "  modules <- GetModules(seurat_obj[[cell_type]])\n",
    "  color_df <- modules %>% subset(module!='grey') %>%\n",
    "  select(c(module, color)) %>% distinct %>%\n",
    "  mutate(module=droplevels(module)) %>%\n",
    "  rename(c(group=module, colour=color))\n",
    "  mods <- levels(modules$module); mods <- mods[mods != 'grey']\n",
    "\n",
    "  color_df$group <- factor(as.character(color_df$group), levels=mods)\n",
    "\n",
    "  # helper function to wrap text\n",
    "  wrapText <- function(x, len) {\n",
    "    sapply(x, function(y) paste(strwrap(y, len), collapse = \"\\n\"), USE.NAMES = FALSE)\n",
    "  }\n",
    "\n",
    "\n",
    "  # subset selected terms\n",
    "  selected_terms <- subset(enrich_df[[cell_type]], Term %in% pathways_to_annotate$pathway)\n",
    "\n",
    "  selected_terms$group <- factor(\n",
    "  as.character(selected_terms$module),\n",
    "  levels = mods\n",
    "  )\n",
    "\n",
    "  # set max pval\n",
    "\n",
    "  quantile(-log(selected_terms$P.value), 0.95)\n",
    "  max_p <- 10\n",
    "\n",
    "  selected_terms$logp <- -log(selected_terms$P.value)\n",
    "  selected_terms$logp <- ifelse(selected_terms$logp > max_p, max_p, selected_terms$logp)\n",
    "\n",
    "  # remove GO Term ID\n",
    "  selected_terms$Term <- str_replace(selected_terms$Term, \" \\\\s*\\\\([^\\\\)]+\\\\)\", \"\")\n",
    "\n",
    "  # selected_terms <- selected_terms %>%\n",
    "  #                   arrange(group) %>%\n",
    "\n",
    "  selected_terms <- selected_terms %>%\n",
    "                    group_by(group) %>%\n",
    "                    arrange(group) %>%\n",
    "                    slice_max(n = 3, order_by = Combined.Score) %>%\n",
    "                    ungroup()\n",
    "  \n",
    "  # selected_terms <- selected_terms %>%\n",
    "  #                   arrange(group) %>%\n",
    "\n",
    "\n",
    "  selected_terms$wrap <- wrapText(selected_terms$Term, 35)\n",
    "\n",
    "  selected_terms$Term <- factor(\n",
    "  as.character(selected_terms$Term),\n",
    "  levels = rev(unique(as.character(selected_terms$Term)))\n",
    "  )\n",
    "\n",
    "  # GO Term dot plot\n",
    "\n",
    "  p <- selected_terms %>%\n",
    "  ggplot(aes(x = group, y = Term, color =logp, size=log(Combined.Score))) +\n",
    "  geom_point() +\n",
    "  scale_color_stepsn(colors=rev(magma(256))) +\n",
    "  RotatedAxis() + xlab('') + ylab('') +\n",
    "  theme(\n",
    "    axis.title.x = element_blank(),\n",
    "    axis.title.y = element_blank(),\n",
    "    panel.border = element_rect(size=1, color='black', fill=NA),\n",
    "    axis.line.x = element_blank(),\n",
    "    axis.line.y = element_blank(),\n",
    "    axis.text.x = element_blank(),\n",
    "    axis.ticks.x = element_blank(),\n",
    "    plot.margin = margin(0,0,0,0),\n",
    "    panel.grid = element_line(size=0.25, color='lightgrey')\n",
    "  )\n",
    "\n",
    "\n",
    "  # make the colorbar as its own heatmap\n",
    "  color_df$var <- 1\n",
    "  cp <- color_df$colour; names(cp) <- color_df$group\n",
    "  colorbar <- color_df %>%\n",
    "  ggplot(aes(x=group, y=var, fill=group)) +\n",
    "  geom_tile() +\n",
    "  scale_fill_manual(values=cp) +\n",
    "  coord_equal() +\n",
    "  NoLegend() + RotatedAxis() +\n",
    "  theme(\n",
    "    plot.title=element_blank(),\n",
    "    axis.line=element_blank(),\n",
    "    axis.ticks.y =element_blank(),\n",
    "    axis.text.y = element_blank(),\n",
    "    axis.title = element_blank(),\n",
    "    plot.margin=margin(0,0,0,0),\n",
    "  )\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, '_selected_GO_terms.pdf'), width=15, height=8)\n",
    "  print(p / colorbar) #+ plot_layout(heights=c(20,1))\n",
    "  dev.off()\n",
    "\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "d5bae359",
   "metadata": {},
   "source": [
    "# **Differential module eigengene (DME) analysis**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c5b93666",
   "metadata": {},
   "source": [
    "We next conduct differential module eigengene (DME) analysis to identify modules that are up- or down-regulated in AD versus control.\n",
    "\n",
    "We emply the FindDMEs function, a specialized adaptation of Seurat's FindMarkers function, typically employing the Mann-Whitney U test to compare the two groups, although other tests can be specified using the test.use parameter. FindDMEs requires a list of barcodes for each group. In this context, the analysis focuses on cells from the EX cluster, aligning with the group targeted in the network analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e343be3",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "group_col <- 'cell_type'\n",
    "groups <- celltypes\n",
    "DMEs <- list()\n",
    "\n",
    "dat_dir = paste0(\"../results/hdWGCNA/DMEs/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(dat_dir)) {\n",
    "  dir.create(dat_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Differential Module Eigengene Analysis in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "\n",
    "  # get cell barcodes for this pseudotime bin in control & AD\n",
    "  g1 <- seurat_obj[[cell_type]]@meta.data[seurat_obj[[cell_type]]$pathology.group %in% c('early', 'late'),] %>% rownames\n",
    "  g2 <- seurat_obj[[cell_type]]@meta.data[seurat_obj[[cell_type]]$pathology.group == 'no',] %>% rownames\n",
    "\n",
    "  print(length(g1))\n",
    "\n",
    "  DMEs[[cell_type]] <- FindDMEs(\n",
    "          seurat_obj[[cell_type]],\n",
    "          barcodes1 = g1,\n",
    "          barcodes2 = g2,\n",
    "          test.use='wilcox',\n",
    "          wgcna_name=toupper(substr(cell_type, 1, 3)), # the name of the hdWGCNA experiment,,\n",
    "          harmonized=TRUE\n",
    "      )\n",
    "  \n",
    "  DMEs[[cell_type]]$group <- cell_type\n",
    "\n",
    "  # fix infs:\n",
    "  DMEs[[cell_type]]$avg_log2FC <- ifelse(abs(DMEs[[cell_type]]$avg_log2FC) == Inf, 0, DMEs[[cell_type]]$avg_log2FC)\n",
    "\n",
    "  write.csv(DMEs[[cell_type]], row.names=FALSE, quote=FALSE, file=paste0(dat_dir, cell_type, '_AD_DMEs.csv'))\n",
    "}\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4ed4d48c",
   "metadata": {},
   "source": [
    "We can now visualize the results using the hdWGNCA functions `PlotDMEsLollipop` or `PlotDMEsVolcano`. First we make a lollipop plot to visualize the DME results.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d11980c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/DMEs/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Lollipop for DMEs in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, '_dmes_lollipop.pdf'), width=10, height=10)\n",
    "  p <- PlotDMEsLollipop(\n",
    "    seurat_obj[[cell_type]], \n",
    "    DMEs[[cell_type]], \n",
    "    wgcna_name = toupper(substr(cell_type, 1, 3)), # the name of the hdWGCNA experiment,, \n",
    "    pvalue = \"p_val_adj\"\n",
    "  )\n",
    "  print(p)\n",
    "  dev.off()\n",
    "\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3bc33b36",
   "metadata": {},
   "source": [
    "# **Save Seurat Object**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "431d9ac4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "dat_dir = paste0(\"../results/hdWGCNA/SeuratObject/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(dat_dir)) {\n",
    "  dir.create(dat_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "    print(paste0('Saving hdWGCNA object in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    saveRDS(seurat_obj[[cell_type]], file=paste0(dat_dir, cell_type, '_hdWGCNA_object.rds'))\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "2f414a000012e1980f5fd1119ed6b9bf8cc73b51e37b52eb52a16ff857effd1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
