{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "62042f97-a76e-4b34-805f-3f1ac81e7288",
   "metadata": {},
   "outputs": [],
   "source": [
    "import rpy2\n",
    "import scipy\n",
    "import logging\n",
    "import warnings\n",
    "import anndata2ri\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import seaborn as sb\n",
    "import decoupler as dc\n",
    "import scrublet as scr\n",
    "import decoupler as dc\n",
    "from scipy import sparse\n",
    "from anndata import AnnData\n",
    "from tabnanny import verbose\n",
    "import matplotlib.pyplot as plt\n",
    "from gsva_prep import prep_gsva\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "from typing import Optional, Union\n",
    "from matplotlib.pyplot import rcParams\n",
    "from functions import pathway_analyses\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from sklearn.model_selection import train_test_split\n",
    "from pytorch_lightning.loggers import TensorBoardLogger\n",
    "from rpy2.robjects.conversion import localconverter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c36a2bac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sys_dpi(width, height, diag):\n",
    "    '''\n",
    "    obtain dpi of system\n",
    "    \n",
    "    w: width in pixels (if unsure, go vist `whatismyscreenresolution.net`)\n",
    "    h: height in pixels\n",
    "    d: diagonal in inches\n",
    "    '''\n",
    "    w_inches = (diag**2/ (1 + height**2/width**2))**0.5\n",
    "    return round(width/w_inches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "447b6d1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.8.0\n",
      "scanpy      1.9.3\n",
      "-----\n",
      "PIL                         9.5.0\n",
      "absl                        NA\n",
      "anndata2ri                  1.1\n",
      "appnope                     0.1.3\n",
      "asttokens                   NA\n",
      "attr                        23.1.0\n",
      "backcall                    0.2.0\n",
      "cffi                        1.15.1\n",
      "comm                        0.1.3\n",
      "cycler                      0.10.0\n",
      "cython_runtime              NA\n",
      "dateutil                    2.8.2\n",
      "debugpy                     1.6.7\n",
      "decorator                   5.1.1\n",
      "decoupler                   1.4.0\n",
      "deprecate                   0.3.2\n",
      "dot_parser                  NA\n",
      "executing                   1.2.0\n",
      "fsspec                      2023.6.0\n",
      "functions                   NA\n",
      "google                      NA\n",
      "gsva_prep                   NA\n",
      "h5py                        3.9.0\n",
      "igraph                      0.10.4\n",
      "ipykernel                   6.23.2\n",
      "ipywidgets                  8.0.6\n",
      "jedi                        0.18.2\n",
      "jinja2                      3.1.2\n",
      "joblib                      1.2.0\n",
      "kiwisolver                  1.4.4\n",
      "leidenalg                   0.9.1\n",
      "llvmlite                    0.40.1rc1\n",
      "louvain                     0.8.0\n",
      "markupsafe                  2.1.3\n",
      "matplotlib                  3.7.1\n",
      "mpl_toolkits                NA\n",
      "mpmath                      1.3.0\n",
      "natsort                     8.4.0\n",
      "numba                       0.57.0\n",
      "numpy                       1.24.3\n",
      "opt_einsum                  v3.3.0\n",
      "packaging                   23.1\n",
      "pandas                      2.0.2\n",
      "parso                       0.8.3\n",
      "patsy                       0.5.3\n",
      "pexpect                     4.8.0\n",
      "pickleshare                 0.7.5\n",
      "pkg_resources               NA\n",
      "platformdirs                3.7.0\n",
      "prompt_toolkit              3.0.38\n",
      "psutil                      5.9.5\n",
      "ptyprocess                  0.7.0\n",
      "pure_eval                   0.2.2\n",
      "pycparser                   2.21\n",
      "pydev_ipython               NA\n",
      "pydevconsole                NA\n",
      "pydevd                      2.9.5\n",
      "pydevd_file_utils           NA\n",
      "pydevd_plugins              NA\n",
      "pydevd_tracing              NA\n",
      "pydot                       1.4.2\n",
      "pygments                    2.15.1\n",
      "pyparsing                   3.1.0\n",
      "pytorch_lightning           1.7.7\n",
      "pytz                        2023.3\n",
      "pytz_deprecation_shim       NA\n",
      "rich                        NA\n",
      "rpy2                        3.5.12\n",
      "scipy                       1.10.1\n",
      "scrublet                    NA\n",
      "seaborn                     0.12.2\n",
      "session_info                1.0.0\n",
      "setuptools                  68.0.0\n",
      "six                         1.16.0\n",
      "sklearn                     1.2.2\n",
      "sparse                      0.14.0\n",
      "stack_data                  0.6.2\n",
      "statsmodels                 0.14.0\n",
      "sympy                       1.12\n",
      "tensorboard                 2.13.0\n",
      "texttable                   1.6.7\n",
      "threadpoolctl               3.1.0\n",
      "torch                       2.0.1\n",
      "torchmetrics                0.11.4\n",
      "tornado                     6.3.2\n",
      "tqdm                        4.65.0\n",
      "traitlets                   5.9.0\n",
      "typing_extensions           NA\n",
      "tzlocal                     NA\n",
      "wcwidth                     0.2.6\n",
      "yaml                        6.0\n",
      "zmq                         25.1.0\n",
      "zoneinfo                    NA\n",
      "-----\n",
      "IPython             8.14.0\n",
      "jupyter_client      8.2.0\n",
      "jupyter_core        5.3.1\n",
      "-----\n",
      "Python 3.10.6 (main, Oct 24 2022, 11:04:34) [Clang 12.0.0 ]\n",
      "macOS-10.16-x86_64-i386-64bit\n",
      "-----\n",
      "Session information updated at 2023-07-07 15:10\n"
     ]
    }
   ],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# # Automatically convert rpy2 outputs to pandas dataframes\n",
    "# pandas2ri.activate()\n",
    "# anndata2ri.activate()\n",
    "# %load_ext rpy2.ipython\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "rcParams['figure.dpi'] = get_sys_dpi(1512, 982, 14.125)\n",
    "#rcParams['figure.figsize']=(4,4) #rescale figures\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "10be21f0-000d-48d7-9e97-4051ec9da0bd",
   "metadata": {},
   "source": [
    "# **hdWGCNA Pseudotime**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8b87d243",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "data": {
      "text/plain": [
       "In addition: Warning messages:\n",
       "1: package ‘IRanges’ was built under R version 4.3.1 \n",
       "2: package ‘GenomeInfoDb’ was built under R version 4.3.1 \n",
       "3: package ‘GSVA’ was built under R version 4.3.1 \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "suppressPackageStartupMessages({\n",
    "    library(WGCNA)\n",
    "    library(Matrix)\n",
    "    library(viridis)\n",
    "    library(harmony)\n",
    "    library(ggpubr)\n",
    "    library(tictoc)\n",
    "    library(RColorBrewer)\n",
    "    library(Hmisc)\n",
    "    library(corrplot)\n",
    "    library(grid)\n",
    "    library(gridExtra)\n",
    "    library(igraph)\n",
    "    library(ggrepel)\n",
    "    library(conflicted)\n",
    "    library(readxl)\n",
    "\n",
    "\n",
    "    # single-cell analysis package\n",
    "    library(Seurat)\n",
    "    library(monocle3)\n",
    "    library(SeuratWrappers)\n",
    "\n",
    "    # plotting and data science packages\n",
    "    library(tidyverse)\n",
    "    library(cowplot)\n",
    "    library(patchwork)\n",
    "\n",
    "    # co-expression network analysis packages:\n",
    "    library(WGCNA)\n",
    "    library(hdWGCNA)\n",
    "\n",
    "    # gene enrichment packages\n",
    "    library(enrichR)\n",
    "    library(GeneOverlap)\n",
    "\n",
    "\n",
    "    library(GSEABase)\n",
    "    library(GSVA) \n",
    "# needs to be run every time you start R and want to use %>%\n",
    "})\n",
    "\n",
    "# using the cowplot theme for ggplot\n",
    "theme_set(theme_cowplot())\n",
    "\n",
    "# set random seed for reproducibility\n",
    "set.seed(12345)\n",
    "\n",
    "# optionally enable multithreading\n",
    "# enableWGCNAThreads(nThreads = 4)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "af425574",
   "metadata": {},
   "source": [
    "## **Prepare data**\n",
    "\n",
    "Now, we load the preprocessed and annotated data for downstream analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "37a91025",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_prefix = 'allen_mtg'\n",
    "map_meta = True\n",
    "deg_method =  'DESeq2-Wald'\n",
    "test_names = ['late_vs_early', 'early_vs_no', 'late_vs_no', 'ad_vs_no']\n",
    "filter_genes = \"TRUE\"\n",
    "subject_ids_for_study = {'allen_mtg': 'individualID',\n",
    "                        'leng_sfg': 'PatientID',\n",
    "                        'leng_etc': 'PatientID'}\n",
    "\n",
    "subject_id = subject_ids_for_study[save_prefix]     # for leng this is `PatientID` for mathys is 'Subject', and allen is 'individualID'\n",
    "gene_celltype_threshold = 0.10          # determines number of cells the gene must be expressed in \n",
    "covariates = ['None']                   # list of covariates to be accounted for in regression.\n",
    "gene_selection = 'custom'               # specifies the gene selection method when setting up seurat object for WGCNA. The th\n",
    "celltypes = [\"OPC\"]  #[\"Excitatory\", \"Inhibitory\", \"Astrocyte\", \"Microglia\", \"Oligodendrocyte\", \"OPC\"]\n",
    "metadata = f'../data/raw/{save_prefix}/{save_prefix}_metadata.csv' \n",
    "meta = pd.read_csv(metadata, encoding_errors='ignore')\n",
    "\n",
    "gene_selection = 'custom'   # specifies the gene selection method when setting up seurat object for WGCNA. The th\n",
    "                            # Posible values are \"custom\", \"fraction\", \"variable\"\n",
    "                            # If custom, a list of genes must be passed.\n",
    "\n",
    "gene_set_select = 'diff_exp'      # If gene_selection = 'custom'. This specifies how to obtain the list of\n",
    "                                  # genes to pass into `SetupForWGCNA`. # The posible values are \"diff_exp\", \"overlap\", \"all\""
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "0b826349",
   "metadata": {},
   "source": [
    "## **Load Seurat object for Network Visualizations**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "031b3dd6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Loading data for hdWGCNA Experiment in OPC\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "An object of class Seurat \n",
      "65852 features across 2838 samples within 1 assay \n",
      "Active assay: originalexp (65852 features, 2000 variable features)\n",
      " 3 dimensional reductions calculated: pca, harmony, umap\n",
      "[1] \"loaded data\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i subject_id -i gene_celltype_threshold -i celltypes -i test_names -i save_prefix -i gene_selection -i gene_set_select -o seurat_obj\n",
    "\n",
    "seurat_obj <- list()\n",
    "\n",
    "for (cell_type in celltypes) {\n",
    "    \n",
    "    print(paste0('Loading data for hdWGCNA Experiment in ', toupper(cell_type)))\n",
    "    tryCatch({\n",
    "        seurat_obj[[cell_type]] <- readRDS(paste0(\"../results/hdWGCNA/SeuratObject/\", save_prefix, '/', cell_type, '_hdWGCNA_object.rds'))\n",
    "        print(seurat_obj[[cell_type]])\n",
    "    \n",
    "    }, error = function(e){\n",
    "        NULL\n",
    "    }, message = function(m){\n",
    "        print(paste0('Could not load data for ', toupper(cell_type)))\n",
    "    })\n",
    "\n",
    "}\n",
    "\n",
    "print('loaded data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8c7e8bda",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9b2ac7d7",
   "metadata": {},
   "source": [
    "# **Monocle3 pseudotime analysis**\n",
    "\n",
    "Here we perform trajectory analysis of the data. \n",
    "\n",
    "Although we've identified several modules associated with the pathological alterations, it is unclear if these pathways play a role in the progression, rather than just a single state, of the disease. This way, we can capture the pathological progression that is  insufficiently captured in singlular cell states. \n",
    "\n",
    "Current single cell lineage tracing algorithms now allow us to take a single tissue sample (with cells presumably at varying stages of disease progression/degradation within the single sample) and recreate a “pseudo-time” scale whereby we infer which cells are early on in the disease process versus those much closer to a final degenerative step based on the evolution of their gene expression profiles. \n",
    "\n",
    "Here, we apply one such method to the different cells from normal and AD cohorts to better elucidate genes potentially relevant to progression of AD.\n",
    "\n",
    "Since the data contains different disease stages, we can investigate the pathological trajectories. This analysis is centred around the concept of 'pseudotime'. In pseudotime analysis a latent variable is inferred based on which the cells are ordered. This latent variable is supposed to measure the differentiation state along a trajectory.\n",
    "\n",
    "Pseudotime analysis is complicated when there are multiple trajectories in the data. In this case, the trajectory structure in the data must first be found before pseudotime can be inferred along each trajectory. The analysis is then called 'trajectory inference'.\n",
    "\n",
    "Once the pseudotime variable is inferred, we can test for modules that vary continuously along pseudotime. These modules are seen as being associated with the trajectory, and may play a regulatory role in the potential pathological trajectory that the analysis found.\n",
    "\n",
    "Here, we measure the trajectory from `no-pathology` group to `late-pathology` group in a cell-type-specific manner, just as described in the [**Morabito. et. al. 2021**](https://www.nature.com/articles/s41588-021-00894-z). We also investigate which modules vary along pseudotime.\n",
    "\n",
    "Based on a recent comparison of pseudotime methods [**Saelens et al., 2018**](https://doi.org/10.1038/s41587-019-0071-9), we have selected the top performing 'Monocle3'.\n",
    "\n",
    "- `'Monocle3'` performed best for complex tree structures and was used in the [**Morabito. et. al. 2021**](https://www.nature.com/articles/s41588-021-00894-z) paper.\n",
    "\n",
    "As the complexity of trajectories are generally not known, it is adviseable to compare trajectory inference outputs.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8537d71c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Running UMAP on OPC\"\n",
      "[1] \"OPC pseudotime analysis with monocle3\"\n",
      "  |======================================================================| 100%\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric\n",
       "To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'\n",
       "This message will be shown once per session\n",
       "15:10:49 UMAP embedding parameters a = 1.577 b = 0.8951\n",
       "15:10:49 Read 2838 rows and found 30 numeric columns\n",
       "15:10:49 Using Annoy for neighbor search, n_neighbors = 15\n",
       "15:10:49 Building Annoy index with metric = cosine, n_trees = 50\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "15:10:49 Writing NN index file to temp file /var/folders/70/42xwxjh94x5f164djlw1hwpc0000gv/T//RtmpShB2ET/file2d8f145f07ad\n",
       "15:10:49 Searching Annoy index using 1 thread, search_k = 1500\n",
       "15:10:49 Annoy recall = 100%\n",
       "15:10:50 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 15\n",
       "15:10:52 Initializing from normalized Laplacian + noise (using irlba)\n",
       "15:10:52 Commencing optimization for 500 epochs, with 65970 positive edges\n",
       "Using method 'umap'\n",
       "0%   10   20   30   40   50   60   70   80   90   100%\n",
       "[----|----|----|----|----|----|----|----|----|----|\n",
       "**************************************************|\n",
       "15:10:55 Optimization finished\n",
       "In addition: Warning messages:\n",
       "1: In cluster_cells_make_graph(data = data, weight = weight, cell_names = cell_names,  :\n",
       "  The nearest neighbors includes the point itself, k must be smaller than\n",
       "the total number of points - 1 (all other points) - 1 (itself)! Total number of points is 22\n",
       "2: In cluster_cells_make_graph(data = data, weight = weight, cell_names = cell_names,  :\n",
       "  The nearest neighbors includes the point itself, k must be smaller than\n",
       "the total number of points - 1 (all other points) - 1 (itself)! Total number of points is 26\n",
       "3: In min(data_df$weight[data_df$weight > 0]) :\n",
       "  no non-missing arguments to min; returning Inf\n",
       "4: ggrepel: 26 unlabeled data points (too many overlaps). Consider increasing max.overlaps \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj -o cds\n",
    "\n",
    "fig_dir = paste0('../results/hdWGCNA/Pseudotime/Graph/', save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "cds <- list()\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Running UMAP on ', toupper(cell_type)))\n",
    "\n",
    "  seurat_obj[[cell_type]] <- RunUMAP(seurat_obj[[cell_type]], \n",
    "                                      reduction='harmony',\n",
    "                                      n.neighbors=15,  \n",
    "                                      dims=1:30,\n",
    "                                      min.dist=0.1)\n",
    "\n",
    "  print(paste0(toupper(cell_type), ' pseudotime analysis with monocle3'))\n",
    "\n",
    "  # convert the seurat object to CDS\n",
    "  cds[[cell_type]] <- as.cell_data_set(seurat_obj[[cell_type]])\n",
    "\n",
    "  # run the monocle clustering\n",
    "  cds[[cell_type]] <- cluster_cells(cds=cds[[cell_type]], reduction_method='UMAP', \n",
    "                                    cluster_method = c(\"louvain\"))\n",
    "\n",
    "  # learn graph for pseudotime\n",
    "  cds[[cell_type]] <- learn_graph(cds[[cell_type]])\n",
    "\n",
    "  # plot the pseudotime graph:\n",
    "  p1 <- plot_cells(\n",
    "    cds = cds[[cell_type]],\n",
    "    color_cells_by = \"pathology.group\",\n",
    "    show_trajectory_graph = TRUE,\n",
    "    label_principal_points = TRUE\n",
    "  ) \n",
    "\n",
    "  # plot the UMAP partitions from the clustering algorithm\n",
    "  p2 <- plot_cells(\n",
    "    cds = cds[[cell_type]],\n",
    "    color_cells_by = \"partition\",\n",
    "    show_trajectory_graph = FALSE\n",
    "  )\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, '_umap_monocle.pdf'),  width=8, height=4)\n",
    "  print(p1 + p2)\n",
    "  dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "db743ad8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )\n",
    "\n",
    "cds =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: cds[cell_type]\n",
    "                     \n",
    "                            for cell_type in cds.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "38555106",
   "metadata": {},
   "source": [
    "## **Learn Graph**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "311f315f",
   "metadata": {},
   "source": [
    "The `Monocle3` `learn_graph` function builds a principal graph in a dimensionally reduced space of the dataset. \n",
    "\n",
    "Here we can see the principal graph overlaid on the UMAP. The principal graph will serve as the basis of the pseudotime trajectories. \n",
    "\n",
    "Next, we have to select the starting node, or principal node, of the graph as the origin of the pseudotime. \n",
    "\n",
    "We choose the principal node by manually inspecting the pseudotime graph for all celltypes pdf file. We choose a node that largely overlaps with cells in the -`no-pathology`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fbff6c48",
   "metadata": {},
   "outputs": [],
   "source": [
    "principal_node = {}\n",
    "\n",
    "for cell_type in celltypes:\n",
    "\n",
    "    principal_node[cell_type] = input(f'please choose principal node for {cell_type}: ').split(',')\n",
    "    principal_node[cell_type] = [node.capitalize() for node in principal_node[cell_type]]\n",
    "\n",
    "principal_node =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: robjects.vectors.StrVector(principal_node[cell_type])\n",
    "                     \n",
    "                            for cell_type in celltypes\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f4c52efb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "$OPC\n",
      "[1] \"Y_239\"\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%R -i principal_node \n",
    "\n",
    "print(principal_node)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f4b23841",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Setting principal node on OPC\"\n",
      "[1] \"Ordering cells ...... \"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Cells aren't colored in a way that allows them to be grouped.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj -i cds\n",
    "\n",
    "fig_dir = paste0('../results/hdWGCNA/Pseudotime/Trajectory/', save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  # get principal node & order cells\n",
    "  print(paste0('Setting principal node on ', toupper(cell_type)))\n",
    "  print(paste0('Ordering cells ...... '))\n",
    "\n",
    "  cds[[cell_type]] <- order_cells(cds[[cell_type]], root_pr_nodes = as.vector(principal_node[[cell_type]]))\n",
    "\n",
    "  # plot psedutime\n",
    "  p <- plot_cells(\n",
    "    cds = cds[[cell_type]],\n",
    "    color_cells_by = \"pseudotime\",\n",
    "    show_trajectory_graph = TRUE,\n",
    "    label_roots = TRUE,\n",
    "    label_leaves = FALSE,\n",
    "    label_branch_points = FALSE,\n",
    "  ) + umap_theme()\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, '_umap_pseudotime.pdf'),  width=7, height=5, useDingbats=FALSE)\n",
    "  print(p)\n",
    "  dev.off()\n",
    "\n",
    "  # # add pseudotime to seurat object:\n",
    "  seurat_obj[[cell_type]]$pseudotime <- pseudotime(cds[[cell_type]])\n",
    "\n",
    "  seurat_obj[[cell_type]]$UMAP1 <- seurat_obj[[cell_type]]@reductions$umap@cell.embeddings[,1]\n",
    "  seurat_obj[[cell_type]]$UMAP2 <- seurat_obj[[cell_type]]@reductions$umap@cell.embeddings[,2]\n",
    "\n",
    "  p1 <- seurat_obj[[cell_type]]@meta.data %>%\n",
    "    ggplot(aes(x=UMAP1, y=UMAP2, color=pseudotime)) +\n",
    "    ggrastr::rasterise(geom_point(size=1), dpi=500, scale=0.75) +\n",
    "    coord_equal() +\n",
    "    scale_color_gradientn(colors=plasma(256), na.value='grey') +\n",
    "    umap_theme()\n",
    "\n",
    "  p2 <- seurat_obj[[cell_type]]@meta.data %>%\n",
    "    ggplot(aes(x=UMAP1, y=UMAP2, color=pathology.group)) +\n",
    "    ggrastr::rasterise(geom_point(size=1), dpi=500, scale=0.75) +\n",
    "    coord_equal() +\n",
    "    #scale_color_gradientn(colors=plasma(256), na.value='grey') +\n",
    "    umap_theme()    \n",
    "\n",
    "  # assemble with patchwork\n",
    "  pdf(paste0(fig_dir, 'pathology_', cell_type, '_umap_pseudotime.pdf'),  width=8, height=4, useDingbats=FALSE)\n",
    "  print(p1 + p2)\n",
    "  dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f416d390",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e00f5d04",
   "metadata": {},
   "source": [
    "## **Module dynamics**\n",
    "\n",
    "Here we use the hdWGCNA function `PlotModuleTrajectory` to visualize how the module eigengenes change throughout the pseudotime trajectories for each co-expression module. \n",
    "\n",
    "This function requires the name of the column in the Seurat object’s meta data where the pseudotime information is stored.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "eca4578b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[conflicted] Will prefer dplyr::select over any other package.\n",
      "[1] \"Plotting module dynamics for OPC\"\n",
      "Adding missing grouping variables: `pseudotime_bins_50`\n",
      "[1] \"OPC-M1\"\n",
      "[1] \"OPC-M2\"\n",
      "[1] \"OPC-M3\"\n",
      "[1] \"OPC-M4\"\n",
      "[1] \"OPC-M5\"\n",
      "[1] \"OPC-M6\"\n",
      "[1] \"OPC-M7\"\n",
      "[1] \"OPC-M8\"\n",
      "`geom_smooth()` using method = 'loess' and formula = 'y ~ x'\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Using bin as id variables\n",
       "In addition: Warning message:\n",
       "In min(xx[xx > upper]) : no non-missing arguments to min; returning Inf\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "conflicts_prefer(dplyr::select)\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/Pseudotime/ModuleDynamics/\", save_prefix, '/')\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Plotting module dynamics for ', toupper(cell_type)))\n",
    "\n",
    "\n",
    "  # # Save plot to PDF\n",
    "  # p  <- PlotModuleTrajectory(\n",
    "  #   seurat_obj[[cell_type]],\n",
    "  #   pseudotime_col = 'pseudotime'\n",
    "  # )\n",
    "\n",
    "  # pdf(paste0(fig_dir, cell_type, '_module_dynamics.pdf'),  width=4, height=4, useDingbats=FALSE)\n",
    "  # print(p)\n",
    "  # dev.off()\n",
    "\n",
    "  # ME scores along the pseuodotime trajectory\n",
    "\n",
    "\n",
    "  # cut pseudotime into bins using\n",
    "  seurat_obj[[cell_type]] <- BinPseudotime(seurat_obj[[cell_type]], n_bins=c(10, 50, 100))\n",
    "\n",
    "  # add MEs to seurat metadata\n",
    "  MEs <- GetMEs(seurat_obj[[cell_type]], harmonized=TRUE)\n",
    "  modules <- GetModules(seurat_obj[[cell_type]])\n",
    "  mods <- levels(modules$module)\n",
    "  mods <- mods[mods!='grey']\n",
    "  module_colors <- modules %>% dplyr::select(c(module, color)) %>% distinct()\n",
    "\n",
    "  rownames(module_colors) <- module_colors$module\n",
    "  mod_colors <- module_colors[mods, 'color']\n",
    "  seurat_obj[[cell_type]]@meta.data <- cbind(seurat_obj[[cell_type]]@meta.data, MEs)\n",
    "\n",
    "  # compute the average ME score in each bin:\n",
    "  features <- mods\n",
    "\n",
    "\n",
    "  # Retrieve metadata of the specified cell type from the Seurat object\n",
    "  meta_data <- seurat_obj[[cell_type]]@meta.data\n",
    "  # Repair duplicate column names using make.names()\n",
    "  names(meta_data) <- make.unique(names(meta_data))\n",
    "  seurat_obj[[cell_type]]@meta.data <- meta_data \n",
    "\n",
    "\n",
    "  avg_scores <- seurat_obj[[cell_type]]@meta.data %>%\n",
    "    group_by(pseudotime_bins_50) %>%\n",
    "    select(all_of(features)) %>%\n",
    "    summarise_all(mean)\n",
    "\n",
    "  colnames(avg_scores)[1] <- 'bin'\n",
    "\n",
    "  plot_df <- reshape2::melt(avg_scores)\n",
    "\n",
    "  p1 <- ggplot(plot_df, aes(x = as.numeric(bin), y = value, color=variable)) +\n",
    "    geom_point() +\n",
    "    geom_smooth() +\n",
    "    scale_color_manual(values=mod_colors)+\n",
    "    xlab('Pseudotime') +\n",
    "    ylab('Module Eigengene') +\n",
    "    theme(\n",
    "      axis.ticks.x = element_blank(),\n",
    "      axis.text.x = element_blank()\n",
    "    ) +\n",
    "    NoLegend()\n",
    "\n",
    "  plot_list <- ModuleFeaturePlot(seurat_obj[[cell_type]], order=TRUE, raster=TRUE, raster_dpi=400, alpha=0.7, title=FALSE)\n",
    "\n",
    "  patch <- wrap_plots(plot_list, ncol=1)\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, '_module_dynamics.pdf'), width=6, height=6)\n",
    "  print(p1 + facet_wrap(~variable, ncol=3, scales='free')) | patch\n",
    "  dev.off()\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d00bdc4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "fc1be299",
   "metadata": {},
   "source": [
    "Based on these dynamics, we can see which co-expression modules are turning their expression programs on or off throughout the transition from stem cells to mature erythrocytes.\n",
    "\n",
    "We can also compare module dynamics for multiple trajectories simultaneously by passing more than one pseudotime_col parameters to PlotModuleTrajectory."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "23952eca",
   "metadata": {},
   "source": [
    "### **Proportion of diagnosis in each pseudotime bin**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f113ceed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "`geom_smooth()` using method = 'loess' and formula = 'y ~ x'\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "In addition: Warning messages:\n",
       "1: Removed 50 rows containing non-finite values (`stat_smooth()`). \n",
       "2: Removed 50 rows containing missing values (`geom_point()`). \n",
       "3: Removed 8 rows containing missing values (`geom_smooth()`). \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/Pseudotime/ModuleDynamics/\", save_prefix, '/')\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "color_scheme_celltype <- list()\n",
    "color_scheme_celltype[['Excitatory']] = 'aquamarine3'\n",
    "color_scheme_celltype[['Inhibitory']] = 'darkolivegreen3'\n",
    "color_scheme_celltype[['Astrocyte']] = 'brown3'\n",
    "color_scheme_celltype[['Oligodendrocyte']] = 'darkorchid3'\n",
    "color_scheme_celltype[['Microglia']] = 'deepskyblue3'\n",
    "color_scheme_celltype[['OPC']] = 'deeppink3'\n",
    "\n",
    "\n",
    "\n",
    "for (cell_type in celltypes) {\n",
    "  tryCatch({\n",
    "    bins <- levels(seurat_obj[[cell_type]]@meta.data$pseudotime_bins_50)\n",
    "    proportion_df <- data.frame()\n",
    "\n",
    "    for (i in 1:length(bins)) {\n",
    "      bin <- bins[i]\n",
    "\n",
    "      cur_meta <- subset(seurat_obj[[cell_type]]@meta.data, pseudotime_bins_50 == bin)\n",
    "      cur_meta$pathology.group <- ifelse(cur_meta$pathology.group != 'no', 'AD', cur_meta$pathology.group)\n",
    "\n",
    "      cur_df <- data.frame(table(cur_meta$pathology.group) / nrow(cur_meta))\n",
    "      cur_df <- cur_df %>% dplyr::rename(c(Diagnosis = Var1, proportion = Freq))\n",
    "      cur_df$bin_num <- i\n",
    "      cur_df$bin <- bin\n",
    "      proportion_df <- rbind(proportion_df, cur_df)\n",
    "    }\n",
    "\n",
    "    proportion_cor <- cor.test(\n",
    "      subset(proportion_df, Diagnosis == 'AD') %>% .$bin_num,\n",
    "      subset(proportion_df, Diagnosis == 'AD') %>% .$proportion,\n",
    "      method = 'pearson'\n",
    "    )\n",
    "\n",
    "    p <- ggplot(proportion_df, aes(x = bin_num, y = proportion, fill = Diagnosis, color = Diagnosis)) +\n",
    "      ylab('scaled feature') +\n",
    "      geom_point() +\n",
    "      ylim(0.5, 1)\n",
    "\n",
    "    p <- p +\n",
    "      geom_smooth(se = FALSE) +\n",
    "      scale_color_manual(values = c('gray', 'seagreen'))\n",
    "\n",
    "    p <- p +\n",
    "      theme(\n",
    "        axis.ticks.x = element_blank(),\n",
    "        axis.text.x = element_blank()\n",
    "      )\n",
    "\n",
    "    pdf(paste0(fig_dir, cell_type, '_diagnosis_proportion.pdf'), width = 5, height = 2)\n",
    "    print(p)\n",
    "    dev.off()\n",
    "  }, error = function(e) {\n",
    "    message(\"Error processing celltype: \", cell_type)\n",
    "  })\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "963ba601",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3bc33b36",
   "metadata": {},
   "source": [
    "# **Save Seurat Object**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "431d9ac4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Saving hdWGCNA object in hdWGCNA Experiment for OPC\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "dat_dir = paste0(\"../results/hdWGCNA/SeuratObject/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(dat_dir)) {\n",
    "  dir.create(dat_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "    print(paste0('Saving hdWGCNA object in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    saveRDS(seurat_obj[[cell_type]], file=paste0(dat_dir, cell_type, '_hdWGCNA_object.rds'))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "125010ce",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "2f414a000012e1980f5fd1119ed6b9bf8cc73b51e37b52eb52a16ff857effd1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
