{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62042f97-a76e-4b34-805f-3f1ac81e7288",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import rpy2\n",
    "import logging\n",
    "import warnings\n",
    "import anndata\n",
    "import anndata2ri\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import decoupler as dc\n",
    "from anndata import AnnData\n",
    "from gsva_prep import prep_gsva\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "from matplotlib.pyplot import rcParams\n",
    "from functions import pathway_analyses\n",
    "from sklearn.model_selection import train_test_split\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "447b6d1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "048c2e09",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "suppressPackageStartupMessages({\n",
    "    library(WGCNA)\n",
    "    library(Matrix)\n",
    "    library(viridis)\n",
    "    library(harmony)\n",
    "    library(ggpubr)\n",
    "    library(tictoc)\n",
    "    library(RColorBrewer)\n",
    "    library(Hmisc)\n",
    "    library(corrplot)\n",
    "    library(grid)\n",
    "    library(gridExtra)\n",
    "    library(igraph)\n",
    "    library(ggrepel)\n",
    "    library(readxl)\n",
    "    library(conflicted)\n",
    "\n",
    "    # single-cell analysis package\n",
    "    library(Seurat)\n",
    "\n",
    "    # plotting and data science packages\n",
    "    library(tidyverse)\n",
    "    library(cowplot)\n",
    "    library(patchwork)\n",
    "\n",
    "    # co-expression network analysis packages:\n",
    "    library(WGCNA)\n",
    "    library(hdWGCNA)\n",
    "\n",
    "    # gene enrichment packages\n",
    "    library(enrichR)\n",
    "    library(GeneOverlap)\n",
    "\n",
    "\n",
    "    library(GSEABase)\n",
    "    library(GSVA) \n",
    "# needs to be run every time you start R and want to use %>%\n",
    "})\n",
    "\n",
    "# using the cowplot theme for ggplot\n",
    "theme_set(theme_cowplot())\n",
    "\n",
    "# set random seed for reproducibility\n",
    "set.seed(12345)\n",
    "\n",
    "# optionally enable multithreading\n",
    "# enableWGCNAThreads(nThreads = 4)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "d40204d0-56d5-4984-a2af-1075b64098f1",
   "metadata": {},
   "source": [
    "### **Prepare data**\n",
    "\n",
    "load the preprocessed and annotated data for differential pathway analysis.\n",
    "\n",
    "Make sure the anndata has the count data in either `.layers` or in `.X`. if count data is not in `.layers` it will be assumed that `.X ` contains the counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53693e9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_prefix = 'leng_etc' # this takes the format '{StudyName}_{ThreeLetterAccronymForBrainRegion}'\n",
    "get_cell_types = False   # whether to reformat cell_type annotation\n",
    "\n",
    "adata_annot = sc.read_h5ad(f'../data/raw/{save_prefix}/{save_prefix}_raw_anndata.h5ad')\n",
    "adata_annot.obs_names_make_unique()\n",
    "adata_annot.var_names_make_unique()\n",
    "\n",
    "if 'counts' not in adata_annot.layers.keys():\n",
    "    print('\"counts\" not in layers...')\n",
    "    print('analysis requires unnormalized count data...')\n",
    "    print('fetching count data from \"adata.X\"...')\n",
    "    adata_annot.layers['counts'] = adata_annot.X.copy()\n",
    "else:\n",
    "    adata_annot.X = adata_annot.layers['counts'].copy()\n",
    "    \n",
    "del adata_annot.obsm, adata_annot.layers, adata_annot.varm, adata_annot.uns, adata_annot.obsp\n",
    "\n",
    "adata_annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bea0b9c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_annot.obs.cell_type.value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e1a784ba",
   "metadata": {},
   "source": [
    "Specify the way to map cells to appropriate cell-types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41fdf8ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltypes = [\"Excitatory\", \"Inhibitory\", \"Astrocyte\", \"Microglia\", \"Oligodendrocyte\", \"OPC\", 'Endothelial'] \n",
    "\n",
    "mapping = {'leng_etc':\n",
    "           \n",
    "           {'Exc': 'Excitatory', \n",
    "            'Inh': 'Inhibitory', \n",
    "            'Astro': 'Astrocyte',\n",
    "            'Endo': 'Endothelial', \n",
    "            'Micro': 'Microglia', \n",
    "            'OPC': 'OPC', \n",
    "            'Oligo': 'Oligodendrocyte'},\n",
    "\n",
    "           'leng_sfg':\n",
    "           \n",
    "           {'Exc': 'Excitatory', \n",
    "            'Inh': 'Inhibitory', \n",
    "            'Astro': 'Astrocyte',\n",
    "            'Endo': 'Endothelial', \n",
    "            'Micro': 'Microglia', \n",
    "            'OPC': 'OPC', \n",
    "            'Oligo': 'Oligodendrocyte'},\n",
    "            \n",
    "            'seaad_mtg':\n",
    "           \n",
    "           {'Excitatory': 'Excitatory', \n",
    "            'Inhibitory': 'Inhibitory',\n",
    "            'Astrocyte': 'Astrocyte',\n",
    "            'Microglia': 'Microglia', \n",
    "            'Endothelial': 'Endothelial', \n",
    "            'OPC': 'OPC',\n",
    "            'Oligodendrocyte': 'Oligodendrocyte'},\n",
    "            }\n",
    "\n",
    "cell_column = {'leng_etc': 'clusterCellType',\n",
    "               'leng_sfg': 'clusterCellType',\n",
    "               'seaad_mtg': 'cell_type',\n",
    "               }"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "95e3f93b",
   "metadata": {},
   "source": [
    "### **Additional Parameters**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb37e8af",
   "metadata": {},
   "outputs": [],
   "source": [
    "control_group = 'no'                        # name of the control group in metadata \n",
    "map_meta = True\n",
    "\n",
    "### There are different pseudobulking strategies\n",
    "pseudobulking_strategy = 'blanchardbulk'      # options ['blanchardbulk', 'metacell', 'standardbulk'] See details in subsequent sections\n",
    "\n",
    "pseudobulk_method = 'agg_x_norm'            # Only required if `pseudobulking_strategy=='custombulk'`` method for aggregrating cells from replicates into pseudobulks. ['agg_x_norm', 'norm_x_agg']\n",
    "gene_selection = 'fraction'                 # Only required if `pseudobulking_strategy=='metacell'``. specifies the gene selection method when setting up seurat object for WGCNA. \n",
    "geneSet = ''                                # Only required if `pseudobulking_strategy=='metacell'`` and `gene_selction='custom'`\n",
    "\n",
    "technical_covariate = 'SampleBatch'         # Technical covariate to be included as random effect not of interest (regressed out by duplicatecorrection)\n",
    "duplicate_correction = 'FALSE'              # whether to run duplicate corretion in Limma to block out technical_covariate\n",
    "gene_celltype_threshold = 0.10              # determines number of cells the gene must be expressed in\n",
    "filter_genes_from = 'pseudobulk'            # whether to filter genees after aggregating pseudobulks or from single_cell object \n",
    "pathway_gene_threshold = 0.33               # determines number of genes that must be present in that pathway\n",
    "\n",
    "\n",
    "metadata = f'../data/raw/{save_prefix}/{save_prefix}_metadata.csv' # Metatdata location\n",
    "meta = pd.read_csv(metadata, encoding_errors='ignore')\n",
    "\n",
    "test_names = ['early_vs_no', 'late_vs_early', 'late_vs_no', 'ad_vs_no'] # test categories\n",
    "\n",
    "covaraites_for_study = {'leng_sfg': ['None'],  # ['ageDeath.cat',],\n",
    "                        'leng_etc': ['None'],  # ['ageDeath.cat',]\n",
    "                        'seaad_mtg': ['None'],\n",
    "                        }\n",
    "\n",
    "subject_ids_for_study = {'leng_sfg': 'PatientID',\n",
    "                        'leng_etc': 'PatientID',\n",
    "                        'seaad_mtg': 'individualID',}\n",
    "\n",
    "bio_covariates = covaraites_for_study[save_prefix]  # list of covariates to be accounted for in regression.\n",
    "subject_id = subject_ids_for_study[save_prefix]     # for leng this is `PatientID` for mathys is 'Subject', and allen is 'individualID'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e120422",
   "metadata": {},
   "outputs": [],
   "source": [
    "if get_cell_types:\n",
    "    adata_annot.obs['cell_type'] = adata_annot.obs[cell_column[save_prefix]].map(mapping[save_prefix])\n",
    "\n",
    "adata_annot = adata_annot[adata_annot.obs.cell_type.isin(celltypes)]\n",
    "\n",
    "# map the pathology group to teh subject id in .obs\n",
    "adata_annot = adata_annot[adata_annot.obs.cell_type.isin(celltypes)]\n",
    "adata_annot.obs['pathology.group'] = adata_annot.obs[subject_id].map(dict(zip(meta[subject_id].astype(str), meta['pathology.group'])))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "738c2a7b",
   "metadata": {},
   "source": [
    "### **Send parameters to R interface with rpy2**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4acb9d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i adata_annot -i subject_id -i gene_celltype_threshold -i celltypes -i test_names -i save_prefix -i gene_selection\n",
    "\n",
    "print(adata_annot)\n",
    "print('loaded data into memory for recursive use')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5d8a01b",
   "metadata": {},
   "source": [
    "## **2.4 Systematic differential analysis of pathway activity**\n",
    "\n",
    "Pathway activity scores were computed in accordance with protocols outlined in [**Joel W. Blanchard et. al.**](https://doi.org/10.1038/s41586-022-05439-w).\n",
    "\n",
    "\n",
    "We perform differential pathway activity analysis using pathways obtained from [**Gene Ontology biological processes (2018 edition**](https://maayanlab.cloud/Enrichr/#libraries)."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c568575c",
   "metadata": {},
   "source": [
    "### **Load and Process Pathways**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12fa36cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "bp = pathway_analyses.read_pathways('../data/pathway_databases/GO_Biological_Process_2018.txt')\n",
    "\n",
    "go_bp_paths = bp.set_index(0)\n",
    "go_bp_paths.fillna(\"\", inplace=True)\n",
    "go_bp_paths_dict = go_bp_paths.to_dict(orient='index')\n",
    "\n",
    "\n",
    "gene_set_by_path = {key: [val for val in value.values() if val != \"\"] for key, value in go_bp_paths_dict.items()}\n",
    "gene_set_by_path = pd.DataFrame.from_dict(gene_set_by_path, orient='index').transpose()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee6e036a",
   "metadata": {},
   "source": [
    "### **Pseudobulk Aggregation and Filtering**\n",
    "\n",
    "\n",
    "Briefly, we first computed cell-type-level normalized gene expression profiles for each individual using the ACTIONet normalization procedure (Mohammadi et al., 2020). We additionally provide other `pseudobulking_strategies`, using the `metacell` approach in the `hdWGCNA`, or the standard pseudobulk (`standardbulk`) method provided in custom script."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88227d4c",
   "metadata": {},
   "source": [
    "#### If **`pseubulking_strategy == 'blanchardbulk'`**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14d360bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "if pseudobulking_strategy=='blanchardbulk':\n",
    "\n",
    "    adata_annot.layers['counts'] = adata_annot.X.copy()\n",
    "    expressed_genes_per_celltype = pathway_analyses.filter_expressed_genes_by_celltype(adata_annot, threshold=gene_celltype_threshold, \n",
    "                                                                                    filter_genes_from = filter_genes_from,\n",
    "                                                                                    subject_id = subject_id)\n",
    "\n",
    "    avs_logcounts_cellxind  = pathway_analyses.get_ind_level_ave(adata_annot, subject_id, pseudobulk_method, \n",
    "                                                                expressed_genes_per_celltype, filter_genes_at_threshold=True)\n",
    "\n",
    "del adata_annot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e84ad3cc",
   "metadata": {},
   "source": [
    "#### If **`pseubulking_strategy == 'metacell'` (TIME CONSUMING)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04c388df",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i pseudobulking_strategy -o meta_obj -o expressed_genes -i geneSet\n",
    "\n",
    "if (pseudobulking_strategy == 'metacell'){\n",
    "  seurat_obj <- as.Seurat(adata_annot, counts = \"X\", data = \"X\")\n",
    "\n",
    "  print(seurat_obj)\n",
    "  # Perform dimensionality reduction and plot\n",
    "\n",
    "  seurat_obj <- FindVariableFeatures(seurat_obj)\n",
    "  seurat_obj <- ScaleData(seurat_obj)\n",
    "  seurat_obj <- RunPCA(seurat_obj)\n",
    "  seurat_obj <- RunHarmony(seurat_obj, group.by.vars = subject_id)\n",
    "  seurat_obj <- RunUMAP(seurat_obj, reduction='harmony', n.neighbors=15, dims=1:30, min.dist=0.1)\n",
    "\n",
    "  # create a hdWGCNA experiment for each celltype\n",
    "  seurat_dat <- seurat_obj\n",
    "  meta_obj <- list()\n",
    "  expressed_genes <- list()\n",
    "\n",
    "  for (cur_cell_type in celltypes){\n",
    "\n",
    "    seurat_obj <- subset(seurat_dat, cell_type == cur_cell_type)\n",
    "\n",
    "    print(paste0('Creating hdWGNA Experiment for ', toupper(cur_cell_type)))\n",
    "\n",
    "    if (gene_selection == 'custom') {\n",
    "      seurat_obj <- SetupForWGCNA(\n",
    "        seurat_obj,\n",
    "        gene_select = \"custom\",                                            # the gene selection approach\n",
    "        gene_list = as.vector(geneSet),    # list of genes to be included\n",
    "        group.by = 'cell_type',                                            # grouping parameter\n",
    "        wgcna_name = toupper(substr(cur_cell_type, 1, 3))                  # the name of the hdWGCNA experiment\n",
    "      )\n",
    "    } else {\n",
    "      seurat_obj <- SetupForWGCNA(\n",
    "        seurat_obj,\n",
    "        gene_select = \"fraction\",                               # the gene selection approach\n",
    "        fraction = gene_celltype_threshold,                     # fraction of cells for gene inclusion\n",
    "        group.by = 'cell_type',                                 # grouping parameter\n",
    "        wgcna_name = toupper(substr(cur_cell_type, 1, 3))       # the name of the hdWGCNA experiment\n",
    "      )\n",
    "    }\n",
    "\n",
    "    print(paste0('Constructing MetaCells in hdWGCNA Experiment for ', toupper(cur_cell_type)))\n",
    "    \n",
    "    seurat_obj <- MetacellsByGroups(\n",
    "            seurat_obj = seurat_obj,\n",
    "            group.by = c(\"cell_type\", subject_id), # specify the columns in seurat_obj@meta.data to group by\n",
    "            reduction = 'harmony',  # select the dimensionality reduction to perform KNN on\n",
    "            k = 25, # nearest-neighbors parameter\n",
    "            max_shared = 10, # maximum number of shared cells between two metacells\n",
    "            ident.group = 'cell_type', # set the Idents of the metacell seurat object\n",
    "            wgcna_name = toupper(substr(cur_cell_type, 1, 3)),        # the name of the hdWGCNA experiment\n",
    "            )\n",
    "\n",
    "    # normalize metacell expression matrix:\n",
    "    seurat_obj  <- NormalizeMetacells(seurat_obj, wgcna_name = toupper(substr(cur_cell_type, 1, 3)))\n",
    "\n",
    "    meta_obj[[cur_cell_type]] <- as.SingleCellExperiment(GetMetacellObject(seurat_obj,\n",
    "                                                      wgcna_name = toupper(substr(cur_cell_type, 1, 3))))   \n",
    "\n",
    "     expressed_genes[[cur_cell_type]] <- GetWGCNAGenes(seurat_obj)                                       \n",
    "  }\n",
    "\n",
    "  rm(seurat_obj)\n",
    "  rm(seurat_dat)\n",
    "  rm(adata_annot)\n",
    "} else {\n",
    "  meta_obj <- NULL\n",
    "  expressed_genes <- NULL\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9804a8e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "if pseudobulking_strategy=='metacell':\n",
    "    \n",
    "    max([len(expressed_genes[v]) for v in expressed_genes.keys()])\n",
    "    expressed_genes = {k: list(v) + [np.nan]*(max_length - len(v)) for k, v in expressed_genes.items()}\n",
    "    expressed_genes_per_celltype = pd.DataFrame(expressed_genes, columns=expressed_genes.keys())\n",
    "\n",
    "    avs_logcounts_cellxind = {}\n",
    "    for cell_type in celltypes:\n",
    "        avs_logcounts_cellxind[cell_type] = pd.DataFrame(meta_obj[cell_type].layers['logcounts'].toarray().T, \n",
    "                                                        columns=meta_obj[cell_type].obs_names, \n",
    "                                                        index=meta_obj[cell_type].var_names)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "223a241c",
   "metadata": {},
   "source": [
    "#### If **`pseubulking_strategy == 'standardbulk'`**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "716259e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i pseudobulking_strategy -o pseudobulk_data\n",
    "\n",
    "if (pseudobulking_strategy == 'standardbulk'){\n",
    "\n",
    "    library(scuttle)\n",
    "    library(Matrix)\n",
    "    library(ensembldb)\n",
    "    library(EnsDb.Hsapiens.v86)\n",
    "    conflicts_prefer(GenomicRanges::setdiff)\n",
    "\n",
    "    source(\"../scripts/functions/sconline_code.R\")\n",
    "\n",
    "    min_size_limit = 15     # minimum acceptable size of the pseudocells. usually 10 or 15.\n",
    "    nPCs = 30               # Number of pcs to use to create the similarity network. Used only if human or mouse\n",
    "    organism = 'Human'\n",
    "\n",
    "    #generating the embedding space\n",
    "\n",
    "    # exp_seurat = .extraExport2SeuratFn(adata_annot) %>%\n",
    "    # Seurat::NormalizeData() %>%\n",
    "    #     FindVariableFeatures() %>% \n",
    "    #     ScaleData() %>% \n",
    "    #     RunPCA(verbose=F)\n",
    "  assay(adata_annot, 'counts') = assay(adata_annot, 'X')\n",
    "  pseudobulk_data = suppressWarnings(.sconline.PseudobulkGeneration(argList = NULL, \n",
    "                                  # The columns in the pheonData that will be used to parse the expression data \n",
    "                                  # and generate the pseudocell/pseudobulk data\n",
    "                                    parsing.col.names = c(subject_id, 'cell_type'), \n",
    "                                  # average pseudocell size.\n",
    "                                    pseudocell.size = NULL,\n",
    "                                    inputExpData = adata_annot,\n",
    "                                  # minimum acceptable size (ie, #cells) for each pseudobulk\n",
    "                                    min_size_limit = min_size_limit,\n",
    "                                  # in case we want to run the function outside sconline space\n",
    "                                    inputPhenoData = as.data.frame(colData(adata_annot)),\n",
    "                                  # the embedding space to be used for the generation of the pseudobulk.\n",
    "                                  # only needed when pseudocell.size is not null\n",
    "                                  # inputEmbedding = embedding_data, \n",
    "                                  # the dimension of the embedding space for the construction of pseudobulk data\n",
    "                                    nPCs = nPCs, \n",
    "                                    ncores = 3,\n",
    "                                    rand_pseudobulk_mod = F,\n",
    "                                    organism = organism))\n",
    "\n",
    "} else {\n",
    "  pseudobulk_data <- NULL\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bda13023",
   "metadata": {},
   "source": [
    "##### **Get pathways**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73724f73",
   "metadata": {},
   "source": [
    "**Obtain filtered genesets**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "280ab24c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_set_per_celltype = pathway_analyses.filter_lowly_exp_genes(expressed_genes_per_celltype, gene_set_by_path, \n",
    "                                                                threshold=pathway_gene_threshold)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb16b283",
   "metadata": {},
   "source": [
    "#### **Visualize Covariates on UMAP**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "958234a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adatas = []\n",
    "\n",
    "# for cell_type in celltypes:\n",
    "#     temp = AnnData(X=avs_logcounts_cellxind[cell_type].T)\n",
    "#     temp.var_names = list(np.arange(0, len(avs_logcounts_cellxind[cell_type].index)))\n",
    "#     temp.obs_names = avs_logcounts_cellxind[cell_type].columns\n",
    "#     temp.obs['cell_type'] = cell_type\n",
    "#     temp.obs[subject_id] = avs_logcounts_cellxind[cell_type].columns\n",
    "#     for covariate in bio_covariates:\n",
    "#         temp.obs[covariate] = temp.obs[subject_id].astype(str).map(dict(zip(meta[subject_id].astype(str), meta[covariate])))\n",
    "#     temp.obs_names = [f'{cell_type}_{x}' for x in avs_logcounts_cellxind[cell_type].columns]\n",
    "#     adatas.append(temp.copy())\n",
    "#     #del temp\n",
    "\n",
    "# adata_concat = anndata.concat(adatas, join='outer')\n",
    "# sc.pp.pca(adata_concat)\n",
    "\n",
    "# sc.set_figure_params(scanpy=True,\n",
    "#                     dpi_save=300, frameon=False, \n",
    "#                     vector_friendly=True, fontsize=14, figsize=(7, 5), \n",
    "#                     color_map=None, format='pdf', facecolor=None,\n",
    "#                     transparent=False, ipython_format='png2x')\n",
    "                     \n",
    "                     \n",
    "# #pdata.obs['pmi'] = pdata.obs['pmi'].astype('float')\n",
    "\n",
    "# dat_dir = f'../results/Covariates/'\n",
    "\n",
    "# if not os.path.exists(dat_dir):\n",
    "#     os.makedirs(dat_dir)\n",
    "\n",
    "# if not os.path.exists('figures/'):\n",
    "#     os.makedirs(\"figures/\")\n",
    "\n",
    "# sc.pl.pca(adata_concat[adata_concat.obs.cell_type.isin(celltypes)], color=['cell_type'] + bio_covariates,\n",
    "#          ncols=3, size=300, save=f'_{save_prefix}_covariates.pdf', components=['1,2'])\n",
    "\n",
    "# os.rename(f'../scripts/figures/pca_{save_prefix}_covariates.pdf',\n",
    "#          f'../results/Covariates/{save_prefix}_covariates.pdf')\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1398d57d",
   "metadata": {},
   "source": [
    "### **Prep data for Gene Set Variation Analysis in R**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11d61179",
   "metadata": {},
   "outputs": [],
   "source": [
    "for cell_type in celltypes:\n",
    "    prep_gsva(avs_logcounts_cellxind[cell_type], gene_set_per_celltype[cell_type], \n",
    "              tempdir=f'../results/gsva/{save_prefix}/{cell_type}/', verbose=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "652c6839",
   "metadata": {},
   "source": [
    "### **Estimate Pathway Activity Scores**\n",
    "\n",
    "After estimating individual-cell-type-level normalized gene expression profile averages, we next calculate pathway activity scores, as described in [**Joel W. Blanchard et. al.**](https://doi.org/10.1038/s41586-022-05439-w) and previously implemented in the R package (Gene Set Variational Analysis) GSVA (v.1.42.0)41. \n",
    "\n",
    "Briefly, GSVA estimates a normalized relative expression level per gene across samples. This expression level is then rank-ordered for each sample and aggregated into gene sets by calculating sample-wise enrichment scores using a Kolmogorov–Smirnov-like rank statistic.\n",
    "\n",
    "As described in [**Joel W. Blanchard et. al.**](https://doi.org/10.1038/s41586-022-05439-w), we used the parameters to evaluate the \n",
    "GSVA function:\n",
    "\n",
    "- `mx.diff=TRUE` \n",
    "- `kcdf=c(\"Gaussian\")`\n",
    "- `min.sz=5` \n",
    "- `max.sz=150`\n",
    "\n",
    "Here, we instead set `max.sz=500`, as described in the [**GSVA Tutorial**](http://www.bioconductor.org/packages/release/bioc/vignettes/GSVA/inst/doc/GSVA.html#5_Quantification_of_pathway_activity_in_bulk_microarray_and_RNA-seq_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d82607c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o gsva_scores -i save_prefix -i celltypes\n",
    "\n",
    "library(GSEABase)\n",
    "library(GSVA)\n",
    "library(tidyr)\n",
    "library(dplyr)\n",
    "\n",
    "# create an empty list to store the output for each celltype\n",
    "gsva_scores = list()\n",
    "\n",
    "# iterate over celltypes and perform gsva\n",
    "for (celltype in celltypes) {\n",
    "    # get gene sets\n",
    "    geneSets = getGmt(file.path('../results/gsva/', save_prefix, '/', celltype, 'gs.gmt'))\n",
    "    geneSets = geneIds(geneSets)\n",
    "\n",
    "    # read in expression data\n",
    "    mat = as.matrix(read.csv(file.path('../results/gsva/', save_prefix, '/', celltype, 'expr.csv'), header=TRUE, row.names=1, check.names=FALSE))\n",
    "    \n",
    "    # run gsva\n",
    "    print(paste0(\"Estimating Pathway Activity Scores in \", celltype, \" Cells\"))\n",
    "    print(\"........\")\n",
    "    exc_gsva = gsva(mat, geneSets, mx.diff=TRUE, kcdf=c(\"Gaussian\"), min.sz=5, max.sz = 500)\n",
    "    \n",
    "    # Convert matrix to data frame and gather into \"long\" format\n",
    "    exc_gsva <- as.data.frame(exc_gsva, stringsAsFactors = FALSE)\n",
    "\n",
    "    # store the output in a list\n",
    "    gsva_scores[[celltype]] = exc_gsva\n",
    "\n",
    "    write.csv(gsva_scores[[celltype]], file = file.path('../results/gsva/', save_prefix, '/', celltype, 'gsva_scores.csv'))\n",
    "}\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "13bc27b5",
   "metadata": {},
   "source": [
    "### **Differential pathway activity analysis**\n",
    "\n",
    "To investigate differential pathway activity across different cell types, we use a multivariate linear model approach, as described in [**Joel W. Blanchard et. al.**](https://doi.org/10.1038/s41586-022-05439-w), using pathway activity scores. \n",
    "\n",
    "The design matrix will appear as consequently: \n",
    "\n",
    "`pathway activity ~ β0 × AD + β1 × Coavriate1 + β2 × Covariate2 + ...`\n",
    "\n",
    "The linear models were fitted using the `lmfit()` function and t-statistics were computed using the `eBayes()` function from the Limma R package (v.3.50.3). \n",
    "\n",
    "To prioritize candidate AD-dysregulated processes, pathways with snRNA-seq evidence of association with AD at a `nominal P value of 0.05` are considered to be potential candidates. This procedure will result in prioritized candidate pathways for all cell types. \n",
    "\n",
    "The linear models were fitted using the lmfit() function and t-statistics were computed using the eBayes() function from the Limma R package (v.3.50.3)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f4001f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o all_data -i control_group -i metadata -i subject_id -i bio_covariates -i save_prefix -i technical_covariate -i duplicate_correction -i test_names\n",
    "\n",
    "library(limma)\n",
    "library(magrittr)\n",
    "library(tidyr)\n",
    "conflicts_prefer(base::unname)\n",
    "\n",
    "source('../scripts/functions/pathway_analyses.R')\n",
    "\n",
    "summary = read.csv(metadata, check.names = FALSE)\n",
    "#control_group = paste(control_group, 'pathology', sep='-')\n",
    "summary$AD = ifelse(summary$pathology.group == control_group, 0, 1)\n",
    "rownames(summary) = summary[, subject_id]\n",
    "\n",
    "\n",
    "\n",
    "gsva_scores = list()\n",
    "# iterate over celltypes and perform limma\n",
    "for (celltype in celltypes) {\n",
    "\n",
    "    # read in pathway scores\n",
    "    pathway_scores = as.matrix(read.csv(file.path('../results/gsva/', save_prefix, '/', celltype, 'gsva_scores.csv'), header=TRUE, row.names=1, check.names=FALSE))\n",
    "\n",
    "    # store the output in a list\n",
    "    gsva_scores[[celltype]] <- t(pathway_scores)\n",
    "} \n",
    "\n",
    "all_data = list()\n",
    "\n",
    "all_data[['gsva_out']] = gsva_scores\n",
    "\n",
    "# get linear model fits\n",
    "print('getting linear model fits...')\n",
    "fits = get_fits(gsva_scores, summary, bio_covariates, technical_covariate)\n",
    "all_data[['fits_all']] = fits\n",
    "\n",
    "# get matrix of scores for heatmap\n",
    "print('get matrix of scores')\n",
    "scores = get_scores(fits)\n",
    "\n",
    "# Create a new empty list to store the reorganized scores\n",
    "new_scores <- list()\n",
    "\n",
    "# Loop over the different tests\n",
    "tests <- test_names\n",
    "for (test in tests) {\n",
    "  \n",
    "  # Create a new empty list for this test\n",
    "  new_scores[[test]] <- list('all' = list())\n",
    "  \n",
    "  # Loop over the different cell types\n",
    "  for (cell_type in celltypes) {\n",
    "    \n",
    "    # Copy the scores for this cell type and test\n",
    "    new_scores[[test]][['all']][[cell_type]] <- scores[['all']][[cell_type]][[test]]\n",
    "    \n",
    "  }\n",
    "}\n",
    "\n",
    "# save new_scores in list\n",
    "all_data[['scores_all']] = new_scores\n",
    "\n",
    "print('filter by score 1.3')\n",
    "# Extract unique gene names from scores object\n",
    "all_data[['scores_filtered']] = list()\n",
    "\n",
    "for(i in names(new_scores)){\n",
    "    names = unique(unname(unlist(lapply(names(new_scores[[i]]$all), function(x) rownames(new_scores[[i]]$all[[x]])))))\n",
    "\n",
    "    # Get a matrix of gene set scores for all samples\n",
    "    mat = get_matrix(new_scores[[i]]$all, names)\n",
    "    \n",
    "    # Keep only rows with absolute values > 1.3 in at least one sample\n",
    "    mat = mat[unname(rowSums(abs(mat)>1.3)>0), ]\n",
    "    \n",
    "    # if(top_20==TRUE){\n",
    "    #     index = unique(unname(unlist(lapply(colnames(mat), function(x) order(abs(mat[[x]]),decreasing = T)[1:20]))))\n",
    "    #     mat = mat[index,]\n",
    "    # }\n",
    "    all_data[['scores_filtered']][[i]] = mat\n",
    "}\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4b52d66a",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26ed113c",
   "metadata": {},
   "outputs": [],
   "source": [
    "valid_pathways = {}\n",
    "all_pathways = {}\n",
    "\n",
    "for test_name in test_names:\n",
    "    valid_pathways[test_name] = {}\n",
    "    all_pathways[test_name] = pd.DataFrame()\n",
    "    # Loop through each unique cell type in the cell_type column of adata_annot and filter statistically significant pathways\n",
    "    for cell_type in celltypes:\n",
    "        try:\n",
    "            # Create a dataframe for the current cell type from a matrix in all_data, rename and set index.\n",
    "            valid_pathways[test_name][cell_type] = pd.DataFrame(all_data['scores_filtered'][test_name][cell_type])\n",
    "            all_data['scores_all'][test_name]['all'][cell_type].index.name = 'pathway'\n",
    "            valid_pathways[test_name][cell_type].index.name = 'pathway'\n",
    "            valid_pathways[test_name][cell_type].rename(columns={cell_type: 'score_adj'}, inplace=True)\n",
    "\n",
    "            # Merge current cell type dataframe with its corresponding scores_all dataframe on \"pathway\" and remove missing rows\n",
    "            valid_pathways[test_name][cell_type] = valid_pathways[test_name][cell_type].merge(all_data['scores_all'][test_name]['all'][cell_type],\n",
    "                                                     how='outer', on='pathway')\n",
    "            valid_pathways[test_name][cell_type].dropna(axis=0, inplace=True)\n",
    "\n",
    "            # Filter rows with P.Value less than 0.05 from the current cell type dataframe and sort by P.Value    \n",
    "            # valid_pathways[test_name][cell_type] = valid_pathways[test_name][cell_type][valid_pathways[test_name][cell_type]['P.Value']<0.05]\n",
    "            valid_pathways[test_name][cell_type].sort_values(by='P.Value', inplace=True)\n",
    "            \n",
    "            # Append the current cell type's dataframe to the all_pathways dataframe and sort by pathway\n",
    "            all_pathways[test_name] = pd.concat([all_pathways[test_name], valid_pathways[test_name][cell_type]], axis=0)\n",
    "            all_pathways[test_name].sort_values(by='pathway', inplace=True)\n",
    "        except KeyError:\n",
    "            continue"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "48d98f9c",
   "metadata": {},
   "source": [
    "### Renaming convention"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2ba11c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in a CSV file containing the renaming convention for pathway names\n",
    "# Remove duplicate names in the renaming convention dataframe\n",
    "renaming_convention = pd.read_csv('../data/pathway_databases/renaming_convention.csv', dtype=str)\n",
    "renaming_convention.drop_duplicates(subset='names', inplace=True)\n",
    "\n",
    "for test_name in test_names:\n",
    "    # Reset the index of the all_pathways dataframe to allow merging with renamed convention\n",
    "    all_pathways[test_name].reset_index(inplace=True)\n",
    "    all_pathways[test_name] = pd.merge(all_pathways[test_name], renaming_convention[['names', 'shortened', 'highlight']], how='left', on='names')\n",
    "    all_pathways[test_name].set_index(keys='pathway', inplace=True)\n",
    "    all_pathways[test_name] =  all_pathways[test_name].astype(object).replace(np.nan, 'None')\n",
    "\n",
    "\n",
    "    dat_dir = f'../results/{test_name}/{save_prefix}/Data/'\n",
    "\n",
    "    if not os.path.exists(dat_dir):\n",
    "        os.makedirs(dat_dir)\n",
    "\n",
    "    # Write the all_pathways dataframe to a CSV file\n",
    "    all_pathways[test_name].to_csv(dat_dir+f'differentially_expressed_pathways.csv')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "f7dd5249",
   "metadata": {},
   "source": [
    "### Preliminary Heat Map Visualization"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9894567c",
   "metadata": {},
   "source": [
    "Plot Select Cell-type-specific Pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d330986",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for test_name in test_names:\n",
    "#     print(f'heatmap for select cell-type-specific pathways in {test_name} test')\n",
    "#     list_of_paths_to_annotate = list(all_pathways[test_name].shortened[all_pathways[test_name].highlight=='yes'])\n",
    "#     df = all_pathways[test_name][all_pathways[test_name]['P.Value']<0.05]\n",
    "\n",
    "#     fig_dir = f'../results/{test_name}/{save_prefix}/HeatMap/'\n",
    "#     if not os.path.exists(fig_dir):\n",
    "#         os.makedirs(fig_dir)\n",
    "        \n",
    "#     save_path = fig_dir+'_cell_type_specific_diff_exp_paths.pdf'\n",
    "\n",
    "#     pathway_analyses.plot_and_select_top_deps(df, list_of_paths_to_annotate, save_prefix=save_prefix,\n",
    "#                          save_path=save_path, filter=True, cell_type_specific=True, \n",
    "#                          test_name=test_name, cell_types=celltypes)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "74a487d8",
   "metadata": {},
   "source": [
    "Plot Select Shared Pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "936ac4de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for test_name in test_names:\n",
    "#     print(f'heatmap for select shared pathways in {test_name} test')\n",
    "#     list_of_paths_to_annotate = list(all_pathways[test_name].shortened[all_pathways[test_name].highlight=='yes'])\n",
    "#     df = all_pathways[test_name][all_pathways[test_name]['P.Value']<0.05]\n",
    "\n",
    "\n",
    "#     fig_dir = f'../results/{test_name}/{save_prefix}/HeatMap/'\n",
    "#     if not os.path.exists(fig_dir):\n",
    "#         os.makedirs(fig_dir)\n",
    "        \n",
    "#     save_path = fig_dir+'filtered_broad_diff_exp_paths.pdf'\n",
    "\n",
    "#     pathway_analyses.plot_and_select_top_deps(df, list_of_paths_to_annotate, save_prefix=save_prefix,\n",
    "#                          save_path=save_path, filter=True, cell_type_specific=False, \n",
    "#                          test_name=test_name, cell_types=celltypes)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "2f414a000012e1980f5fd1119ed6b9bf8cc73b51e37b52eb52a16ff857effd1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
