{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62042f97-a76e-4b34-805f-3f1ac81e7288",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import rpy2\n",
    "import logging\n",
    "import warnings\n",
    "import anndata2ri\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import seaborn as sb\n",
    "from tabnanny import verbose\n",
    "import matplotlib.pyplot as plt\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "from matplotlib.pyplot import rcParams\n",
    "from functions import pathway_analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c36a2bac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sys_dpi(width, height, diag):\n",
    "    '''\n",
    "    obtain dpi of system\n",
    "    \n",
    "    w: width in pixels (if unsure, go vist `whatismyscreenresolution.net`)\n",
    "    h: height in pixels\n",
    "    d: diagonal in inches\n",
    "    '''\n",
    "    w_inches = (diag**2/ (1 + height**2/width**2))**0.5\n",
    "    return round(width/w_inches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "447b6d1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "rcParams['figure.dpi'] = get_sys_dpi(1512, 982, 14.125)\n",
    "#rcParams['figure.figsize']=(4,4) #rescale figures\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdb62c2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "suppressPackageStartupMessages({\n",
    "  \n",
    "library(EnhancedVolcano)\n",
    "library(circlize)\n",
    "library(GSEABase)\n",
    "library(GSVA) \n",
    "library(ComplexHeatmap)\n",
    "library(UpSetR)\n",
    "\n",
    "library(VennDiagram)\n",
    "library(gridExtra)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(tidyverse)\n",
    "library(future.apply)\n",
    "library(ggpubr)\n",
    "library(reshape2)\n",
    "library(tictoc)\n",
    "\n",
    "library(patchwork)\n",
    "library(ggridges)\n",
    "library(RColorBrewer)\n",
    "library(eulerr)\n",
    "library(GeneOverlap)\n",
    "library(readxl)\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)\n",
    "library(tidyr)\n",
    "\n",
    "library(readxl)\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)\n",
    "library(tidyr)\n",
    "\n",
    "})\n",
    "\n",
    "# # using the cowplot theme for ggplot\n",
    "# theme_set(theme_cowplot())\n",
    "\n",
    "# # set random seed for reproducibility\n",
    "# set.seed(12345)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04f3655b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types = [\"Excitatory\", \"Inhibitory\", \"Astrocyte\", \"Microglia\", \"Oligodendrocyte\", \"OPC\",] # cell types on which meta analysis will be computed\n",
    "\n",
    "studies = ['seaad_mtg', 'leng_etc', 'leng_sfg'] # list of studies in the meta-analysis\n",
    "\n",
    "tests = ['early_vs_no', 'late_vs_early', 'ad_vs_no', 'late_vs_no'] # list of test names to consider\n",
    "\n",
    "go_node_depths = [1, 2, 3, 4, 5, 6, 7, 8] # list of depths to obtain gene ontology biological process parents\n",
    "\n",
    "deg_methods =  ['DESeq2-Wald']\n",
    "\n",
    "color_scheme_celltype = {}\n",
    "\n",
    "color_scheme_celltype['Excitatory'] = 'aquamarine3'\n",
    "color_scheme_celltype['Inhibitory'] = 'darkolivegreen3'\n",
    "color_scheme_celltype['Astrocyte'] = 'brown3'\n",
    "color_scheme_celltype['Oligodendrocyte'] = 'darkorchid3'\n",
    "color_scheme_celltype['Microglia'] = 'deepskyblue3'\n",
    "color_scheme_celltype['OPC'] = 'deeppink3'\n",
    "\n",
    "\n",
    "color_scheme_celltype =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: color_scheme_celltype[cell_type]\n",
    "                     \n",
    "                            for cell_type in cell_types\n",
    "                        }\n",
    "                    )\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3a87dcee",
   "metadata": {},
   "source": [
    "## **Parse Gene Core ontology (OBO Format)**\n",
    "\n",
    "This is the basic version of the GO, filtered such that the graph is guaranteed to be acyclic and annotations can be propagated up the graph. The relations included are is a, part of, regulates, negatively regulates and positively regulates. This version excludes relationships that cross the 3 GO hierarchies.\n",
    "\n",
    "The parsing is done with [**goscripts**](https://pmoris.github.io/goscripts/_build/html/source/README.html), a python script and package for Gene Ontology enrichment analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2ca7297",
   "metadata": {},
   "outputs": [],
   "source": [
    "from goscripts.goscripts.obo_tools import importOBO, completeChildHierarchy, buildGOtree\n",
    "\n",
    "GOdict = importOBO('../data/pathway_databases/go_terms.obo', ignore_part_of=True)\n",
    "# Generate the entire GO tree’s parent structure by walking through the hierarchy of each GO entry"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "69ec4739",
   "metadata": {},
   "source": [
    "Next we build the entire GO tree parent and child structure for only `Biological Processes with GO ID GO:0008150` by walking through the hierarchy of each GO entry using `buildGOtree` and setting the `root_node = ['GO:0008150']`.\n",
    "\n",
    "This function assigns all higher order ancestors (recursive parents) for each GO object, assigns immediate children, assigns all recursively found children, assign depth to each node."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2694be19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# here we set the root node as \"Biological Processes with GO ID GO:0008150\"\n",
    "buildGOtree(GOdict, root_nodes=['GO:0008150']) \n",
    "#completeChildHierarchy(GOdict) # Generates the entire GO tree’s child structure."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "10be21f0-000d-48d7-9e97-4051ec9da0bd",
   "metadata": {},
   "source": [
    "## **Reading in the data**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "d40204d0-56d5-4984-a2af-1075b64098f1",
   "metadata": {},
   "source": [
    "We load the preprocessed and annotated pathway differential scores for each study across tests specifies in the previosu cells:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8f8162f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load all pathways\n",
    "pathway_scores = {}\n",
    "\n",
    "for study in studies:\n",
    "    pathway_scores[study] = {}\n",
    "    for test in tests:\n",
    "        pathway_scores[study][test] = pd.read_csv(f'../results/{test}/{study}/Data/differentially_expressed_pathways.csv')\n",
    "        pathway_scores[study][test]['shortened'] = pathway_scores[study][test]['pathway']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a57e27f",
   "metadata": {},
   "outputs": [],
   "source": [
    "for study in studies:\n",
    "\n",
    "    filtered = pathway_scores[study]['ad_vs_no'][pathway_scores[study]['ad_vs_no']['P.Value']<0.05]\n",
    "    collated_df = pd.DataFrame(filtered.groupby(filtered.pathway).agg({'score_adj': list, 'celltype': list, \n",
    "                        'logFC': list, 'P.Value': list, 'shortened': list, 'highlight': list}))\n",
    "    # filter pathways only expressed in one cell type\n",
    "    unique = collated_df[\"celltype\"].apply(len) == 1\n",
    "    shared = collated_df[\"celltype\"].apply(len) > 1\n",
    "\n",
    "    print(f'{study.upper()} has {sum(unique)/len(unique)*100}% UNIQUE PATHWAYS and {sum(shared)/len(unique)*100} SHARED PATHWAYS')\n",
    "\n",
    "\n",
    "    for cell_type in cell_types:\n",
    "        total_paths = len(pathway_scores[study]['ad_vs_no'][(pathway_scores[study]['ad_vs_no']['P.Value']<0.05)&\n",
    "                                                        (pathway_scores[study]['ad_vs_no']['celltype']==cell_type)])\n",
    "        \n",
    "        up_paths = len(pathway_scores[study]['ad_vs_no'][(pathway_scores[study]['ad_vs_no']['P.Value']<0.05)&\n",
    "                                                        (pathway_scores[study]['ad_vs_no']['celltype']==cell_type)&\n",
    "                                                        (pathway_scores[study]['ad_vs_no']['logFC']>0)])\n",
    "\n",
    "        down_paths = len(pathway_scores[study]['ad_vs_no'][(pathway_scores[study]['ad_vs_no']['P.Value']<0.05)&\n",
    "                                                        (pathway_scores[study]['ad_vs_no']['celltype']==cell_type)&\n",
    "                                                        (pathway_scores[study]['ad_vs_no']['logFC']<0)])\n",
    "\n",
    "        \n",
    "\n",
    "        print(f'in {study.upper()} there are a total of {total_paths} dysregulated pathways in {cell_type}')\n",
    "        print(f'in {study.upper()} {down_paths/total_paths*100}% pathways in {cell_type} are DOWNreguled')\n",
    "        print(f'in {study.upper()} {up_paths/total_paths*100}% pathways in {cell_type} are UPreguled')\n",
    "        print(\"***********\")    \n",
    "    print('================================================')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "6acf32aa",
   "metadata": {},
   "source": [
    "### **Get parent GO name at a certain node**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf7233aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_parent_at_depth(goid, depth, GOdict):\n",
    "    \n",
    "    \"\"\"\n",
    "    Given a Gene Ontology (GO) term ID and a desired depth in the GO hierarchy, this function returns the name of the\n",
    "    first parent GO term at that depth.\n",
    "\n",
    "    Parameters:\n",
    "    -----------\n",
    "    - goid (str): A string representing the GO term ID.\n",
    "    - depth (int): An integer representing the desired depth in the GO hierarchy.\n",
    "\n",
    "    Returns:\n",
    "    --------\n",
    "    - str: The name of the parent GO term at the desired depth in the GO hierarchy.\n",
    "\n",
    "    \"\"\"\n",
    "        \n",
    "    for parent in GOdict.get(goid).recursive_parents:\n",
    "        if GOdict.get(parent).depth == depth:\n",
    "            return GOdict.get(parent).name\n",
    "        else:\n",
    "            continue"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "28581ab2",
   "metadata": {},
   "source": [
    "### **Obtain And Save Overall Overlap**\n",
    "\n",
    "This section computes the overall overlap of pathways across all studies and generates a complete hierarchy of pathway nodes, including the parent name at each node."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d55f9453",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "for test_name in tests:\n",
    "    eval_string = []\n",
    "    for i, study in enumerate(pathway_scores.keys()):\n",
    "        eval_string.append(f'set(pathway_scores[\"{study}\"][\"{test_name}\"][pathway_scores[\"{study}\"][\"{test_name}\"][\"P.Value\"]<0.05].pathway)')\n",
    "\n",
    "    eval_string = '&'.join(eval_string)\n",
    "    eval_list = list(eval(eval_string))\n",
    "    df = pd.DataFrame()\n",
    "    df['pathway'] = eval_list\n",
    "    df['GO'] = [pathway.split(' ')[-1].strip('(').strip(')') for pathway in eval_list]\n",
    "    \n",
    "    for depth in go_node_depths:\n",
    "        df[f'parent_at_node_{depth}'] = df['GO'].apply(lambda x: get_parent_at_depth(x, depth, GOdict))\n",
    "\n",
    "    dat_dir = f'../results/{test_name}/Consensus/'\n",
    "\n",
    "    if not os.path.exists(dat_dir):\n",
    "        os.makedirs(dat_dir)\n",
    "    \n",
    "    df.to_csv(dat_dir+'pathway_overlap.csv')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "d354b5f4",
   "metadata": {},
   "source": [
    "Next, we load a list of manually curated annotated pathways that we are interested in plotting. We then assign the parent at each node using the function `get_parent_at_depth()`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fde6d00b",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathways_to_annotate = pd.read_excel('../data/pathway_databases/pathways_to_annotate.xlsx')\n",
    "pathways_to_annotate['GO'] = pathways_to_annotate['pathway'].apply(lambda x: x.split(' ')[-1].strip('(').strip(')'))\n",
    "\n",
    "for depth in go_node_depths:\n",
    "        pathways_to_annotate[f'parent_at_node_{depth}'] = pathways_to_annotate['GO'].apply(lambda x: get_parent_at_depth(x, depth, GOdict))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "893f5292",
   "metadata": {},
   "source": [
    "View list of unique parents at a certain node"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db278b3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathways_to_annotate['parent_at_node_3'].unique()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "7a0d82a4",
   "metadata": {},
   "source": [
    "Based on the pathways of interest, we can curate a list of parent nodes for each level. Only pathways that belong to those parents at each level will be selected for visualization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebb13aeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "parents_to_annotate = {\n",
    "                        'node_1': [None,\n",
    "                                   'immune system process',\n",
    "                                   'cellular process',\n",
    "                                   'metabolic process',\n",
    "                                   'biological regulation',\n",
    "                                   'response to stimulus',\n",
    "                                   'homeostatic process',\n",
    "                                   'localization',\n",
    "                                  ],\n",
    "\n",
    "############### PATHWAYS AT NODE 2 #############################################                                    \n",
    "\n",
    "                        'node_2': [None,\n",
    "                                   'regulation of molecular function',\n",
    "                                   'response to stress',\n",
    "                                   'chemical homeostasis',\n",
    "                                   'regulation of biological process',\n",
    "                                   'organic substance metabolic process',\n",
    "                                   'establishment of localization',\n",
    "                                   'cellular developmental process',\n",
    "                                   'small molecule metabolic process',\n",
    "                                   'cellular metabolic process',\n",
    "                                   'cellular component organization or biogenesis',\n",
    "                                   'developmental growth',\n",
    "                                   'cellular response to stimulus',\n",
    "                                   'cell recognition',\n",
    "                                   'transmembrane transport',\n",
    "                                   'cell adhesion',\n",
    "                                   'catabolic process',\n",
    "                                   'cellular homeostasis',\n",
    "                                   'ensheathment of neurons',\n",
    "                                   'cell motility',\n",
    "                                   'signal transduction',\n",
    "                                   'response to chemical',\n",
    "                                   'regulation of biological quality',\n",
    "                                   'activation of immune response',\n",
    "                                   'regulation of pH',\n",
    "                                   'maintenance of location',\n",
    "                                   'leukocyte activation',\n",
    "                                   'anatomical structure formation involved in morphogenesis',\n",
    "                                   'cell death',\n",
    "                                   'cellular localization',\n",
    "                                   'protein-containing complex localization',\n",
    "                                   'meiotic cell cycle process',\n",
    "                                   'membrane docking',\n",
    "                                   'actin filament-based process'\n",
    "                                   ],\n",
    "\n",
    "############### PATHWAYS AT NODE3  #############################################      \n",
    "#                               \n",
    "                        \"node_3\": [None,\n",
    "                                   'regulation of catalytic activity',\n",
    "                                    'positive regulation of molecular function', \n",
    "                                    'defense response', \n",
    "                                    'lipid homeostasis',\n",
    "                                    'G protein-coupled receptor signaling pathway',\n",
    "                                    'macromolecule metabolic process', \n",
    "                                    'amide metabolic process',\n",
    "                                    'establishment of localization in cell', \n",
    "                                    'cell differentiation',\n",
    "                                    'nucleobase-containing small molecule metabolic process',\n",
    "                                    'cellular component organization', \n",
    "                                    'developmental cell growth',\n",
    "                                    'cellular response to stress', \n",
    "                                    'import across plasma membrane',\n",
    "                                    'transport', 'cell-cell adhesion',\n",
    "                                    'regulation of cellular process', \n",
    "                                    'monoatomic ion homeostasis',\n",
    "                                    'cellular response to chemical stimulus', \n",
    "                                    'axon ensheathment',\n",
    "                                    'cell migration', \n",
    "                                    'cell-cell signaling', \n",
    "                                    'nervous system process',\n",
    "                                    'intracellular transport', \n",
    "                                    'response to salt',\n",
    "                                    'vesicle localization', \n",
    "                                    'regulation of membrane potential',\n",
    "                                    'cell development', \n",
    "                                    'amino acid metabolic process',\n",
    "                                    'organic acid metabolic process',\n",
    "                                    'cell surface receptor signaling pathway',\n",
    "                                    'regulation of immune system process', \n",
    "                                    'cell-cell recognition',\n",
    "                                    'organic hydroxy compound metabolic process',\n",
    "                                    'regulation of cellular pH', \n",
    "                                    'maintenance of protein location',\n",
    "                                    'glial cell activation',\n",
    "                                    'regulation of multicellular organismal process',\n",
    "                                    'cellular component assembly involved in morphogenesis',\n",
    "                                    'regulation of metabolic process', \n",
    "                                    'regulation of localization',\n",
    "                                    'cellular macromolecule localization',\n",
    "                                    'positive regulation of biological process',\n",
    "                                    'negative regulation of molecular function',\n",
    "                                    'regulation of DNA-binding transcription factor activity',\n",
    "                                    'neuron death', \n",
    "                                    'carbohydrate derivative metabolic process',\n",
    "                                    'localization within membrane',\n",
    "                                    'positive regulation of cytosolic calcium ion concentration',\n",
    "                                    'regulation of growth', \n",
    "                                    'regulation of developmental process',\n",
    "                                    'regulation of locomotion', \n",
    "                                    'regulation of response to stimulus',\n",
    "                                    'response to nitrogen compound', \n",
    "                                    'response to wounding',\n",
    "                                    'organelle localization by membrane tethering',\n",
    "                                    ],\n",
    "\n",
    "############### PATHWAYS AT NODE 5 ############################################# \n",
    "#                                    \n",
    "                        \"node_5\": [None,\n",
    "                                   'positive regulation of cyclase activity',\n",
    "                                   'positive regulation of hydrolase activity', \n",
    "                                   'adenylate cyclase-activating G protein-coupled receptor signaling pathway',\n",
    "                                   'adenylate cyclase-inhibiting G protein-coupled receptor signaling pathway',\n",
    "                                   'synaptic vesicle cytoskeletal transport',\n",
    "                                   'nucleoside triphosphate metabolic process',\n",
    "                                   'organelle disassembly',\n",
    "                                   'plasma membrane bounded cell projection organization',\n",
    "                                   'neuron projection extension involved in neuron projection guidance',\n",
    "                                   'mitochondrion transport along microtubule',\n",
    "                                   'cell projection morphogenesis',\n",
    "                                   'calcium ion transmembrane transport',\n",
    "                                   'monoatomic cation transport', \n",
    "                                   'regulated exocytosis',\n",
    "                                   'calcium-mediated signaling', \n",
    "                                   'ribonucleotide metabolic process',\n",
    "                                   'cell-cell junction organization',\n",
    "                                   'intracellular monoatomic cation homeostasis',\n",
    "                                   'response to amyloid-beta', \n",
    "                                   'telencephalon cell migration',\n",
    "                                   'cyclic-nucleotide-mediated signaling', \n",
    "                                   'trans-synaptic signaling',\n",
    "                                   'endocytosis',\n",
    "                                   'regulation of G protein-coupled receptor signaling pathway',\n",
    "                                   'endoplasmic reticulum organization',\n",
    "                                   'proteasomal protein catabolic process', \n",
    "                                   'synapse organization',\n",
    "                                   'glutamine family amino acid metabolic process',\n",
    "                                   'positive regulation of immune response',\n",
    "                                   'mitochondrion organization', \n",
    "                                   'protein processing',\n",
    "                                   'modulation of chemical synaptic transmission',\n",
    "                                   'plasma membrane organization',\n",
    "                                   'negative regulation of cyclase activity',\n",
    "                                   'regulation of protein metabolic process',\n",
    "                                   'negative regulation of cellular component organization',\n",
    "                                   'regulation of monoatomic ion transport',\n",
    "                                   'regulation of exocytosis',\n",
    "                                   'regulation of intracellular signal transduction',\n",
    "                                   'negative regulation of nitrogen compound metabolic process',\n",
    "                                   'regulation of cell projection organization',\n",
    "                                   'negative regulation of signal transduction',\n",
    "                                   'regulation of programmed cell death',\n",
    "                                   'positive regulation of cell communication',\n",
    "                                   'regulation of defense response',\n",
    "                                   'regulation of protein transport',\n",
    "                                   'regulation of intracellular transport',\n",
    "                                   'regulation of nervous system development',\n",
    "                                   'regulation of neural precursor cell proliferation',\n",
    "                                   'negative regulation of cell differentiation',\n",
    "                                    'regulation of response to endoplasmic reticulum stress',\n",
    "                                    'regulation of receptor binding',\n",
    "                                    'positive regulation of macromolecule metabolic process',\n",
    "                                    'monoatomic anion transport',\n",
    "                                    'negative regulation of synaptic transmission',\n",
    "                                    'regulation of transcription regulatory region DNA binding',\n",
    "                                    'negative regulation of immune response',\n",
    "                                    'regulation of cellular biosynthetic process',\n",
    "                                    'protein-containing complex assembly', \n",
    "                                    'chromosome organization',\n",
    "                                    'glycolipid metabolic process', \n",
    "                                    'lipoprotein localization',\n",
    "                                    'negative regulation of catabolic process', \n",
    "                                    'lysosomal transport',\n",
    "                                    'regulation of extracellular matrix organization',\n",
    "                                    'regulation of apoptotic signaling pathway',\n",
    "                                    'regulation of macrophage activation',\n",
    "                                    'postsynaptic specialization organization',\n",
    "                                    'membrane invagination',\n",
    "                                    'regulation of axon extension',\n",
    "                                    'membrane fusion',\n",
    "                                    'regulation of monoatomic ion transmembrane transport',\n",
    "                                    'regulation of adaptive immune response',\n",
    "                                    'regulation of phosphorus metabolic process',\n",
    "                                    'positive regulation of vasculature development',\n",
    "                                    'positive regulation of endocytosis',\n",
    "                                    'positive regulation of biosynthetic process',\n",
    "                                    'regulation of cellular catabolic process',\n",
    "                                    'positive regulation of cell growth',\n",
    "                                    'extracellular matrix organization',\n",
    "                                    'regulation of cell adhesion mediated by integrin',\n",
    "                                    'regulation of endocytosis',\n",
    "                                    'regulation of nervous system process',\n",
    "                                    'positive regulation of secretion by cell',\n",
    "                                    'regulation of lipid transport',\n",
    "                                    'regulation of glutamate receptor signaling pathway',\n",
    "                                    'regulation of lipid metabolic process',\n",
    "                                    'regulation of neuron differentiation',\n",
    "                                    'positive regulation of cell motility', \n",
    "                                    'regulation of secretion',\n",
    "                                    'positive regulation of protein localization',\n",
    "                                    'positive regulation of cellular component organization',\n",
    "                                    'membrane assembly', \n",
    "                                    'presynaptic membrane organization',\n",
    "                                    'protein localization to cell junction',\n",
    "                                    'regulation of monoatomic ion transmembrane transporter activity',\n",
    "                                    'regulation of vasculature development',\n",
    "                                    'regulation of extent of cell growth',\n",
    "                                    'regulation of response to oxidative stress',\n",
    "                                    'regulation of DNA metabolic process',\n",
    "                                    'regulation of neuron death', \n",
    "                                    'regulation of cell migration',\n",
    "                                    'regulation of protein localization',\n",
    "                                    'regulation of protein-containing complex assembly',\n",
    "                                    'regulation of cell junction assembly',\n",
    "                                    'regulation of synapse organization', \n",
    "                                    'axo-dendritic transport',\n",
    "                                    'negative regulation of cell cycle',\n",
    "                                    'DNA damage checkpoint signaling',\n",
    "                                    'vesicle-mediated transport in synapse',\n",
    "                                    'actin filament-based transport',\n",
    "                                    ],\n",
    "\n",
    "############### PATHWAYS AT NODE 6 #############################################                                    \n",
    "\n",
    "                        \"node_6\": [None,\n",
    "                                   'positive regulation of adenylate cyclase activity',\n",
    "                                   # 'regulation of endopeptidase activity',  \n",
    "                                   'axonal transport',\n",
    "                                   # 'ribonucleoside triphosphate metabolic process',\n",
    "                                   # 'mitochondrion disassembly', \n",
    "                                   'neuron projection development',\n",
    "                                   # 'plasma membrane bounded cell projection morphogenesis',\n",
    "                                   'calcium ion transmembrane import into cytosol',\n",
    "                                   'metal ion transport', \n",
    "                                   # 'cyclic nucleotide metabolic process',\n",
    "                                   # 'nucleotide catabolic process',\n",
    "                                   'anterograde trans-synaptic signaling',\n",
    "                                   'receptor-mediated endocytosis',\n",
    "                                   'negative regulation of G protein-coupled receptor signaling pathway',\n",
    "                                   'protein quality control for misfolded or incompletely synthesized proteins',\n",
    "                                   'synapse assembly', \n",
    "                                   # 'dicarboxylic acid catabolic process',\n",
    "                                   # 'dicarboxylic acid metabolic process',\n",
    "                                   'negative regulation of protein metabolic process',\n",
    "                                   'regulation of plasma membrane bounded cell projection organization',\n",
    "                                   'negative regulation of calcium ion transport',\n",
    "                                   'regulation of metal ion transport',\n",
    "                                   'regulation of regulated secretory pathway',\n",
    "                                   'negative regulation of intracellular signal transduction',\n",
    "                                   'regulation of protein kinase activity',\n",
    "                                   'regulation of monoatomic cation transmembrane transport',\n",
    "                                   'negative regulation of cation channel activity',\n",
    "                                   'endoplasmic reticulum calcium ion homeostasis',\n",
    "                                   'negative regulation of programmed cell death',\n",
    "                                   'positive regulation of intracellular signal transduction',\n",
    "                                   #'regulation of establishment of protein localization',\n",
    "                                   'regulation of neurogenesis',\n",
    "                                   'negative regulation of potassium ion transport',\n",
    "                                   'regulation of apoptotic process',\n",
    "                                   'regulation of receptor-mediated endocytosis',\n",
    "                                   'regulation of synaptic transmission, glutamatergic',\n",
    "                                   # 'regulation of RNA biosynthetic process',\n",
    "                                   'neuron projection organization',\n",
    "                                   'regulation of proteasomal protein catabolic process',\n",
    "                                   'regulation of phagocytosis', \n",
    "                                   #'protein localization to vacuole',\n",
    "                                   'regulation of calcium-mediated signaling',\n",
    "                                   # 'positive regulation of extracellular matrix organization',\n",
    "                                   # 'positive regulation of programmed cell death',\n",
    "                                   'regulation of microglial cell activation',\n",
    "                                   'mitochondrial calcium ion homeostasis',\n",
    "                                   'positive regulation of potassium ion transport',\n",
    "                                   'positive regulation of calcium ion transport',\n",
    "                                   'postsynaptic density organization',\n",
    "                                   'plasma membrane invagination',\n",
    "                                   'regulation of acute inflammatory response',\n",
    "                                   'regulation of phosphate metabolic process',\n",
    "                                   'regulation of neuroinflammatory response',\n",
    "                                   # 'apoptotic mitochondrial changes',\n",
    "                                   'positive regulation of amide metabolic process',\n",
    "                                   'positive regulation of phagocytosis',\n",
    "                                   'regulation of astrocyte differentiation',\n",
    "                                   'regulation of autophagy',\n",
    "                                   'positive regulation of gene expression',\n",
    "                                   'regulation of sterol transport', \n",
    "                                   # 'regulation of neuron migration',\n",
    "                                   'positive regulation of cell projection organization',\n",
    "                                   'regulation of protein localization to cell surface',\n",
    "                                   'positive regulation of sodium ion transport',\n",
    "                                   'regulation of sodium ion transmembrane transporter activity',\n",
    "                                   'regulation of synapse assembly',\n",
    "                                   'regulation of glycoprotein metabolic process',\n",
    "                                   'regulation of amyloid precursor protein catabolic process',\n",
    "                                   'regulation of programmed necrotic cell death',\n",
    "                                   'sarcoplasmic reticulum calcium ion transport',\n",
    "                                   # 'negative regulation of cell cycle process',\n",
    "                                   # 'organelle membrane fusion',\n",
    "                                   ]\n",
    "                       }"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4d96ef39",
   "metadata": {},
   "source": [
    "Filter the overall list of pathways of interest using the curated list of parents described above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d51206c",
   "metadata": {},
   "outputs": [],
   "source": [
    "list_of_paths_to_annotate = pathways_to_annotate[pathways_to_annotate.parent_at_node_6.isin(parents_to_annotate['node_6'])]['pathway'].to_list()\n",
    "#list_of_paths_to_annotate = pathways_to_annotate['pathway'].to_list()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "0c5fa283",
   "metadata": {},
   "source": [
    "### **Heatmap showing select diffentially expressed pathways accross all cell_types in each study**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "169c12e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i list_of_paths_to_annotate -i tests -i studies -o mat -i cell_types -i color_scheme_celltype\n",
    "\n",
    "\n",
    "source('../scripts/functions/pathway_analyses.R')\n",
    "\n",
    "# import the renamed-pathway names\n",
    "\n",
    "print('importing pathway renaming convention')\n",
    "\n",
    "path_names_go <- as.data.frame(read_excel('../data/pathway_databases/pathways_to_annotate.xlsx'))\n",
    "# path_names_go <- data.frame(pathway = unique(path_names_go$pathway))\n",
    "rownames(path_names_go) <- path_names_go$pathway\n",
    "\n",
    "# names of pathways that will be annotated\n",
    "names = as.vector(list_of_paths_to_annotate)\n",
    "names = rownames(path_names_go)\n",
    "\n",
    "for (study in studies) {\n",
    "\n",
    "  for (test_name in tests) {\n",
    "    # Read in pathway data\n",
    "    pathway_scores <- as.data.frame(read.csv(sprintf('../results/%s/%s/Data/differentially_expressed_pathways.csv', test_name, study), header=TRUE))\n",
    "    pathway_scores <- pathway_scores[pathway_scores$pathway %in% rownames(path_names_go), ]\n",
    "    \n",
    "    # select only significant pathways\n",
    "    pathway_scores[abs(pathway_scores$score_adj) < 1.3, \"score_adj\"] <- 0\n",
    "    pathway_scores_subset <- pathway_scores[, c(\"pathway\", \"celltype\", \"score_adj\")]\n",
    "    pivot_table <- pathway_scores_subset %>% pivot_wider(names_from = \"celltype\", values_from = \"score_adj\")\n",
    "\n",
    "    pivot_table <- as.data.frame(pivot_table)\n",
    "\n",
    "    # Set the 'pathway' column as the row names (index)\n",
    "    row.names(pivot_table) <- pivot_table$pathway\n",
    "    \n",
    "    pivot_table$pathway <- NULL\n",
    "    ct_order <- intersect(cell_types, colnames(pivot_table))\n",
    "    mat <- pivot_table[, ct_order] \n",
    "    mat <-  mat %>%\n",
    "            mutate_all(~coalesce(., 0))\n",
    "    \n",
    "    # draw heatmap for all GO pathways\n",
    "    unique = mat[unname(rowSums(abs(mat)>1.3)==1),] # obtain cell_type_specific pathways.\n",
    "    shared_2 = mat[unname(rowSums(abs(mat)>1.3)>1),] # obtain shared pathways\n",
    "\n",
    "    h1 = get_go_hmap(unique, names, path_names_go, 'shortened_name', cell_type_order=ct_order)\n",
    "    h2 = get_go_hmap(shared_2, names, path_names_go, 'shortened_name', cell_type_order=ct_order)\n",
    "\n",
    "\n",
    "    fig_dir = paste0('../results/', test_name, '/', study, '/HeatMap/')\n",
    "\n",
    "    if (!dir.exists(fig_dir)) {\n",
    "    dir.create(fig_dir, recursive=TRUE)\n",
    "    }\n",
    "\n",
    "    pdf(paste0(fig_dir, 'filtered_diff_exp_paths.pdf'), height = 10, width = 8.5)\n",
    "    print(h1%v%h2)\n",
    "    dev.off()\n",
    "  }\n",
    "}\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "7eab8933",
   "metadata": {},
   "source": [
    "### **Barcharts showing the distribution of up- and down-regulated pathways in each celltype for each study**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc759623",
   "metadata": {},
   "outputs": [],
   "source": [
    "for test_name in tests:\n",
    "    \n",
    "    fig, axs = plt.subplots(nrows=1, ncols=3, figsize=(11, 6), sharex=False, sharey = True, layout='constrained')\n",
    "    axs = axs.ravel()\n",
    "\n",
    "    print(f'Directionality of pathways in all studies in {test_name.upper()}')\n",
    "\n",
    "    for ind, study in enumerate(studies):\n",
    "        df = pathway_scores[study][test_name][pathway_scores[study][test_name].celltype.isin(cell_types)]\n",
    "        df = df[abs(df.score_adj)>1.3]\n",
    "        df['Direction'] = df.score_adj.apply(lambda x: 'down' if x<0 else \"up\")\n",
    "\n",
    "        sb.countplot(data=df, x='celltype', hue='Direction', hue_order=['down', 'up'],\n",
    "                     ax=axs[ind], linewidth=4, order=cell_types, \n",
    "                     palette={'up': 'black', 'down': 'grey'},\n",
    "                     )\n",
    "\n",
    "        axs[ind].set_title(study, fontdict={'fontsize': 10})\n",
    "        axs[ind].yaxis.label.set_fontsize(10)\n",
    "        axs[ind].set_xticklabels(cell_types, rotation=90,)\n",
    "        axs[ind].spines[['right', 'top']].set_visible(False)\n",
    "           \n",
    "\n",
    "    fig_dir = f'../results/{test_name}/Consensus/Barchat/'\n",
    "\n",
    "    if not os.path.exists(fig_dir):\n",
    "        os.makedirs(fig_dir)\n",
    "\n",
    "    plt.savefig(fig_dir+'directionality_distribution_of_pathways.pdf')\n",
    "    plt.show()\n",
    "    "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "0f2a06bb",
   "metadata": {},
   "source": [
    "### **Barcharts showing the GSVA enrichment score of pathways that are differentially expressed with different directionality across any group of cell_types in each study**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf691426",
   "metadata": {},
   "outputs": [],
   "source": [
    "for test_name in tests:\n",
    "    \n",
    "    fig, axs = plt.subplots(nrows=1, ncols=3, figsize=(11, 11), sharex=False, sharey = True, layout='constrained')\n",
    "    axs = axs.ravel()\n",
    "\n",
    "    print(f'Directionality of pathways in all studies in {test_name.upper()}')\n",
    "\n",
    "    for ind, study in enumerate(studies):\n",
    "        df = pathway_scores[study][test_name][pathway_scores[study][test_name].celltype.isin(cell_types)]\n",
    "        df = df[abs(df.score_adj)>1.3]\n",
    "        df['Direction'] = df.score_adj.apply(lambda x: 'down' if x<0 else \"up\")\n",
    "        df = df[df.pathway.isin([key for key, value in dict(df['pathway'].value_counts() > 1).items() if value == True])]\n",
    "        keys = dict(df.groupby('pathway')['Direction'].value_counts()).keys()\n",
    "        \n",
    "        pathways = [key[0] for key in keys if ((key[0], \"down\") in keys) and ((key[0], \"up\") in keys)]\n",
    "\n",
    "\n",
    "        df = df[df.pathway.isin(pathways)]\n",
    "\n",
    "        print(f'Number of pathways with opposing directionality in {study.upper()}: {len(df)}')\n",
    "\n",
    "        if len(df[df.pathway.isin(list_of_paths_to_annotate)])>0:\n",
    "            \n",
    "            df = df[df.pathway.isin(list_of_paths_to_annotate)]\n",
    "            df['pathway'] = df['pathway'].map(dict(zip(pathways_to_annotate.pathway,\n",
    "                                                       pathways_to_annotate.shortened_name)))\n",
    "           \n",
    "        sb.barplot(data=df, x='pathway', y='logFC', hue='celltype', \n",
    "                   hue_order = sorted(list(df.celltype.unique()), key=lambda x: cell_types.index(x)),\n",
    "                     ax=axs[ind], linewidth=4, #order=cell_types, \n",
    "                     palette={'Excitatory': [x/255 for x in [102, 204, 168]], \n",
    "                              'Inhibitory':  [x/255 for x in [164, 204, 89]],\n",
    "                              'Astrocyte':  [x/255 for x in [204, 51, 51]],\n",
    "                              'Microglia':  [x/255 for x in [0, 153, 204]],\n",
    "                              'Oligodendrocyte':  [x/255 for x in [153, 48, 204]],\n",
    "                              'OPC':  [x/255 for x in [204, 15, 117]],\n",
    "                              \n",
    "                              },\n",
    "                     )\n",
    "\n",
    "\n",
    "        axs[ind].set_title(study, fontdict={'fontsize': 10})\n",
    "        axs[ind].yaxis.label.set_fontsize(10)\n",
    "        axs[ind].set_xticklabels(df.pathway.unique(), rotation=90,)\n",
    "        axs[ind].spines[['right', 'top']].set_visible(False)\n",
    "\n",
    "\n",
    "    fig_dir = f'../results/{test_name}/Consensus/BarChat/'\n",
    "\n",
    "    if not os.path.exists(fig_dir):\n",
    "        os.makedirs(fig_dir)\n",
    "         \n",
    "    plt.savefig(fig_dir+'directionality_of_select_pathways.pdf')\n",
    "    plt.show()\n",
    "    "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "94d806ec",
   "metadata": {},
   "source": [
    "### **Upset plot to show how many pathways are shared & how many are cell type specific in each study**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6de3127",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o upset_df \n",
    "\n",
    "\n",
    "all_combinations <- lapply(2:length(cell_types), function(k) combn(cell_types, k, simplify = FALSE))\n",
    "\n",
    "pathway_scores <- list()\n",
    "\n",
    "for (study in studies) {\n",
    "  pathway_scores[[study]] <- list()\n",
    "    for (test in tests) {\n",
    "    # Read in pathway data from a CSV file for the current test and study\n",
    "    all_pathways <- read.csv(sprintf('../results/%s/%s/Data/differentially_expressed_pathways.csv', test, study), header=TRUE)\n",
    "    \n",
    "    # Filter the pathway data to select only significant pathways\n",
    "    pathway_scores[[study]][[test]] <- all_pathways[abs(all_pathways[\"score_adj\"]) > 1.3, ]\n",
    "  }\n",
    "}\n",
    "\n",
    "upset_df <- list() \n",
    "for(study in studies){\n",
    "  upset_df[[study]] <- list()\n",
    "  for(test_name in tests){ \n",
    "    scores <- pathway_scores[[study]][[test_name]] # subset the pathway scores dataframe to only include pathways that are common to all cell types in the current test\n",
    "    scores$celltype <- as.character(scores$celltype) \n",
    "    scores$pathway <- factor(as.character(scores$pathway), levels=unique(scores$pathway))  \n",
    "\n",
    "    upset_list <- list()  \n",
    "\n",
    "    for(cur_celltype in cell_types){ \n",
    "      test_df <- scores %>% subset(celltype == cur_celltype) \n",
    "      upset_list[[cur_celltype]] <- table(test_df$pathway) \n",
    "    }\n",
    "\n",
    "    upset_df[[study]][[test_name]] <- as.data.frame(Reduce(cbind, upset_list))\n",
    "    colnames(upset_df[[study]][[test_name]]) <- cell_types\n",
    "\n",
    "    # remove entries with all 0\n",
    "    upset_df[[study]][[test_name]] <- upset_df[[study]][[test_name]][rowSums(upset_df[[study]][[test_name]]) > 0,]\n",
    "  \n",
    "    # convert to binary\n",
    "    upset_df[[study]][[test_name]][upset_df[[study]][[test_name]] > 0] <- 1\n",
    "\n",
    "    fig_dir = paste0('../results/', test_name, '/', study, '/Upset/')\n",
    "\n",
    "    if (!dir.exists(fig_dir)) {\n",
    "      dir.create(fig_dir, recursive=TRUE)\n",
    "    }\n",
    "    \n",
    "    col_names <- colnames(upset_df[[study]][[test_name]])\n",
    "\n",
    "    cur_color_scheme <- unlist(color_scheme_celltype)[match(col_names, names(color_scheme_celltype))]\n",
    "\n",
    "    cur_color_scheme <- cur_color_scheme[order(colSums(upset_df[[study]][[test_name]]), decreasing = TRUE)]\n",
    "    \n",
    "    unique <- upset_df[[study]][[test_name]][rowSums(upset_df[[study]][[test_name]])<=1, ]\n",
    "    shared <- upset_df[[study]][[test_name]][rowSums(upset_df[[study]][[test_name]])>=2, ]\n",
    "\n",
    "    svg(paste0(fig_dir, 'upset_individual_unique_', study, '.svg'), width=8, height=4)\n",
    "    print(upset(unique,\n",
    "              sets=as.character(cell_types),\n",
    "              #group.by='sets',\n",
    "              order.by='freq',\n",
    "              #cutoff=4,\n",
    "              sets.bar.color=cur_color_scheme,\n",
    "            ))\n",
    "\n",
    "    \n",
    "    col_names <- colnames(unique)\n",
    "\n",
    "    cur_color_scheme <- unlist(color_scheme_celltype)[match(col_names, names(color_scheme_celltype))]\n",
    "\n",
    "    cur_color_scheme <- cur_color_scheme[order(colSums(unique), decreasing = TRUE)]\n",
    "\n",
    "    pdf(paste0(fig_dir, 'upset_individual_unique_', study, '.pdf'), width=8, height=4)\n",
    "    print(upset(unique,\n",
    "              sets=as.character(cell_types),\n",
    "              #group.by='sets',\n",
    "              order.by='freq',\n",
    "              #cutoff=4,\n",
    "              sets.bar.color=cur_color_scheme,\n",
    "            ))\n",
    "\n",
    "    dev.off()\n",
    "\n",
    "    svg(paste0(fig_dir, 'upset_individual_shared_', study, '.svg'), width=8, height=4)\n",
    "    print(upset(shared,\n",
    "              sets=as.character(cell_types),\n",
    "              #group.by='sets',\n",
    "              order.by='freq',\n",
    "              #cutoff=4,\n",
    "              sets.bar.color=cur_color_scheme,\n",
    "            ))\n",
    "    \n",
    "    col_names <- colnames(shared)\n",
    "\n",
    "    cur_color_scheme <- unlist(color_scheme_celltype)[match(col_names, names(color_scheme_celltype))]\n",
    "\n",
    "    cur_color_scheme <- cur_color_scheme[order(colSums(shared), decreasing = TRUE)]\n",
    "\n",
    "    pdf(paste0(fig_dir, 'upset_individual_shared_', study, '.pdf'), width=8, height=4)\n",
    "    print(upset(shared,\n",
    "              sets=as.character(cell_types),\n",
    "              #group.by='sets',\n",
    "              order.by='freq',\n",
    "              #cutoff=4,\n",
    "              sets.bar.color=cur_color_scheme,\n",
    "            ))\n",
    "\n",
    "    dev.off()\n",
    "  }\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "66a4bc88",
   "metadata": {},
   "source": [
    "### **Heatmap Visualizing Shared Differential Pathway Expression Across All Studies**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0da2c0e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathways_to_annotate = pd.read_excel('../data/pathway_databases/pathways_to_annotate.xlsx')\n",
    "filtered_pathways = pathways_to_annotate['pathway'].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d78f39a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o common_pathways -o valid_test_names\n",
    "\n",
    "\n",
    "source('../scripts/functions/pathway_analyses.R')\n",
    "\n",
    "cell_types = c(\"Excitatory\", \"Inhibitory\", \"Astrocyte\", \"Microglia\", \"Oligodendrocyte\", \"OPC\")\n",
    "\n",
    "print('importing pathway renaming convention')\n",
    "\n",
    "path_names_go <- as.data.frame(read_excel('../data/pathway_databases/pathways_to_annotate.xlsx'))\n",
    "# path_names_go <- data.frame(pathway = unique(path_names_go$pathway))\n",
    "rownames(path_names_go) <- path_names_go$pathway\n",
    "\n",
    "names = as.vector(list_of_paths_to_annotate)\n",
    "names = rownames(path_names_go)\n",
    "\n",
    "pathway_scores <- list()\n",
    "\n",
    "for (study in studies) {\n",
    "  pathway_scores[[study]] <- list()\n",
    "    for (test_name in tests) {\n",
    "    all_pathways <- read.csv(sprintf('../results/%s/%s/Data/differentially_expressed_pathways.csv', test_name, study), header=TRUE)\n",
    "    pathway_scores[[study]][[test_name]] <- all_pathways[abs(all_pathways[\"score_adj\"]) > 1.3, ]\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "common_pathways <- list()\n",
    "\n",
    "for (test_name in tests) {\n",
    "  \n",
    "  cur_study_pathways <- list()\n",
    "  \n",
    "  for (study in studies) {\n",
    "    cur_study_pathways[[study]] <- pathway_scores[[study]][[test_name]]\n",
    "\n",
    "    cur_study_pathways[[study]] <- as.character(unique(cur_study_pathways[[study]]$pathway))\n",
    "  }\n",
    "  \n",
    "  common_pathways[[test_name]] <- Reduce(intersect, cur_study_pathways)\n",
    "  common_pathways[[test_name]] <- intersect(common_pathways[[test_name]], names)\n",
    "\n",
    "  if (length(common_pathways[[test_name]])==0){\n",
    "    common_pathways[[test_name]] <- NULL\n",
    "  }\n",
    "  \n",
    "}\n",
    "\n",
    "valid_test_names <- names(common_pathways)\n",
    "\n",
    "\n",
    "for (test_name in valid_test_names) {\n",
    "\n",
    "  tryCatch({\n",
    "\n",
    "  h <- list()\n",
    "  ha <- NULL\n",
    "\n",
    "  for (study in studies) {\n",
    "\n",
    "    scores <- pathway_scores[[study]][[test_name]][pathway_scores[[study]][[test_name]]$pathway %in% common_pathways[[test_name]], ]\n",
    "\n",
    "    \n",
    "    # Filter the pathway data to select only significant pathways\n",
    "    scores[abs(scores$score_adj) < 1.3, \"score_adj\"] <- 0\n",
    "    pathway_scores_subset <- scores[, c(\"pathway\", \"celltype\", \"score_adj\")]\n",
    "    pivot_table <- pathway_scores_subset %>% pivot_wider(names_from = \"celltype\", values_from = \"score_adj\")\n",
    "\n",
    "    pivot_table <- as.data.frame(pivot_table)\n",
    "\n",
    "    row.names(pivot_table) <- pivot_table$pathway\n",
    "\n",
    "    pivot_table$pathway <- NULL\n",
    "    ct_order <- intersect(cell_types, colnames(pivot_table))\n",
    "    mat <- pivot_table[, ct_order]\n",
    "    mat <- mat[rownames(mat) %in% path_names_go$pathway, ] \n",
    "    mat <-  mat %>%\n",
    "        mutate_all(~coalesce(., 0))\n",
    "    # draw heatmap for all GO pathways\n",
    "    print('drawing heatmaps - GO')\n",
    "    h[[study]] = get_go_hmap(mat, names, path_names_go, \n",
    "                            'shortened_name', cell_type_order=ct_order,\n",
    "                            reorder=TRUE)\n",
    "\n",
    "  }\n",
    "\n",
    "  fig_dir = paste0('../results/', test_name, '/Consensus/HeatMap/')\n",
    "\n",
    "  if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "  }\n",
    "\n",
    "  pdf(paste0(fig_dir, 'filtered_multi_study_pathway_overlap_R.pdf'), height = 21, width = 55.5)\n",
    "  print(Reduce('+', h))\n",
    "  dev.off()\n",
    "\n",
    "}, error = function(e) {\n",
    "    # Print the error message\n",
    "    cat(\"Error occurred for test:\", test_name, \"\\n\")\n",
    "    message(e)\n",
    "    # Continue to the next iteration\n",
    "    return()\n",
    "  })\n",
    "}\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b197f930",
   "metadata": {},
   "source": [
    "### **Venn Diagrams showing number of overlapping pathways across all studiess in each cell type**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5306a6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "\n",
    "pathway_scores <- list()\n",
    "studies <- studies\n",
    "all_pathways <- list()\n",
    "for (study in studies) {\n",
    "  pathway_scores[[study]] <- list()\n",
    "  all_pathways[[study]] <- list()\n",
    "  for (test in tests) {\n",
    "    all_pathways[[study]][[test]] <- read.csv(sprintf('../results/%s/%s/Data/differentially_expressed_pathways.csv', test, study), header=TRUE)\n",
    "    \n",
    "    pathway_scores[[study]][[test]] <- all_pathways[[study]][[test]][abs(all_pathways[[study]][[test]][\"score_adj\"]) > 1.3, ]\n",
    "  }\n",
    "}\n",
    "\n",
    "library(GeneOverlap)\n",
    "\n",
    "\n",
    "for(test_name in valid_test_names){\n",
    "    \n",
    "  for(cur_celltype in cell_types){\n",
    "    \n",
    "      # get celltype degs for this celltype:\n",
    "      overlap_df <- data.frame()\n",
    "      pathway_set_list <- list()\n",
    "      for (study in studies){\n",
    "        cur_celltype_pathways <- pathway_scores[[study]][[test_name]] %>% subset(celltype == cur_celltype)\n",
    "        pathway_set_list[[toupper(tail(strsplit(study, \"_\")[[1]], -1))]] = as.character(unique(cur_celltype_pathways$pathway))\n",
    "        for (study2 in studies){\n",
    "          if (study!=study2){\n",
    "            cur_celltype_pathways2 <- pathway_scores[[study2]][[test_name]] %>% subset(celltype == cur_celltype)\n",
    "            genome.size <- nrow(all_pathways[[study]][[test]] %>% subset(celltype == cur_celltype)) + nrow(all_pathways[[study2]][[test]] %>% subset(celltype == cur_celltype))\n",
    "\n",
    "            gene_set1 <- as.character(unique(cur_celltype_pathways$pathway))\n",
    "            gene_set2 <- as.character(unique(cur_celltype_pathways2$pathway))\n",
    "\n",
    "            cur_overlap <- testGeneOverlap(newGeneOverlap(gene_set1, gene_set2, genome.size = genome.size))\n",
    "\n",
    "            temp_df <- data.frame(\n",
    "              'odds.ratio' = cur_overlap@odds.ratio,\n",
    "              'pval' = cur_overlap@pval,\n",
    "              'Jaccard' = cur_overlap@Jaccard,\n",
    "              'size_intersection' = length(cur_overlap@intersection),\n",
    "              'combination' = paste(study, study2, sep = \"_\")  # Concatenate study names\n",
    "            )\n",
    "            \n",
    "            overlap_df <- rbind(overlap_df, temp_df)\n",
    "          }\n",
    "        }\n",
    "      }\n",
    "\n",
    "      fig_dir = paste0('../results/', test_name, '/Consensus/Venn/')\n",
    "\n",
    "      if (!dir.exists(fig_dir)) {\n",
    "      dir.create(fig_dir, recursive=TRUE)\n",
    "      }\n",
    "\n",
    "      write.csv(overlap_df, paste0(fig_dir, 'overlap_stats_', cur_celltype, \".csv\"), row.names = FALSE)\n",
    "\n",
    "      venn_list <- list()\n",
    "      for(i in 1:length(pathway_set_list)){\n",
    "        cur_pathways <- pathway_set_list[[i]]\n",
    "    \n",
    "        other_indices <- (i:length(pathway_set_list))[i:length(pathway_set_list) != i]\n",
    "    \n",
    "        # number of pathways in cur set and not other sets:\n",
    "        venn_list[[names(pathway_set_list)[i]]] <- sum(!(cur_pathways %in% Reduce(union, \n",
    "                                                        pathway_set_list[other_indices])))\n",
    "    \n",
    "    \n",
    "        for(j in other_indices){\n",
    "          other_pathways <- pathway_set_list[[j]]\n",
    "          venn_list[[paste0(names(pathway_set_list)[i],'&',\n",
    "                    names(pathway_set_list)[j])]] <- length(intersect(cur_pathways, other_pathways))\n",
    "        }\n",
    "      }\n",
    "    \n",
    "      # number of pathways in common between all sets:\n",
    "      venn_list[[paste0(names(pathway_set_list)[1], '&',\n",
    "                        names(pathway_set_list)[2], '&', \n",
    "                        names(pathway_set_list)[3])]] <- length(Reduce(intersect, pathway_set_list))\n",
    "    \n",
    "\n",
    "\n",
    "      VennDiag <- euler(unlist(venn_list))\n",
    "      svg(paste0(fig_dir, 'venn_',cur_celltype,'.svg'), width=6, height=6)\n",
    "  \n",
    "      print(plot(\n",
    "        VennDiag, quantities=TRUE, counts = TRUE, font=5, cex=1, alpha=0.8,\n",
    "        fill=rep(color_scheme_celltype[[cur_celltype]], 3)\n",
    "      ))\n",
    "\n",
    "      pdf(paste0(fig_dir, 'venn_',cur_celltype,'.pdf'), width=6, height=6)  \n",
    "      print(plot(\n",
    "        VennDiag, quantities=TRUE, counts = TRUE, font=5, cex=1, alpha=0.8,\n",
    "        fill=rep(color_scheme_celltype[[cur_celltype]], 3)\n",
    "      ))\n",
    "      dev.off()\n",
    "\n",
    "    }\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3582d536",
   "metadata": {},
   "source": [
    "### **Circular HeatMap for DEGs in all pathways for each test**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2ee3a8cd",
   "metadata": {},
   "source": [
    "#### **Plot Genes In Each Differentially Expressed Pathway For Each Study**\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "6c174b4c",
   "metadata": {},
   "source": [
    "**First, obtain a list of genes that are in differentially expressed in pathways for each cell type across all studies and tests**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72278aca",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o geneSets\n",
    "\n",
    "\n",
    "geneSets <- list()\n",
    "pathway_names <- list()\n",
    "for(test_name in tests){\n",
    "\n",
    "  geneSets[[test_name]] <- list()\n",
    "  pathway_names[[test_name]] <- list()\n",
    "\n",
    "  for (study in studies){\n",
    "\n",
    "    all_pathways <- read.csv(sprintf('../results/%s/%s/Data/differentially_expressed_pathways.csv', test_name, study), header=TRUE)\n",
    "    \n",
    "    all_pathways <- all_pathways[abs(all_pathways[\"score_adj\"]) > 1.3, ]\n",
    "\n",
    "    pathway_names[[test_name]][[study]] <- list()\n",
    "    geneSets[[test_name]][[study]] <- list()\n",
    "\n",
    "    for (cur_celltype in cell_types){\n",
    "\n",
    "        subset_df <- all_pathways %>% subset(celltype == cur_celltype)\n",
    "        pathway_names[[test_name]][[study]][[cur_celltype]] <- subset_df$pathway\n",
    "\n",
    "        # get gene sets\n",
    "        geneSets[[test_name]][[study]][[cur_celltype]] <- getGmt(file.path('../results/gsva/', study, '/', cur_celltype, 'gs.gmt'))\n",
    "        geneSets[[test_name]][[study]][[cur_celltype]] <- geneIds(geneSets[[test_name]][[study]][[cur_celltype]])\n",
    "\n",
    "        # Check if test_name exists\n",
    "        if (test_name %in% valid_test_names) {\n",
    "          geneSets[[test_name]][[study]][[cur_celltype]] <-  geneSets[[test_name]][[study]][[cur_celltype]][names(geneSets[[test_name]][[study]][[cur_celltype]]) %in% pathway_names[[test_name]][[study]][[cur_celltype]]]\n",
    "          } else {\n",
    "          geneSets[[test_name]][[study]][[cur_celltype]] <- character(0)  # Set an empty vector if no scores exist\n",
    "        }\n",
    "        geneSets[[test_name]][[study]][[cur_celltype]] <- unlist(geneSets[[test_name]][[study]][[cur_celltype]], recursive = FALSE)\n",
    "    }\n",
    "  }\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a50da718",
   "metadata": {},
   "source": [
    "**Next, obtain the dataframe of P-values and LogFC for each gene in the genesetfor each cell type across all studies and tests**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09cffecf",
   "metadata": {},
   "outputs": [],
   "source": [
    "degs = {}\n",
    "merged_degs = {}\n",
    "DEG_dist = {}\n",
    "\n",
    "for study in studies:\n",
    "\n",
    "    degs[study] = {}\n",
    "    merged_degs[study] = {}\n",
    "    DEG_dist[study] = {}\n",
    "\n",
    "    for test in tests:\n",
    "      \n",
    "        degs[study][test] = {}\n",
    "        merged_degs[study][test] = {}\n",
    "        DEG_dist[study][test] = {}\n",
    "\n",
    "        for method in deg_methods:\n",
    "         \n",
    "            degs[study][test][method] = {}\n",
    "            merged_degs[study][test][method] = {}\n",
    "            merged_df = pd.DataFrame(columns=['gene'])\n",
    " \n",
    "            deg_distribution = {}\n",
    "            deg_distribution['Cell Type'] = []\n",
    "            deg_distribution['Gene Program'] = []\n",
    "            deg_distribution['Up'] = []\n",
    "            deg_distribution['Down'] = []\n",
    "\n",
    "            for cell_type in cell_types:\n",
    "                degs[study][test][method][cell_type] = pd.read_excel(f'../results/{test}/{study}/DEG/{method}_degs.xlsx', \n",
    "                                                                            sheet_name=cell_type,)\n",
    "                \n",
    "                df = degs[study][test][method][cell_type]\n",
    "                df = df[df.gene.isin(geneSets[test][study][cell_type])]\n",
    "                deg_distribution['Cell Type'].append(cell_type)\n",
    "                deg_distribution['Gene Program'].append(len(df))\n",
    "                deg_distribution['Up'].append(len(df[(df.avg_logFC>0.25)&(df.p_val_adj<0.01)]))\n",
    "                deg_distribution['Down'].append(len(df[(df.avg_logFC<0.25)&(df.p_val_adj<0.01)]))\n",
    "                df = df.rename(columns={'avg_logFC': f'{cell_type}_logFC', 'p_val_adj': f'{cell_type}_Pvalue'})\n",
    "                \n",
    "                merged_df = pd.merge(merged_df, df[['gene', f'{cell_type}_logFC', f'{cell_type}_Pvalue']], on='gene', how='outer')\n",
    "        \n",
    "            merged_degs[study][test][method]['logFC'] = merged_df[['gene'] + list(merged_df.filter(like='logFC').columns)]\n",
    "            merged_degs[study][test][method]['Pvals'] = merged_df[['gene'] + list(merged_df.filter(like='Pvalue').columns)]\n",
    "            DEG_dist[study][test][method] = pd.DataFrame(deg_distribution)\n",
    "\n",
    "# convert nested dataframe into Rpy2 object \n",
    "\n",
    "all_degs =  robjects.ListVector(\n",
    "                        {\n",
    "                            study: robjects.ListVector(\n",
    "                                {\n",
    "                                    test_name: robjects.ListVector(\n",
    "                                        {\n",
    "                                            method: robjects.ListVector(\n",
    "    \n",
    "                                                {stat: pandas2ri.py2rpy(merged_degs[study][test_name][method][stat])\n",
    "                                                 \n",
    "                                                    for stat in ['logFC', 'Pvals']\n",
    "                                                }\n",
    "                                            )\n",
    "                                            for method in deg_methods\n",
    "                                        }\n",
    "                                    )\n",
    "                                    for test_name in tests\n",
    "                                }\n",
    "                            )\n",
    "                            for study in studies\n",
    "                        }\n",
    "                    )\n",
    "\n",
    "\n",
    "DEG_dist =  robjects.ListVector(\n",
    "                        {\n",
    "                            study: robjects.ListVector(\n",
    "                                {\n",
    "                                    test_name: robjects.ListVector(\n",
    "                                        {\n",
    "                                            method: robjects.ListVector(\n",
    "    \n",
    "                                                pandas2ri.py2rpy(DEG_dist[study][test_name][method])\n",
    "                                                 \n",
    "                                            )\n",
    "                                            for method in deg_methods\n",
    "                                        }\n",
    "                                    )\n",
    "                                    for test_name in tests\n",
    "                                }\n",
    "                            )\n",
    "                            for study in studies\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e03e59ee",
   "metadata": {},
   "source": [
    "**Obtain Table of DEG distributions in perturbed gene programs**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53e31654",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i all_degs -i deg_methods -i DEG_dist\n",
    "\n",
    "library(knitr)\n",
    "library(kableExtra)\n",
    "library(magrittr)\n",
    "\n",
    "for (study in studies){\n",
    "  for (test_name in tests){\n",
    "\n",
    "    fig_dir = paste0('../results/', test_name, '/', study, '/DEG_distribution/')\n",
    "\n",
    "    if (!dir.exists(fig_dir)) {\n",
    "    dir.create(fig_dir, recursive=TRUE)\n",
    "    }\n",
    "\n",
    "    if (test_name %in% valid_test_names) {\n",
    "      for (method in deg_methods){\n",
    "        rownames(DEG_dist[[study]][[test_name]][[method]]) = DEG_dist[[study]][[test_name]][[method]][, 'Cell Type']\n",
    "        # Create the table using kable and style it with kableExtra\n",
    "        kable(DEG_dist[[study]][[test_name]][[method]][, c(\"Gene Program\", \"Up\", \"Down\")], \"html\", booktabs = TRUE) %>%\n",
    "          kable_styling(bootstrap_options = c(\"striped\", \"hover\"), full_width = F) %>%\n",
    "          save_kable(paste0(fig_dir, 'deg_distribution_', method, '.pdf'))\n",
    "\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "# Print the new data frame to check\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9f40134b",
   "metadata": {},
   "source": [
    "**Next, Make Circos plot using this filtered data**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fdd0ac6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "\n",
    "library(viridis)\n",
    "\n",
    "for (study in studies){\n",
    "  for (test_name in tests){\n",
    "\n",
    "    if (test_name %in% valid_test_names) {\n",
    "      for (method in deg_methods){\n",
    "\n",
    "        # for (stat in c('logFC', 'Pvals')){\n",
    "        \n",
    "        mat1 = all_degs[[study]][[test_name]][[method]][['logFC']]\n",
    "        mat2 = all_degs[[study]][[test_name]][[method]][['Pvals']]\n",
    "\n",
    "        rownames(mat1) <- mat1[, \"gene\"]\n",
    "        rownames(mat2) <- mat2[, \"gene\"]\n",
    "\n",
    "        mat1[is.na(mat1)] <- 0\n",
    "        mat2[is.na(mat2)] <- 1\n",
    "\n",
    "        mat1 <- mat1[, -which(colnames(mat1) == \"gene\")]\n",
    "        mat2 <- -log10(mat2[, -which(colnames(mat2) == \"gene\")])\n",
    "\n",
    "        # Step 1: Create a logical matrix with the same dimensions as mat1 and mat2 based on the conditions\n",
    "        mat3_logical1 <- ((mat1) > 0.25) & (mat2 < -log10(0.01))\n",
    "        mat3_logical2 <- ((mat1) < -0.25) & (mat2 < -log10(0.01))\n",
    "\n",
    "        # Step 2: Create an empty matrix with the same dimensions as mat1 and mat2\n",
    "        mat3 <- matrix(0, nrow = nrow(mat1), ncol = ncol(mat1))\n",
    "\n",
    "        # Step 3: Assign 1 to the elements that satisfy the conditions in mat3\n",
    "        mat3[mat3_logical1] <- 1\n",
    "        mat3[mat3_logical2] <- -1\n",
    "\n",
    "        # Set column and row names\n",
    "        colnames(mat3) <- colnames(mat1)\n",
    "        rownames(mat3) <- rownames(mat1)\n",
    "\n",
    "        fig_dir = paste0('../results/', test_name, '/', study, '/DEG/')\n",
    "\n",
    "        if (!dir.exists(fig_dir)) {\n",
    "        dir.create(fig_dir, recursive=TRUE)\n",
    "        }\n",
    "\n",
    "        pdf(paste0(fig_dir, 'circular_heatmap_', method, '.pdf'), width=8, height=8)\n",
    "\n",
    "        # col_fun1 = colorRamp2(c(min(mat1), 0, max(mat1)), c(\"blue\", \"white\", \"red\"))\n",
    "        # col_fun2 = colorRamp2(c(min(mat2), max(mat2)), c(\"white\", \"green\"))\n",
    "\n",
    "        col_fun1 = colorRamp2(c(-5, 0, 5), c(\"blue\", \"white\", \"red\"))\n",
    "        col_fun2 = colorRamp2(c(0, 6), c(\"white\", \"green\"))\n",
    "        col_fun3 = colorRamp2(c(-1, 0, 1), c('purple', \"lightgrey\", \"yellow\"))\n",
    "        \n",
    "        # circos.initialize()\n",
    "        circos.par(gap.after = 10)\n",
    "        circos.heatmap.initialize(mat1)\n",
    "\n",
    "        circos.heatmap(mat1, split = NULL, col = col_fun1, track.height = 0.2, bg.border = \"black\", bg.lwd = 2,)\n",
    "        circos.heatmap(mat2, split = NULL, col = col_fun2, track.height = 0.2, bg.border = \"black\", bg.lwd = 2,)\n",
    "        circos.heatmap(mat3, split = NULL, col = col_fun3, track.height = 0.2)\n",
    "\n",
    "        circos.track(track.index = get.current.track.index(), panel.fun = function(x, y) {\n",
    "            if (CELL_META$sector.numeric.index == 1) { # the last sector\n",
    "                cn = lapply(colnames(mat1), function(col_name) toupper(substr(col_name, 1, 3)))\n",
    "                n = length(cn)\n",
    "\n",
    "                for (i in 1:nrow(mat1)) {\n",
    "                  for (j in 1:ncol(mat1)) {\n",
    "                    if (abs(mat1[i, j]) > 0.25 && mat2[i, j] > -log10(0.01)) {\n",
    "                        circos.text(x[i], y[j], \"X\", col = \"black\", cex = 5.5)\n",
    "                    }\n",
    "                  }\n",
    "                }\n",
    "                circos.text(rep(CELL_META$cell.xlim[2], n) + convert_x(1, \"mm\"), \n",
    "                            1:n - 0.5, cn, \n",
    "                            cex = 0.5, adj = c(0, 0.5), facing = \"inside\")\n",
    "            }\n",
    "        }, bg.border = NA)\n",
    "\n",
    "\n",
    "        # circos.track(track.index = get.current.track.index(), panel.fun = function(x, y) {\n",
    "        # if(CELL_META$sector.numeric.index == 1) { # the last sector\n",
    "        #     cn = lapply(colnames(mat1), function(col_name) toupper(substr(col_name, 1, 3)))\n",
    "        #     n = length(cn)\n",
    "        #     circos.text(rep(CELL_META$cell.xlim[2], n) + convert_x(1, \"mm\"), \n",
    "        #         1:n - 0.5, cn, \n",
    "        #         cex = 0.5, adj = c(0, 0.5), facing = \"inside\")\n",
    "        #   }\n",
    "        # }, bg.border = NA)\n",
    "\n",
    "\n",
    "        lgd1 <- Legend(col_fun = col_fun1, title = \"log(Fold Change)\",\n",
    "                      direction = \"horizontal\", title_position = \"topcenter\", \n",
    "                      legend_width = unit(2, \"cm\"), legend_height = unit(0.1, \"cm\"))\n",
    "        # grid.draw(lgd1)\n",
    "\n",
    "        lgd2 <- Legend(col_fun = col_fun2, title = \"-log10(FDR)\", \n",
    "                      direction = \"horizontal\", title_position = \"topcenter\",\n",
    "                      legend_width = unit(2, \"cm\"), legend_height = unit(0.1, \"cm\"))\n",
    "\n",
    "\n",
    "        lgd3 <- Legend(col_fun = col_fun3, title = \"DEG\", \n",
    "              direction = \"horizontal\", title_position = \"topcenter\",\n",
    "              legend_width = unit(2, \"cm\"), legend_height = unit(0.1, \"cm\"))\n",
    "\n",
    "        # grid.draw(lgd2)\n",
    "\n",
    "        plgd <- packLegend(lgd1, lgd2, direction = \"vertical\")\n",
    "\n",
    "        grid.draw(plgd)\n",
    "        print(circos.par(gap.after = c(90)))\n",
    "        circos.clear()\n",
    "\n",
    "        dev.off()\n",
    "             \n",
    "      } \n",
    "    } \n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ebce269",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "2f414a000012e1980f5fd1119ed6b9bf8cc73b51e37b52eb52a16ff857effd1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
