{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import rpy2\n",
    "import scipy\n",
    "import logging\n",
    "import warnings\n",
    "import anndata2ri\n",
    "import collections\n",
    "import tables\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import seaborn as sb\n",
    "import decoupler as dc\n",
    "import scrublet as scr\n",
    "from scipy import sparse\n",
    "import scipy.sparse as sp_sparse\n",
    "import anndata\n",
    "from anndata import AnnData\n",
    "from tabnanny import verbose\n",
    "import matplotlib.pyplot as plt\n",
    "from gsva_prep import prep_gsva\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "from typing import Optional, Union\n",
    "from matplotlib.pyplot import rcParams\n",
    "from functions import pathway_analyses\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from sklearn.model_selection import train_test_split\n",
    "from pytorch_lightning.loggers import TensorBoardLogger\n",
    "from rpy2.robjects.conversion import localconverter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sys_dpi(width, height, diag):\n",
    "    '''\n",
    "    obtain dpi of system\n",
    "    \n",
    "    w: width in pixels (if unsure, go vist `whatismyscreenresolution.net`)\n",
    "    h: height in pixels\n",
    "    d: diagonal in inches\n",
    "    '''\n",
    "    w_inches = (diag**2/ (1 + height**2/width**2))**0.5\n",
    "    return round(width/w_inches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.8.0\n",
      "scanpy      1.9.3\n",
      "-----\n",
      "PIL                         9.5.0\n",
      "absl                        NA\n",
      "anndata2ri                  1.1\n",
      "appnope                     0.1.3\n",
      "asttokens                   NA\n",
      "attr                        23.1.0\n",
      "backcall                    0.2.0\n",
      "cffi                        1.15.1\n",
      "comm                        0.1.3\n",
      "cpuinfo                     NA\n",
      "cycler                      0.10.0\n",
      "cython_runtime              NA\n",
      "dateutil                    2.8.2\n",
      "debugpy                     1.6.7\n",
      "decorator                   5.1.1\n",
      "decoupler                   1.4.0\n",
      "deprecate                   0.3.2\n",
      "dot_parser                  NA\n",
      "executing                   1.2.0\n",
      "fsspec                      2023.6.0\n",
      "functions                   NA\n",
      "google                      NA\n",
      "gsva_prep                   NA\n",
      "h5py                        3.9.0\n",
      "igraph                      0.10.4\n",
      "ipykernel                   6.23.2\n",
      "ipywidgets                  8.0.6\n",
      "jedi                        0.18.2\n",
      "jinja2                      3.1.2\n",
      "joblib                      1.2.0\n",
      "kiwisolver                  1.4.4\n",
      "leidenalg                   0.9.1\n",
      "llvmlite                    0.40.1rc1\n",
      "louvain                     0.8.0\n",
      "markupsafe                  2.1.3\n",
      "matplotlib                  3.7.1\n",
      "mpl_toolkits                NA\n",
      "mpmath                      1.3.0\n",
      "natsort                     8.4.0\n",
      "numba                       0.57.0\n",
      "numexpr                     2.8.4\n",
      "numpy                       1.24.3\n",
      "opt_einsum                  v3.3.0\n",
      "packaging                   23.1\n",
      "pandas                      2.0.2\n",
      "parso                       0.8.3\n",
      "patsy                       0.5.3\n",
      "pexpect                     4.8.0\n",
      "pickleshare                 0.7.5\n",
      "pkg_resources               NA\n",
      "platformdirs                3.7.0\n",
      "prompt_toolkit              3.0.38\n",
      "psutil                      5.9.5\n",
      "ptyprocess                  0.7.0\n",
      "pure_eval                   0.2.2\n",
      "pycparser                   2.21\n",
      "pydev_ipython               NA\n",
      "pydevconsole                NA\n",
      "pydevd                      2.9.5\n",
      "pydevd_file_utils           NA\n",
      "pydevd_plugins              NA\n",
      "pydevd_tracing              NA\n",
      "pydot                       1.4.2\n",
      "pygments                    2.15.1\n",
      "pyparsing                   3.1.0\n",
      "pytorch_lightning           1.7.7\n",
      "pytz                        2023.3\n",
      "pytz_deprecation_shim       NA\n",
      "rich                        NA\n",
      "rpy2                        3.5.12\n",
      "scipy                       1.10.1\n",
      "scrublet                    NA\n",
      "seaborn                     0.12.2\n",
      "session_info                1.0.0\n",
      "setuptools                  68.0.0\n",
      "six                         1.16.0\n",
      "sklearn                     1.2.2\n",
      "sparse                      0.14.0\n",
      "stack_data                  0.6.2\n",
      "statsmodels                 0.14.0\n",
      "sympy                       1.12\n",
      "tables                      3.8.0\n",
      "tensorboard                 2.13.0\n",
      "texttable                   1.6.7\n",
      "threadpoolctl               3.1.0\n",
      "torch                       2.0.1\n",
      "torchmetrics                0.11.4\n",
      "tornado                     6.3.2\n",
      "tqdm                        4.65.0\n",
      "traitlets                   5.9.0\n",
      "typing_extensions           NA\n",
      "tzlocal                     NA\n",
      "wcwidth                     0.2.6\n",
      "yaml                        6.0\n",
      "zmq                         25.1.0\n",
      "zoneinfo                    NA\n",
      "-----\n",
      "IPython             8.14.0\n",
      "jupyter_client      8.2.0\n",
      "jupyter_core        5.3.1\n",
      "-----\n",
      "Python 3.10.6 (main, Oct 24 2022, 11:04:34) [Clang 12.0.0 ]\n",
      "macOS-10.16-x86_64-i386-64bit\n",
      "-----\n",
      "Session information updated at 2023-07-27 18:29\n"
     ]
    }
   ],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# # Automatically convert rpy2 outputs to pandas dataframes\n",
    "# pandas2ri.activate()\n",
    "# anndata2ri.activate()\n",
    "# %load_ext rpy2.ipython\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "rcParams['figure.dpi'] = get_sys_dpi(1512, 982, 14.125)\n",
    "#rcParams['figure.figsize']=(4,4) #rescale figures\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_dendrogram(model, **kwargs):\n",
    "    # Create linkage matrix and then plot the dendrogram\n",
    "\n",
    "    # create the counts of samples under each node\n",
    "    counts = np.zeros(model.children_.shape[0])\n",
    "    n_samples = len(model.labels_)\n",
    "    for i, merge in enumerate(model.children_):\n",
    "        current_count = 0\n",
    "        for child_idx in merge:\n",
    "            if child_idx < n_samples:\n",
    "                current_count += 1  # leaf node\n",
    "            else:\n",
    "                current_count += counts[child_idx - n_samples]\n",
    "        counts[i] = current_count\n",
    "\n",
    "    linkage_matrix = np.column_stack(\n",
    "        [model.children_, model.distances_, counts]\n",
    "    ).astype(float)\n",
    "\n",
    "    # Plot the corresponding dendrogram\n",
    "    dendrogram(linkage_matrix, **kwargs)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Load input data in raw form and save in `.h5ad` format**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **[Grubman et. al. 2019](https://doi.org/10.1038/s41593-019-0539-4) (Entorhinal Cortex (ETC))**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load AD barcodes\n",
    "\n",
    "metadata = pd.read_csv(\"../data/raw/grubman_etc/scRNA_metadata.tsv\", sep='\\t')\n",
    "batches = list(metadata.batch.unique())\n",
    "gene_names = list(pd.read_csv(\"../data/raw/grubman_etc/mtx/AD1_AD2_genes.tsv\", sep=\"\\t\", names=['ENS', 'genes'])['genes'])\n",
    "count_data = sc.read_csv(\"../data/raw/grubman_etc/scRNA_rawCounts.tsv\", delimiter=\"\\t\").T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "grubman_adata = count_data.copy()\n",
    "grubman_adata.obs = metadata.copy()\n",
    "\n",
    "grubman_adata.obs['cell_type'] = grubman_adata.obs.cellType.map({'oligo': 'Oligodendrocyte', \n",
    "                                                                 'astro': 'Astrocyte',\n",
    "                                                                 'OPC': 'OPC',\n",
    "                                                                 'neuron': 'Neuron',\n",
    "                                                                 'Endo': 'Endothelial',\n",
    "                                                                 'mg': 'Microglia'})\n",
    "\n",
    "grubman_adata = grubman_adata[~grubman_adata.obs['cell_type'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/_core/anndata.py:1235: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata/_core/anndata.py:1235: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n"
     ]
    }
   ],
   "source": [
    "grubman_adata.write_h5ad(\"../data/raw/grubman_etc/grubman_etc_raw_anndata.h5ad\")\n",
    "del grubman_adata, count_data"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **[Leng et. al. 2021](https://www.synapse.org/#!Synapse:syn21788402/wiki/601825) (Entorhinal Cortex (ETC) and Superior Frontal Gyrus (SFG))**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:py.warnings:/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata2ri/r2py.py:106: FutureWarning: X.dtype being converted to np.float32 from float64. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.\n",
      "  return AnnData(exprs, obs, var, uns, obsm or None, layers=layers)\n",
      "\n",
      "WARNING:py.warnings:/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/anndata2ri/r2py.py:106: FutureWarning: X.dtype being converted to np.float32 from float64. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.\n",
      "  return AnnData(exprs, obs, var, uns, obsm or None, layers=layers)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# processed data obtained \n",
    "readRDS = robjects.r['readRDS']\n",
    "df_etc = readRDS('../data/raw/leng_etc/sce.EC.scAlign.assigned.rds')\n",
    "adata_leng_etc = df_etc\n",
    "adata_leng_etc.obs['cell_type'] = adata_leng_etc.obs['clusterCellType'].map({'Exc': 'Excitatory',\n",
    "                                                                            'Inh': 'Inhibitory',\n",
    "                                                                            'Astro': 'Astrocyte',\n",
    "                                                                            'Endo': 'Endothelial', \n",
    "                                                                            'Micro': 'Microglia', \n",
    "                                                                            'OPC': 'OPC', \n",
    "                                                                            'Oligo': 'Oligodendrocyte'})\n",
    "# processed data obtained \n",
    "readRDS = robjects.r['readRDS']\n",
    "df_sfg = readRDS('../data/raw/leng_sfg/sce.SFG.scAlign.assigned.rds')\n",
    "adata_leng_sfg = df_sfg\n",
    "adata_leng_sfg.obs['cell_type'] = adata_leng_sfg.obs['clusterCellType'].map({'Exc': 'Excitatory',\n",
    "                                                                            'Inh': 'Inhibitory',\n",
    "                                                                            'Astro': 'Astrocyte',\n",
    "                                                                            'Endo': 'Endothelial', \n",
    "                                                                            'Micro': 'Microglia', \n",
    "                                                                            'OPC': 'OPC', \n",
    "                                                                            'Oligo': 'Oligodendrocyte'})\n",
    "\n",
    "adata_leng_etc.write_h5ad('../data/raw/leng_etc/leng_etc_raw_anndata.h5ad')\n",
    "adata_leng_sfg.write_h5ad('../data/raw/leng_sfg/leng_sfg_raw_anndata.h5ad')\n",
    "del adata_leng_etc, adata_leng_sfg, df_etc, df_sfg"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Define pathology groups using heirachical clustering of ADNC and CDR scores.**\n",
    "\n",
    "Individuals are grouped into no, early, and late-pathology groups based on the `AD neuropathological change (ADNC)` score, which represents assessment of amyloid-β deposits (‘A’), staging of neurofibrillary tangles (‘B’) and scoring of neuritic plaques (‘C’) [**Thomas J Montine et al., 2012**](https://pubmed.ncbi.nlm.nih.gov/22101365/), and the `Clinical Dementia rating (CDr)` which reflects the degree of cognitive impairment [**C P Hughes**](https://pubmed.ncbi.nlm.nih.gov/7104545/).\n",
    "\n",
    "This done to maintain consistency with the clustering used in Mathys et al, where the nine clinico-pathological traits (Supplementary Table 3) where used to group individuals into AD-pathology groups, segregated into two subgroups that correspond to the pathological progression of AD: `‘early-pathology’1 (amyloid burden, but modest neurofibrillary tangles and modest cognitive impairment)`, and `and ‘late-pathology’ (higher amyloid burden, and also increased neurofibrillary tangles, global pathology, and cognitive impairment)`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# meta = pd.read_csv('../data/raw/leng_etc/leng_etc_metadata.csv')\n",
    "\n",
    "# X = pd.DataFrame()\n",
    "# temp = pd.DataFrame(meta['ADNC_score']).apply(lambda x: x.str.split(','), axis=1, result_type='broadcast')\n",
    "# X[\"A_score\"] = temp.applymap(lambda x: x[0][-1])\n",
    "# X[\"B_score\"] = temp.applymap(lambda x: x[1][-1])\n",
    "# X[\"C_score\"] = temp.applymap(lambda x: x[2][-1])\n",
    "# X['CDR_before_death'] = meta['CDR_before_death']\n",
    "\n",
    "\n",
    "# # setting distance_threshold=0 ensures we compute the full tree.\n",
    "# model = AgglomerativeClustering(distance_threshold=0, n_clusters=None)\n",
    "\n",
    "# model = model.fit(X)\n",
    "# plt.title(\"Hierarchical Clustering Dendrogram\")\n",
    "# # plot the top three levels of the dendrogram\n",
    "# plot_dendrogram(model, truncate_mode=\"level\", p=3)\n",
    "# plt.xlabel(\"Number of points in node (or index of point if no parenthesis).\")\n",
    "# plt.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **[Mathys et. al. 2019](https://doi.org/10.1038/s41586-019-1195-2) (Prefrontal Cortex)**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load filtered data (Mathys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#adata_mathys_pfc = sc.read_mtx(\"../data/raw/mathys_pfc/mathys_pfc_count_matrix.mtx\").T\n",
    "\n",
    "adata_mathys_pfc = sc.read_mtx(\"../data/raw/mathys_pfc/filtered_count_matrix.mtx\").T\n",
    "\n",
    "obs_names = pd.read_csv(\"../data/raw/mathys_pfc/filtered_column_metadata.txt\", sep='\\t')\n",
    "obs_names.set_index('TAG', inplace=True)\n",
    "obs_names.index.rename('index', inplace=True)\n",
    "adata_mathys_pfc.obs = obs_names\n",
    "\n",
    "gene_names = pd.read_csv(\"../data/raw/mathys_pfc/filtered_gene_row_names.txt\", sep='\\t', header=None)\n",
    "adata_mathys_pfc.var_names = gene_names[0]\n",
    "\n",
    "sample_key = pd.read_csv(\"../data/raw/mathys_pfc/snRNAseqPFC_BA10_Sample_key.csv\")\n",
    "id_mapping = pd.read_csv(\"../data/raw/mathys_pfc/snRNAseqPFC_BA10_id_mapping.csv\")\n",
    "\n",
    "merged_key = id_mapping.merge(sample_key, how=\"outer\", on='projid')\n",
    "merged_key['libraryid'] = merged_key['fastq'].str.split('-').apply(lambda x: x[1].split(\"_\")[0])\n",
    "merged_key.drop_duplicates(subset='libraryid', keep='first', inplace=True)\n",
    "merged_key.reset_index(inplace=True)\n",
    "merged_key.drop('index', axis=1, inplace=True)\n",
    "\n",
    "adata_mathys_pfc.obs[list(merged_key.columns)] = None\n",
    "\n",
    "for val in merged_key['sample']:\n",
    "    adata_mathys_pfc.obs.loc[adata_mathys_pfc.obs.index.str.endswith(f'.{val}'), list(merged_key.columns)] = merged_key[merged_key['sample']==val].values\n",
    "\n",
    "\n",
    "# main epidemiological and pathological characteristics of the participants\n",
    "epi_n_patho = pd.read_excel(\"../data/raw/mathys_pfc/41586_2019_1195_MOESM3_ESM.xlsx\", sheet_name=1)\n",
    "\n",
    "# clinico-pathological variables.\n",
    "clin_n_patho = pd.read_excel(\"../data/raw/mathys_pfc/41586_2019_1195_MOESM5_ESM.xlsx\", sheet_name=0)\n",
    "\n",
    "# clinical, epidemiological and pathological metadata\n",
    "metadata = epi_n_patho.merge(clin_n_patho, how=\"outer\", on=\"Subject\")\n",
    "\n",
    "# ID mapping from Mathys\n",
    "mapping = pd.read_csv(\"../data/raw/mathys_pfc/snRNAseqPFC_BA10_id_mapping.csv\", sep=\",\")\n",
    "mapping['sampleid'] = mapping['fastq'].str.split('-').apply(lambda x: x[1].split(\"_\")[0])\n",
    "\n",
    "metadata = metadata.merge(mapping, how='outer', on='Subject').sort_values(by='sampleid')\n",
    "metadata.drop_duplicates(subset='sampleid', inplace=True)\n",
    "metadata.reset_index(inplace=True)\n",
    "metadata.drop('index', axis=1, inplace=True)\n",
    "\n",
    "sampleids = list(metadata.sampleid.unique())\n",
    "\n",
    "\n",
    "adata_mathys_pfc.obs[list(metadata.columns)] = None\n",
    "columns = list(metadata.columns) \n",
    "\n",
    "sid = metadata.sampleid.astype(int).min()\n",
    "\n",
    "for ind, val in enumerate(sampleids):\n",
    "    adata_mathys_pfc.obs.loc[adata_mathys_pfc.obs.libraryid==val, columns] = metadata[metadata['sampleid']==val].values\n",
    "\n",
    "adata_mathys_pfc.obs = adata_mathys_pfc.obs.astype('category').copy()\n",
    "\n",
    "adata_mathys_pfc.obs_names_make_unique()\n",
    "adata_mathys_pfc.var_names_make_unique()\n",
    "\n",
    "obs_columns = list(adata_mathys_pfc.obs.columns)\n",
    "adata_mathys_pfc.obs[obs_columns] = adata_mathys_pfc.obs[obs_columns].astype(str)\n",
    "\n",
    "var_columns = list(adata_mathys_pfc.var.columns)\n",
    "adata_mathys_pfc.var[var_columns] = adata_mathys_pfc.var[var_columns].astype(str)\n",
    "\n",
    "adata_mathys_pfc.obs['cell_type'] = adata_mathys_pfc.obs['broad.cell.type'].map({'Ex': 'Excitatory',\n",
    "                                                                            'In': 'Inhibitory',\n",
    "                                                                            'Ast': 'Astrocyte',\n",
    "                                                                            'End': 'Endothelial', \n",
    "                                                                            'Mic': 'Microglia', \n",
    "                                                                            'Opc': 'OPC', \n",
    "                                                                            'Oli': 'Oligodendrocyte',\n",
    "                                                                            'Per': 'Pericyte'})\n",
    "\n",
    "adata_mathys_pfc.write_h5ad('../data/raw/mathys_pfc/mathys_pfc_raw_anndata.h5ad')\n",
    "\n",
    "del epi_n_patho, clin_n_patho, metadata, mapping, sampleids, merged_key, obs_names, gene_names, adata_mathys_pfc\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **[Human Multiple Cortical Areas SMART-seq reference](https://portal.brain-map.org/atlases-and-data/rnaseq/human-multiple-cortical-areas-smart-seq)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 47432 × 50281\n",
       "    obs: 'exp_component_name', 'specimen_type', 'cluster_color', 'cluster_order', 'cluster_label', 'class_color', 'class_order', 'class_label', 'subclass_color', 'subclass_order', 'subclass_label', 'full_genotype_color', 'full_genotype_order', 'full_genotype_label', 'donor_sex_color', 'donor_sex_order', 'donor_sex_label', 'region_color', 'region_order', 'region_label', 'cortical_layer_color', 'cortical_layer_order', 'cortical_layer_label', 'cell_type_accession_color', 'cell_type_accession_order', 'cell_type_accession_label', 'cell_type_alias_color', 'cell_type_order', 'cell_type_alias_label', 'cell_type_alt_alias_color', 'cell_type_alt_alias_order', 'cell_type_alt_alias_label', 'cell_type_designation_color', 'cell_type_designation_order', 'cell_type_designation_label', 'external_donor_name_color', 'external_donor_name_order', 'external_donor_name_label', 'outlier_call', 'outlier_type'\n",
       "    var: 'chromosome', 'entrez_id', 'gene_name', 'mouse_homologenes'"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts = sc.read_csv('../data/raw/allen_mca/matrix.csv')\n",
    "counts = sparse.csr_matrix(counts.X)\n",
    "adata_reference = AnnData(counts)\n",
    "\n",
    "del counts\n",
    "\n",
    "adata_reference.obs = pd.read_csv('../data/raw/allen_mca/metadata.csv', index_col='sample_name')\n",
    "adata_reference.var = pd.read_csv('../data/raw/allen_mca/human_MTG_2018-06-14_genes-rows.csv', index_col='gene')\n",
    "# adata_reference.obs_names = adata_reference.obs.sample_name.copy()\n",
    "tsne_cord = pd.read_csv('../data/raw/allen_mca/tsne.csv')\n",
    "adata_reference = adata_reference[adata_reference.obs_names.isin(list(tsne_cord.sample_name))]\n",
    "# adata_reference.obsm['X_tsne'] = tsne_cord[['tsne_1', 'tsne_2']].to_numpy()\n",
    "adata_reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/c1/8tpkntr176v0pt5_bk6zhrp40000gn/T/ipykernel_6446/1401974434.py:2: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_reference.obs[obs_columns] = adata_reference.obs[obs_columns].astype(str)\n"
     ]
    }
   ],
   "source": [
    "obs_columns = list(adata_reference.obs.columns)\n",
    "adata_reference.obs[obs_columns] = adata_reference.obs[obs_columns].astype(str)\n",
    "\n",
    "var_columns = list(adata_reference.var.columns)\n",
    "adata_reference.var[var_columns] = adata_reference.var[var_columns].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "adata_reference.obs['cell_type'] = \"\"\n",
    "\n",
    "adata_reference.obs.loc[~adata_reference.obs['class_label'].str.startswith('Non-'), 'cell_type'] = \\\n",
    "    adata_reference.obs['class_label'][~adata_reference.obs['class_label'].str.startswith(\"Non\")].map({\"Glutamatergic\": \"Excitatory\",\n",
    "                                                                                                        \"GABAergic\": \"Inhibitory\"})\n",
    "\n",
    "adata_reference.obs.loc[adata_reference.obs['class_label'].str.contains(\"Non-\"), 'cell_type'] = adata_reference.obs.loc[adata_reference.obs['class_label'].str.contains(\"Non-\"), 'subclass_label']\n",
    "\n",
    "adata_reference.obs.loc[adata_reference.obs.cell_type=='Microglia-PVM', 'cell_type'] = 'Microglia'\n",
    "\n",
    "adata_reference.obs.cell_type = adata_reference.obs.cell_type.astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_reference.write_h5ad(f'../data/raw/allen_mca/allen_mca_raw_anndata.h5ad', compression='gzip')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **[Lau et. al. 2002](https://www.pnas.org/doi/full/10.1073/pnas.2008762117) (Prefrontal Cortex (PTC))**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_dir = '../data/raw/lau_pfc/'\n",
    "file_list = list(set(['_'.join(file.split('_')[:-1]) for file in os.listdir(path_to_dir) if ~('lau' in file)]))\n",
    "\n",
    "adata_lau_pfc = dict()\n",
    "for file in file_list:\n",
    "    adata = sc.read_mtx('../data/raw/lau_pfc/'+file+'_matrix.mtx.gz').T\n",
    "    barcodes = pd.read_csv('../data/raw/lau_pfc/'+file+'_barcodes.tsv.gz', sep='\\t', names=['barcode'], header=None, dtype=str, index_col=0)\n",
    "    adata.obs = barcodes.astype(str)\n",
    "    adata.obs['Subject'] = file.split(\"_\")[-1]\n",
    "    genes = pd.read_csv('../data/raw/lau_pfc/'+file+'_features.tsv.gz', sep='\\t', names=['1', 'gene_name', '3'], header=None, dtype=str, index_col=1)\n",
    "    adata.var = genes.astype(str)\n",
    "\n",
    "    adata_lau_pfc[file] = adata\n",
    "    adata_lau_pfc[file].obs_names_make_unique()\n",
    "    adata_lau_pfc[file].var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_lau_pfc_concat = anndata.concat([adata_lau_pfc[key] for key in adata_lau_pfc.keys()], join='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filtered out 356 cells that have less than 200 genes expressed\n",
      "filtered out 3833 cells that have more than 20000 counts\n"
     ]
    }
   ],
   "source": [
    "adata_lau_pfc_concat.var['mt'] = adata_lau_pfc_concat.var_names.str.startswith('MT-')\n",
    "\n",
    "sc.pp.calculate_qc_metrics(adata_lau_pfc_concat, qc_vars=['mt'], inplace=True)\n",
    "sc.pp.filter_cells(adata_lau_pfc_concat, min_genes=200)\n",
    "sc.pp.filter_cells(adata_lau_pfc_concat, max_counts=20000)\n",
    "\n",
    "adata_lau_pfc_concat = adata_lau_pfc_concat[adata_lau_pfc_concat.obs[\"pct_counts_mt\"] <= 20].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 169514 × 33538\n",
       "    obs: 'Subject', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'n_genes', 'n_counts'\n",
       "    var: 'mt', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_lau_pfc_concat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_lau_pfc_concat.write_h5ad('../data/raw/lau_pfc/lau_pfc_raw_anndata.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **[Zhou et al. 2020](https://www.synapse.org/#!Synapse:syn21670836) (Prefrontal Cortex (PFC))**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_dir = '../data/raw/zhou_pfc/Data/Matrix_files/'\n",
    "file_list = list(set(['_'.join(file.split('_')[:-1]) for file in os.listdir(path_to_dir)]))\n",
    "\n",
    "adata_zhou_pfc = dict()\n",
    "for file in file_list:\n",
    "    adata = sc.read_mtx('../data/raw/zhou_pfc/Data/Matrix_files/'+file+'_matrix.mtx.gz').T\n",
    "    barcodes = pd.read_csv('../data/raw/zhou_pfc/Data/Matrix_files/'+file+'_barcodes.tsv.gz', sep='\\t', names=['barcode'], header=None, dtype=str, index_col=0)\n",
    "    adata.obs = barcodes.astype(str)\n",
    "    adata.obs['Subject'] = file.split(\"_\")[-1]\n",
    "    genes = pd.read_csv('../data/raw/zhou_pfc/Data/Matrix_files/'+file+'_features.tsv.gz', sep='\\t', names=['1', 'gene_name', '3'], header=None, dtype=str, index_col=1)\n",
    "    adata.var = genes.astype(str)\n",
    "    adata.obs_names = file + '_' + adata.obs_names\n",
    "    \n",
    "    adata_zhou_pfc[file] = adata\n",
    "    adata_zhou_pfc[file].obs_names_make_unique()\n",
    "    adata_zhou_pfc[file].var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/c1/8tpkntr176v0pt5_bk6zhrp40000gn/T/ipykernel_6446/268278942.py:18: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_zhou_pfc_concat.obs['cell_type'] = adata_zhou_pfc_concat.obs['Label'].astype('category')\n"
     ]
    }
   ],
   "source": [
    "adata_zhou_pfc_concat = anndata.concat([adata_zhou_pfc[key] for key in adata_zhou_pfc.keys()], join='inner')\n",
    "adata_zhou_pfc_concat.obs.index  = [ind.split('-')[0] for ind in adata_zhou_pfc_concat.obs.index]\n",
    "\n",
    "cell_ids = pd.read_excel('../data/raw/zhou_pfc/Data/Metadata/clusters_cellID.xlsx', sheet_name='All_nuclei')\n",
    "\n",
    "adata_zhou_pfc_concat.obs['Label'] = adata_zhou_pfc_concat.obs.index.map(dict(zip(cell_ids['Barcodes'], cell_ids['Label'])))\n",
    "adata_zhou_pfc_concat.obs['Label'] = adata_zhou_pfc_concat.obs['Label'].map({'OPC': 'OPC',\n",
    "                                                                            'Ex0': 'Excitatory', \n",
    "                                                                            'Oli1': \"Oligoddendrocyte\", \n",
    "                                                                            'Astro': 'Astrocyte', \n",
    "                                                                            'Micro': 'Microglia', \n",
    "                                                                            'In': \"Inhibitory\", \n",
    "                                                                           'Oli0': 'Oligodendrocyte',\n",
    "                                                                           'Endo': 'Endothelial',\n",
    "                                                                           'Ex1': 'Excitatory'})\n",
    "\n",
    "adata_zhou_pfc_concat = adata_zhou_pfc_concat[~adata_zhou_pfc_concat.obs['Label'].isna()]\n",
    "adata_zhou_pfc_concat.obs['cell_type'] = adata_zhou_pfc_concat.obs['Label'].astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_zhou_pfc_concat.write_h5ad('../data/raw/zhou_pfc/zhou_pfc_raw_anndata.h5ad')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **[Allen Institute SEA-AD 2023](https://portal.brain-map.org/explore/seattle-alzheimers-disease) (Middle Temporal Gyrus (MTG))**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Using Raw Filtered Data**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#no_pathology = [\"H21.33.003\", \"H21.33.004\", \"H21.33.023\", \"H20.33.044\"]\n",
    "no_pathology = [\"H21.33.003\", \"H21.33.004\", \"H21.33.023\",  \"H20.33.044\" ]\n",
    "#early_pathology = [\"H21.33.044\", \"H21.33.005\"]\n",
    "early_pathology = [\"H21.33.005\", \"H20.33.015\", \"H20.33.040\"]\n",
    "# late_pathology = [\"H20.33.020\", \"H20.33.004\", \"H21.33.029\"]\n",
    "# late_pathology = [\"H20.33.020\", \"H21.33.009\"]\n",
    "late_pathology = [\"H20.33.017\", \"H20.33.020\", \"H21.33.029\"]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('/Users/tadeoye/Documents/Research codes/scRNA_seq_meta_analysis/data/raw/SEA-AD/filtered_count_matrix/SEAAD_MTG_RNAseq_final-nuclei.2022-08-18.h5ad')\n",
    "#remove cells from reference patient\n",
    "adata = adata[adata.obs['Donor ID'].isin(no_pathology + early_pathology + late_pathology)]\n",
    "\n",
    "sc.pp.filter_cells(adata, min_genes=500)\n",
    "\n",
    "adata.obs_names_make_unique()\n",
    "adata.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['individualID'] = adata.obs['Donor ID'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['cell_type'] = adata.obs['Subclass'].copy()\n",
    "adata.obs['cell_type'] = adata.obs['cell_type'].astype(str)\n",
    "adata.obs.loc[adata.obs['Class'].str.startswith(\"Neuronal:\"), 'cell_type'] = adata.obs['Class'][adata.obs['Class'].str.startswith(\"Neuronal:\")].map({\"Neuronal: Glutamatergic\": \"Excitatory\",\n",
    "                                                                \"Neuronal: GABAergic\": \"Inhibitory\"})\n",
    "\n",
    "adata.obs.loc[adata.obs.cell_type=='Microglia-PVM', 'cell_type'] = 'Microglia'\n",
    "adata.obs['cell_type'] = adata.obs['cell_type'].astype('category')\n",
    "adata = adata[adata.obs['cell_type'].isin(['Astrocyte', 'Endothelial', 'Excitatory', 'Inhibitory', 'Microglia', 'OPC', 'Oligodendrocyte'])]\n",
    "\n",
    "adata = adata[~adata.obs.cell_type.isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping = {**dict(zip(no_pathology, [\"no\"]*len(no_pathology))), \n",
    "           **dict(zip(early_pathology, [\"early\"]*len(early_pathology))),\n",
    "           **dict(zip(late_pathology, [\"late\"]*len(late_pathology)))}\n",
    "\n",
    "adata.obs['pathology.group'] = adata.obs.individualID.map(mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = adata.obs.drop_duplicates('individualID')\n",
    "metadata.to_csv('../data/raw/seaad_mtg/seaad_mtg_metadata.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('../data/raw/seaad_mtg/seaad_mtg_raw_anndata.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.X.astype(np.int32)\n",
    "\n",
    "adata.write_h5ad('../data/raw/seaad_mtg/seaad_mtg_raw_anndata.h5ad')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Using Raw Pre-Mapped Data**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'name', 'versionNumber', 'contentType', 'dataFileSizeBytes',\n",
       "       'createdBy', 'createdOn', 'modifiedBy', 'modifiedOn', 'parentId',\n",
       "       'synapseURL', 'dataFileMD5Hex', 'assay_x', 'consortium', 'dataSubtype',\n",
       "       'dataType', 'fileFormat', 'grant', 'individualID_x', 'isModelSystem',\n",
       "       'isMultiSpecimen', 'isMultiome', 'libraryPrep', 'nucleicAcidSource_x',\n",
       "       'organ_x', 'platform', 'readLength', 'resourceType', 'runType', 'sex_x',\n",
       "       'species_x', 'specimenID', 'study', 'tissue_x', 'path', 'error',\n",
       "       'individualID_y', 'specimenIdSource', 'organ_y', 'tissue_y',\n",
       "       'BrodmannArea', 'sampleStatus', 'tissueWeight', 'tissueVolume',\n",
       "       'nucleicAcidSource_y', 'cellType', 'fastingState', 'isPostMortem',\n",
       "       'samplingAge', 'samplingAgeUnits', 'visitNumber', 'assay_y',\n",
       "       'individualID', 'dataset', 'individualIdSource', 'species_y', 'sex_y',\n",
       "       'race', 'ethnicity', 'yearsEducation', 'ageDeath', 'causeDeath',\n",
       "       'mannerDeath', 'apoe4Status', 'apoeGenotype', 'pmi', 'pH',\n",
       "       'brainWeight', 'diagnosis', 'diagnosisCriteria', 'CERAD', 'Braak',\n",
       "       'Cognitive status', 'Age of dementia diagnosis',\n",
       "       'Consensus clinical diagnosis', 'CASI score', 'Interval from last CASI',\n",
       "       'MMSE score', 'Interval from last MMSE', 'MoCA score',\n",
       "       'Interval from last MoCA', 'ADNC', 'Thal phase', 'Overall CAA Score',\n",
       "       'Lewy body disease pathology', 'Microinfarcts in screening sections',\n",
       "       'Atherosclerosis', 'Arteriolosclerosis', 'LATE-NC stage'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biospecimen_metadata = pd.read_csv('../data/raw/SEA-AD/SEA-AD_metadata/SEA-AD_biospecimen_metadata.csv')\n",
    "biospecimen_metadata = biospecimen_metadata[biospecimen_metadata.assay.str.lower()=='snrnaseq']\n",
    "individual_metadata = pd.read_csv('../data/raw/SEA-AD/SEA-AD_metadata/SEA-AD_individual_metadata.csv')\n",
    "manifest = pd.read_csv('../data/raw/SEA-AD/raw_feature_bc_matrices/manifest_1682425120001270000.csv')\n",
    "\n",
    "all_metadata = manifest.merge(biospecimen_metadata, on='specimenID', how='outer')\n",
    "all_metadata['individualID'] = all_metadata['individualID_y'] \n",
    "all_metadata = all_metadata.merge(individual_metadata, on = 'individualID', how='outer')\n",
    "\n",
    "all_metadata.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_metadata = all_metadata[all_metadata.tissue_x == 'middle temporal gyrus']\n",
    "all_metadata.to_csv('../data/raw/SEA-AD/SEA-AD_metadata/allen_mtg_metadata.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/70/42xwxjh94x5f164djlw1hwpc0000gv/T/ipykernel_69748/1182312209.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filtered_metadata['pathology.group'] = filtered_metadata.individualID.map(mapping)\n",
      "/var/folders/70/42xwxjh94x5f164djlw1hwpc0000gv/T/ipykernel_69748/1182312209.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filtered_metadata['path'] = filtered_metadata['path'].str.replace('/Users/temitopeleke/Documents/Research Documents/Research codes/mathys_reproduce/', '../')\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "mapping = {**dict(zip(no_pathology, [\"no\"]*len(no_pathology))), \n",
    "           **dict(zip(early_pathology, [\"early\"]*len(early_pathology))),\n",
    "           **dict(zip(late_pathology, [\"late\"]*len(late_pathology)))}\n",
    "\n",
    "all_groups = [*no_pathology, *early_pathology, *late_pathology]\n",
    "filtered_metadata = all_metadata[all_metadata.individualID.isin(all_groups)]\n",
    "filtered_metadata['pathology.group'] = filtered_metadata.individualID.map(mapping)\n",
    "filtered_metadata['path'] = filtered_metadata['path'].str.replace('/Users/temitopeleke/Documents/Research Documents/Research codes/mathys_reproduce/', '../')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filtered out 6054 cells that have less than 500 genes expressed\n"
     ]
    }
   ],
   "source": [
    "\n",
    "CountMatrix = collections.namedtuple('CountMatrix', ['feature_ref', 'barcodes', 'matrix'])\n",
    " \n",
    "def get_matrix_from_h5(filename):\n",
    "    with tables.open_file(filename, 'r') as f:\n",
    "        mat_group = f.get_node(f.root, 'matrix')\n",
    "        barcodes = f.get_node(mat_group, 'barcodes').read()\n",
    "        data = getattr(mat_group, 'data').read()\n",
    "        indices = getattr(mat_group, 'indices').read()\n",
    "        indptr = getattr(mat_group, 'indptr').read()\n",
    "        shape = getattr(mat_group, 'shape').read()\n",
    "        matrix = sp_sparse.csc_matrix((data, indices, indptr), shape=shape)\n",
    "         \n",
    "        feature_ref = {}\n",
    "        feature_group = f.get_node(mat_group, 'features')\n",
    "        feature_ids = getattr(feature_group, 'id').read()\n",
    "        feature_names = getattr(feature_group, 'name').read()\n",
    "        feature_types = getattr(feature_group, 'feature_type').read()\n",
    "        feature_ref['id'] = feature_ids\n",
    "        feature_ref['name'] = feature_names\n",
    "        feature_ref['feature_type'] = feature_types\n",
    "        tag_keys = getattr(feature_group, '_all_tag_keys').read()\n",
    "        for key in tag_keys:\n",
    "            key = key.decode(\"utf-8\")\n",
    "            feature_ref[key] = getattr(feature_group, key).read()\n",
    "         \n",
    "        return CountMatrix(feature_ref, barcodes, matrix)\n",
    " \n",
    "\n",
    "adatas = []\n",
    "columns = list(filtered_metadata.columns)\n",
    "\n",
    "for path in list(filtered_metadata.path)[0:]:\n",
    "\n",
    "    filtered_matrix_h5 = path\n",
    "    filtered_feature_bc_matrix = get_matrix_from_h5(filtered_matrix_h5)\n",
    "\n",
    "    adata = AnnData(filtered_feature_bc_matrix.matrix.T)\n",
    "    #adata.X = filtered_feature_bc_matrix.matrix\n",
    "    adata.obs_names = [name.split(\"-\")[0] for name in filtered_feature_bc_matrix.barcodes.astype('U').astype(str)]\n",
    "    adata.var_names = filtered_feature_bc_matrix.feature_ref['name'].astype('U').astype(str)\n",
    "    adata.var[\"id\"] = filtered_feature_bc_matrix.feature_ref['id'].astype('U').astype(str)\n",
    "    adata.var['genome'] = filtered_feature_bc_matrix.feature_ref['genome'].astype('U').astype(str)\n",
    "    adata.var['feature_type'] = filtered_feature_bc_matrix.feature_ref['feature_type'].astype('U').astype(str)\n",
    "    adata.obs[columns] = filtered_metadata.loc[filtered_metadata.path == path, columns].iloc[0]\n",
    "\n",
    "    cell_classes = pd.read_csv(f'../data/raw/SEA-AD/cell_classes/{filtered_metadata.loc[filtered_metadata.path==path, \"specimenID\"].iloc[0]}.csv')\n",
    "    cell_classes['barcodes'] = cell_classes['sample_id'].apply(lambda x: x.split(\"-\")[0])\n",
    "\n",
    "    adata = adata[cell_classes['barcodes']]\n",
    "\n",
    "    adata.obs[list(cell_classes.columns)] = 'None'\n",
    "\n",
    "    adata.obs['Class'] = adata.obs.index.map(dict(zip(cell_classes.barcodes, cell_classes.Class)))\n",
    "    adata.obs['Subclass'] = adata.obs.index.map(dict(zip(cell_classes.barcodes, cell_classes.Subclass)))\n",
    "    adata.obs['Supertype'] = adata.obs.index.map(dict(zip(cell_classes.barcodes, cell_classes.Supertype)))\n",
    "    adata.obs['cell_labels'] = adata.obs['Subclass'].copy()\n",
    "    adata.obs.loc[adata.obs['Class'].str.startswith(\"Neuronal:\"), 'cell_labels'] = adata.obs['Class'][adata.obs['Class'].str.startswith(\"Neuronal:\")].map({\"Neuronal: Glutamatergic\": \"Excitatory\",\n",
    "                                                                 \"Neuronal: GABAergic\": \"Inhibitory\"})\n",
    "    \n",
    "    adata.obs.loc[adata.obs.cell_labels=='Microglia-PVM', 'cell_labels'] = 'Microglia'\n",
    "\n",
    "    # Initial removal of low-quality nuclei\n",
    "    # SEA-AD nuclei with fewer than 500 genes detected were removed upstream\n",
    "         \n",
    "    sc.pp.filter_cells(adata, min_genes=500)\n",
    "    \n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "\n",
    "    adatas.append(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_allen_mtg = anndata.concat(adatas, join='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 123222 × 65852\n",
       "    obs: 'ID', 'name', 'versionNumber', 'contentType', 'dataFileSizeBytes', 'createdBy', 'createdOn', 'modifiedBy', 'modifiedOn', 'parentId', 'synapseURL', 'dataFileMD5Hex', 'assay_x', 'consortium', 'dataSubtype', 'dataType', 'fileFormat', 'grant', 'individualID_x', 'isModelSystem', 'isMultiSpecimen', 'isMultiome', 'libraryPrep', 'nucleicAcidSource_x', 'organ_x', 'platform', 'readLength', 'resourceType', 'runType', 'sex_x', 'species_x', 'specimenID', 'study', 'tissue_x', 'path', 'error', 'individualID_y', 'specimenIdSource', 'organ_y', 'tissue_y', 'BrodmannArea', 'sampleStatus', 'tissueWeight', 'tissueVolume', 'nucleicAcidSource_y', 'cellType', 'fastingState', 'isPostMortem', 'samplingAge', 'samplingAgeUnits', 'visitNumber', 'assay_y', 'individualID', 'dataset', 'individualIdSource', 'species_y', 'sex_y', 'race', 'ethnicity', 'yearsEducation', 'ageDeath', 'causeDeath', 'mannerDeath', 'apoe4Status', 'apoeGenotype', 'pmi', 'pH', 'brainWeight', 'diagnosis', 'diagnosisCriteria', 'CERAD', 'Braak', 'Cognitive status', 'Age of dementia diagnosis', 'Consensus clinical diagnosis', 'CASI score', 'Interval from last CASI', 'MMSE score', 'Interval from last MMSE', 'MoCA score', 'Interval from last MoCA', 'ADNC', 'Thal phase', 'Overall CAA Score', 'Lewy body disease pathology', 'Microinfarcts in screening sections', 'Atherosclerosis', 'Arteriolosclerosis', 'LATE-NC stage', 'pathology.group', 'sample_id', 'Class', 'Subclass', 'Supertype', 'barcodes', 'cell_labels', 'n_genes'"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_allen_mtg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/70/42xwxjh94x5f164djlw1hwpc0000gv/T/ipykernel_69748/3914187288.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filtered_metadata.drop_duplicates(subset='individualID', keep='first', inplace=True)\n"
     ]
    }
   ],
   "source": [
    "adata_allen_mtg.write_h5ad('../data/raw/allen_mtg/allen_mtg_raw_anndata.h5ad')\n",
    "filtered_metadata.to_csv('../data/raw/allen_mtg/filtered_allen_mtg_metadata.csv')\n",
    "filtered_metadata.drop_duplicates(subset='individualID', keep='first', inplace=True)\n",
    "filtered_metadata.to_csv('../data/raw/allen_mtg/allen_mtg_metadata.csv')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Raw RE-MAPPED Data**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome, tadeoye_usf!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import synapseclient \n",
    "\n",
    "syn = synapseclient.Synapse()\n",
    "\n",
    "syn.login()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "results = syn.tableQuery('select * from syn11346063')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[WARNING] /Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/synapseclient/table.py:596: DtypeWarning: Columns (25,29,30,31,41) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(filepath,\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## Get corresponding FastQs to re-mapp.\n",
    "\n",
    "manifest = results.asDataFrame().copy()\n",
    "manifest.study = manifest.study.apply(lambda x: x[0] if (type(x)==list) and (len(x)>0)  else x)\n",
    "manifest.assay = manifest.assay.apply(lambda x: x[0] if (type(x)==list) and (len(x)>0)  else x)\n",
    "manifest = manifest[(manifest.study=='SEA-AD') & (manifest.assay==\"snrnaSeq\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv('../data/raw/allen_mtg/filtered_allen_mtg_metadata.csv')\n",
    "filtered_manifest = manifest[(manifest.specimenID.isin(metadata.specimenID.to_list())) & (manifest.name.str.endswith('fastq.gz'))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>specimenID</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31112893_18_629c707f-a6e5-41b3-8dfd-cd04b3076cdc</th>\n",
       "      <td>syn31112893</td>\n",
       "      <td>M1TX_201022_159_F01</td>\n",
       "      <td>NW_TX0132-6_S01_L003_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31113100_18_5ce461cc-6baa-48aa-bfbe-ba2bda5fce3d</th>\n",
       "      <td>syn31113100</td>\n",
       "      <td>M1TX_201104_155_F01</td>\n",
       "      <td>NW_TX0104-6_S01_L003_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31113284_18_4efe1e3b-ed10-4a5a-9da1-01069cddc8ab</th>\n",
       "      <td>syn31113284</td>\n",
       "      <td>M1TX_201104_157_C01</td>\n",
       "      <td>NW_TX0093-10_S01_L003_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31114433_18_0578ea64-f710-46e0-a2f6-f724fd997966</th>\n",
       "      <td>syn31114433</td>\n",
       "      <td>M1TX_210713_106_C01</td>\n",
       "      <td>NW_TX0155-6_S01_L003_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31114446_18_23ea0e54-09d6-40d1-b91e-9fbe5fe6861f</th>\n",
       "      <td>syn31114446</td>\n",
       "      <td>M1TX_210713_106_D01</td>\n",
       "      <td>NW_TX0155-7_S01_L003_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31114519_18_07e90c5f-2be0-4647-b95b-eeaaaf46496e</th>\n",
       "      <td>syn31114519</td>\n",
       "      <td>M2TX_210309_207_B01</td>\n",
       "      <td>NW_TX0119-6_S01_L003_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31114538_18_e1c24f33-f732-4e1e-8fbb-5d501478b3ea</th>\n",
       "      <td>syn31114538</td>\n",
       "      <td>M2TX_210305_205_F01</td>\n",
       "      <td>NW_TX0127-6_S01_L003_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31114575_18_ba0c435d-d7cb-4a13-a606-8f704b497867</th>\n",
       "      <td>syn31114575</td>\n",
       "      <td>M1TX_210811_167_H01</td>\n",
       "      <td>NW_TX0165-9_S01_L003_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31114606_18_214f8746-7d32-4d16-bda8-9adc9a740204</th>\n",
       "      <td>syn31114606</td>\n",
       "      <td>M1TX_210322_113_E01</td>\n",
       "      <td>NW_TX0122-8_S01_L003_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31114624_18_a6e1a4c4-8d17-4bcb-899a-c5368b45dbfb</th>\n",
       "      <td>syn31114624</td>\n",
       "      <td>M2TX_210305_204_B01</td>\n",
       "      <td>NW_TX0120-4_S01_L003_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31114628_18_52b2059c-1363-469a-97ec-f99474bf8d43</th>\n",
       "      <td>syn31114628</td>\n",
       "      <td>M2TX_210309_208_C01</td>\n",
       "      <td>NW_TX0128-2_S01_L003_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31114634_18_5dfbfc26-4281-4bad-ad0e-8e3b020231cd</th>\n",
       "      <td>syn31114634</td>\n",
       "      <td>M1TX_210811_167_G01</td>\n",
       "      <td>NW_TX0168-3_S01_L003_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31114666_18_48abe9f0-1556-47bd-a8db-e4f8b0e9770d</th>\n",
       "      <td>syn31114666</td>\n",
       "      <td>M2TX_210426_203_B01</td>\n",
       "      <td>NW_TX0132-2_S01_L003_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31114748_18_2baa3af0-0581-489b-88d8-6629203ee3a4</th>\n",
       "      <td>syn31114748</td>\n",
       "      <td>M2TX_210426_204_E01</td>\n",
       "      <td>NW_TX0166-2_S01_L003_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31115143_18_0f2de2f9-38bd-4151-8ee5-a0e24961e12b</th>\n",
       "      <td>syn31115143</td>\n",
       "      <td>M1TX_210322_111_D01</td>\n",
       "      <td>NW_TX0122-7_S01_L003_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31115243_18_bcca1184-3fe0-44ca-979f-d67ff47edda8</th>\n",
       "      <td>syn31115243</td>\n",
       "      <td>M1TX_210322_110_A01</td>\n",
       "      <td>NW_TX0129-5_S01_L003_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31115587_18_415ecafb-9b23-4d9c-b540-3989d82cc2de</th>\n",
       "      <td>syn31115587</td>\n",
       "      <td>M1TX_210914_157_C01</td>\n",
       "      <td>NW_TX0178-1_S01_L003_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31115590_18_d0974081-8ba2-44f5-b9e8-a4fe3a3391b3</th>\n",
       "      <td>syn31115590</td>\n",
       "      <td>M1TX_210914_157_D01</td>\n",
       "      <td>NW_TX0178-2_S01_L003_I1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31121182_17_ef75d951-6793-45f2-978a-dc37ad2dbc22</th>\n",
       "      <td>syn31121182</td>\n",
       "      <td>M1TX_201104_155_F01</td>\n",
       "      <td>NW_TX0104-6_S01_L003_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31121192_17_1be1a723-33c9-48de-b42b-f9ee7bf1d3df</th>\n",
       "      <td>syn31121192</td>\n",
       "      <td>M1TX_201022_159_F01</td>\n",
       "      <td>NW_TX0132-6_S01_L003_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31121285_17_ec07e313-cb53-40eb-b90e-f1c8a5b48119</th>\n",
       "      <td>syn31121285</td>\n",
       "      <td>M1TX_201104_157_C01</td>\n",
       "      <td>NW_TX0093-10_S01_L003_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31141312_16_8b580aa6-a0d8-493a-97e3-25b0545016f1</th>\n",
       "      <td>syn31141312</td>\n",
       "      <td>M2TX_210309_207_B01</td>\n",
       "      <td>NW_TX0119-6_S01_L003_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31141337_16_5f2b1379-a6a6-44fa-ada7-bf6236bc4d3d</th>\n",
       "      <td>syn31141337</td>\n",
       "      <td>M1TX_210811_167_H01</td>\n",
       "      <td>NW_TX0165-9_S01_L003_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31141343_16_46b458b9-5791-4e32-a3c2-6209c23fe47d</th>\n",
       "      <td>syn31141343</td>\n",
       "      <td>M1TX_210713_106_C01</td>\n",
       "      <td>NW_TX0155-6_S01_L003_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31141345_16_ed5839c3-fd0c-4965-8592-d3e4844f6a4a</th>\n",
       "      <td>syn31141345</td>\n",
       "      <td>M2TX_210305_205_F01</td>\n",
       "      <td>NW_TX0127-6_S01_L003_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31141351_16_905d1193-68a8-4f7a-a700-9cf150574099</th>\n",
       "      <td>syn31141351</td>\n",
       "      <td>M2TX_210305_204_B01</td>\n",
       "      <td>NW_TX0120-4_S01_L003_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31141356_16_3a67ca8d-2413-41ab-af54-9bffebe5dd5e</th>\n",
       "      <td>syn31141356</td>\n",
       "      <td>M1TX_210713_106_D01</td>\n",
       "      <td>NW_TX0155-7_S01_L003_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31141365_16_0b4630c0-1044-4d62-a781-61f7bb878ec0</th>\n",
       "      <td>syn31141365</td>\n",
       "      <td>M2TX_210309_208_C01</td>\n",
       "      <td>NW_TX0128-2_S01_L003_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31141389_16_4205b1fd-fc8f-4068-aa60-42fe2d01923d</th>\n",
       "      <td>syn31141389</td>\n",
       "      <td>M2TX_210426_204_E01</td>\n",
       "      <td>NW_TX0166-2_S01_L003_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31141390_16_5dfb0f9b-8d35-44f0-bb4e-e7eea63a4ce8</th>\n",
       "      <td>syn31141390</td>\n",
       "      <td>M1TX_210811_167_G01</td>\n",
       "      <td>NW_TX0168-3_S01_L003_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31141400_16_a4f0746c-48e9-4a7d-bc80-fe4051eccb59</th>\n",
       "      <td>syn31141400</td>\n",
       "      <td>M2TX_210426_203_B01</td>\n",
       "      <td>NW_TX0132-2_S01_L003_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31141461_15_5d361ba2-3417-4a53-be85-a58ff974a33c</th>\n",
       "      <td>syn31141461</td>\n",
       "      <td>M1TX_210322_113_E01</td>\n",
       "      <td>NW_TX0122-8_S01_L003_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31141507_15_c2f061f8-11ba-41d1-9e5b-bf89c1bc1609</th>\n",
       "      <td>syn31141507</td>\n",
       "      <td>M1TX_210322_111_D01</td>\n",
       "      <td>NW_TX0122-7_S01_L003_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31141761_15_8bb0e74b-9013-4c57-abeb-8c883392d16f</th>\n",
       "      <td>syn31141761</td>\n",
       "      <td>M1TX_210322_110_A01</td>\n",
       "      <td>NW_TX0129-5_S01_L003_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31147577_14_92e2139c-31dc-4a78-b261-ea1056dc19ac</th>\n",
       "      <td>syn31147577</td>\n",
       "      <td>M1TX_210914_157_C01</td>\n",
       "      <td>NW_TX0178-1_S01_L003_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31148893_14_f70d8ec0-31ef-4f6c-9d69-7d1405b601f0</th>\n",
       "      <td>syn31148893</td>\n",
       "      <td>M1TX_210914_157_D01</td>\n",
       "      <td>NW_TX0178-2_S01_L003_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31180975_13_153fcf35-d9e7-4f26-a5dc-e0983fd4775b</th>\n",
       "      <td>syn31180975</td>\n",
       "      <td>M1TX_201104_155_F01</td>\n",
       "      <td>NW_TX0104-6_S01_L003_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31396149_10_e4f3ff0f-99c1-40f3-adb6-1ffdc2bb3c6e</th>\n",
       "      <td>syn31396149</td>\n",
       "      <td>M1TX_201104_157_C01</td>\n",
       "      <td>NW_TX0093-10_S01_L003_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31410644_9_84dcf6b8-9268-4832-8cf0-d77357b4cd58</th>\n",
       "      <td>syn31410644</td>\n",
       "      <td>M1TX_210713_106_C01</td>\n",
       "      <td>NW_TX0155-6_S01_L003_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31422193_9_f052569e-84ad-4a6b-885e-689e29d02d26</th>\n",
       "      <td>syn31422193</td>\n",
       "      <td>M2TX_210305_205_F01</td>\n",
       "      <td>NW_TX0127-6_S01_L003_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31425585_8_fdec2721-ab41-4da1-ba4c-65dd8407c436</th>\n",
       "      <td>syn31425585</td>\n",
       "      <td>M1TX_210811_167_H01</td>\n",
       "      <td>NW_TX0165-9_S01_L003_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31427690_8_1c1bbcb8-d6cd-4937-8072-53eadba110cf</th>\n",
       "      <td>syn31427690</td>\n",
       "      <td>M1TX_210811_167_G01</td>\n",
       "      <td>NW_TX0168-3_S01_L003_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31436641_8_be797a3f-c12c-45ec-bb96-742c28b074cd</th>\n",
       "      <td>syn31436641</td>\n",
       "      <td>M2TX_210305_204_B01</td>\n",
       "      <td>NW_TX0120-4_S01_L003_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31436642_8_627b1d0d-e750-4516-a628-ef64099ec674</th>\n",
       "      <td>syn31436642</td>\n",
       "      <td>M2TX_210426_204_E01</td>\n",
       "      <td>NW_TX0166-2_S01_L003_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31436660_7_56f03e7f-58ea-4719-ae19-ad893ac09658</th>\n",
       "      <td>syn31436660</td>\n",
       "      <td>M2TX_210309_207_B01</td>\n",
       "      <td>NW_TX0119-6_S01_L003_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31436661_7_30eabfde-f755-4d04-b423-41667a285bcf</th>\n",
       "      <td>syn31436661</td>\n",
       "      <td>M2TX_210309_208_C01</td>\n",
       "      <td>NW_TX0128-2_S01_L003_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31436706_7_a7a6a269-fb58-4503-b4fe-6777bf627e9d</th>\n",
       "      <td>syn31436706</td>\n",
       "      <td>M2TX_210426_203_B01</td>\n",
       "      <td>NW_TX0132-2_S01_L003_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31436929_6_88c8fd34-ff64-4909-9719-69c3e9fbb7c6</th>\n",
       "      <td>syn31436929</td>\n",
       "      <td>M1TX_210322_110_A01</td>\n",
       "      <td>NW_TX0129-5_S01_L003_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31436949_6_5d919be9-947c-4bdb-b67c-c9b4cc5053ca</th>\n",
       "      <td>syn31436949</td>\n",
       "      <td>M1TX_210322_111_D01</td>\n",
       "      <td>NW_TX0122-7_S01_L003_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31437196_5_60306370-8e99-472e-a966-c2e728c9dcaa</th>\n",
       "      <td>syn31437196</td>\n",
       "      <td>M1TX_210713_106_D01</td>\n",
       "      <td>NW_TX0155-7_S01_L003_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31447481_3_f204462b-e3d3-4ad1-9400-12a85e3378e7</th>\n",
       "      <td>syn31447481</td>\n",
       "      <td>M1TX_210914_157_D01</td>\n",
       "      <td>NW_TX0178-2_S01_L003_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31448938_3_75d667df-bf7b-460c-9ea8-b558c17c0aea</th>\n",
       "      <td>syn31448938</td>\n",
       "      <td>M1TX_210322_113_E01</td>\n",
       "      <td>NW_TX0122-8_S01_L003_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31451467_2_766c3fa8-e961-4ee4-b8f1-6de382714dad</th>\n",
       "      <td>syn31451467</td>\n",
       "      <td>M1TX_201022_159_F01</td>\n",
       "      <td>NW_TX0132-6_S01_L003_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31451986_2_1a2da913-2bff-466b-9711-34d1afc07a19</th>\n",
       "      <td>syn31451986</td>\n",
       "      <td>M1TX_210914_157_C01</td>\n",
       "      <td>NW_TX0178-1_S01_L003_R2_001.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           id  \\\n",
       "31112893_18_629c707f-a6e5-41b3-8dfd-cd04b3076cdc  syn31112893   \n",
       "31113100_18_5ce461cc-6baa-48aa-bfbe-ba2bda5fce3d  syn31113100   \n",
       "31113284_18_4efe1e3b-ed10-4a5a-9da1-01069cddc8ab  syn31113284   \n",
       "31114433_18_0578ea64-f710-46e0-a2f6-f724fd997966  syn31114433   \n",
       "31114446_18_23ea0e54-09d6-40d1-b91e-9fbe5fe6861f  syn31114446   \n",
       "31114519_18_07e90c5f-2be0-4647-b95b-eeaaaf46496e  syn31114519   \n",
       "31114538_18_e1c24f33-f732-4e1e-8fbb-5d501478b3ea  syn31114538   \n",
       "31114575_18_ba0c435d-d7cb-4a13-a606-8f704b497867  syn31114575   \n",
       "31114606_18_214f8746-7d32-4d16-bda8-9adc9a740204  syn31114606   \n",
       "31114624_18_a6e1a4c4-8d17-4bcb-899a-c5368b45dbfb  syn31114624   \n",
       "31114628_18_52b2059c-1363-469a-97ec-f99474bf8d43  syn31114628   \n",
       "31114634_18_5dfbfc26-4281-4bad-ad0e-8e3b020231cd  syn31114634   \n",
       "31114666_18_48abe9f0-1556-47bd-a8db-e4f8b0e9770d  syn31114666   \n",
       "31114748_18_2baa3af0-0581-489b-88d8-6629203ee3a4  syn31114748   \n",
       "31115143_18_0f2de2f9-38bd-4151-8ee5-a0e24961e12b  syn31115143   \n",
       "31115243_18_bcca1184-3fe0-44ca-979f-d67ff47edda8  syn31115243   \n",
       "31115587_18_415ecafb-9b23-4d9c-b540-3989d82cc2de  syn31115587   \n",
       "31115590_18_d0974081-8ba2-44f5-b9e8-a4fe3a3391b3  syn31115590   \n",
       "31121182_17_ef75d951-6793-45f2-978a-dc37ad2dbc22  syn31121182   \n",
       "31121192_17_1be1a723-33c9-48de-b42b-f9ee7bf1d3df  syn31121192   \n",
       "31121285_17_ec07e313-cb53-40eb-b90e-f1c8a5b48119  syn31121285   \n",
       "31141312_16_8b580aa6-a0d8-493a-97e3-25b0545016f1  syn31141312   \n",
       "31141337_16_5f2b1379-a6a6-44fa-ada7-bf6236bc4d3d  syn31141337   \n",
       "31141343_16_46b458b9-5791-4e32-a3c2-6209c23fe47d  syn31141343   \n",
       "31141345_16_ed5839c3-fd0c-4965-8592-d3e4844f6a4a  syn31141345   \n",
       "31141351_16_905d1193-68a8-4f7a-a700-9cf150574099  syn31141351   \n",
       "31141356_16_3a67ca8d-2413-41ab-af54-9bffebe5dd5e  syn31141356   \n",
       "31141365_16_0b4630c0-1044-4d62-a781-61f7bb878ec0  syn31141365   \n",
       "31141389_16_4205b1fd-fc8f-4068-aa60-42fe2d01923d  syn31141389   \n",
       "31141390_16_5dfb0f9b-8d35-44f0-bb4e-e7eea63a4ce8  syn31141390   \n",
       "31141400_16_a4f0746c-48e9-4a7d-bc80-fe4051eccb59  syn31141400   \n",
       "31141461_15_5d361ba2-3417-4a53-be85-a58ff974a33c  syn31141461   \n",
       "31141507_15_c2f061f8-11ba-41d1-9e5b-bf89c1bc1609  syn31141507   \n",
       "31141761_15_8bb0e74b-9013-4c57-abeb-8c883392d16f  syn31141761   \n",
       "31147577_14_92e2139c-31dc-4a78-b261-ea1056dc19ac  syn31147577   \n",
       "31148893_14_f70d8ec0-31ef-4f6c-9d69-7d1405b601f0  syn31148893   \n",
       "31180975_13_153fcf35-d9e7-4f26-a5dc-e0983fd4775b  syn31180975   \n",
       "31396149_10_e4f3ff0f-99c1-40f3-adb6-1ffdc2bb3c6e  syn31396149   \n",
       "31410644_9_84dcf6b8-9268-4832-8cf0-d77357b4cd58   syn31410644   \n",
       "31422193_9_f052569e-84ad-4a6b-885e-689e29d02d26   syn31422193   \n",
       "31425585_8_fdec2721-ab41-4da1-ba4c-65dd8407c436   syn31425585   \n",
       "31427690_8_1c1bbcb8-d6cd-4937-8072-53eadba110cf   syn31427690   \n",
       "31436641_8_be797a3f-c12c-45ec-bb96-742c28b074cd   syn31436641   \n",
       "31436642_8_627b1d0d-e750-4516-a628-ef64099ec674   syn31436642   \n",
       "31436660_7_56f03e7f-58ea-4719-ae19-ad893ac09658   syn31436660   \n",
       "31436661_7_30eabfde-f755-4d04-b423-41667a285bcf   syn31436661   \n",
       "31436706_7_a7a6a269-fb58-4503-b4fe-6777bf627e9d   syn31436706   \n",
       "31436929_6_88c8fd34-ff64-4909-9719-69c3e9fbb7c6   syn31436929   \n",
       "31436949_6_5d919be9-947c-4bdb-b67c-c9b4cc5053ca   syn31436949   \n",
       "31437196_5_60306370-8e99-472e-a966-c2e728c9dcaa   syn31437196   \n",
       "31447481_3_f204462b-e3d3-4ad1-9400-12a85e3378e7   syn31447481   \n",
       "31448938_3_75d667df-bf7b-460c-9ea8-b558c17c0aea   syn31448938   \n",
       "31451467_2_766c3fa8-e961-4ee4-b8f1-6de382714dad   syn31451467   \n",
       "31451986_2_1a2da913-2bff-466b-9711-34d1afc07a19   syn31451986   \n",
       "\n",
       "                                                           specimenID  \\\n",
       "31112893_18_629c707f-a6e5-41b3-8dfd-cd04b3076cdc  M1TX_201022_159_F01   \n",
       "31113100_18_5ce461cc-6baa-48aa-bfbe-ba2bda5fce3d  M1TX_201104_155_F01   \n",
       "31113284_18_4efe1e3b-ed10-4a5a-9da1-01069cddc8ab  M1TX_201104_157_C01   \n",
       "31114433_18_0578ea64-f710-46e0-a2f6-f724fd997966  M1TX_210713_106_C01   \n",
       "31114446_18_23ea0e54-09d6-40d1-b91e-9fbe5fe6861f  M1TX_210713_106_D01   \n",
       "31114519_18_07e90c5f-2be0-4647-b95b-eeaaaf46496e  M2TX_210309_207_B01   \n",
       "31114538_18_e1c24f33-f732-4e1e-8fbb-5d501478b3ea  M2TX_210305_205_F01   \n",
       "31114575_18_ba0c435d-d7cb-4a13-a606-8f704b497867  M1TX_210811_167_H01   \n",
       "31114606_18_214f8746-7d32-4d16-bda8-9adc9a740204  M1TX_210322_113_E01   \n",
       "31114624_18_a6e1a4c4-8d17-4bcb-899a-c5368b45dbfb  M2TX_210305_204_B01   \n",
       "31114628_18_52b2059c-1363-469a-97ec-f99474bf8d43  M2TX_210309_208_C01   \n",
       "31114634_18_5dfbfc26-4281-4bad-ad0e-8e3b020231cd  M1TX_210811_167_G01   \n",
       "31114666_18_48abe9f0-1556-47bd-a8db-e4f8b0e9770d  M2TX_210426_203_B01   \n",
       "31114748_18_2baa3af0-0581-489b-88d8-6629203ee3a4  M2TX_210426_204_E01   \n",
       "31115143_18_0f2de2f9-38bd-4151-8ee5-a0e24961e12b  M1TX_210322_111_D01   \n",
       "31115243_18_bcca1184-3fe0-44ca-979f-d67ff47edda8  M1TX_210322_110_A01   \n",
       "31115587_18_415ecafb-9b23-4d9c-b540-3989d82cc2de  M1TX_210914_157_C01   \n",
       "31115590_18_d0974081-8ba2-44f5-b9e8-a4fe3a3391b3  M1TX_210914_157_D01   \n",
       "31121182_17_ef75d951-6793-45f2-978a-dc37ad2dbc22  M1TX_201104_155_F01   \n",
       "31121192_17_1be1a723-33c9-48de-b42b-f9ee7bf1d3df  M1TX_201022_159_F01   \n",
       "31121285_17_ec07e313-cb53-40eb-b90e-f1c8a5b48119  M1TX_201104_157_C01   \n",
       "31141312_16_8b580aa6-a0d8-493a-97e3-25b0545016f1  M2TX_210309_207_B01   \n",
       "31141337_16_5f2b1379-a6a6-44fa-ada7-bf6236bc4d3d  M1TX_210811_167_H01   \n",
       "31141343_16_46b458b9-5791-4e32-a3c2-6209c23fe47d  M1TX_210713_106_C01   \n",
       "31141345_16_ed5839c3-fd0c-4965-8592-d3e4844f6a4a  M2TX_210305_205_F01   \n",
       "31141351_16_905d1193-68a8-4f7a-a700-9cf150574099  M2TX_210305_204_B01   \n",
       "31141356_16_3a67ca8d-2413-41ab-af54-9bffebe5dd5e  M1TX_210713_106_D01   \n",
       "31141365_16_0b4630c0-1044-4d62-a781-61f7bb878ec0  M2TX_210309_208_C01   \n",
       "31141389_16_4205b1fd-fc8f-4068-aa60-42fe2d01923d  M2TX_210426_204_E01   \n",
       "31141390_16_5dfb0f9b-8d35-44f0-bb4e-e7eea63a4ce8  M1TX_210811_167_G01   \n",
       "31141400_16_a4f0746c-48e9-4a7d-bc80-fe4051eccb59  M2TX_210426_203_B01   \n",
       "31141461_15_5d361ba2-3417-4a53-be85-a58ff974a33c  M1TX_210322_113_E01   \n",
       "31141507_15_c2f061f8-11ba-41d1-9e5b-bf89c1bc1609  M1TX_210322_111_D01   \n",
       "31141761_15_8bb0e74b-9013-4c57-abeb-8c883392d16f  M1TX_210322_110_A01   \n",
       "31147577_14_92e2139c-31dc-4a78-b261-ea1056dc19ac  M1TX_210914_157_C01   \n",
       "31148893_14_f70d8ec0-31ef-4f6c-9d69-7d1405b601f0  M1TX_210914_157_D01   \n",
       "31180975_13_153fcf35-d9e7-4f26-a5dc-e0983fd4775b  M1TX_201104_155_F01   \n",
       "31396149_10_e4f3ff0f-99c1-40f3-adb6-1ffdc2bb3c6e  M1TX_201104_157_C01   \n",
       "31410644_9_84dcf6b8-9268-4832-8cf0-d77357b4cd58   M1TX_210713_106_C01   \n",
       "31422193_9_f052569e-84ad-4a6b-885e-689e29d02d26   M2TX_210305_205_F01   \n",
       "31425585_8_fdec2721-ab41-4da1-ba4c-65dd8407c436   M1TX_210811_167_H01   \n",
       "31427690_8_1c1bbcb8-d6cd-4937-8072-53eadba110cf   M1TX_210811_167_G01   \n",
       "31436641_8_be797a3f-c12c-45ec-bb96-742c28b074cd   M2TX_210305_204_B01   \n",
       "31436642_8_627b1d0d-e750-4516-a628-ef64099ec674   M2TX_210426_204_E01   \n",
       "31436660_7_56f03e7f-58ea-4719-ae19-ad893ac09658   M2TX_210309_207_B01   \n",
       "31436661_7_30eabfde-f755-4d04-b423-41667a285bcf   M2TX_210309_208_C01   \n",
       "31436706_7_a7a6a269-fb58-4503-b4fe-6777bf627e9d   M2TX_210426_203_B01   \n",
       "31436929_6_88c8fd34-ff64-4909-9719-69c3e9fbb7c6   M1TX_210322_110_A01   \n",
       "31436949_6_5d919be9-947c-4bdb-b67c-c9b4cc5053ca   M1TX_210322_111_D01   \n",
       "31437196_5_60306370-8e99-472e-a966-c2e728c9dcaa   M1TX_210713_106_D01   \n",
       "31447481_3_f204462b-e3d3-4ad1-9400-12a85e3378e7   M1TX_210914_157_D01   \n",
       "31448938_3_75d667df-bf7b-460c-9ea8-b558c17c0aea   M1TX_210322_113_E01   \n",
       "31451467_2_766c3fa8-e961-4ee4-b8f1-6de382714dad   M1TX_201022_159_F01   \n",
       "31451986_2_1a2da913-2bff-466b-9711-34d1afc07a19   M1TX_210914_157_C01   \n",
       "\n",
       "                                                                                   name  \n",
       "31112893_18_629c707f-a6e5-41b3-8dfd-cd04b3076cdc   NW_TX0132-6_S01_L003_I1_001.fastq.gz  \n",
       "31113100_18_5ce461cc-6baa-48aa-bfbe-ba2bda5fce3d   NW_TX0104-6_S01_L003_I1_001.fastq.gz  \n",
       "31113284_18_4efe1e3b-ed10-4a5a-9da1-01069cddc8ab  NW_TX0093-10_S01_L003_I1_001.fastq.gz  \n",
       "31114433_18_0578ea64-f710-46e0-a2f6-f724fd997966   NW_TX0155-6_S01_L003_I1_001.fastq.gz  \n",
       "31114446_18_23ea0e54-09d6-40d1-b91e-9fbe5fe6861f   NW_TX0155-7_S01_L003_I1_001.fastq.gz  \n",
       "31114519_18_07e90c5f-2be0-4647-b95b-eeaaaf46496e   NW_TX0119-6_S01_L003_I1_001.fastq.gz  \n",
       "31114538_18_e1c24f33-f732-4e1e-8fbb-5d501478b3ea   NW_TX0127-6_S01_L003_I1_001.fastq.gz  \n",
       "31114575_18_ba0c435d-d7cb-4a13-a606-8f704b497867   NW_TX0165-9_S01_L003_I1_001.fastq.gz  \n",
       "31114606_18_214f8746-7d32-4d16-bda8-9adc9a740204   NW_TX0122-8_S01_L003_I1_001.fastq.gz  \n",
       "31114624_18_a6e1a4c4-8d17-4bcb-899a-c5368b45dbfb   NW_TX0120-4_S01_L003_I1_001.fastq.gz  \n",
       "31114628_18_52b2059c-1363-469a-97ec-f99474bf8d43   NW_TX0128-2_S01_L003_I1_001.fastq.gz  \n",
       "31114634_18_5dfbfc26-4281-4bad-ad0e-8e3b020231cd   NW_TX0168-3_S01_L003_I1_001.fastq.gz  \n",
       "31114666_18_48abe9f0-1556-47bd-a8db-e4f8b0e9770d   NW_TX0132-2_S01_L003_I1_001.fastq.gz  \n",
       "31114748_18_2baa3af0-0581-489b-88d8-6629203ee3a4   NW_TX0166-2_S01_L003_I1_001.fastq.gz  \n",
       "31115143_18_0f2de2f9-38bd-4151-8ee5-a0e24961e12b   NW_TX0122-7_S01_L003_I1_001.fastq.gz  \n",
       "31115243_18_bcca1184-3fe0-44ca-979f-d67ff47edda8   NW_TX0129-5_S01_L003_I1_001.fastq.gz  \n",
       "31115587_18_415ecafb-9b23-4d9c-b540-3989d82cc2de   NW_TX0178-1_S01_L003_I1_001.fastq.gz  \n",
       "31115590_18_d0974081-8ba2-44f5-b9e8-a4fe3a3391b3   NW_TX0178-2_S01_L003_I1_001.fastq.gz  \n",
       "31121182_17_ef75d951-6793-45f2-978a-dc37ad2dbc22   NW_TX0104-6_S01_L003_R1_001.fastq.gz  \n",
       "31121192_17_1be1a723-33c9-48de-b42b-f9ee7bf1d3df   NW_TX0132-6_S01_L003_R1_001.fastq.gz  \n",
       "31121285_17_ec07e313-cb53-40eb-b90e-f1c8a5b48119  NW_TX0093-10_S01_L003_R1_001.fastq.gz  \n",
       "31141312_16_8b580aa6-a0d8-493a-97e3-25b0545016f1   NW_TX0119-6_S01_L003_R1_001.fastq.gz  \n",
       "31141337_16_5f2b1379-a6a6-44fa-ada7-bf6236bc4d3d   NW_TX0165-9_S01_L003_R1_001.fastq.gz  \n",
       "31141343_16_46b458b9-5791-4e32-a3c2-6209c23fe47d   NW_TX0155-6_S01_L003_R1_001.fastq.gz  \n",
       "31141345_16_ed5839c3-fd0c-4965-8592-d3e4844f6a4a   NW_TX0127-6_S01_L003_R1_001.fastq.gz  \n",
       "31141351_16_905d1193-68a8-4f7a-a700-9cf150574099   NW_TX0120-4_S01_L003_R1_001.fastq.gz  \n",
       "31141356_16_3a67ca8d-2413-41ab-af54-9bffebe5dd5e   NW_TX0155-7_S01_L003_R1_001.fastq.gz  \n",
       "31141365_16_0b4630c0-1044-4d62-a781-61f7bb878ec0   NW_TX0128-2_S01_L003_R1_001.fastq.gz  \n",
       "31141389_16_4205b1fd-fc8f-4068-aa60-42fe2d01923d   NW_TX0166-2_S01_L003_R1_001.fastq.gz  \n",
       "31141390_16_5dfb0f9b-8d35-44f0-bb4e-e7eea63a4ce8   NW_TX0168-3_S01_L003_R1_001.fastq.gz  \n",
       "31141400_16_a4f0746c-48e9-4a7d-bc80-fe4051eccb59   NW_TX0132-2_S01_L003_R1_001.fastq.gz  \n",
       "31141461_15_5d361ba2-3417-4a53-be85-a58ff974a33c   NW_TX0122-8_S01_L003_R1_001.fastq.gz  \n",
       "31141507_15_c2f061f8-11ba-41d1-9e5b-bf89c1bc1609   NW_TX0122-7_S01_L003_R1_001.fastq.gz  \n",
       "31141761_15_8bb0e74b-9013-4c57-abeb-8c883392d16f   NW_TX0129-5_S01_L003_R1_001.fastq.gz  \n",
       "31147577_14_92e2139c-31dc-4a78-b261-ea1056dc19ac   NW_TX0178-1_S01_L003_R1_001.fastq.gz  \n",
       "31148893_14_f70d8ec0-31ef-4f6c-9d69-7d1405b601f0   NW_TX0178-2_S01_L003_R1_001.fastq.gz  \n",
       "31180975_13_153fcf35-d9e7-4f26-a5dc-e0983fd4775b   NW_TX0104-6_S01_L003_R2_001.fastq.gz  \n",
       "31396149_10_e4f3ff0f-99c1-40f3-adb6-1ffdc2bb3c6e  NW_TX0093-10_S01_L003_R2_001.fastq.gz  \n",
       "31410644_9_84dcf6b8-9268-4832-8cf0-d77357b4cd58    NW_TX0155-6_S01_L003_R2_001.fastq.gz  \n",
       "31422193_9_f052569e-84ad-4a6b-885e-689e29d02d26    NW_TX0127-6_S01_L003_R2_001.fastq.gz  \n",
       "31425585_8_fdec2721-ab41-4da1-ba4c-65dd8407c436    NW_TX0165-9_S01_L003_R2_001.fastq.gz  \n",
       "31427690_8_1c1bbcb8-d6cd-4937-8072-53eadba110cf    NW_TX0168-3_S01_L003_R2_001.fastq.gz  \n",
       "31436641_8_be797a3f-c12c-45ec-bb96-742c28b074cd    NW_TX0120-4_S01_L003_R2_001.fastq.gz  \n",
       "31436642_8_627b1d0d-e750-4516-a628-ef64099ec674    NW_TX0166-2_S01_L003_R2_001.fastq.gz  \n",
       "31436660_7_56f03e7f-58ea-4719-ae19-ad893ac09658    NW_TX0119-6_S01_L003_R2_001.fastq.gz  \n",
       "31436661_7_30eabfde-f755-4d04-b423-41667a285bcf    NW_TX0128-2_S01_L003_R2_001.fastq.gz  \n",
       "31436706_7_a7a6a269-fb58-4503-b4fe-6777bf627e9d    NW_TX0132-2_S01_L003_R2_001.fastq.gz  \n",
       "31436929_6_88c8fd34-ff64-4909-9719-69c3e9fbb7c6    NW_TX0129-5_S01_L003_R2_001.fastq.gz  \n",
       "31436949_6_5d919be9-947c-4bdb-b67c-c9b4cc5053ca    NW_TX0122-7_S01_L003_R2_001.fastq.gz  \n",
       "31437196_5_60306370-8e99-472e-a966-c2e728c9dcaa    NW_TX0155-7_S01_L003_R2_001.fastq.gz  \n",
       "31447481_3_f204462b-e3d3-4ad1-9400-12a85e3378e7    NW_TX0178-2_S01_L003_R2_001.fastq.gz  \n",
       "31448938_3_75d667df-bf7b-460c-9ea8-b558c17c0aea    NW_TX0122-8_S01_L003_R2_001.fastq.gz  \n",
       "31451467_2_766c3fa8-e961-4ee4-b8f1-6de382714dad    NW_TX0132-6_S01_L003_R2_001.fastq.gz  \n",
       "31451986_2_1a2da913-2bff-466b-9711-34d1afc07a19    NW_TX0178-1_S01_L003_R2_001.fastq.gz  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_manifest[['id', 'specimenID', 'name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "2f414a000012e1980f5fd1119ed6b9bf8cc73b51e37b52eb52a16ff857effd1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
